// @arkade-os/sdk/dist/cjs/wallet/serviceWorker/wallet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServiceWorkerWallet = exports.ServiceWorkerReadonlyWallet = void 0;
const base_1 = require("@scure/base");
const utils_1 = require("../../worker/browser/utils");
const repositories_1 = require("../../repositories");
const wallet_message_handler_1 = require("./wallet-message-handler");
const utils_2 = require("../utils");
const isPrivateKeyIdentity = (identity) => {
    return typeof identity.toHex === "function";
};
class ServiceWorkerReadonlyAssetManager {
    constructor(sendMessage, messageTag) {
        this.sendMessage = sendMessage;
        this.messageTag = messageTag;
    }
    async getAssetDetails(assetId) {
        const message = {
            tag: this.messageTag,
            type: "GET_ASSET_DETAILS",
            id: (0, utils_2.getRandomId)(),
            payload: { assetId },
        };
        const response = await this.sendMessage(message);
        return response.payload.assetDetails;
    }
}
class ServiceWorkerAssetManager extends ServiceWorkerReadonlyAssetManager {
    async issue(params) {
        const message = {
            tag: this.messageTag,
            type: "ISSUE",
            id: (0, utils_2.getRandomId)(),
            payload: { params },
        };
        const response = await this.sendMessage(message);
        return response.payload.result;
    }
    async reissue(params) {
        const message = {
            tag: this.messageTag,
            type: "REISSUE",
            id: (0, utils_2.getRandomId)(),
            payload: { params },
        };
        const response = await this.sendMessage(message);
        return response.payload.txid;
    }
    async burn(params) {
        const message = {
            tag: this.messageTag,
            type: "BURN",
            id: (0, utils_2.getRandomId)(),
            payload: { params },
        };
        const response = await this.sendMessage(message);
        return response.payload.txid;
    }
}
const initializeMessageBus = (serviceWorker, config, timeoutMs = 2000) => {
    const initCmd = {
        tag: "INITIALIZE_MESSAGE_BUS",
        id: (0, utils_2.getRandomId)(),
        config: { ...config, timeoutMs },
    };
    return new Promise((resolve, reject) => {
        const cleanup = () => {
            navigator.serviceWorker.removeEventListener("message", onMessage);
            clearTimeout(timeoutId);
        };
        const onMessage = (event) => {
            const response = event.data;
            if (response?.id !== initCmd.id)
                return;
            cleanup();
            if (response.error) {
                reject(response.error);
            }
            else {
                resolve();
            }
        };
        const timeoutId = setTimeout(() => {
            cleanup();
            reject(new Error("MessageBus timed out!"));
        }, timeoutMs);
        navigator.serviceWorker.addEventListener("message", onMessage);
        serviceWorker.postMessage(initCmd);
    });
};
class ServiceWorkerReadonlyWallet {
    get assetManager() {
        return this._readonlyAssetManager;
    }
    constructor(serviceWorker, identity, walletRepository, contractRepository, messageTag) {
        this.serviceWorker = serviceWorker;
        this.messageTag = messageTag;
        this.initConfig = null;
        this.initWalletPayload = null;
        this.reinitPromise = null;
        this.identity = identity;
        this.walletRepository = walletRepository;
        this.contractRepository = contractRepository;
        this._readonlyAssetManager = new ServiceWorkerReadonlyAssetManager((msg) => this.sendMessage(msg), messageTag);
    }
    static async create(options) {
        const walletRepository = options.storage?.walletRepository ??
            new repositories_1.IndexedDBWalletRepository();
        const contractRepository = options.storage?.contractRepository ??
            new repositories_1.IndexedDBContractRepository();
        const messageTag = options.walletUpdaterTag ?? wallet_message_handler_1.DEFAULT_MESSAGE_TAG;
        // Create the wallet instance
        const wallet = new ServiceWorkerReadonlyWallet(options.serviceWorker, options.identity, walletRepository, contractRepository, messageTag);
        const publicKey = await options.identity
            .compressedPublicKey()
            .then(base_1.hex.encode);
        const initConfig = {
            key: { publicKey },
            arkServerUrl: options.arkServerUrl,
            arkServerPublicKey: options.arkServerPublicKey,
            delegatorUrl: options.delegatorUrl,
        };
        // Bootstrap the MessageBus in the service worker
        await initializeMessageBus(options.serviceWorker, {
            wallet: initConfig.key,
            arkServer: {
                url: initConfig.arkServerUrl,
                publicKey: initConfig.arkServerPublicKey,
            },
            delegatorUrl: initConfig.delegatorUrl,
            timeoutMs: options.messageBusTimeoutMs,
        }, options.messageBusTimeoutMs);
        // Initialize the wallet handler
        const initMessage = {
            tag: messageTag,
            type: "INIT_WALLET",
            id: (0, utils_2.getRandomId)(),
            payload: initConfig,
        };
        await wallet.sendMessage(initMessage);
        wallet.initConfig = {
            wallet: initConfig.key,
            arkServer: {
                url: initConfig.arkServerUrl,
                publicKey: initConfig.arkServerPublicKey,
            },
            delegatorUrl: initConfig.delegatorUrl,
        };
        wallet.initWalletPayload = initConfig;
        wallet.messageBusTimeoutMs = options.messageBusTimeoutMs;
        return wallet;
    }
    /**
     * Simplified setup method that handles service worker registration,
     * identity creation, and wallet initialization automatically.
     *
     * @example
     * ```typescript
     * // One-liner setup - handles everything automatically!
     * const wallet = await ServiceWorkerReadonlyWallet.setup({
     *   serviceWorkerPath: '/service-worker.js',
     *   arkServerUrl: 'https://mutinynet.arkade.sh'
     * });
     *
     * // With custom readonly identity
     * const identity = ReadonlySingleKey.fromPublicKey('your_public_key_hex');
     * const wallet = await ServiceWorkerReadonlyWallet.setup({
     *   serviceWorkerPath: '/service-worker.js',
     *   arkServerUrl: 'https://mutinynet.arkade.sh',
     *   identity
     * });
     * ```
     */
    static async setup(options) {
        // Register and setup the service worker
        const serviceWorker = await (0, utils_1.setupServiceWorker)({
            path: options.serviceWorkerPath,
            activationTimeoutMs: options.serviceWorkerActivationTimeoutMs,
        });
        // Use the existing create method
        return await ServiceWorkerReadonlyWallet.create({
            ...options,
            serviceWorker,
        });
    }
    sendMessageDirect(request) {
        return new Promise((resolve, reject) => {
            const cleanup = () => {
                clearTimeout(timeoutId);
                navigator.serviceWorker.removeEventListener("message", messageHandler);
            };
            const timeoutId = setTimeout(() => {
                cleanup();
                reject(new Error(`Service worker message timed out (${request.type})`));
            }, 30000);
            const messageHandler = (event) => {
                const response = event.data;
                if (request.id !== response.id) {
                    return;
                }
                cleanup();
                if (response.error) {
                    reject(response.error);
                }
                else {
                    resolve(response);
                }
            };
            navigator.serviceWorker.addEventListener("message", messageHandler);
            this.serviceWorker.postMessage(request);
        });
    }
    // send a message, retrying up to 2 times if the service worker was
    // killed and restarted by the OS (mobile browsers do this aggressively)
    async sendMessage(request) {
        const maxRetries = 2;
        for (let attempt = 0;; attempt++) {
            try {
                return await this.sendMessageDirect(request);
            }
            catch (error) {
                const isNotInitialized = typeof error?.message === "string" &&
                    error.message.includes("MessageBus not initialized");
                if (!isNotInitialized || attempt >= maxRetries) {
                    throw error;
                }
                await this.reinitialize();
            }
        }
    }
    async reinitialize() {
        if (this.reinitPromise)
            return this.reinitPromise;
        this.reinitPromise = (async () => {
            if (!this.initConfig || !this.initWalletPayload) {
                throw new Error("Cannot re-initialize: missing configuration");
            }
            await initializeMessageBus(this.serviceWorker, this.initConfig, this.messageBusTimeoutMs);
            const initMessage = {
                tag: this.messageTag,
                type: "INIT_WALLET",
                id: (0, utils_2.getRandomId)(),
                payload: this.initWalletPayload,
            };
            await this.sendMessageDirect(initMessage);
        })().finally(() => {
            this.reinitPromise = null;
        });
        return this.reinitPromise;
    }
    async clear() {
        const message = {
            id: (0, utils_2.getRandomId)(),
            tag: this.messageTag,
            type: "CLEAR",
        };
        // Clear page-side storage to maintain parity with SW
        try {
            const address = await this.getAddress();
            await this.walletRepository.deleteVtxos(address);
        }
        catch (_) {
            console.warn("Failed to clear vtxos from wallet repository");
        }
        await this.sendMessage(message);
    }
    async getAddress() {
        const message = {
            id: (0, utils_2.getRandomId)(),
            tag: this.messageTag,
            type: "GET_ADDRESS",
        };
        try {
            const response = await this.sendMessage(message);
            return response.payload.address;
        }
        catch (error) {
            throw new Error(`Failed to get address: ${error}`);
        }
    }
    async getBoardingAddress() {
        const message = {
            id: (0, utils_2.getRandomId)(),
            tag: this.messageTag,
            type: "GET_BOARDING_ADDRESS",
        };
        try {
            const response = await this.sendMessage(message);
            return response.payload.address;
        }
        catch (error) {
            throw new Error(`Failed to get boarding address: ${error}`);
        }
    }
    async getBalance() {
        const message = {
            id: (0, utils_2.getRandomId)(),
            tag: this.messageTag,
            type: "GET_BALANCE",
        };
        try {
            const response = await this.sendMessage(message);
            return response.payload;
        }
        catch (error) {
            throw new Error(`Failed to get balance: ${error}`);
        }
    }
    async getBoardingUtxos() {
        const message = {
            id: (0, utils_2.getRandomId)(),
            tag: this.messageTag,
            type: "GET_BOARDING_UTXOS",
        };
        try {
            const response = await this.sendMessage(message);
            return response.payload.utxos;
        }
        catch (error) {
            throw new Error(`Failed to get boarding UTXOs: ${error}`);
        }
    }
    async getStatus() {
        const message = {
            id: (0, utils_2.getRandomId)(),
            tag: this.messageTag,
            type: "GET_STATUS",
        };
        try {
            const response = await this.sendMessage(message);
            return response.payload;
        }
        catch (error) {
            throw new Error(`Failed to get status: ${error}`);
        }
    }
    async getTransactionHistory() {
        const message = {
            id: (0, utils_2.getRandomId)(),
            tag: this.messageTag,
            type: "GET_TRANSACTION_HISTORY",
        };
        try {
            const response = await this.sendMessage(message);
            return response.payload
                .transactions;
        }
        catch (error) {
            throw new Error(`Failed to get transaction history: ${error}`);
        }
    }
    async getVtxos(filter) {
        const message = {
            id: (0, utils_2.getRandomId)(),
            tag: this.messageTag,
            type: "GET_VTXOS",
            payload: { filter },
        };
        try {
            const response = await this.sendMessage(message);
            return response.payload.vtxos;
        }
        catch (error) {
            throw new Error(`Failed to get vtxos: ${error}`);
        }
    }
    async reload() {
        const message = {
            id: (0, utils_2.getRandomId)(),
            tag: this.messageTag,
            type: "RELOAD_WALLET",
        };
        try {
            const response = await this.sendMessage(message);
            return response.payload.reloaded;
        }
        catch (error) {
            throw new Error(`Failed to reload wallet: ${error}`);
        }
    }
    async getContractManager() {
        const wallet = this;
        const sendContractMessage = async (message) => {
            return wallet.sendMessage(message);
        };
        const messageTag = this.messageTag;
        const manager = {
            async createContract(params) {
                const message = {
                    type: "CREATE_CONTRACT",
                    id: (0, utils_2.getRandomId)(),
                    tag: messageTag,
                    payload: params,
                };
                try {
                    const response = await sendContractMessage(message);
                    return response.payload
                        .contract;
                }
                catch (e) {
                    throw new Error("Failed to create contract");
                }
            },
            async getContracts(filter) {
                const message = {
                    type: "GET_CONTRACTS",
                    id: (0, utils_2.getRandomId)(),
                    tag: messageTag,
                    payload: { filter },
                };
                try {
                    const response = await sendContractMessage(message);
                    return response.payload.contracts;
                }
                catch (e) {
                    throw new Error("Failed to get contracts");
                }
            },
            async getContractsWithVtxos(filter) {
                const message = {
                    type: "GET_CONTRACTS_WITH_VTXOS",
                    id: (0, utils_2.getRandomId)(),
                    tag: messageTag,
                    payload: { filter },
                };
                try {
                    const response = await sendContractMessage(message);
                    return response.payload
                        .contracts;
                }
                catch (e) {
                    throw new Error("Failed to get contracts with vtxos");
                }
            },
            async updateContract(script, updates) {
                const message = {
                    type: "UPDATE_CONTRACT",
                    id: (0, utils_2.getRandomId)(),
                    tag: messageTag,
                    payload: { script, updates },
                };
                try {
                    const response = await sendContractMessage(message);
                    return response.payload
                        .contract;
                }
                catch (e) {
                    throw new Error("Failed to update contract");
                }
            },
            async setContractState(script, state) {
                const message = {
                    type: "UPDATE_CONTRACT",
                    id: (0, utils_2.getRandomId)(),
                    tag: messageTag,
                    payload: { script, updates: { state } },
                };
                try {
                    await sendContractMessage(message);
                    return;
                }
                catch (e) {
                    throw new Error("Failed to update contract state");
                }
            },
            async deleteContract(script) {
                const message = {
                    type: "DELETE_CONTRACT",
                    id: (0, utils_2.getRandomId)(),
                    tag: messageTag,
                    payload: { script },
                };
                try {
                    await sendContractMessage(message);
                    return;
                }
                catch (e) {
                    throw new Error("Failed to delete contract");
                }
            },
            async getSpendablePaths(options) {
                const message = {
                    type: "GET_SPENDABLE_PATHS",
                    id: (0, utils_2.getRandomId)(),
                    tag: messageTag,
                    payload: { options },
                };
                try {
                    const response = await sendContractMessage(message);
                    return response.payload
                        .paths;
                }
                catch (e) {
                    throw new Error("Failed to get spendable paths");
                }
            },
            async getAllSpendingPaths(options) {
                const message = {
                    type: "GET_ALL_SPENDING_PATHS",
                    id: (0, utils_2.getRandomId)(),
                    tag: messageTag,
                    payload: { options },
                };
                try {
                    const response = await sendContractMessage(message);
                    return response.payload
                        .paths;
                }
                catch (e) {
                    throw new Error("Failed to get all spending paths");
                }
            },
            onContractEvent(callback) {
                const messageHandler = (event) => {
                    const response = event.data;
                    if (response.type !== "CONTRACT_EVENT") {
                        return;
                    }
                    if (response.tag !== messageTag) {
                        return;
                    }
                    callback(response.payload.event);
                };
                navigator.serviceWorker.addEventListener("message", messageHandler);
                return () => {
                    navigator.serviceWorker.removeEventListener("message", messageHandler);
                };
            },
            async isWatching() {
                const message = {
                    type: "IS_CONTRACT_MANAGER_WATCHING",
                    id: (0, utils_2.getRandomId)(),
                    tag: messageTag,
                };
                try {
                    const response = await sendContractMessage(message);
                    return response
                        .payload.isWatching;
                }
                catch (e) {
                    throw new Error("Failed to check if contract manager is watching");
                }
            },
            dispose() {
                return;
            },
            [Symbol.dispose]() {
                // no-op
                return;
            },
        };
        return manager;
    }
}
exports.ServiceWorkerReadonlyWallet = ServiceWorkerReadonlyWallet;
class ServiceWorkerWallet extends ServiceWorkerReadonlyWallet {
    constructor(serviceWorker, identity, walletRepository, contractRepository, messageTag, hasDelegator) {
        super(serviceWorker, identity, walletRepository, contractRepository, messageTag);
        this.serviceWorker = serviceWorker;
        this.identity = identity;
        this.walletRepository = walletRepository;
        this.contractRepository = contractRepository;
        this._assetManager = new ServiceWorkerAssetManager((msg) => this.sendMessage(msg), messageTag);
        this.hasDelegator = hasDelegator;
    }
    get assetManager() {
        return this._assetManager;
    }
    static async create(options) {
        const walletRepository = options.storage?.walletRepository ??
            new repositories_1.IndexedDBWalletRepository();
        const contractRepository = options.storage?.contractRepository ??
            new repositories_1.IndexedDBContractRepository();
        // Extract identity and check if it can expose private key
        const identity = isPrivateKeyIdentity(options.identity)
            ? options.identity
            : null;
        if (!identity) {
            throw new Error("ServiceWorkerWallet.create() requires a Identity that can expose a single private key");
        }
        // Extract private key for service worker initialization
        const privateKey = identity.toHex();
        const messageTag = options.walletUpdaterTag ?? wallet_message_handler_1.DEFAULT_MESSAGE_TAG;
        // Create the wallet instance
        const wallet = new ServiceWorkerWallet(options.serviceWorker, identity, walletRepository, contractRepository, messageTag, !!options.delegatorUrl);
        const initConfig = {
            key: { privateKey },
            arkServerUrl: options.arkServerUrl,
            arkServerPublicKey: options.arkServerPublicKey,
            delegatorUrl: options.delegatorUrl,
        };
        await initializeMessageBus(options.serviceWorker, {
            wallet: initConfig.key,
            arkServer: {
                url: initConfig.arkServerUrl,
                publicKey: initConfig.arkServerPublicKey,
            },
            delegatorUrl: initConfig.delegatorUrl,
            timeoutMs: options.messageBusTimeoutMs,
        }, options.messageBusTimeoutMs);
        // Initialize the service worker with the config
        const initMessage = {
            tag: messageTag,
            type: "INIT_WALLET",
            id: (0, utils_2.getRandomId)(),
            payload: initConfig,
        };
        // Initialize the service worker
        await wallet.sendMessage(initMessage);
        wallet.initConfig = {
            wallet: initConfig.key,
            arkServer: {
                url: initConfig.arkServerUrl,
                publicKey: initConfig.arkServerPublicKey,
            },
            delegatorUrl: initConfig.delegatorUrl,
        };
        wallet.initWalletPayload = initConfig;
        wallet.messageBusTimeoutMs = options.messageBusTimeoutMs;
        return wallet;
    }
    /**
     * Simplified setup method that handles service worker registration,
     * identity creation, and wallet initialization automatically.
     *
     * @example
     * ```typescript
     * // One-liner setup - handles everything automatically!
     * const wallet = await ServiceWorkerWallet.setup({
     *   serviceWorkerPath: '/service-worker.js',
     *   arkServerUrl: 'https://mutinynet.arkade.sh'
     * });
     *
     * // With custom identity
     * const identity = SingleKey.fromHex('your_private_key_hex');
     * const wallet = await ServiceWorkerWallet.setup({
     *   serviceWorkerPath: '/service-worker.js',
     *   arkServerUrl: 'https://mutinynet.arkade.sh',
     *   identity
     * });
     * ```
     */
    static async setup(options) {
        // Register and setup the service worker
        const serviceWorker = await (0, utils_1.setupServiceWorker)({
            path: options.serviceWorkerPath,
            activationTimeoutMs: options.serviceWorkerActivationTimeoutMs,
        });
        // Use the existing create method
        return ServiceWorkerWallet.create({
            ...options,
            serviceWorker,
        });
    }
    async sendBitcoin(params) {
        const message = {
            id: (0, utils_2.getRandomId)(),
            tag: this.messageTag,
            type: "SEND_BITCOIN",
            payload: params,
        };
        try {
            const response = await this.sendMessage(message);
            return response.payload.txid;
        }
        catch (error) {
            throw new Error(`Failed to send bitcoin: ${error}`);
        }
    }
    async settle(params, callback) {
        const message = {
            id: (0, utils_2.getRandomId)(),
            tag: this.messageTag,
            type: "SETTLE",
            payload: { params },
        };
        try {
            return new Promise((resolve, reject) => {
                const messageHandler = (event) => {
                    const response = event.data;
                    if (response.id !== message.id) {
                        return;
                    }
                    if (response.error) {
                        navigator.serviceWorker.removeEventListener("message", messageHandler);
                        reject(response.error);
                        return;
                    }
                    switch (response.type) {
                        case "SETTLE_EVENT":
                            if (callback) {
                                callback(response.payload);
                            }
                            break;
                        case "SETTLE_SUCCESS":
                            navigator.serviceWorker.removeEventListener("message", messageHandler);
                            resolve(response.payload.txid);
                            break;
                        default:
                            console.error(`Unexpected response type for SETTLE request: ${response.type}`);
                    }
                };
                navigator.serviceWorker.addEventListener("message", messageHandler);
                this.serviceWorker.postMessage(message);
            });
        }
        catch (error) {
            throw new Error(`Settlement failed: ${error}`);
        }
    }
    async send(...recipients) {
        const message = {
            tag: this.messageTag,
            type: "SEND",
            id: (0, utils_2.getRandomId)(),
            payload: { recipients },
        };
        try {
            const response = await this.sendMessage(message);
            return response.payload.txid;
        }
        catch (error) {
            throw new Error(`Send failed: ${error}`);
        }
    }
    async getDelegatorManager() {
        if (!this.hasDelegator) {
            return undefined;
        }
        const wallet = this;
        const messageTag = this.messageTag;
        const manager = {
            async delegate(vtxos, destination, delegateAt) {
                const message = {
                    tag: messageTag,
                    type: "DELEGATE",
                    id: (0, utils_2.getRandomId)(),
                    payload: {
                        vtxoOutpoints: vtxos.map((v) => ({
                            txid: v.txid,
                            vout: v.vout,
                        })),
                        destination,
                        delegateAt: delegateAt?.getTime(),
                    },
                };
                try {
                    const response = await wallet.sendMessage(message);
                    const payload = response.payload;
                    return {
                        delegated: payload.delegated,
                        failed: payload.failed.map((f) => ({
                            outpoints: f.outpoints,
                            error: f.error,
                        })),
                    };
                }
                catch (error) {
                    throw new Error(`Delegation failed: ${error}`);
                }
            },
            async getDelegateInfo() {
                const message = {
                    type: "GET_DELEGATE_INFO",
                    id: (0, utils_2.getRandomId)(),
                    tag: messageTag,
                };
                try {
                    const response = await wallet.sendMessage(message);
                    return response.payload.info;
                }
                catch (e) {
                    throw new Error("Failed to get delegate info");
                }
            },
        };
        return manager;
    }
}
exports.ServiceWorkerWallet = ServiceWorkerWallet;
