// @arkade-os/sdk/dist/cjs/contracts/handlers/delegate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DelegateContractHandler = void 0;
const base_1 = require("@scure/base");
const delegate_1 = require("../../script/delegate");
const default_1 = require("../../script/default");
const helpers_1 = require("./helpers");
/**
 * Handler for delegate wallet VTXOs.
 *
 * Delegate contracts extend the default tapscript with an additional delegate path:
 * - forfeit: (Alice + Server) multisig for collaborative spending
 * - exit: (Alice) + CSV timelock for unilateral exit
 * - delegate: (Alice + Delegate + Server) multisig for delegated renewal
 */
exports.DelegateContractHandler = {
    type: "delegate",
    createScript(params) {
        const typed = this.deserializeParams(params);
        return new delegate_1.DelegateVtxo.Script(typed);
    },
    serializeParams(params) {
        return {
            pubKey: base_1.hex.encode(params.pubKey),
            serverPubKey: base_1.hex.encode(params.serverPubKey),
            delegatePubKey: base_1.hex.encode(params.delegatePubKey),
            csvTimelock: (0, helpers_1.timelockToSequence)(params.csvTimelock).toString(),
        };
    },
    deserializeParams(params) {
        const csvTimelock = params.csvTimelock
            ? (0, helpers_1.sequenceToTimelock)(Number(params.csvTimelock))
            : default_1.DefaultVtxo.Script.DEFAULT_TIMELOCK;
        return {
            pubKey: base_1.hex.decode(params.pubKey),
            serverPubKey: base_1.hex.decode(params.serverPubKey),
            delegatePubKey: base_1.hex.decode(params.delegatePubKey),
            csvTimelock,
        };
    },
    selectPath(script, contract, context) {
        if (context.collaborative) {
            return { leaf: script.forfeit() };
        }
        const sequence = contract.params.csvTimelock
            ? Number(contract.params.csvTimelock)
            : undefined;
        if (!(0, helpers_1.isCsvSpendable)(context, sequence)) {
            return null;
        }
        return {
            leaf: script.exit(),
            sequence,
        };
    },
    getAllSpendingPaths(script, contract, context) {
        const paths = [];
        if (context.collaborative) {
            paths.push({ leaf: script.forfeit() });
        }
        const exitPath = { leaf: script.exit() };
        if (contract.params.csvTimelock) {
            exitPath.sequence = Number(contract.params.csvTimelock);
        }
        paths.push(exitPath);
        // Delegate path (Alice + Delegate + Server) — collaborative only
        if (context.collaborative) {
            paths.push({ leaf: script.delegate() });
        }
        return paths;
    },
    getSpendablePaths(script, contract, context) {
        const paths = [];
        if (context.collaborative) {
            paths.push({ leaf: script.forfeit() });
        }
        const exitSequence = contract.params.csvTimelock
            ? Number(contract.params.csvTimelock)
            : undefined;
        if ((0, helpers_1.isCsvSpendable)(context, exitSequence)) {
            const exitPath = { leaf: script.exit() };
            if (exitSequence !== undefined) {
                exitPath.sequence = exitSequence;
            }
            paths.push(exitPath);
        }
        return paths;
    },
};
