// @arkade-os/sdk/dist/esm/worker/expo/asyncStorageTaskQueue.js
/**
 * AsyncStorage-backed TaskQueue for Expo/React Native.
 *
 * Persists inbox, outbox, and an optional config blob to AsyncStorage
 * so that data survives process restarts and can be shared between
 * foreground and background execution contexts.
 */
export class AsyncStorageTaskQueue {
    constructor(storage, prefix = "ark:task-queue") {
        this.storage = storage;
        this.inboxKey = `${prefix}:inbox`;
        this.outboxKey = `${prefix}:outbox`;
        this.configKey = `${prefix}:config`;
    }
    // ── Inbox ────────────────────────────────────────────────────────
    async addTask(task) {
        const tasks = await this.readList(this.inboxKey);
        tasks.push(task);
        await this.writeList(this.inboxKey, tasks);
    }
    async removeTask(id) {
        const tasks = await this.readList(this.inboxKey);
        await this.writeList(this.inboxKey, tasks.filter((t) => t.id !== id));
    }
    async getTasks(type) {
        const tasks = await this.readList(this.inboxKey);
        if (type) {
            return tasks.filter((t) => t.type === type);
        }
        return tasks;
    }
    async clearTasks() {
        await this.storage.removeItem(this.inboxKey);
    }
    // ── Outbox ───────────────────────────────────────────────────────
    async pushResult(result) {
        const results = await this.readList(this.outboxKey);
        results.push(result);
        await this.writeList(this.outboxKey, results);
    }
    async getResults() {
        return this.readList(this.outboxKey);
    }
    async acknowledgeResults(ids) {
        const idSet = new Set(ids);
        const results = await this.readList(this.outboxKey);
        await this.writeList(this.outboxKey, results.filter((r) => !idSet.has(r.id)));
    }
    // ── Config persistence (for background handler rehydration) ──────
    /**
     * Persist a config blob alongside the queue data.
     * Used by {@link ExpoWallet.setup} to store the wallet parameters
     * that the background handler needs to reconstruct providers.
     */
    async persistConfig(config) {
        await this.storage.setItem(this.configKey, JSON.stringify(config));
    }
    /**
     * Load the persisted config blob.
     * Used by the background handler to rehydrate wallet dependencies.
     */
    async loadConfig() {
        const raw = await this.storage.getItem(this.configKey);
        return raw ? JSON.parse(raw) : null;
    }
    // ── Helpers ──────────────────────────────────────────────────────
    async readList(key) {
        const raw = await this.storage.getItem(key);
        return raw ? JSON.parse(raw) : [];
    }
    async writeList(key, list) {
        await this.storage.setItem(key, JSON.stringify(list));
    }
}
