// @arkade-os/sdk/dist/cjs/script/delegate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DelegateVtxo = void 0;
const default_1 = require("./default");
const tapscript_1 = require("./tapscript");
const base_1 = require("./base");
const base_2 = require("@scure/base");
/**
 * DelegateVtxo extends DefaultVtxo with an extra delegator path
 */
var DelegateVtxo;
(function (DelegateVtxo) {
    /**
     * DelegateVtxo.Script extends DefaultVtxo.Script and adds a delegate path.
     * @example
     * ```typescript
     * const vtxoScript = new DelegateVtxo.Script({
     *     pubKey: new Uint8Array(32),
     *     serverPubKey: new Uint8Array(32),
     *     delegatePubKey: new Uint8Array(32),
     * });
     *
     * console.log("script pub key:", vtxoScript.pkScript)
     * ```
     */
    class Script extends base_1.VtxoScript {
        constructor(options) {
            const defaultVtxo = new default_1.DefaultVtxo.Script(options);
            const { delegatePubKey, pubKey, serverPubKey } = options;
            const delegateScript = tapscript_1.MultisigTapscript.encode({
                pubkeys: [pubKey, delegatePubKey, serverPubKey],
            }).script;
            super([...defaultVtxo.scripts, delegateScript]);
            this.options = options;
            this.defaultVtxo = defaultVtxo;
            this.delegateScript = base_2.hex.encode(delegateScript);
        }
        forfeit() {
            return this.findLeaf(this.defaultVtxo.forfeitScript);
        }
        exit() {
            return this.findLeaf(this.defaultVtxo.exitScript);
        }
        delegate() {
            return this.findLeaf(this.delegateScript);
        }
    }
    DelegateVtxo.Script = Script;
})(DelegateVtxo || (exports.DelegateVtxo = DelegateVtxo = {}));
