// @arkade-os/sdk/dist/cjs/extension/asset/assetRef.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AssetRef = void 0;
const base_1 = require("@scure/base");
const types_1 = require("./types");
const assetId_1 = require("./assetId");
const utils_1 = require("./utils");
class AssetRef {
    constructor(ref) {
        this.ref = ref;
    }
    get type() {
        return this.ref.type;
    }
    static fromId(assetId) {
        return new AssetRef({ type: types_1.AssetRefType.ByID, assetId });
    }
    static fromGroupIndex(groupIndex) {
        return new AssetRef({ type: types_1.AssetRefType.ByGroup, groupIndex });
    }
    static fromString(s) {
        let buf;
        try {
            buf = base_1.hex.decode(s);
        }
        catch {
            throw new Error("invalid asset ref format, must be hex");
        }
        return AssetRef.fromBytes(buf);
    }
    static fromBytes(buf) {
        if (!buf || buf.length === 0) {
            throw new Error("missing asset ref");
        }
        const reader = new utils_1.BufferReader(buf);
        return AssetRef.fromReader(reader);
    }
    serialize() {
        const writer = new utils_1.BufferWriter();
        this.serializeTo(writer);
        return writer.toBytes();
    }
    toString() {
        return base_1.hex.encode(this.serialize());
    }
    static fromReader(reader) {
        const type = reader.readByte();
        let ref;
        switch (type) {
            case types_1.AssetRefType.ByID: {
                const assetId = assetId_1.AssetId.fromReader(reader);
                ref = new AssetRef({ type: types_1.AssetRefType.ByID, assetId });
                break;
            }
            case types_1.AssetRefType.ByGroup: {
                if (reader.remaining() < 2) {
                    throw new Error("invalid asset ref length");
                }
                const groupIndex = reader.readUint16LE();
                ref = new AssetRef({ type: types_1.AssetRefType.ByGroup, groupIndex });
                break;
            }
            case types_1.AssetRefType.Unspecified:
                throw new Error("asset ref type unspecified");
            default:
                throw new Error(`asset ref type unknown ${type}`);
        }
        return ref;
    }
    serializeTo(writer) {
        writer.writeByte(this.ref.type);
        switch (this.ref.type) {
            case types_1.AssetRefType.ByID:
                this.ref.assetId.serializeTo(writer);
                break;
            case types_1.AssetRefType.ByGroup:
                writer.writeUint16LE(this.ref.groupIndex);
                break;
        }
    }
}
exports.AssetRef = AssetRef;
