// @arkade-os/sdk/dist/cjs/extension/asset/assetGroup.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AssetGroup = void 0;
const base_1 = require("@scure/base");
const types_1 = require("./types");
const assetId_1 = require("./assetId");
const assetRef_1 = require("./assetRef");
const assetInput_1 = require("./assetInput");
const assetOutput_1 = require("./assetOutput");
const metadata_1 = require("./metadata");
const utils_1 = require("./utils");
/**
 * An asset group contains inputs/outputs and all data related to a given asset id.
 */
class AssetGroup {
    constructor(assetId, controlAsset, inputs, outputs, metadata) {
        this.assetId = assetId;
        this.controlAsset = controlAsset;
        this.inputs = inputs;
        this.outputs = outputs;
        this.metadataList = new metadata_1.MetadataList(metadata);
    }
    static create(assetId, controlAsset, inputs, outputs, metadata) {
        const ag = new AssetGroup(assetId, controlAsset, inputs, outputs, metadata);
        ag.validate();
        return ag;
    }
    // from hex encoded
    static fromString(s) {
        let buf;
        try {
            buf = base_1.hex.decode(s);
        }
        catch {
            throw new Error("invalid format, must be hex");
        }
        return AssetGroup.fromBytes(buf);
    }
    static fromBytes(buf) {
        if (!buf || buf.length === 0) {
            throw new Error("missing asset group");
        }
        const reader = new utils_1.BufferReader(buf);
        return AssetGroup.fromReader(reader);
    }
    // an issuance is a group with null assetId
    isIssuance() {
        return this.assetId === null;
    }
    // a reissuance is a group that is not an issuance
    // but where the sum of the outputs is greater than the sum of the inputs
    isReissuance() {
        const sumReducer = (s, { amount }) => s + amount;
        const sumOutputs = this.outputs.reduce(sumReducer, 0n);
        const sumInputs = this.inputs
            .map((i) => ({
            amount: i.input.type === types_1.AssetInputType.Local ? i.input.amount : 0n,
        }))
            .reduce(sumReducer, 0n);
        return !this.isIssuance() && sumInputs < sumOutputs;
    }
    serialize() {
        this.validate();
        const writer = new utils_1.BufferWriter();
        this.serializeTo(writer);
        return writer.toBytes();
    }
    validate() {
        if (this.inputs.length === 0 && this.outputs.length === 0) {
            throw new Error("empty asset group");
        }
        if (this.isIssuance()) {
            if (this.inputs.length !== 0) {
                throw new Error("issuance must have no inputs");
            }
        }
        else {
            if (this.controlAsset !== null) {
                throw new Error("only issuance can have a control asset");
            }
        }
    }
    toBatchLeafAssetGroup(intentTxid) {
        const leafInput = assetInput_1.AssetInput.createIntent(base_1.hex.encode(intentTxid), 0, 0);
        return new AssetGroup(this.assetId, this.controlAsset, [leafInput], this.outputs, this.metadataList.items);
    }
    toString() {
        return base_1.hex.encode(this.serialize());
    }
    static fromReader(reader) {
        const presence = reader.readByte();
        let assetId = null;
        let controlAsset = null;
        let metadata = [];
        if (presence & types_1.MASK_ASSET_ID) {
            assetId = assetId_1.AssetId.fromReader(reader);
        }
        if (presence & types_1.MASK_CONTROL_ASSET) {
            controlAsset = assetRef_1.AssetRef.fromReader(reader);
        }
        if (presence & types_1.MASK_METADATA) {
            metadata = metadata_1.MetadataList.fromReader(reader).items;
        }
        const inputs = assetInput_1.AssetInputs.fromReader(reader);
        const outputs = assetOutput_1.AssetOutputs.fromReader(reader);
        const ag = new AssetGroup(assetId, controlAsset, inputs.inputs, outputs.outputs, metadata);
        ag.validate();
        return ag;
    }
    serializeTo(writer) {
        let presence = 0;
        if (this.assetId !== null) {
            presence |= types_1.MASK_ASSET_ID;
        }
        if (this.controlAsset !== null) {
            presence |= types_1.MASK_CONTROL_ASSET;
        }
        if (this.metadataList.length > 0) {
            presence |= types_1.MASK_METADATA;
        }
        writer.writeByte(presence);
        if (presence & types_1.MASK_ASSET_ID) {
            this.assetId.serializeTo(writer);
        }
        if (presence & types_1.MASK_CONTROL_ASSET) {
            this.controlAsset.serializeTo(writer);
        }
        if (presence & types_1.MASK_METADATA) {
            this.metadataList.serializeTo(writer);
        }
        writer.writeVarUint(this.inputs.length);
        for (const input of this.inputs) {
            input.serializeTo(writer);
        }
        writer.writeVarUint(this.outputs.length);
        for (const output of this.outputs) {
            output.serializeTo(writer);
        }
    }
}
exports.AssetGroup = AssetGroup;
