// @arkade-os/sdk/dist/cjs/extension/asset/assetId.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AssetId = void 0;
const base_1 = require("@scure/base");
const types_1 = require("./types");
const utils_1 = require("./utils");
/**
 * AssetId represents the id of an asset.
 * @param txid - the genesis transaction id (decoded from hex)
 * @param groupIndex - the asset group index in the genesis transaction
 */
class AssetId {
    constructor(txid, groupIndex) {
        this.txid = txid;
        this.groupIndex = groupIndex;
    }
    static create(txid, groupIndex) {
        if (!txid) {
            throw new Error("missing txid");
        }
        let buf;
        try {
            buf = base_1.hex.decode(txid);
        }
        catch {
            throw new Error("invalid txid format, must be hex");
        }
        if (buf.length !== types_1.TX_HASH_SIZE) {
            throw new Error(`invalid txid length: got ${buf.length} bytes, want ${types_1.TX_HASH_SIZE} bytes`);
        }
        const assetId = new AssetId(buf, groupIndex);
        assetId.validate();
        return assetId;
    }
    static fromString(s) {
        let buf;
        try {
            buf = base_1.hex.decode(s);
        }
        catch {
            throw new Error("invalid asset id format, must be hex");
        }
        return AssetId.fromBytes(buf);
    }
    static fromBytes(buf) {
        if (!buf || buf.length === 0) {
            throw new Error("missing asset id");
        }
        if (buf.length !== types_1.ASSET_ID_SIZE) {
            throw new Error(`invalid asset id length: got ${buf.length} bytes, want ${types_1.ASSET_ID_SIZE} bytes`);
        }
        const reader = new utils_1.BufferReader(buf);
        return AssetId.fromReader(reader);
    }
    serialize() {
        const writer = new utils_1.BufferWriter();
        this.serializeTo(writer);
        return writer.toBytes();
    }
    toString() {
        return base_1.hex.encode(this.serialize());
    }
    validate() {
        if ((0, utils_1.isZeroBytes)(this.txid)) {
            throw new Error("empty txid");
        }
        if (!Number.isInteger(this.groupIndex) ||
            this.groupIndex < 0 ||
            this.groupIndex > 0xffff) {
            throw new Error(`invalid group index: ${this.groupIndex}, must be in range [0, 65535]`);
        }
    }
    static fromReader(reader) {
        if (reader.remaining() < types_1.ASSET_ID_SIZE) {
            throw new Error(`invalid asset id length: got ${reader.remaining()}, want ${types_1.ASSET_ID_SIZE}`);
        }
        const txid = reader.readSlice(types_1.TX_HASH_SIZE);
        const index = reader.readUint16LE();
        const assetId = new AssetId(txid, index);
        assetId.validate();
        return assetId;
    }
    serializeTo(writer) {
        writer.write(this.txid);
        writer.writeUint16LE(this.groupIndex);
    }
}
exports.AssetId = AssetId;
