// @arkade-os/sdk/dist/cjs/extension/asset/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Packet = exports.AssetGroup = exports.MetadataList = exports.Metadata = exports.AssetOutputs = exports.AssetOutput = exports.AssetInputs = exports.AssetInput = exports.AssetRef = exports.AssetId = exports.AssetRefType = exports.AssetInputType = void 0;
var types_1 = require("./types");
Object.defineProperty(exports, "AssetInputType", { enumerable: true, get: function () { return types_1.AssetInputType; } });
Object.defineProperty(exports, "AssetRefType", { enumerable: true, get: function () { return types_1.AssetRefType; } });
var assetId_1 = require("./assetId");
Object.defineProperty(exports, "AssetId", { enumerable: true, get: function () { return assetId_1.AssetId; } });
var assetRef_1 = require("./assetRef");
Object.defineProperty(exports, "AssetRef", { enumerable: true, get: function () { return assetRef_1.AssetRef; } });
var assetInput_1 = require("./assetInput");
Object.defineProperty(exports, "AssetInput", { enumerable: true, get: function () { return assetInput_1.AssetInput; } });
Object.defineProperty(exports, "AssetInputs", { enumerable: true, get: function () { return assetInput_1.AssetInputs; } });
var assetOutput_1 = require("./assetOutput");
Object.defineProperty(exports, "AssetOutput", { enumerable: true, get: function () { return assetOutput_1.AssetOutput; } });
Object.defineProperty(exports, "AssetOutputs", { enumerable: true, get: function () { return assetOutput_1.AssetOutputs; } });
var metadata_1 = require("./metadata");
Object.defineProperty(exports, "Metadata", { enumerable: true, get: function () { return metadata_1.Metadata; } });
Object.defineProperty(exports, "MetadataList", { enumerable: true, get: function () { return metadata_1.MetadataList; } });
var assetGroup_1 = require("./assetGroup");
Object.defineProperty(exports, "AssetGroup", { enumerable: true, get: function () { return assetGroup_1.AssetGroup; } });
var packet_1 = require("./packet");
Object.defineProperty(exports, "Packet", { enumerable: true, get: function () { return packet_1.Packet; } });
