// @arkade-os/sdk/dist/cjs/extension/asset/packet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Packet = void 0;
const base_1 = require("@scure/base");
const types_1 = require("./types");
const assetGroup_1 = require("./assetGroup");
const utils_1 = require("./utils");
/**
 * Packet represents a collection of asset groups.
 * It encodes/decodes as raw bytes only — OP_RETURN framing is handled by the Extension module.
 */
class Packet {
    constructor(groups) {
        this.groups = groups;
    }
    static create(groups) {
        const p = new Packet(groups);
        p.validate();
        return p;
    }
    /**
     * fromBytes parses a Packet from raw bytes.
     */
    static fromBytes(buf) {
        return Packet.fromReader(new utils_1.BufferReader(buf));
    }
    /**
     * fromString parses a Packet from a raw hex string (not an OP_RETURN script).
     */
    static fromString(s) {
        if (!s) {
            throw new Error("missing packet data");
        }
        let buf;
        try {
            buf = base_1.hex.decode(s);
        }
        catch {
            throw new Error("invalid packet format, must be hex");
        }
        return Packet.fromBytes(buf);
    }
    /**
     * type returns the TLV packet type tag. Implements ExtensionPacket interface.
     */
    type() {
        return Packet.PACKET_TYPE;
    }
    leafTxPacket(intentTxid) {
        const leafGroups = this.groups.map((group) => group.toBatchLeafAssetGroup(intentTxid));
        return new Packet(leafGroups);
    }
    /**
     * serialize encodes the packet as raw bytes (varint group count + group data).
     * Does NOT include OP_RETURN, ARK magic, or TLV type/length — those are
     * added by the Extension module.
     */
    serialize() {
        if (this.groups.length === 0) {
            return new Uint8Array(0);
        }
        const writer = new utils_1.BufferWriter();
        writer.writeVarUint(this.groups.length);
        for (const group of this.groups) {
            group.serializeTo(writer);
        }
        return writer.toBytes();
    }
    /**
     * toString returns the hex-encoded raw packet bytes.
     */
    toString() {
        return base_1.hex.encode(this.serialize());
    }
    validate() {
        if (this.groups.length === 0) {
            throw new Error("missing assets");
        }
        const seenAssetIds = new Set();
        for (const group of this.groups) {
            if (group.assetId !== null) {
                const key = group.assetId.toString();
                if (seenAssetIds.has(key)) {
                    throw new Error(`duplicate asset group for asset ${key}`);
                }
                seenAssetIds.add(key);
            }
            if (group.controlAsset !== null &&
                group.controlAsset.ref.type === types_1.AssetRefType.ByGroup &&
                group.controlAsset.ref.groupIndex >= this.groups.length) {
                throw new Error(`invalid control asset group index, ${group.controlAsset.ref.groupIndex} out of range [0, ${this.groups.length - 1}]`);
            }
        }
    }
    static fromReader(reader) {
        const count = Number(reader.readVarUint());
        const groups = [];
        for (let i = 0; i < count; i++) {
            groups.push(assetGroup_1.AssetGroup.fromReader(reader));
        }
        if (reader.remaining() > 0) {
            throw new Error(`invalid packet length, left ${reader.remaining()} unknown bytes to read`);
        }
        const packet = new Packet(groups);
        packet.validate();
        return packet;
    }
}
exports.Packet = Packet;
/** PACKET_TYPE is the 1-byte TLV type tag used in the Extension envelope. */
Packet.PACKET_TYPE = 0;
