// @arkade-os/sdk/dist/cjs/extension/asset/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MASK_METADATA = exports.MASK_CONTROL_ASSET = exports.MASK_ASSET_ID = exports.AssetRefType = exports.AssetInputType = exports.ASSET_VERSION = exports.ASSET_ID_SIZE = exports.TX_HASH_SIZE = void 0;
exports.TX_HASH_SIZE = 32;
exports.ASSET_ID_SIZE = 34;
exports.ASSET_VERSION = 0x01;
var AssetInputType;
(function (AssetInputType) {
    AssetInputType[AssetInputType["Unspecified"] = 0] = "Unspecified";
    AssetInputType[AssetInputType["Local"] = 1] = "Local";
    AssetInputType[AssetInputType["Intent"] = 2] = "Intent";
})(AssetInputType || (exports.AssetInputType = AssetInputType = {}));
var AssetRefType;
(function (AssetRefType) {
    AssetRefType[AssetRefType["Unspecified"] = 0] = "Unspecified";
    AssetRefType[AssetRefType["ByID"] = 1] = "ByID";
    AssetRefType[AssetRefType["ByGroup"] = 2] = "ByGroup";
})(AssetRefType || (exports.AssetRefType = AssetRefType = {}));
// Presence byte masks for AssetGroup
exports.MASK_ASSET_ID = 0x01;
exports.MASK_CONTROL_ASSET = 0x02;
exports.MASK_METADATA = 0x04;
