// @arkade-os/sdk/dist/esm/repositories/realm/contractRepository.js
/**
 * Realm-based implementation of ContractRepository.
 *
 * Consumers must open Realm with the schemas from `./schemas.ts` and pass
 * the instance to the constructor.
 *
 * Realm handles schema creation on open, so `ensureInit()` is a no-op.
 * The consumer owns the Realm lifecycle — `[Symbol.asyncDispose]` is a no-op.
 */
export class RealmContractRepository {
    constructor(realm) {
        this.realm = realm;
        this.version = 1;
    }
    // ── Lifecycle ──────────────────────────────────────────────────────
    async ensureInit() {
        // Realm handles schema on open — nothing to initialise.
    }
    async [Symbol.asyncDispose]() {
        // no-op — consumer owns the Realm lifecycle
    }
    // ── Clear ──────────────────────────────────────────────────────────
    async clear() {
        await this.ensureInit();
        this.realm.write(() => {
            this.realm.delete(this.realm.objects("ArkContract"));
        });
    }
    // ── Contract management ────────────────────────────────────────────
    async getContracts(filter) {
        await this.ensureInit();
        let results = this.realm.objects("ArkContract");
        if (filter) {
            const filterParts = [];
            const filterArgs = [];
            let argIndex = 0;
            argIndex = this.addFilterCondition(filterParts, filterArgs, "script", filter.script, argIndex);
            argIndex = this.addFilterCondition(filterParts, filterArgs, "state", filter.state, argIndex);
            argIndex = this.addFilterCondition(filterParts, filterArgs, "type", filter.type, argIndex);
            if (filterParts.length > 0) {
                const query = filterParts.join(" AND ");
                results = results.filtered(query, ...filterArgs);
            }
        }
        return [...results].map(contractObjectToDomain);
    }
    async saveContract(contract) {
        await this.ensureInit();
        this.realm.write(() => {
            this.realm.create("ArkContract", {
                script: contract.script,
                address: contract.address,
                type: contract.type,
                state: contract.state,
                paramsJson: JSON.stringify(contract.params),
                createdAt: contract.createdAt,
                expiresAt: contract.expiresAt ?? null,
                label: contract.label ?? null,
                metadataJson: contract.metadata
                    ? JSON.stringify(contract.metadata)
                    : null,
            }, "modified");
        });
    }
    async deleteContract(script) {
        await this.ensureInit();
        this.realm.write(() => {
            const toDelete = this.realm
                .objects("ArkContract")
                .filtered("script == $0", script);
            this.realm.delete(toDelete);
        });
    }
    // ── Helpers ─────────────────────────────────────────────────────────
    addFilterCondition(parts, args, column, value, argIndex) {
        if (value === undefined)
            return argIndex;
        if (Array.isArray(value)) {
            if (value.length === 0)
                return argIndex;
            const conditions = value.map((_, i) => {
                return `${column} == $${argIndex + i}`;
            });
            parts.push(`(${conditions.join(" OR ")})`);
            args.push(...value);
            return argIndex + value.length;
        }
        else {
            parts.push(`${column} == $${argIndex}`);
            args.push(value);
            return argIndex + 1;
        }
    }
}
// ── Realm object → Domain converter ──────────────────────────────────────
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function contractObjectToDomain(obj) {
    const contract = {
        script: obj.script,
        address: obj.address,
        type: obj.type,
        state: obj.state,
        params: JSON.parse(obj.paramsJson),
        createdAt: obj.createdAt,
    };
    if (obj.expiresAt !== null && obj.expiresAt !== undefined) {
        contract.expiresAt = obj.expiresAt;
    }
    if (obj.label !== null && obj.label !== undefined) {
        contract.label = obj.label;
    }
    if (obj.metadataJson !== null && obj.metadataJson !== undefined) {
        contract.metadata = JSON.parse(obj.metadataJson);
    }
    return contract;
}
