// @arkade-os/sdk/dist/esm/repositories/realm/walletRepository.js
import { serializeVtxo, serializeUtxo, deserializeVtxo, deserializeUtxo, } from '../serialization.js';
/**
 * Realm-based implementation of WalletRepository.
 *
 * Consumers must open Realm with the schemas from `./schemas.ts` and pass
 * the instance to the constructor.
 *
 * Realm handles schema creation on open, so `ensureInit()` is a no-op.
 * The consumer owns the Realm lifecycle — `[Symbol.asyncDispose]` is a no-op.
 */
export class RealmWalletRepository {
    constructor(realm) {
        this.realm = realm;
        this.version = 1;
    }
    // ── Lifecycle ──────────────────────────────────────────────────────
    async ensureInit() {
        // Realm handles schema on open — nothing to initialise.
    }
    async [Symbol.asyncDispose]() {
        // no-op — consumer owns the Realm lifecycle
    }
    // ── Clear ──────────────────────────────────────────────────────────
    async clear() {
        await this.ensureInit();
        this.realm.write(() => {
            this.realm.delete(this.realm.objects("ArkVtxo"));
            this.realm.delete(this.realm.objects("ArkUtxo"));
            this.realm.delete(this.realm.objects("ArkTransaction"));
            this.realm.delete(this.realm.objects("ArkWalletState"));
        });
    }
    // ── VTXO management ────────────────────────────────────────────────
    async getVtxos(address) {
        await this.ensureInit();
        const results = this.realm
            .objects("ArkVtxo")
            .filtered("address == $0", address);
        return [...results].map(vtxoObjectToDomain);
    }
    async saveVtxos(address, vtxos) {
        await this.ensureInit();
        this.realm.write(() => {
            for (const vtxo of vtxos) {
                const s = serializeVtxo(vtxo);
                this.realm.create("ArkVtxo", {
                    pk: `${s.txid}:${s.vout}`,
                    address,
                    txid: s.txid,
                    vout: s.vout,
                    value: s.value,
                    tapTree: s.tapTree,
                    forfeitCb: s.forfeitTapLeafScript.cb,
                    forfeitS: s.forfeitTapLeafScript.s,
                    intentCb: s.intentTapLeafScript.cb,
                    intentS: s.intentTapLeafScript.s,
                    statusJson: JSON.stringify(s.status),
                    virtualStatusJson: JSON.stringify(s.virtualStatus),
                    createdAt: typeof s.createdAt === "string"
                        ? s.createdAt
                        : s.createdAt instanceof Date
                            ? s.createdAt.toISOString()
                            : new Date(s.createdAt).toISOString(),
                    isUnrolled: s.isUnrolled ?? false,
                    isSpent: s.isSpent === undefined ? null : s.isSpent,
                    spentBy: s.spentBy ?? null,
                    settledBy: s.settledBy ?? null,
                    arkTxId: s.arkTxId ?? null,
                    extraWitnessJson: s.extraWitness
                        ? JSON.stringify(s.extraWitness)
                        : null,
                    assetsJson: s.assets ? JSON.stringify(s.assets) : null,
                }, "modified");
            }
        });
    }
    async deleteVtxos(address) {
        await this.ensureInit();
        this.realm.write(() => {
            const toDelete = this.realm
                .objects("ArkVtxo")
                .filtered("address == $0", address);
            this.realm.delete(toDelete);
        });
    }
    // ── UTXO management ────────────────────────────────────────────────
    async getUtxos(address) {
        await this.ensureInit();
        const results = this.realm
            .objects("ArkUtxo")
            .filtered("address == $0", address);
        return [...results].map(utxoObjectToDomain);
    }
    async saveUtxos(address, utxos) {
        await this.ensureInit();
        this.realm.write(() => {
            for (const utxo of utxos) {
                const s = serializeUtxo(utxo);
                this.realm.create("ArkUtxo", {
                    pk: `${s.txid}:${s.vout}`,
                    address,
                    txid: s.txid,
                    vout: s.vout,
                    value: s.value,
                    tapTree: s.tapTree,
                    forfeitCb: s.forfeitTapLeafScript.cb,
                    forfeitS: s.forfeitTapLeafScript.s,
                    intentCb: s.intentTapLeafScript.cb,
                    intentS: s.intentTapLeafScript.s,
                    statusJson: JSON.stringify(s.status),
                    extraWitnessJson: s.extraWitness
                        ? JSON.stringify(s.extraWitness)
                        : null,
                }, "modified");
            }
        });
    }
    async deleteUtxos(address) {
        await this.ensureInit();
        this.realm.write(() => {
            const toDelete = this.realm
                .objects("ArkUtxo")
                .filtered("address == $0", address);
            this.realm.delete(toDelete);
        });
    }
    // ── Transaction history ────────────────────────────────────────────
    async getTransactionHistory(address) {
        await this.ensureInit();
        const results = this.realm
            .objects("ArkTransaction")
            .filtered("address == $0", address);
        const txs = [...results].map(txObjectToDomain);
        txs.sort((a, b) => a.createdAt - b.createdAt);
        return txs;
    }
    async saveTransactions(address, txs) {
        await this.ensureInit();
        this.realm.write(() => {
            for (const tx of txs) {
                this.realm.create("ArkTransaction", {
                    pk: `${address}:${tx.key.boardingTxid}:${tx.key.commitmentTxid}:${tx.key.arkTxid}`,
                    address,
                    boardingTxid: tx.key.boardingTxid,
                    commitmentTxid: tx.key.commitmentTxid,
                    arkTxid: tx.key.arkTxid,
                    type: tx.type,
                    amount: tx.amount,
                    settled: tx.settled,
                    createdAt: tx.createdAt,
                    assetsJson: tx.assets
                        ? JSON.stringify(tx.assets)
                        : null,
                }, "modified");
            }
        });
    }
    async deleteTransactions(address) {
        await this.ensureInit();
        this.realm.write(() => {
            const toDelete = this.realm
                .objects("ArkTransaction")
                .filtered("address == $0", address);
            this.realm.delete(toDelete);
        });
    }
    // ── Wallet state ───────────────────────────────────────────────────
    async getWalletState() {
        await this.ensureInit();
        const results = this.realm
            .objects("ArkWalletState")
            .filtered("key == $0", "state");
        const items = [...results];
        if (items.length === 0)
            return null;
        const obj = items[0];
        const state = {};
        if (obj.lastSyncTime !== null && obj.lastSyncTime !== undefined) {
            state.lastSyncTime = obj.lastSyncTime;
        }
        if (obj.settingsJson) {
            state.settings = JSON.parse(obj.settingsJson);
        }
        return state;
    }
    async saveWalletState(state) {
        await this.ensureInit();
        this.realm.write(() => {
            this.realm.create("ArkWalletState", {
                key: "state",
                lastSyncTime: state.lastSyncTime ?? null,
                settingsJson: state.settings
                    ? JSON.stringify(state.settings)
                    : null,
            }, "modified");
        });
    }
}
// ── Realm object → Domain converters ─────────────────────────────────────
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function vtxoObjectToDomain(obj) {
    const serialized = {
        txid: obj.txid,
        vout: obj.vout,
        value: obj.value,
        tapTree: obj.tapTree,
        forfeitTapLeafScript: {
            cb: obj.forfeitCb,
            s: obj.forfeitS,
        },
        intentTapLeafScript: {
            cb: obj.intentCb,
            s: obj.intentS,
        },
        status: JSON.parse(obj.statusJson),
        virtualStatus: JSON.parse(obj.virtualStatusJson),
        createdAt: new Date(obj.createdAt),
        isUnrolled: obj.isUnrolled,
        isSpent: obj.isSpent === null ? undefined : obj.isSpent,
        spentBy: obj.spentBy ?? undefined,
        settledBy: obj.settledBy ?? undefined,
        arkTxId: obj.arkTxId ?? undefined,
        extraWitness: obj.extraWitnessJson
            ? JSON.parse(obj.extraWitnessJson)
            : undefined,
        assets: obj.assetsJson ? JSON.parse(obj.assetsJson) : undefined,
    };
    return deserializeVtxo(serialized);
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function utxoObjectToDomain(obj) {
    const serialized = {
        txid: obj.txid,
        vout: obj.vout,
        value: obj.value,
        tapTree: obj.tapTree,
        forfeitTapLeafScript: {
            cb: obj.forfeitCb,
            s: obj.forfeitS,
        },
        intentTapLeafScript: {
            cb: obj.intentCb,
            s: obj.intentS,
        },
        status: JSON.parse(obj.statusJson),
        extraWitness: obj.extraWitnessJson
            ? JSON.parse(obj.extraWitnessJson)
            : undefined,
    };
    return deserializeUtxo(serialized);
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function txObjectToDomain(obj) {
    const tx = {
        key: {
            boardingTxid: obj.boardingTxid,
            commitmentTxid: obj.commitmentTxid,
            arkTxid: obj.arkTxid,
        },
        type: obj.type,
        amount: obj.amount,
        settled: obj.settled,
        createdAt: obj.createdAt,
    };
    if (obj.assetsJson) {
        tx.assets = JSON.parse(obj.assetsJson);
    }
    return tx;
}
