// @arkade-os/sdk/dist/cjs/worker/expo/taskRunner.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.runTasks = runTasks;
const utils_1 = require("../../wallet/utils");
/**
 * Run all pending tasks from the queue through matching processors.
 *
 * For each task in the inbox:
 * 1. Find the processor whose `taskType` matches `task.type`.
 * 2. Execute it, producing a {@link TaskResult}.
 * 3. Push the result to the outbox and remove the task from the inbox.
 *
 * Tasks with no matching processor produce a `"noop"` result.
 * Processor errors produce a `"failed"` result with the error message.
 */
async function runTasks(queue, processors, deps) {
    const tasks = await queue.getTasks();
    const processorMap = new Map(processors.map((p) => [p.taskType, p]));
    const results = [];
    for (const task of tasks) {
        const processor = processorMap.get(task.type);
        let partial;
        if (!processor) {
            partial = {
                taskItemId: task.id,
                type: task.type,
                status: "noop",
            };
        }
        else {
            try {
                partial = await processor.execute(task, deps);
            }
            catch (error) {
                partial = {
                    taskItemId: task.id,
                    type: task.type,
                    status: "failed",
                    data: {
                        error: error instanceof Error
                            ? error.message
                            : String(error),
                    },
                };
            }
        }
        const result = {
            ...partial,
            id: (0, utils_1.getRandomId)(),
            executedAt: Date.now(),
        };
        await queue.pushResult(result);
        await queue.removeTask(task.id);
        results.push(result);
    }
    return results;
}
