// @arkade-os/sdk/dist/cjs/wallet/expo/background.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defineExpoBackgroundTask = defineExpoBackgroundTask;
exports.registerExpoBackgroundTask = registerExpoBackgroundTask;
exports.unregisterExpoBackgroundTask = unregisterExpoBackgroundTask;
const base_1 = require("@scure/base");
const taskRunner_1 = require("../../worker/expo/taskRunner");
const processors_1 = require("../../worker/expo/processors");
const default_1 = require("../../script/default");
const expoArk_1 = require("../../providers/expoArk");
const expoIndexer_1 = require("../../providers/expoIndexer");
const utils_1 = require("../utils");
function requireTaskManager() {
    try {
        return require("expo-task-manager");
    }
    catch {
        throw new Error("expo-task-manager is required for background tasks. " +
            "Install it with: npx expo install expo-task-manager");
    }
}
function requireBackgroundTask() {
    try {
        return require("expo-background-task");
    }
    catch {
        throw new Error("expo-background-task is required for background tasks. " +
            "Install it with: npx expo install expo-background-task");
    }
}
/**
 * Define the Expo background task handler.
 *
 * **Must be called at module/global scope** (before React mounts).
 * Internally calls `TaskManager.defineTask()`.
 *
 * @example
 * ```ts
 * // At the top of your app entry file
 * import { defineExpoBackgroundTask } from "@arkade-os/sdk/wallet/expo";
 * import { AsyncStorageTaskQueue } from "@arkade-os/sdk/worker/expo";
 * import AsyncStorage from "@react-native-async-storage/async-storage";
 *
 * const taskQueue = new AsyncStorageTaskQueue(AsyncStorage);
 * defineExpoBackgroundTask("ark-background-poll", {
 *     taskQueue,
 *     walletRepository: new IndexedDBWalletRepository(),
 *     contractRepository: new IndexedDBContractRepository(),
 * });
 * ```
 */
function defineExpoBackgroundTask(taskName, options) {
    const TaskManager = requireTaskManager();
    const BackgroundTask = requireBackgroundTask();
    const { taskQueue, walletRepository, contractRepository, processors = [processors_1.contractPollProcessor], } = options;
    TaskManager.defineTask(taskName, async () => {
        try {
            const config = await taskQueue.loadConfig();
            if (!config) {
                // No config persisted yet — ExpoWallet.setup() hasn't run.
                // Nothing to do.
                return BackgroundTask.BackgroundTaskResult.Success;
            }
            // Reconstruct providers
            const indexerProvider = new expoIndexer_1.ExpoIndexerProvider(config.arkServerUrl);
            const arkProvider = new expoArk_1.ExpoArkProvider(config.arkServerUrl);
            // Reconstruct default offchainTapscript as fallback
            // for VTXOs not associated with a contract.
            const defaultTapscript = new default_1.DefaultVtxo.Script({
                pubKey: base_1.hex.decode(config.pubkeyHex),
                serverPubKey: base_1.hex.decode(config.serverPubKeyHex),
                csvTimelock: {
                    value: BigInt(config.exitTimelockValue),
                    type: config.exitTimelockType,
                },
            });
            await (0, taskRunner_1.runTasks)(taskQueue, processors, {
                walletRepository,
                contractRepository,
                indexerProvider,
                arkProvider,
                extendVtxo: (vtxo, contract) => {
                    if (contract) {
                        return (0, utils_1.extendVtxoFromContract)(vtxo, contract);
                    }
                    return (0, utils_1.extendVirtualCoin)({ offchainTapscript: defaultTapscript }, vtxo);
                },
            });
            // Acknowledge outbox results (no foreground to consume them)
            const results = await taskQueue.getResults();
            if (results.length > 0) {
                await taskQueue.acknowledgeResults(results.map((r) => r.id));
            }
            // Re-seed the contract-poll task for the next OS wake
            const existing = await taskQueue.getTasks(processors_1.CONTRACT_POLL_TASK_TYPE);
            if (existing.length === 0) {
                const task = {
                    id: (0, utils_1.getRandomId)(),
                    type: processors_1.CONTRACT_POLL_TASK_TYPE,
                    data: {},
                    createdAt: Date.now(),
                };
                await taskQueue.addTask(task);
            }
            return BackgroundTask.BackgroundTaskResult.Success;
        }
        catch (error) {
            console.error("[ark-sdk] Background task failed:", error instanceof Error ? error.message : error);
            return BackgroundTask.BackgroundTaskResult.Failed;
        }
    });
}
/**
 * Activate the OS-level background task scheduler.
 *
 * Call this after {@link defineExpoBackgroundTask} (typically inside
 * {@link ExpoWallet.setup} or in a React component after wallet init).
 *
 * @param minimumInterval - Minimum interval in minutes (default 15).
 */
async function registerExpoBackgroundTask(taskName, options) {
    const BackgroundTask = requireBackgroundTask();
    await BackgroundTask.registerTaskAsync(taskName, {
        minimumInterval: (options?.minimumInterval ?? 15) * 60,
    });
}
/**
 * Unregister the background task from the OS scheduler.
 */
async function unregisterExpoBackgroundTask(taskName) {
    const BackgroundTask = requireBackgroundTask();
    await BackgroundTask.unregisterTaskAsync(taskName);
}
