// @arkade-os/sdk/dist/cjs/wallet/expo/wallet.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExpoWallet = void 0;
const base_1 = require("@scure/base");
const wallet_1 = require("../wallet");
const ark_1 = require("../../providers/ark");
const taskRunner_1 = require("../../worker/expo/taskRunner");
const processors_1 = require("../../worker/expo/processors");
const utils_1 = require("../utils");
const default_1 = require("../../script/default");
/**
 * Expo/React Native wallet with built-in background task processing.
 *
 * Wraps a standard {@link Wallet} and adds a lightweight task queue
 * for keeping contract/VTXO state fresh while the app is active and
 * across Expo BackgroundTask wakes.
 *
 * @example
 * ```ts
 * import { ExpoWallet } from "@arkade-os/sdk/wallet/expo";
 * import { AsyncStorageTaskQueue } from "@arkade-os/sdk/worker/expo";
 *
 * const wallet = await ExpoWallet.setup({
 *     identity: SingleKey.fromHex(privateKey),
 *     arkServerUrl,
 *     esploraUrl,
 *     storage: { walletRepository, contractRepository },
 *     background: {
 *         taskName: "ark-background-poll",
 *         taskQueue: new AsyncStorageTaskQueue(AsyncStorage),
 *         foregroundIntervalMs: 20_000,
 *         minimumBackgroundInterval: 15,
 *     },
 * });
 *
 * const balance = await wallet.getBalance();
 * ```
 */
class ExpoWallet {
    constructor(wallet, taskQueue, processors, deps, taskName, foregroundIntervalMs) {
        this.wallet = wallet;
        this.taskQueue = taskQueue;
        this.processors = processors;
        this.deps = deps;
        this.identity = wallet.identity;
        this.arkProvider = wallet.arkProvider;
        this.indexerProvider = wallet.indexerProvider;
        this.taskName = taskName;
        if (foregroundIntervalMs && foregroundIntervalMs > 0) {
            this.startForegroundPolling(foregroundIntervalMs);
        }
    }
    /**
     * Create an ExpoWallet with background task support.
     *
     * 1. Creates the inner {@link Wallet} via `Wallet.create()`.
     * 2. Wires up processors (defaults to {@link contractPollProcessor}).
     * 3. Persists background config for the background handler (if the queue supports it).
     * 4. Seeds the task queue with a `contract-poll` task.
     * 5. Registers the background task with the OS scheduler (if `minimumBackgroundInterval` is set).
     * 6. Starts foreground polling if `foregroundIntervalMs` is set.
     */
    static async setup(config) {
        const wallet = await wallet_1.Wallet.create(config);
        const processors = config.background.processors ?? [
            processors_1.contractPollProcessor,
        ];
        const deps = {
            walletRepository: wallet.walletRepository,
            contractRepository: wallet.contractRepository,
            indexerProvider: wallet.indexerProvider,
            arkProvider: wallet.arkProvider,
            extendVtxo: (vtxo, contract) => {
                if (contract) {
                    return (0, utils_1.extendVtxoFromContract)(vtxo, contract);
                }
                return (0, utils_1.extendVirtualCoin)(wallet, vtxo);
            },
        };
        const { taskQueue } = config.background;
        // Persist wallet params so the background handler can rehydrate
        // without a network call. Only works with AsyncStorageTaskQueue.
        if ("persistConfig" in taskQueue) {
            const arkServerUrl = config.arkServerUrl ??
                (wallet.arkProvider instanceof ark_1.RestArkProvider
                    ? wallet.arkProvider.serverUrl
                    : undefined);
            if (arkServerUrl) {
                const timelock = wallet.offchainTapscript.options.csvTimelock ??
                    default_1.DefaultVtxo.Script.DEFAULT_TIMELOCK;
                const bgConfig = {
                    arkServerUrl,
                    pubkeyHex: base_1.hex.encode(wallet.offchainTapscript.options.pubKey),
                    serverPubKeyHex: base_1.hex.encode(wallet.offchainTapscript.options.serverPubKey),
                    exitTimelockValue: timelock.value.toString(),
                    exitTimelockType: timelock.type,
                };
                await taskQueue.persistConfig(bgConfig);
            }
        }
        const expoWallet = new ExpoWallet(wallet, taskQueue, processors, deps, config.background.taskName, config.background.foregroundIntervalMs);
        // Seed the queue so the first tick (or background wake) has work to do
        await expoWallet.seedContractPollTask();
        // Activate OS-level background scheduling
        if (config.background.minimumBackgroundInterval) {
            try {
                const { registerExpoBackgroundTask } = await Promise.resolve().then(() => __importStar(require("./background")));
                await registerExpoBackgroundTask(config.background.taskName, {
                    minimumInterval: config.background.minimumBackgroundInterval,
                });
            }
            catch {
                // expo-background-task not installed — foreground-only mode
            }
        }
        return expoWallet;
    }
    // ── Foreground polling ───────────────────────────────────────────
    startForegroundPolling(intervalMs) {
        this.foregroundIntervalId = setInterval(() => {
            this.runForegroundPoll().catch(console.error);
        }, intervalMs);
    }
    async runForegroundPoll() {
        await (0, taskRunner_1.runTasks)(this.taskQueue, this.processors, this.deps);
        // Consume results immediately (no background handoff needed)
        const results = await this.taskQueue.getResults();
        if (results.length > 0) {
            await this.taskQueue.acknowledgeResults(results.map((r) => r.id));
        }
        // Re-seed for the next tick
        await this.seedContractPollTask();
    }
    async seedContractPollTask() {
        const existing = await this.taskQueue.getTasks(processors_1.CONTRACT_POLL_TASK_TYPE);
        if (existing.length > 0)
            return;
        const task = {
            id: (0, utils_1.getRandomId)(),
            type: processors_1.CONTRACT_POLL_TASK_TYPE,
            data: {},
            createdAt: Date.now(),
        };
        await this.taskQueue.addTask(task);
    }
    // ── Lifecycle ────────────────────────────────────────────────────
    /**
     * Stop foreground polling and unregister the background task.
     */
    async dispose() {
        if (this.foregroundIntervalId) {
            clearInterval(this.foregroundIntervalId);
            this.foregroundIntervalId = undefined;
        }
        try {
            const { unregisterExpoBackgroundTask } = await Promise.resolve().then(() => __importStar(require("./background")));
            await unregisterExpoBackgroundTask(this.taskName);
        }
        catch {
            // expo-background-task not installed — nothing to unregister
        }
    }
    // ── IWallet delegation ───────────────────────────────────────────
    getAddress() {
        return this.wallet.getAddress();
    }
    getBoardingAddress() {
        return this.wallet.getBoardingAddress();
    }
    getBalance() {
        return this.wallet.getBalance();
    }
    getVtxos(filter) {
        return this.wallet.getVtxos(filter);
    }
    getBoardingUtxos() {
        return this.wallet.getBoardingUtxos();
    }
    getTransactionHistory() {
        return this.wallet.getTransactionHistory();
    }
    getContractManager() {
        return this.wallet.getContractManager();
    }
    getDelegatorManager() {
        return this.wallet.getDelegatorManager();
    }
    sendBitcoin(params) {
        return this.wallet.sendBitcoin(params);
    }
    settle(params, eventCallback) {
        return this.wallet.settle(params, eventCallback);
    }
    send(...recipients) {
        return this.wallet.send(...recipients);
    }
    get assetManager() {
        return this.wallet.assetManager;
    }
}
exports.ExpoWallet = ExpoWallet;
