// @arkade-os/sdk/dist/esm/contracts/handlers/delegate.js
import { hex } from "@scure/base";
import { DelegateVtxo } from '../../script/delegate.js';
import { DefaultVtxo } from '../../script/default.js';
import { isCsvSpendable, sequenceToTimelock, timelockToSequence, } from './helpers.js';
/**
 * Handler for delegate wallet VTXOs.
 *
 * Delegate contracts extend the default tapscript with an additional delegate path:
 * - forfeit: (Alice + Server) multisig for collaborative spending
 * - exit: (Alice) + CSV timelock for unilateral exit
 * - delegate: (Alice + Delegate + Server) multisig for delegated renewal
 */
export const DelegateContractHandler = {
    type: "delegate",
    createScript(params) {
        const typed = this.deserializeParams(params);
        return new DelegateVtxo.Script(typed);
    },
    serializeParams(params) {
        return {
            pubKey: hex.encode(params.pubKey),
            serverPubKey: hex.encode(params.serverPubKey),
            delegatePubKey: hex.encode(params.delegatePubKey),
            csvTimelock: timelockToSequence(params.csvTimelock).toString(),
        };
    },
    deserializeParams(params) {
        const csvTimelock = params.csvTimelock
            ? sequenceToTimelock(Number(params.csvTimelock))
            : DefaultVtxo.Script.DEFAULT_TIMELOCK;
        return {
            pubKey: hex.decode(params.pubKey),
            serverPubKey: hex.decode(params.serverPubKey),
            delegatePubKey: hex.decode(params.delegatePubKey),
            csvTimelock,
        };
    },
    selectPath(script, contract, context) {
        if (context.collaborative) {
            return { leaf: script.forfeit() };
        }
        const sequence = contract.params.csvTimelock
            ? Number(contract.params.csvTimelock)
            : undefined;
        if (!isCsvSpendable(context, sequence)) {
            return null;
        }
        return {
            leaf: script.exit(),
            sequence,
        };
    },
    getAllSpendingPaths(script, contract, context) {
        const paths = [];
        if (context.collaborative) {
            paths.push({ leaf: script.forfeit() });
        }
        const exitPath = { leaf: script.exit() };
        if (contract.params.csvTimelock) {
            exitPath.sequence = Number(contract.params.csvTimelock);
        }
        paths.push(exitPath);
        // Delegate path (Alice + Delegate + Server) — collaborative only
        if (context.collaborative) {
            paths.push({ leaf: script.delegate() });
        }
        return paths;
    },
    getSpendablePaths(script, contract, context) {
        const paths = [];
        if (context.collaborative) {
            paths.push({ leaf: script.forfeit() });
        }
        const exitSequence = contract.params.csvTimelock
            ? Number(contract.params.csvTimelock)
            : undefined;
        if (isCsvSpendable(context, exitSequence)) {
            const exitPath = { leaf: script.exit() };
            if (exitSequence !== undefined) {
                exitPath.sequence = exitSequence;
            }
            paths.push(exitPath);
        }
        return paths;
    },
};
