// @arkade-os/sdk/dist/esm/extension/asset/assetRef.js
import { hex } from "@scure/base";
import { AssetRefType } from './types.js';
import { AssetId } from './assetId.js';
import { BufferReader, BufferWriter } from './utils.js';
export class AssetRef {
    constructor(ref) {
        this.ref = ref;
    }
    get type() {
        return this.ref.type;
    }
    static fromId(assetId) {
        return new AssetRef({ type: AssetRefType.ByID, assetId });
    }
    static fromGroupIndex(groupIndex) {
        return new AssetRef({ type: AssetRefType.ByGroup, groupIndex });
    }
    static fromString(s) {
        let buf;
        try {
            buf = hex.decode(s);
        }
        catch {
            throw new Error("invalid asset ref format, must be hex");
        }
        return AssetRef.fromBytes(buf);
    }
    static fromBytes(buf) {
        if (!buf || buf.length === 0) {
            throw new Error("missing asset ref");
        }
        const reader = new BufferReader(buf);
        return AssetRef.fromReader(reader);
    }
    serialize() {
        const writer = new BufferWriter();
        this.serializeTo(writer);
        return writer.toBytes();
    }
    toString() {
        return hex.encode(this.serialize());
    }
    static fromReader(reader) {
        const type = reader.readByte();
        let ref;
        switch (type) {
            case AssetRefType.ByID: {
                const assetId = AssetId.fromReader(reader);
                ref = new AssetRef({ type: AssetRefType.ByID, assetId });
                break;
            }
            case AssetRefType.ByGroup: {
                if (reader.remaining() < 2) {
                    throw new Error("invalid asset ref length");
                }
                const groupIndex = reader.readUint16LE();
                ref = new AssetRef({ type: AssetRefType.ByGroup, groupIndex });
                break;
            }
            case AssetRefType.Unspecified:
                throw new Error("asset ref type unspecified");
            default:
                throw new Error(`asset ref type unknown ${type}`);
        }
        return ref;
    }
    serializeTo(writer) {
        writer.writeByte(this.ref.type);
        switch (this.ref.type) {
            case AssetRefType.ByID:
                this.ref.assetId.serializeTo(writer);
                break;
            case AssetRefType.ByGroup:
                writer.writeUint16LE(this.ref.groupIndex);
                break;
        }
    }
}
