// @arkade-os/sdk/dist/esm/extension/asset/assetInput.js
import { hex } from "@scure/base";
import { AssetInputType, TX_HASH_SIZE } from './types.js';
import { BufferReader, BufferWriter, isZeroBytes } from './utils.js';
/**
 * AssetInput represents an input of an asset group.
 * a local input references a real transaction input and specify the amount in satoshis.
 * an intent input references an external intent transaction. It is created by the server to handle batch leaf transaction.
 */
export class AssetInput {
    constructor(input) {
        this.input = input;
    }
    get vin() {
        return this.input.vin;
    }
    get amount() {
        return this.input.amount;
    }
    static create(vin, amount) {
        const input = new AssetInput({
            type: AssetInputType.Local,
            vin,
            amount: typeof amount === "number" ? BigInt(amount) : amount,
        });
        input.validate();
        return input;
    }
    static createIntent(txid, vin, amount) {
        if (!txid || txid.length === 0) {
            throw new Error("missing input intent txid");
        }
        let buf;
        try {
            buf = hex.decode(txid);
        }
        catch {
            throw new Error("invalid input intent txid format, must be hex");
        }
        if (buf.length !== TX_HASH_SIZE) {
            throw new Error("invalid input intent txid length");
        }
        const input = new AssetInput({
            type: AssetInputType.Intent,
            txid: buf,
            vin,
            amount: typeof amount === "number" ? BigInt(amount) : amount,
        });
        input.validate();
        return input;
    }
    static fromString(s) {
        let buf;
        try {
            buf = hex.decode(s);
        }
        catch {
            throw new Error("invalid format, must be hex");
        }
        return AssetInput.fromBytes(buf);
    }
    static fromBytes(buf) {
        const reader = new BufferReader(buf);
        return AssetInput.fromReader(reader);
    }
    serialize() {
        const writer = new BufferWriter();
        this.serializeTo(writer);
        return writer.toBytes();
    }
    toString() {
        return hex.encode(this.serialize());
    }
    validate() {
        switch (this.input.type) {
            case AssetInputType.Local:
                break;
            case AssetInputType.Intent:
                if (isZeroBytes(this.input.txid)) {
                    throw new Error("missing input intent txid");
                }
                break;
        }
    }
    static fromReader(reader) {
        const type = reader.readByte();
        let input;
        switch (type) {
            case AssetInputType.Local: {
                const vin = reader.readUint16LE();
                const amount = reader.readVarUint();
                input = new AssetInput({
                    type: AssetInputType.Local,
                    vin,
                    amount,
                });
                break;
            }
            case AssetInputType.Intent: {
                if (reader.remaining() < TX_HASH_SIZE) {
                    throw new Error("invalid input intent txid length");
                }
                const txid = reader.readSlice(TX_HASH_SIZE);
                const vin = reader.readUint16LE();
                const amount = reader.readVarUint();
                input = new AssetInput({
                    type: AssetInputType.Intent,
                    txid: new Uint8Array(txid),
                    vin,
                    amount,
                });
                break;
            }
            case AssetInputType.Unspecified:
                throw new Error("asset input type unspecified");
            default:
                throw new Error(`asset input type ${type} unknown`);
        }
        input.validate();
        return input;
    }
    serializeTo(writer) {
        writer.writeByte(this.input.type);
        if (this.input.type === AssetInputType.Intent) {
            writer.write(this.input.txid);
        }
        writer.writeUint16LE(this.input.vin);
        writer.writeVarUint(this.input.amount);
    }
}
/**
 * AssetInputs represents a list of asset inputs.
 */
export class AssetInputs {
    constructor(inputs) {
        this.inputs = inputs;
    }
    static create(inputs) {
        const list = new AssetInputs(inputs);
        list.validate();
        return list;
    }
    static fromString(s) {
        if (!s || s.length === 0) {
            throw new Error("missing asset inputs");
        }
        let buf;
        try {
            buf = hex.decode(s);
        }
        catch {
            throw new Error("invalid asset inputs format, must be hex");
        }
        const reader = new BufferReader(buf);
        return AssetInputs.fromReader(reader);
    }
    serialize() {
        const writer = new BufferWriter();
        this.serializeTo(writer);
        return writer.toBytes();
    }
    toString() {
        return hex.encode(this.serialize());
    }
    validate() {
        const seen = new Set();
        let listType = AssetInputType.Unspecified;
        for (const assetInput of this.inputs) {
            assetInput.validate();
            if (listType === AssetInputType.Unspecified) {
                listType = assetInput.input.type;
            }
            else if (listType !== assetInput.input.type) {
                throw new Error("all inputs must be of the same type");
            }
            // verify the same input vin is not duplicated
            if (assetInput.input.type === AssetInputType.Local) {
                if (seen.has(assetInput.input.vin)) {
                    throw new Error(`duplicated input vin ${assetInput.input.vin}`);
                }
                seen.add(assetInput.input.vin);
                continue;
            }
        }
    }
    static fromReader(reader) {
        const count = Number(reader.readVarUint());
        const inputs = [];
        for (let i = 0; i < count; i++) {
            inputs.push(AssetInput.fromReader(reader));
        }
        return AssetInputs.create(inputs);
    }
    serializeTo(writer) {
        writer.writeVarUint(this.inputs.length);
        for (const input of this.inputs) {
            input.serializeTo(writer);
        }
    }
}
