// @arkade-os/sdk/dist/esm/extension/asset/assetId.js
import { hex } from "@scure/base";
import { TX_HASH_SIZE, ASSET_ID_SIZE } from './types.js';
import { BufferReader, BufferWriter, isZeroBytes } from './utils.js';
/**
 * AssetId represents the id of an asset.
 * @param txid - the genesis transaction id (decoded from hex)
 * @param groupIndex - the asset group index in the genesis transaction
 */
export class AssetId {
    constructor(txid, groupIndex) {
        this.txid = txid;
        this.groupIndex = groupIndex;
    }
    static create(txid, groupIndex) {
        if (!txid) {
            throw new Error("missing txid");
        }
        let buf;
        try {
            buf = hex.decode(txid);
        }
        catch {
            throw new Error("invalid txid format, must be hex");
        }
        if (buf.length !== TX_HASH_SIZE) {
            throw new Error(`invalid txid length: got ${buf.length} bytes, want ${TX_HASH_SIZE} bytes`);
        }
        const assetId = new AssetId(buf, groupIndex);
        assetId.validate();
        return assetId;
    }
    static fromString(s) {
        let buf;
        try {
            buf = hex.decode(s);
        }
        catch {
            throw new Error("invalid asset id format, must be hex");
        }
        return AssetId.fromBytes(buf);
    }
    static fromBytes(buf) {
        if (!buf || buf.length === 0) {
            throw new Error("missing asset id");
        }
        if (buf.length !== ASSET_ID_SIZE) {
            throw new Error(`invalid asset id length: got ${buf.length} bytes, want ${ASSET_ID_SIZE} bytes`);
        }
        const reader = new BufferReader(buf);
        return AssetId.fromReader(reader);
    }
    serialize() {
        const writer = new BufferWriter();
        this.serializeTo(writer);
        return writer.toBytes();
    }
    toString() {
        return hex.encode(this.serialize());
    }
    validate() {
        if (isZeroBytes(this.txid)) {
            throw new Error("empty txid");
        }
        if (!Number.isInteger(this.groupIndex) ||
            this.groupIndex < 0 ||
            this.groupIndex > 0xffff) {
            throw new Error(`invalid group index: ${this.groupIndex}, must be in range [0, 65535]`);
        }
    }
    static fromReader(reader) {
        if (reader.remaining() < ASSET_ID_SIZE) {
            throw new Error(`invalid asset id length: got ${reader.remaining()}, want ${ASSET_ID_SIZE}`);
        }
        const txid = reader.readSlice(TX_HASH_SIZE);
        const index = reader.readUint16LE();
        const assetId = new AssetId(txid, index);
        assetId.validate();
        return assetId;
    }
    serializeTo(writer) {
        writer.write(this.txid);
        writer.writeUint16LE(this.groupIndex);
    }
}
