// @arkade-os/sdk/dist/esm/extension/asset/metadata.js
import { schnorr } from "@noble/curves/secp256k1.js";
import { hex } from "@scure/base";
import { compareBytes } from "@scure/btc-signer/utils.js";
import { BufferReader, BufferWriter } from './utils.js';
/**
 * Metadata represents a key-value pair.
 * @param key - the key
 * @param value - the value
 */
export class Metadata {
    constructor(key, value) {
        this.key = key;
        this.value = value;
    }
    static create(key, value) {
        const md = new Metadata(key, value);
        md.validate();
        return md;
    }
    static fromString(s) {
        let buf;
        try {
            buf = hex.decode(s);
        }
        catch {
            throw new Error("invalid metadata format, must be hex");
        }
        return Metadata.fromBytes(buf);
    }
    static fromBytes(buf) {
        if (!buf || buf.length === 0) {
            throw new Error("missing metadata");
        }
        const reader = new BufferReader(buf);
        return Metadata.fromReader(reader);
    }
    serialize() {
        const writer = new BufferWriter();
        this.serializeTo(writer);
        return writer.toBytes();
    }
    toString() {
        return hex.encode(this.serialize());
    }
    get keyString() {
        return new TextDecoder().decode(this.key);
    }
    get valueString() {
        return new TextDecoder().decode(this.value);
    }
    validate() {
        if (this.key.length === 0) {
            throw new Error("missing metadata key");
        }
        if (this.value.length === 0) {
            throw new Error("missing metadata value");
        }
    }
    static fromReader(reader) {
        let key;
        let value;
        try {
            key = reader.readVarSlice();
        }
        catch {
            throw new Error("invalid metadata length");
        }
        try {
            value = reader.readVarSlice();
        }
        catch {
            throw new Error("invalid metadata length");
        }
        const md = new Metadata(key, value);
        md.validate();
        return md;
    }
    serializeTo(writer) {
        writer.writeVarSlice(this.key);
        writer.writeVarSlice(this.value);
    }
}
export class MetadataList {
    constructor(items) {
        this.items = items;
    }
    static fromString(s) {
        let buf;
        try {
            buf = hex.decode(s);
        }
        catch {
            throw new Error("invalid metadata list format");
        }
        return MetadataList.fromBytes(buf);
    }
    static fromBytes(buf) {
        if (!buf || buf.length === 0) {
            throw new Error("missing metadata list");
        }
        const reader = new BufferReader(buf);
        return MetadataList.fromReader(reader);
    }
    static fromReader(reader) {
        const count = Number(reader.readVarUint());
        const items = Array.from({ length: count }, () => Metadata.fromReader(reader));
        return new MetadataList(items);
    }
    serializeTo(writer) {
        writer.writeVarUint(this.items.length);
        for (const item of this) {
            item.serializeTo(writer);
        }
    }
    serialize() {
        const writer = new BufferWriter();
        this.serializeTo(writer);
        return writer.toBytes();
    }
    [Symbol.iterator]() {
        return this.items[Symbol.iterator]();
    }
    get length() {
        return this.items.length;
    }
    hash() {
        if (this.items.length === 0)
            throw new Error("missing metadata list");
        const levels = buildMetadataMerkleTree(this.items);
        return levels[levels.length - 1][0]; // the last level is the root
    }
}
MetadataList.ARK_LEAF_TAG = "ArkadeAssetLeaf";
MetadataList.ARK_BRANCH_TAG = "ArkadeAssetBranch";
MetadataList.ARK_LEAF_VERSION = 0x00;
function computeMetadataLeafHash(md) {
    const writer = new BufferWriter();
    writer.writeByte(MetadataList.ARK_LEAF_VERSION);
    writer.writeVarSlice(md.key);
    writer.writeVarSlice(md.value);
    return schnorr.utils.taggedHash(MetadataList.ARK_LEAF_TAG, writer.toBytes());
}
function computeMetadataBranchHash(a, b) {
    const [smaller, larger] = compareBytes(a, b) === -1 ? [a, b] : [b, a];
    return schnorr.utils.taggedHash(MetadataList.ARK_BRANCH_TAG, smaller, larger);
}
function buildMetadataMerkleTree(leaves) {
    if (leaves.length === 0)
        return [];
    const leafHashes = leaves.map(computeMetadataLeafHash);
    const levels = [leafHashes];
    let current = leafHashes;
    while (current.length > 1) {
        const next = [];
        for (let i = 0; i < current.length; i += 2) {
            if (i + 1 < current.length) {
                next.push(computeMetadataBranchHash(current[i], current[i + 1]));
            }
            else {
                next.push(current[i]);
            }
        }
        levels.push(next);
        current = next;
    }
    return levels;
}
