// @arkade-os/sdk/dist/esm/providers/delegator.js
import { Intent } from '../intent/index.js';
/**
 * REST-based Delegator provider implementation.
 * @example
 * ```typescript
 * const provider = new RestDelegatorProvider('https://delegator.example.com');
 * const info = await provider.getDelegateInfo();
 * await provider.delegate(intent, forfeitTxs);
 * ```
 */
export class RestDelegatorProvider {
    constructor(url) {
        this.url = url;
    }
    async delegate(intent, forfeitTxs, options) {
        const url = `${this.url}/v1/delegate`;
        const response = await fetch(url, {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
            },
            body: JSON.stringify({
                intent: {
                    message: Intent.encodeMessage(intent.message),
                    proof: intent.proof,
                },
                forfeit_txs: forfeitTxs,
                reject_replace: options?.rejectReplace ?? false,
            }),
        });
        if (!response.ok) {
            const errorText = await response.text();
            throw new Error(`Failed to delegate: ${errorText}`);
        }
    }
    async getDelegateInfo() {
        const url = `${this.url}/v1/delegator/info`;
        const response = await fetch(url);
        if (!response.ok) {
            const errorText = await response.text();
            throw new Error(`Failed to get delegate info: ${errorText}`);
        }
        const data = await response.json();
        if (!isDelegateInfo(data)) {
            throw new Error("Invalid delegate info");
        }
        return data;
    }
}
function isDelegateInfo(data) {
    return (!!data &&
        typeof data === "object" &&
        "pubkey" in data &&
        "fee" in data &&
        "delegatorAddress" in data &&
        typeof data.pubkey === "string" &&
        typeof data.fee === "string" &&
        typeof data.delegatorAddress === "string" &&
        data.pubkey !== "" &&
        data.fee !== "" &&
        data.delegatorAddress !== "");
}
