// @arkade-os/sdk/dist/cjs/repositories/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContractRepositoryImpl = exports.WalletRepositoryImpl = exports.rollbackMigration = exports.getMigrationStatus = exports.requiresMigration = exports.migrateWalletRepository = exports.MIGRATION_KEY = void 0;
__exportStar(require("./walletRepository"), exports);
__exportStar(require("./contractRepository"), exports);
__exportStar(require("./inMemory/walletRepository"), exports);
__exportStar(require("./inMemory/contractRepository"), exports);
__exportStar(require("./indexedDB/contractRepository"), exports);
__exportStar(require("./indexedDB/walletRepository"), exports);
var fromStorageAdapter_1 = require("./migrations/fromStorageAdapter");
Object.defineProperty(exports, "MIGRATION_KEY", { enumerable: true, get: function () { return fromStorageAdapter_1.MIGRATION_KEY; } });
Object.defineProperty(exports, "migrateWalletRepository", { enumerable: true, get: function () { return fromStorageAdapter_1.migrateWalletRepository; } });
Object.defineProperty(exports, "requiresMigration", { enumerable: true, get: function () { return fromStorageAdapter_1.requiresMigration; } });
Object.defineProperty(exports, "getMigrationStatus", { enumerable: true, get: function () { return fromStorageAdapter_1.getMigrationStatus; } });
Object.defineProperty(exports, "rollbackMigration", { enumerable: true, get: function () { return fromStorageAdapter_1.rollbackMigration; } });
// Deprecated
var walletRepositoryImpl_1 = require("./migrations/walletRepositoryImpl");
Object.defineProperty(exports, "WalletRepositoryImpl", { enumerable: true, get: function () { return walletRepositoryImpl_1.WalletRepositoryImpl; } });
var contractRepositoryImpl_1 = require("./migrations/contractRepositoryImpl");
Object.defineProperty(exports, "ContractRepositoryImpl", { enumerable: true, get: function () { return contractRepositoryImpl_1.ContractRepositoryImpl; } });
