// @arkade-os/sdk/dist/cjs/repositories/serialization.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deserializeUtxo = exports.deserializeVtxo = exports.deserializeTapLeaf = exports.serializeUtxo = exports.serializeVtxo = exports.serializeTapLeaf = void 0;
const base_1 = require("@scure/base");
const btc_signer_1 = require("@scure/btc-signer");
const serializeTapLeaf = ([cb, s,]) => ({
    cb: base_1.hex.encode(btc_signer_1.TaprootControlBlock.encode(cb)),
    s: base_1.hex.encode(s),
});
exports.serializeTapLeaf = serializeTapLeaf;
const serializeVtxo = (v) => ({
    ...v,
    tapTree: base_1.hex.encode(v.tapTree),
    forfeitTapLeafScript: (0, exports.serializeTapLeaf)(v.forfeitTapLeafScript),
    intentTapLeafScript: (0, exports.serializeTapLeaf)(v.intentTapLeafScript),
    extraWitness: v.extraWitness?.map(base_1.hex.encode),
});
exports.serializeVtxo = serializeVtxo;
const serializeUtxo = (u) => ({
    ...u,
    tapTree: base_1.hex.encode(u.tapTree),
    forfeitTapLeafScript: (0, exports.serializeTapLeaf)(u.forfeitTapLeafScript),
    intentTapLeafScript: (0, exports.serializeTapLeaf)(u.intentTapLeafScript),
    extraWitness: u.extraWitness?.map(base_1.hex.encode),
});
exports.serializeUtxo = serializeUtxo;
const deserializeTapLeaf = (t) => {
    const cb = btc_signer_1.TaprootControlBlock.decode(base_1.hex.decode(t.cb));
    const s = base_1.hex.decode(t.s);
    return [cb, s];
};
exports.deserializeTapLeaf = deserializeTapLeaf;
const deserializeVtxo = (o) => ({
    ...o,
    createdAt: new Date(o.createdAt),
    tapTree: base_1.hex.decode(o.tapTree),
    forfeitTapLeafScript: (0, exports.deserializeTapLeaf)(o.forfeitTapLeafScript),
    intentTapLeafScript: (0, exports.deserializeTapLeaf)(o.intentTapLeafScript),
    extraWitness: o.extraWitness?.map(base_1.hex.decode),
});
exports.deserializeVtxo = deserializeVtxo;
const deserializeUtxo = (o) => ({
    ...o,
    tapTree: base_1.hex.decode(o.tapTree),
    forfeitTapLeafScript: (0, exports.deserializeTapLeaf)(o.forfeitTapLeafScript),
    intentTapLeafScript: (0, exports.deserializeTapLeaf)(o.intentTapLeafScript),
    extraWitness: o.extraWitness?.map(base_1.hex.decode),
});
exports.deserializeUtxo = deserializeUtxo;
