// @arkade-os/sdk/dist/cjs/repositories/indexedDB/contractRepository.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IndexedDBContractRepository = void 0;
const db_1 = require("./db");
const manager_1 = require("./manager");
const schema_1 = require("./schema");
const utils_1 = require("../../worker/browser/utils");
/**
 * IndexedDB-based implementation of ContractRepository.
 *
 * Data is stored as JSON strings in key/value stores.
 */
class IndexedDBContractRepository {
    constructor(dbName = utils_1.DEFAULT_DB_NAME) {
        this.dbName = dbName;
        this.version = 1;
        this.db = null;
    }
    async clear() {
        try {
            const db = await this.getDB();
            return new Promise((resolve, reject) => {
                const transaction = db.transaction([db_1.STORE_CONTRACTS], "readwrite");
                const contractDataStore = transaction.objectStore(db_1.STORE_CONTRACTS);
                const contractsStore = transaction.objectStore(db_1.STORE_CONTRACTS);
                const contractDataRequest = contractDataStore.clear();
                const contractsRequest = contractsStore.clear();
                let completed = 0;
                const checkComplete = () => {
                    completed++;
                    if (completed === 2) {
                        resolve();
                    }
                };
                contractDataRequest.onsuccess = checkComplete;
                contractsRequest.onsuccess = checkComplete;
                contractDataRequest.onerror = () => reject(contractDataRequest.error);
                contractsRequest.onerror = () => reject(contractsRequest.error);
            });
        }
        catch (error) {
            console.error("Failed to clear contract data:", error);
            throw error;
        }
    }
    async getContracts(filter) {
        try {
            const db = await this.getDB();
            const store = db
                .transaction([db_1.STORE_CONTRACTS], "readonly")
                .objectStore(db_1.STORE_CONTRACTS);
            if (!filter || Object.keys(filter).length === 0) {
                return new Promise((resolve, reject) => {
                    const request = store.getAll();
                    request.onerror = () => reject(request.error);
                    request.onsuccess = () => resolve(request.result ?? []);
                });
            }
            const normalizedFilter = normalizeFilter(filter);
            // first by script, primary key
            if (normalizedFilter.has("script")) {
                const scripts = normalizedFilter.get("script");
                const contracts = await Promise.all(scripts.map((script) => new Promise((resolve, reject) => {
                    const req = store.get(script);
                    req.onerror = () => reject(req.error);
                    req.onsuccess = () => resolve(req.result);
                })));
                return this.applyContractFilter(contracts, normalizedFilter);
            }
            // by state, still an index
            if (normalizedFilter.has("state")) {
                const contracts = await this.getContractsByIndexValues(store, "state", normalizedFilter.get("state"));
                return this.applyContractFilter(contracts, normalizedFilter);
            }
            // by type, still an index
            if (normalizedFilter.has("type")) {
                const contracts = await this.getContractsByIndexValues(store, "type", normalizedFilter.get("type"));
                return this.applyContractFilter(contracts, normalizedFilter);
            }
            // any other filtering happens in-memory
            const allContracts = await new Promise((resolve, reject) => {
                const request = store.getAll();
                request.onerror = () => reject(request.error);
                request.onsuccess = () => resolve(request.result ?? []);
            });
            return this.applyContractFilter(allContracts, normalizedFilter);
        }
        catch (error) {
            console.error("Failed to get contracts:", error);
            return [];
        }
    }
    async saveContract(contract) {
        try {
            const db = await this.getDB();
            return new Promise((resolve, reject) => {
                const transaction = db.transaction([db_1.STORE_CONTRACTS], "readwrite");
                const store = transaction.objectStore(db_1.STORE_CONTRACTS);
                const request = store.put(contract);
                request.onerror = () => reject(request.error);
                request.onsuccess = () => resolve();
            });
        }
        catch (error) {
            console.error("Failed to save contract:", error);
            throw error;
        }
    }
    async deleteContract(script) {
        try {
            const db = await this.getDB();
            return new Promise((resolve, reject) => {
                const transaction = db.transaction([db_1.STORE_CONTRACTS], "readwrite");
                const store = transaction.objectStore(db_1.STORE_CONTRACTS);
                const getRequest = store.get(script);
                getRequest.onerror = () => reject(getRequest.error);
                getRequest.onsuccess = () => {
                    const request = store.delete(script);
                    request.onerror = () => reject(request.error);
                    request.onsuccess = () => resolve();
                };
            });
        }
        catch (error) {
            console.error(`Failed to delete contract ${script}:`, error);
            throw error;
        }
    }
    getContractsByIndexValues(store, indexName, values) {
        if (values.length === 0)
            return Promise.resolve([]);
        const index = store.index(indexName);
        const requests = values.map((value) => new Promise((resolve, reject) => {
            const request = index.getAll(value);
            request.onerror = () => reject(request.error);
            request.onsuccess = () => resolve(request.result ?? []);
        }));
        return Promise.all(requests).then((results) => results.flatMap((result) => result));
    }
    applyContractFilter(
    // can filter directly the result of a query
    contracts, filter) {
        return contracts.filter((contract) => {
            if (contract === undefined)
                return false;
            if (filter.has("script") &&
                !filter.get("script")?.includes(contract.script))
                return false;
            if (filter.has("state") &&
                !filter.get("state")?.includes(contract.state))
                return false;
            if (filter.has("type") &&
                !filter.get("type")?.includes(contract.type))
                return false;
            return true;
        });
    }
    async getDB() {
        if (this.db)
            return this.db;
        this.db = await (0, manager_1.openDatabase)(this.dbName, db_1.DB_VERSION, schema_1.initDatabase);
        return this.db;
    }
    async [Symbol.asyncDispose]() {
        if (!this.db)
            return;
        await (0, manager_1.closeDatabase)(this.dbName);
        this.db = null;
    }
}
exports.IndexedDBContractRepository = IndexedDBContractRepository;
const FILTER_FIELDS = ["script", "state", "type"];
// Transform all filter fields into an array of values
function normalizeFilter(filter) {
    const res = new Map();
    FILTER_FIELDS.forEach((current) => {
        if (!filter?.[current])
            return;
        if (Array.isArray(filter[current])) {
            res.set(current, filter[current]);
        }
        else {
            res.set(current, [filter[current]]);
        }
    });
    return res;
}
