// @arkade-os/sdk/dist/cjs/repositories/sqlite/contractRepository.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SQLiteContractRepository = void 0;
/**
 * SQLite-based implementation of ContractRepository.
 *
 * Uses the SQLExecutor interface so consumers can plug in any SQLite driver
 * (expo-sqlite, better-sqlite3, etc.).
 *
 * Tables are created lazily on first operation via `ensureInit()`.
 * The consumer owns the SQLExecutor lifecycle — `[Symbol.asyncDispose]` is a no-op.
 */
class SQLiteContractRepository {
    constructor(db, options) {
        this.db = db;
        this.version = 1;
        this.initPromise = null;
        this.prefix = sanitizePrefix(options?.prefix ?? "ark_");
        this.table = `${this.prefix}contracts`;
    }
    // ── Lifecycle ──────────────────────────────────────────────────────
    ensureInit() {
        if (!this.initPromise) {
            this.initPromise = this.init();
        }
        return this.initPromise;
    }
    async init() {
        await this.db.run(`
            CREATE TABLE IF NOT EXISTS ${this.table} (
                script TEXT PRIMARY KEY,
                address TEXT NOT NULL,
                type TEXT NOT NULL,
                state TEXT NOT NULL,
                params_json TEXT NOT NULL,
                created_at INTEGER NOT NULL,
                expires_at INTEGER,
                label TEXT,
                metadata_json TEXT
            )
        `);
        await this.db.run(`CREATE INDEX IF NOT EXISTS idx_${this.prefix}contracts_type ON ${this.table} (type)`);
        await this.db.run(`CREATE INDEX IF NOT EXISTS idx_${this.prefix}contracts_state ON ${this.table} (state)`);
    }
    async [Symbol.asyncDispose]() {
        // no-op — consumer owns the SQLExecutor lifecycle
    }
    // ── Clear ──────────────────────────────────────────────────────────
    async clear() {
        await this.ensureInit();
        await this.db.run(`DELETE FROM ${this.table}`);
    }
    // ── Contract management ────────────────────────────────────────────
    async getContracts(filter) {
        await this.ensureInit();
        const conditions = [];
        const params = [];
        if (filter) {
            this.addFilterCondition(conditions, params, "script", filter.script);
            this.addFilterCondition(conditions, params, "state", filter.state);
            this.addFilterCondition(conditions, params, "type", filter.type);
        }
        let sql = `SELECT * FROM ${this.table}`;
        if (conditions.length > 0) {
            sql += ` WHERE ${conditions.join(" AND ")}`;
        }
        const rows = await this.db.all(sql, params);
        return rows.map(contractRowToDomain);
    }
    async saveContract(contract) {
        await this.ensureInit();
        await this.db.run(`INSERT OR REPLACE INTO ${this.table}
                (script, address, type, state, params_json,
                 created_at, expires_at, label, metadata_json)
             VALUES (?, ?, ?, ?, ?,
                     ?, ?, ?, ?)`, [
            contract.script,
            contract.address,
            contract.type,
            contract.state,
            JSON.stringify(contract.params),
            contract.createdAt,
            contract.expiresAt ?? null,
            contract.label ?? null,
            contract.metadata ? JSON.stringify(contract.metadata) : null,
        ]);
    }
    async deleteContract(script) {
        await this.ensureInit();
        await this.db.run(`DELETE FROM ${this.table} WHERE script = ?`, [
            script,
        ]);
    }
    // ── Helpers ─────────────────────────────────────────────────────────
    addFilterCondition(conditions, params, column, value) {
        if (value === undefined)
            return;
        if (Array.isArray(value)) {
            if (value.length === 0)
                return;
            const placeholders = value.map(() => "?").join(", ");
            conditions.push(`${column} IN (${placeholders})`);
            params.push(...value);
        }
        else {
            conditions.push(`${column} = ?`);
            params.push(value);
        }
    }
}
exports.SQLiteContractRepository = SQLiteContractRepository;
// ── Row → Domain converter ──────────────────────────────────────────────
const SAFE_PREFIX = /^[a-zA-Z0-9_]+$/;
function sanitizePrefix(prefix) {
    if (!SAFE_PREFIX.test(prefix)) {
        throw new Error(`Invalid table prefix "${prefix}": only letters, digits, and underscores are allowed`);
    }
    return prefix;
}
function contractRowToDomain(row) {
    const contract = {
        script: row.script,
        address: row.address,
        type: row.type,
        state: row.state,
        params: JSON.parse(row.params_json),
        createdAt: row.created_at,
    };
    if (row.expires_at !== null) {
        contract.expiresAt = row.expires_at;
    }
    if (row.label !== null) {
        contract.label = row.label;
    }
    if (row.metadata_json !== null) {
        contract.metadata = JSON.parse(row.metadata_json);
    }
    return contract;
}
