// @arkade-os/sdk/dist/esm/worker/messageBus.js
/// <reference lib="webworker" />
import { getActiveServiceWorker, setupServiceWorkerOnce, } from './browser/service-worker-manager.js';
import { RestArkProvider } from '../providers/ark.js';
import { RestDelegatorProvider } from '../providers/delegator.js';
import { ReadonlySingleKey, SingleKey } from '../identity/index.js';
import { ReadonlyWallet, Wallet } from '../wallet/wallet.js';
import { hex } from "@scure/base";
export class MessageBus {
    constructor(walletRepository, contractRepository, { messageHandlers, tickIntervalMs = 10000, debug = false, buildServices, }) {
        this.walletRepository = walletRepository;
        this.contractRepository = contractRepository;
        this.running = false;
        this.tickTimeout = null;
        this.tickInProgress = false;
        this.debug = false;
        this.initialized = false;
        this.handlers = new Map(messageHandlers.map((u) => [u.messageTag, u]));
        this.tickIntervalMs = tickIntervalMs;
        this.debug = debug;
        this.buildServicesFn = buildServices ?? this.buildServices.bind(this);
    }
    async start() {
        if (this.running)
            return;
        this.running = true;
        if (this.debug)
            console.log("MessageBus starting");
        // Hook message routing
        self.addEventListener("message", this.onMessage.bind(this));
        // activate service worker immediately
        self.addEventListener("install", () => {
            self.skipWaiting();
        });
        // take control of clients immediately
        self.addEventListener("activate", () => {
            self.clients.claim();
            if (this.initialized) {
                this.runTick();
            }
        });
    }
    async stop() {
        if (this.debug)
            console.log("MessageBus stopping");
        this.running = false;
        this.tickInProgress = false;
        this.initialized = false;
        if (this.tickTimeout !== null) {
            self.clearTimeout(this.tickTimeout);
            this.tickTimeout = null;
        }
        self.removeEventListener("message", this.onMessage.bind(this));
        await Promise.all(Array.from(this.handlers.values()).map((updater) => updater.stop()));
    }
    scheduleNextTick() {
        if (!this.running)
            return;
        if (this.tickTimeout !== null)
            return;
        if (this.tickInProgress)
            return;
        this.tickTimeout = self.setTimeout(() => this.runTick(), this.tickIntervalMs);
    }
    async runTick() {
        if (!this.running)
            return;
        if (this.tickInProgress)
            return;
        this.tickInProgress = true;
        if (this.tickTimeout !== null) {
            self.clearTimeout(this.tickTimeout);
            this.tickTimeout = null;
        }
        try {
            const now = Date.now();
            for (const updater of this.handlers.values()) {
                try {
                    const response = await updater.tick(now);
                    if (this.debug)
                        console.log(`[${updater.messageTag}] outgoing tick response:`, response);
                    if (response && response.length > 0) {
                        self.clients
                            .matchAll({
                            includeUncontrolled: true,
                            type: "window",
                        })
                            .then((clients) => {
                            for (const message of response) {
                                clients.forEach((client) => {
                                    client.postMessage(message);
                                });
                            }
                        });
                    }
                }
                catch (err) {
                    if (this.debug)
                        console.error(`[${updater.messageTag}] tick failed`, err);
                }
            }
        }
        finally {
            this.tickInProgress = false;
            this.scheduleNextTick();
        }
    }
    async waitForInit(config) {
        if (this.initialized) {
            // Stop existing handlers before re-initializing.
            // This handles the case where CLEAR was called, which nullifies
            // handler state (readonlyWallet, etc.) without resetting the
            // initialized flag. Without this, handlers never get start()
            // called again and all messages fail with "not initialized".
            //
            // Clear the flag first so onMessage() rejects incoming messages
            // during the stop/start window instead of routing them to
            // half-reset handlers. Restored to true after start() completes.
            this.initialized = false;
            await Promise.all(Array.from(this.handlers.values()).map((h) => h.stop().catch(() => { })));
        }
        const services = await this.buildServicesFn(config);
        // Start all handlers
        for (const updater of this.handlers.values()) {
            if (this.debug)
                console.log(`Starting updater: ${updater.messageTag}`);
            await updater.start(services, {
                walletRepository: this.walletRepository,
            });
        }
        // Kick off scheduler
        this.scheduleNextTick();
        this.initialized = true;
    }
    async buildServices(config) {
        const arkProvider = new RestArkProvider(config.arkServer.url);
        const storage = {
            walletRepository: this.walletRepository,
            contractRepository: this.contractRepository,
        };
        const delegatorProvider = config.delegatorUrl
            ? new RestDelegatorProvider(config.delegatorUrl)
            : undefined;
        if ("privateKey" in config.wallet) {
            const identity = SingleKey.fromHex(config.wallet.privateKey);
            const wallet = await Wallet.create({
                identity,
                arkServerUrl: config.arkServer.url,
                arkServerPublicKey: config.arkServer.publicKey,
                storage,
                delegatorProvider,
            });
            return { wallet, arkProvider, readonlyWallet: wallet };
        }
        else if ("publicKey" in config.wallet) {
            const identity = ReadonlySingleKey.fromPublicKey(hex.decode(config.wallet.publicKey));
            const readonlyWallet = await ReadonlyWallet.create({
                identity,
                arkServerUrl: config.arkServer.url,
                arkServerPublicKey: config.arkServer.publicKey,
                storage,
                delegatorProvider,
            });
            return { readonlyWallet, arkProvider };
        }
        else {
            throw new Error("Missing privateKey or publicKey in configuration object");
        }
    }
    async onMessage(event) {
        const { id, tag, broadcast } = event.data;
        if (tag === "INITIALIZE_MESSAGE_BUS") {
            if (this.debug) {
                console.log("Init Command received");
            }
            await this.waitForInit(event.data.config);
            event.source?.postMessage({ id, tag });
            if (this.debug) {
                console.log("MessageBus initialized");
            }
            return;
        }
        if (!this.initialized) {
            if (this.debug)
                console.warn("Event received before initialization, dropping", event.data);
            // Send error response so the caller's promise rejects instead of
            // hanging forever. This happens when the browser kills and restarts
            // the service worker — the new instance has initialized=false and
            // messages arrive before INITIALIZE_MESSAGE_BUS is re-sent.
            event.source?.postMessage({
                id,
                tag: tag ?? "unknown",
                error: new Error("MessageBus not initialized"),
            });
            return;
        }
        if (!id || !tag) {
            if (this.debug)
                console.error("Invalid message received, missing required fields:", event.data);
            event.source?.postMessage({
                id,
                tag: tag ?? "unknown",
                error: new TypeError("Invalid message received, missing required fields"),
            });
            return;
        }
        if (broadcast) {
            const updaters = Array.from(this.handlers.values());
            const results = await Promise.allSettled(updaters.map((updater) => updater.handleMessage(event.data)));
            results.forEach((result, index) => {
                const updater = updaters[index];
                if (result.status === "fulfilled") {
                    const response = result.value;
                    if (response) {
                        event.source?.postMessage(response);
                    }
                }
                else {
                    if (this.debug)
                        console.error(`[${updater.messageTag}] handleMessage failed`, result.reason);
                    const error = result.reason instanceof Error
                        ? result.reason
                        : new Error(String(result.reason));
                    event.source?.postMessage({
                        id,
                        tag: updater.messageTag,
                        error,
                    });
                }
            });
            return;
        }
        const updater = this.handlers.get(tag);
        if (!updater) {
            if (this.debug)
                console.warn(`[${tag}] unknown message tag, ignoring message`);
            return;
        }
        try {
            const response = await updater.handleMessage(event.data);
            if (this.debug)
                console.log(`[${tag}] outgoing response:`, response);
            if (response) {
                event.source?.postMessage(response);
            }
        }
        catch (err) {
            if (this.debug)
                console.error(`[${tag}] handleMessage failed`, err);
            const error = err instanceof Error ? err : new Error(String(err));
            event.source?.postMessage({ id, tag, error });
        }
    }
    /**
     * Returns the registered SW for the path.
     * It uses the functions in `service-worker-manager.ts` module.
     * @param path
     * @return the Service Worker
     * @throws if not running in a browser environment
     */
    static async getServiceWorker(path) {
        return getActiveServiceWorker(path);
    }
    /**
     * Set up and register the Service Worker, ensuring it's done once at most.
     * It uses the functions in `service-worker-manager.ts` module.
     * @param path
     * @return the Service Worker
     * @throws if not running in a browser environment
     */
    static async setup(path) {
        await setupServiceWorkerOnce(path);
        return getActiveServiceWorker(path);
    }
}
