// @arkade-os/sdk/dist/esm/wallet/expo/background.js
import { hex } from "@scure/base";
import { runTasks } from '../../worker/expo/taskRunner.js';
import { contractPollProcessor, CONTRACT_POLL_TASK_TYPE, } from '../../worker/expo/processors/index.js';
import { DefaultVtxo } from '../../script/default.js';
import { ExpoArkProvider } from '../../providers/expoArk.js';
import { ExpoIndexerProvider } from '../../providers/expoIndexer.js';
import { extendVirtualCoin, extendVtxoFromContract, getRandomId, } from '../utils.js';
function requireTaskManager() {
    try {
        return require("expo-task-manager");
    }
    catch {
        throw new Error("expo-task-manager is required for background tasks. " +
            "Install it with: npx expo install expo-task-manager");
    }
}
function requireBackgroundTask() {
    try {
        return require("expo-background-task");
    }
    catch {
        throw new Error("expo-background-task is required for background tasks. " +
            "Install it with: npx expo install expo-background-task");
    }
}
/**
 * Define the Expo background task handler.
 *
 * **Must be called at module/global scope** (before React mounts).
 * Internally calls `TaskManager.defineTask()`.
 *
 * @example
 * ```ts
 * // At the top of your app entry file
 * import { defineExpoBackgroundTask } from "@arkade-os/sdk/wallet/expo";
 * import { AsyncStorageTaskQueue } from "@arkade-os/sdk/worker/expo";
 * import AsyncStorage from "@react-native-async-storage/async-storage";
 *
 * const taskQueue = new AsyncStorageTaskQueue(AsyncStorage);
 * defineExpoBackgroundTask("ark-background-poll", {
 *     taskQueue,
 *     walletRepository: new IndexedDBWalletRepository(),
 *     contractRepository: new IndexedDBContractRepository(),
 * });
 * ```
 */
export function defineExpoBackgroundTask(taskName, options) {
    const TaskManager = requireTaskManager();
    const BackgroundTask = requireBackgroundTask();
    const { taskQueue, walletRepository, contractRepository, processors = [contractPollProcessor], } = options;
    TaskManager.defineTask(taskName, async () => {
        try {
            const config = await taskQueue.loadConfig();
            if (!config) {
                // No config persisted yet — ExpoWallet.setup() hasn't run.
                // Nothing to do.
                return BackgroundTask.BackgroundTaskResult.Success;
            }
            // Reconstruct providers
            const indexerProvider = new ExpoIndexerProvider(config.arkServerUrl);
            const arkProvider = new ExpoArkProvider(config.arkServerUrl);
            // Reconstruct default offchainTapscript as fallback
            // for VTXOs not associated with a contract.
            const defaultTapscript = new DefaultVtxo.Script({
                pubKey: hex.decode(config.pubkeyHex),
                serverPubKey: hex.decode(config.serverPubKeyHex),
                csvTimelock: {
                    value: BigInt(config.exitTimelockValue),
                    type: config.exitTimelockType,
                },
            });
            await runTasks(taskQueue, processors, {
                walletRepository,
                contractRepository,
                indexerProvider,
                arkProvider,
                extendVtxo: (vtxo, contract) => {
                    if (contract) {
                        return extendVtxoFromContract(vtxo, contract);
                    }
                    return extendVirtualCoin({ offchainTapscript: defaultTapscript }, vtxo);
                },
            });
            // Acknowledge outbox results (no foreground to consume them)
            const results = await taskQueue.getResults();
            if (results.length > 0) {
                await taskQueue.acknowledgeResults(results.map((r) => r.id));
            }
            // Re-seed the contract-poll task for the next OS wake
            const existing = await taskQueue.getTasks(CONTRACT_POLL_TASK_TYPE);
            if (existing.length === 0) {
                const task = {
                    id: getRandomId(),
                    type: CONTRACT_POLL_TASK_TYPE,
                    data: {},
                    createdAt: Date.now(),
                };
                await taskQueue.addTask(task);
            }
            return BackgroundTask.BackgroundTaskResult.Success;
        }
        catch (error) {
            console.error("[ark-sdk] Background task failed:", error instanceof Error ? error.message : error);
            return BackgroundTask.BackgroundTaskResult.Failed;
        }
    });
}
/**
 * Activate the OS-level background task scheduler.
 *
 * Call this after {@link defineExpoBackgroundTask} (typically inside
 * {@link ExpoWallet.setup} or in a React component after wallet init).
 *
 * @param minimumInterval - Minimum interval in minutes (default 15).
 */
export async function registerExpoBackgroundTask(taskName, options) {
    const BackgroundTask = requireBackgroundTask();
    await BackgroundTask.registerTaskAsync(taskName, {
        minimumInterval: (options?.minimumInterval ?? 15) * 60,
    });
}
/**
 * Unregister the background task from the OS scheduler.
 */
export async function unregisterExpoBackgroundTask(taskName) {
    const BackgroundTask = requireBackgroundTask();
    await BackgroundTask.unregisterTaskAsync(taskName);
}
