// @arkade-os/sdk/dist/esm/wallet/expo/wallet.js
import { hex } from "@scure/base";
import { Wallet } from '../wallet.js';
import { RestArkProvider } from '../../providers/ark.js';
import { runTasks } from '../../worker/expo/taskRunner.js';
import { contractPollProcessor, CONTRACT_POLL_TASK_TYPE, } from '../../worker/expo/processors/index.js';
import { extendVirtualCoin, extendVtxoFromContract, getRandomId, } from '../utils.js';
import { DefaultVtxo } from '../../script/default.js';
/**
 * Expo/React Native wallet with built-in background task processing.
 *
 * Wraps a standard {@link Wallet} and adds a lightweight task queue
 * for keeping contract/VTXO state fresh while the app is active and
 * across Expo BackgroundTask wakes.
 *
 * @example
 * ```ts
 * import { ExpoWallet } from "@arkade-os/sdk/wallet/expo";
 * import { AsyncStorageTaskQueue } from "@arkade-os/sdk/worker/expo";
 *
 * const wallet = await ExpoWallet.setup({
 *     identity: SingleKey.fromHex(privateKey),
 *     arkServerUrl,
 *     esploraUrl,
 *     storage: { walletRepository, contractRepository },
 *     background: {
 *         taskName: "ark-background-poll",
 *         taskQueue: new AsyncStorageTaskQueue(AsyncStorage),
 *         foregroundIntervalMs: 20_000,
 *         minimumBackgroundInterval: 15,
 *     },
 * });
 *
 * const balance = await wallet.getBalance();
 * ```
 */
export class ExpoWallet {
    constructor(wallet, taskQueue, processors, deps, taskName, foregroundIntervalMs) {
        this.wallet = wallet;
        this.taskQueue = taskQueue;
        this.processors = processors;
        this.deps = deps;
        this.identity = wallet.identity;
        this.arkProvider = wallet.arkProvider;
        this.indexerProvider = wallet.indexerProvider;
        this.taskName = taskName;
        if (foregroundIntervalMs && foregroundIntervalMs > 0) {
            this.startForegroundPolling(foregroundIntervalMs);
        }
    }
    /**
     * Create an ExpoWallet with background task support.
     *
     * 1. Creates the inner {@link Wallet} via `Wallet.create()`.
     * 2. Wires up processors (defaults to {@link contractPollProcessor}).
     * 3. Persists background config for the background handler (if the queue supports it).
     * 4. Seeds the task queue with a `contract-poll` task.
     * 5. Registers the background task with the OS scheduler (if `minimumBackgroundInterval` is set).
     * 6. Starts foreground polling if `foregroundIntervalMs` is set.
     */
    static async setup(config) {
        const wallet = await Wallet.create(config);
        const processors = config.background.processors ?? [
            contractPollProcessor,
        ];
        const deps = {
            walletRepository: wallet.walletRepository,
            contractRepository: wallet.contractRepository,
            indexerProvider: wallet.indexerProvider,
            arkProvider: wallet.arkProvider,
            extendVtxo: (vtxo, contract) => {
                if (contract) {
                    return extendVtxoFromContract(vtxo, contract);
                }
                return extendVirtualCoin(wallet, vtxo);
            },
        };
        const { taskQueue } = config.background;
        // Persist wallet params so the background handler can rehydrate
        // without a network call. Only works with AsyncStorageTaskQueue.
        if ("persistConfig" in taskQueue) {
            const arkServerUrl = config.arkServerUrl ??
                (wallet.arkProvider instanceof RestArkProvider
                    ? wallet.arkProvider.serverUrl
                    : undefined);
            if (arkServerUrl) {
                const timelock = wallet.offchainTapscript.options.csvTimelock ??
                    DefaultVtxo.Script.DEFAULT_TIMELOCK;
                const bgConfig = {
                    arkServerUrl,
                    pubkeyHex: hex.encode(wallet.offchainTapscript.options.pubKey),
                    serverPubKeyHex: hex.encode(wallet.offchainTapscript.options.serverPubKey),
                    exitTimelockValue: timelock.value.toString(),
                    exitTimelockType: timelock.type,
                };
                await taskQueue.persistConfig(bgConfig);
            }
        }
        const expoWallet = new ExpoWallet(wallet, taskQueue, processors, deps, config.background.taskName, config.background.foregroundIntervalMs);
        // Seed the queue so the first tick (or background wake) has work to do
        await expoWallet.seedContractPollTask();
        // Activate OS-level background scheduling
        if (config.background.minimumBackgroundInterval) {
            try {
                const { registerExpoBackgroundTask } = await import("./background");
                await registerExpoBackgroundTask(config.background.taskName, {
                    minimumInterval: config.background.minimumBackgroundInterval,
                });
            }
            catch {
                // expo-background-task not installed — foreground-only mode
            }
        }
        return expoWallet;
    }
    // ── Foreground polling ───────────────────────────────────────────
    startForegroundPolling(intervalMs) {
        this.foregroundIntervalId = setInterval(() => {
            this.runForegroundPoll().catch(console.error);
        }, intervalMs);
    }
    async runForegroundPoll() {
        await runTasks(this.taskQueue, this.processors, this.deps);
        // Consume results immediately (no background handoff needed)
        const results = await this.taskQueue.getResults();
        if (results.length > 0) {
            await this.taskQueue.acknowledgeResults(results.map((r) => r.id));
        }
        // Re-seed for the next tick
        await this.seedContractPollTask();
    }
    async seedContractPollTask() {
        const existing = await this.taskQueue.getTasks(CONTRACT_POLL_TASK_TYPE);
        if (existing.length > 0)
            return;
        const task = {
            id: getRandomId(),
            type: CONTRACT_POLL_TASK_TYPE,
            data: {},
            createdAt: Date.now(),
        };
        await this.taskQueue.addTask(task);
    }
    // ── Lifecycle ────────────────────────────────────────────────────
    /**
     * Stop foreground polling and unregister the background task.
     */
    async dispose() {
        if (this.foregroundIntervalId) {
            clearInterval(this.foregroundIntervalId);
            this.foregroundIntervalId = undefined;
        }
        try {
            const { unregisterExpoBackgroundTask } = await import("./background");
            await unregisterExpoBackgroundTask(this.taskName);
        }
        catch {
            // expo-background-task not installed — nothing to unregister
        }
    }
    // ── IWallet delegation ───────────────────────────────────────────
    getAddress() {
        return this.wallet.getAddress();
    }
    getBoardingAddress() {
        return this.wallet.getBoardingAddress();
    }
    getBalance() {
        return this.wallet.getBalance();
    }
    getVtxos(filter) {
        return this.wallet.getVtxos(filter);
    }
    getBoardingUtxos() {
        return this.wallet.getBoardingUtxos();
    }
    getTransactionHistory() {
        return this.wallet.getTransactionHistory();
    }
    getContractManager() {
        return this.wallet.getContractManager();
    }
    getDelegatorManager() {
        return this.wallet.getDelegatorManager();
    }
    sendBitcoin(params) {
        return this.wallet.sendBitcoin(params);
    }
    settle(params, eventCallback) {
        return this.wallet.settle(params, eventCallback);
    }
    send(...recipients) {
        return this.wallet.send(...recipients);
    }
    get assetManager() {
        return this.wallet.assetManager;
    }
}
