// @kukks/bitcoin-descriptors/dist/signers.js
// Copyright (c) 2025 Jose-Luis Landabaso - https://bitcoinerlab.com
// Distributed under the MIT software license
import { readUInt32BE } from './scriptUtils.js';
/**
 * Adapts a BIP32Interface node to the HDKey interface expected by @scure/btc-signer.
 * scure's Transaction.sign/signIdx auto-handle Taproot tweaking when given an HDKey.
 */
function toScureHDKey(node) {
    if (!node.privateKey)
        throw new Error('Cannot create HDKey signer from neutered node');
    return {
        publicKey: node.publicKey,
        privateKey: node.privateKey,
        fingerprint: readUInt32BE(node.fingerprint),
        derive(path) {
            return toScureHDKey(node.derivePath(path));
        },
        deriveChild(index) {
            return toScureHDKey(node.derive(index));
        },
        sign(hash) {
            return node.sign(hash);
        }
    };
}
/**
 * Derives the private key from a BIP32 node using tapBip32Derivation info
 * and signs the Taproot input directly with the raw private key.
 * This is needed because @scure/btc-signer's signIdx HDKey path only checks
 * bip32Derivation, not tapBip32Derivation.
 */
function signTapBip32(psbt, index, masterNode) {
    const input = psbt.getInput(index);
    const tapBip32 = input.tapBip32Derivation;
    if (!tapBip32 || !tapBip32.length)
        return false;
    const fp = readUInt32BE(masterNode.fingerprint);
    let signed = false;
    for (const [, { der }] of tapBip32) {
        if (der.fingerprint !== fp)
            continue;
        let node = masterNode;
        for (const childIdx of der.path) {
            node = node.derive(childIdx);
        }
        if (!node.privateKey)
            throw new Error('Cannot sign: derived node has no private key');
        psbt.signIdx(node.privateKey, index);
        signed = true;
    }
    return signed;
}
/**
 * Signs a specific input of a PSBT with an ECPair.
 *
 * Uses @scure/btc-signer's signIdx which automatically handles
 * Taproot key tweaking internally.
 *
 * @param {Object} params - The parameters object
 * @param {PsbtLike} params.psbt - The PSBT to sign
 * @param {number} params.index - The input index to sign
 * @param {ECPairInterface} params.ecpair - The ECPair to sign with
 */
export function signInputECPair({ psbt, index, ecpair }) {
    if (!ecpair.privateKey)
        throw new Error('Missing private key');
    psbt.signIdx(ecpair.privateKey, index);
}
/**
 * Signs all inputs of a PSBT with an ECPair.
 *
 * Uses @scure/btc-signer's sign which automatically handles
 * Taproot key tweaking internally.
 *
 * @param {Object} params - The parameters object
 * @param {PsbtLike} params.psbt - The PSBT to sign
 * @param {ECPairInterface} params.ecpair - The ECPair to sign with
 */
export function signECPair({ psbt, ecpair }) {
    if (!ecpair.privateKey)
        throw new Error('Missing private key');
    const signed = psbt.sign(ecpair.privateKey);
    if (signed === 0) {
        throw new Error('No inputs were signed');
    }
}
/**
 * Signs a specific input of a PSBT with a BIP32 node.
 *
 * Handles Taproot inputs via tapBip32Derivation, and non-Taproot via
 * @scure/btc-signer's signIdx with an HDKey adapter.
 *
 * @param {Object} params - The parameters object
 * @param {PsbtLike} params.psbt - The PSBT to sign
 * @param {number} params.index - The input index to sign
 * @param {BIP32Interface} params.node - The BIP32 node to sign with
 */
export function signInputBIP32({ psbt, index, node }) {
    if (!signTapBip32(psbt, index, node)) {
        psbt.signIdx(toScureHDKey(node), index);
    }
}
/**
 * Signs all inputs of a PSBT with a BIP32 master node.
 *
 * First signs any Taproot inputs via tapBip32Derivation, then uses
 * @scure/btc-signer's sign for remaining non-Taproot inputs.
 *
 * @param {Object} params - The parameters object
 * @param {PsbtLike} params.psbt - The PSBT to sign
 * @param {BIP32Interface} params.masterNode - The BIP32 master node to sign with
 */
export function signBIP32({ psbt, masterNode }) {
    let tapSigned = 0;
    for (let i = 0; i < psbt.inputsLength; i++) {
        if (signTapBip32(psbt, i, masterNode))
            tapSigned++;
    }
    let nonTapSigned = 0;
    try {
        nonTapSigned = psbt.sign(toScureHDKey(masterNode));
    }
    catch {
        // sign() throws 'No inputs signed' when no bip32Derivation matches
    }
    if (tapSigned + nonTapSigned === 0) {
        throw new Error('No inputs were signed');
    }
}
