// @arkade-os/sdk/dist/cjs/wallet/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DUST_AMOUNT = void 0;
exports.extendVirtualCoin = extendVirtualCoin;
exports.extendCoin = extendCoin;
exports.extendVtxoFromContract = extendVtxoFromContract;
exports.getRandomId = getRandomId;
exports.isValidArkAddress = isValidArkAddress;
exports.validateRecipients = validateRecipients;
const __1 = require("..");
const handlers_1 = require("../contracts/handlers");
const base_1 = require("@scure/base");
exports.DUST_AMOUNT = 546; // sats
function extendVirtualCoin(wallet, vtxo) {
    return {
        ...vtxo,
        forfeitTapLeafScript: wallet.offchainTapscript.forfeit(),
        intentTapLeafScript: wallet.offchainTapscript.forfeit(),
        tapTree: wallet.offchainTapscript.encode(),
    };
}
function extendCoin(wallet, utxo) {
    return {
        ...utxo,
        forfeitTapLeafScript: wallet.boardingTapscript.forfeit(),
        intentTapLeafScript: wallet.boardingTapscript.forfeit(),
        tapTree: wallet.boardingTapscript.encode(),
    };
}
function extendVtxoFromContract(vtxo, contract) {
    const handler = handlers_1.contractHandlers.get(contract.type);
    if (!handler) {
        throw new Error(`No handler for contract type '${contract.type}'`);
    }
    const script = handler.createScript(contract.params);
    return {
        ...vtxo,
        forfeitTapLeafScript: script.forfeit(),
        intentTapLeafScript: script.forfeit(),
        tapTree: script.encode(),
    };
}
function getRandomId() {
    const randomValue = crypto.getRandomValues(new Uint8Array(16));
    return base_1.hex.encode(randomValue);
}
function isValidArkAddress(address) {
    try {
        __1.ArkAddress.decode(address);
        return true;
    }
    catch (e) {
        return false;
    }
}
function validateRecipients(recipients, dustAmount) {
    const validatedRecipients = [];
    for (const recipient of recipients) {
        let address;
        try {
            address = __1.ArkAddress.decode(recipient.address);
        }
        catch (e) {
            throw new Error(`Invalid Ark address: ${recipient.address}`);
        }
        const amount = recipient.amount || dustAmount;
        if (amount <= 0) {
            throw new Error("Amount must be positive");
        }
        validatedRecipients.push({
            address: recipient.address,
            assets: recipient.assets ?? [],
            amount,
            script: amount < dustAmount
                ? address.subdustPkScript
                : address.pkScript,
        });
    }
    return validatedRecipients;
}
