// @arkade-os/boltz-swap/dist/chunk-XVJ47GNJ.js
// src/errors.ts
var SwapError = class extends Error {
  isClaimable;
  isRefundable;
  pendingSwap;
  constructor(options = {}) {
    super(options.message ?? "Error during swap.");
    this.name = "SwapError";
    this.isClaimable = options.isClaimable ?? false;
    this.isRefundable = options.isRefundable ?? false;
    this.pendingSwap = options.pendingSwap;
  }
};
var InvoiceExpiredError = class extends SwapError {
  constructor(options = {}) {
    super({ message: "The invoice has expired.", ...options });
    this.name = "InvoiceExpiredError";
  }
};
var InvoiceFailedToPayError = class extends SwapError {
  constructor(options = {}) {
    super({
      message: "The provider failed to pay the invoice",
      ...options
    });
    this.name = "InvoiceFailedToPayError";
  }
};
var InsufficientFundsError = class extends SwapError {
  constructor(options = {}) {
    super({ message: "Not enough funds available", ...options });
    this.name = "InsufficientFundsError";
  }
};
var NetworkError = class extends Error {
  statusCode;
  errorData;
  constructor(message, statusCode, errorData) {
    super(message);
    this.name = "NetworkError";
    this.statusCode = statusCode;
    this.errorData = errorData;
  }
};
var SchemaError = class extends SwapError {
  constructor(options = {}) {
    super({ message: "Invalid API response", ...options });
    this.name = "SchemaError";
  }
};
var SwapExpiredError = class extends SwapError {
  constructor(options = {}) {
    super({ message: "The swap has expired", ...options });
    this.name = "SwapExpiredError";
  }
};
var TransactionFailedError = class extends SwapError {
  constructor(options = {}) {
    super({ message: "The transaction has failed.", ...options });
    this.name = "TransactionFailedError";
  }
};
var PreimageFetchError = class extends SwapError {
  constructor(options = {}) {
    super({
      message: "The payment settled, but fetching the preimage failed.",
      ...options
    });
    this.name = "PreimageFetchError";
  }
};
var TransactionLockupFailedError = class extends SwapError {
  constructor(options = {}) {
    super({ message: "The transaction lockup has failed.", ...options });
    this.name = "TransactionLockupFailedError";
  }
};
var TransactionRefundedError = class extends SwapError {
  constructor(options = {}) {
    super({ message: "The transaction has been refunded.", ...options });
    this.name = "TransactionRefundedError";
  }
};

// src/boltz-swap-provider.ts
import { Transaction } from "@arkade-os/sdk";
import { base64 } from "@scure/base";
var isSubmarineFailedStatus = (status) => {
  return [
    "invoice.failedToPay",
    "transaction.lockupFailed",
    "swap.expired"
  ].includes(status);
};
var isSubmarineFinalStatus = (status) => {
  return [
    "invoice.failedToPay",
    "transaction.claimed",
    "swap.expired"
  ].includes(status);
};
var isSubmarinePendingStatus = (status) => {
  return [
    "swap.created",
    "transaction.mempool",
    "transaction.confirmed",
    "invoice.set",
    "invoice.pending",
    "invoice.paid",
    "transaction.claim.pending"
  ].includes(status);
};
var isSubmarineRefundableStatus = (status) => {
  return [
    "invoice.failedToPay",
    "transaction.lockupFailed",
    "swap.expired"
  ].includes(status);
};
var isSubmarineSuccessStatus = (status) => {
  return status === "transaction.claimed";
};
var isReverseFailedStatus = (status) => {
  return [
    "invoice.expired",
    "transaction.failed",
    "transaction.refunded",
    "swap.expired"
  ].includes(status);
};
var isReverseFinalStatus = (status) => {
  return [
    "transaction.refunded",
    "transaction.failed",
    "invoice.settled",
    // normal status for completed swaps
    "swap.expired"
  ].includes(status);
};
var isReversePendingStatus = (status) => {
  return [
    "swap.created",
    "transaction.mempool",
    "transaction.confirmed"
  ].includes(status);
};
var isReverseClaimableStatus = (status) => {
  return ["transaction.mempool", "transaction.confirmed"].includes(status);
};
var isReverseSuccessStatus = (status) => {
  return status === "invoice.settled";
};
var isChainFailedStatus = (status) => {
  return ["transaction.failed", "swap.expired"].includes(status);
};
var isChainClaimableStatus = (status) => {
  return [
    "transaction.server.mempool",
    "transaction.server.confirmed"
  ].includes(status);
};
var isChainFinalStatus = (status) => {
  return [
    "transaction.refunded",
    "transaction.failed",
    "transaction.claimed",
    // normal status for completed swaps
    "swap.expired"
  ].includes(status);
};
var isChainPendingStatus = (status) => {
  return [
    "swap.created",
    "transaction.mempool",
    "transaction.confirmed",
    "transaction.lockupFailed",
    "transaction.server.mempool",
    "transaction.server.confirmed"
  ].includes(status);
};
var isChainRefundableStatus = (status) => {
  return ["swap.expired"].includes(status);
};
var isChainSignableStatus = (status) => {
  return ["transaction.claim.pending"].includes(status);
};
var isChainSuccessStatus = (status) => {
  return status === "transaction.claimed";
};
var isPendingReverseSwap = (swap) => {
  return swap.type === "reverse";
};
var isPendingSubmarineSwap = (swap) => {
  return swap.type === "submarine";
};
var isPendingChainSwap = (swap) => {
  return swap.type === "chain";
};
var isSubmarineSwapRefundable = (swap) => {
  return isSubmarineRefundableStatus(swap.status) && isPendingSubmarineSwap(swap) && swap.refundable !== false && swap.refunded !== true;
};
var isChainSwapRefundable = (swap) => {
  return isChainRefundableStatus(swap.status) && isPendingChainSwap(swap) && swap.request.from === "ARK";
};
var isReverseSwapClaimable = (swap) => {
  return isReverseClaimableStatus(swap.status) && isPendingReverseSwap(swap);
};
var isChainSwapClaimable = (swap) => {
  return isChainClaimableStatus(swap.status) && isPendingChainSwap(swap);
};
var isTimeoutBlockHeights = (data) => {
  return data && typeof data === "object" && typeof data.refund === "number" && typeof data.unilateralClaim === "number" && typeof data.unilateralRefund === "number" && typeof data.unilateralRefundWithoutReceiver === "number";
};
var isGetReverseSwapTxIdResponse = (data) => {
  return data && typeof data === "object" && typeof data.id === "string" && typeof data.timeoutBlockHeight === "number";
};
var isGetSwapStatusResponse = (data) => {
  return data && typeof data === "object" && typeof data.status === "string" && (data.zeroConfRejected === void 0 || typeof data.zeroConfRejected === "boolean") && (data.transaction === void 0 || data.transaction && typeof data.transaction === "object" && typeof data.transaction.id === "string" && (data.transaction.eta === void 0 || typeof data.transaction.eta === "number") && (data.transaction.hex === void 0 || typeof data.transaction.hex === "string") && (data.transaction.preimage === void 0 || typeof data.transaction.preimage === "string"));
};
var isGetSubmarinePairsResponse = (data) => {
  return data && typeof data === "object" && data.ARK && typeof data.ARK === "object" && data.ARK.BTC && typeof data.ARK.BTC === "object" && typeof data.ARK.BTC.hash === "string" && typeof data.ARK.BTC.rate === "number" && data.ARK.BTC.limits && typeof data.ARK.BTC.limits === "object" && typeof data.ARK.BTC.limits.maximal === "number" && typeof data.ARK.BTC.limits.minimal === "number" && typeof data.ARK.BTC.limits.maximalZeroConf === "number" && data.ARK.BTC.fees && typeof data.ARK.BTC.fees === "object" && typeof data.ARK.BTC.fees.percentage === "number" && typeof data.ARK.BTC.fees.minerFees === "number";
};
var isGetReversePairsResponse = (data) => {
  return data && typeof data === "object" && data.BTC && typeof data.BTC === "object" && data.BTC.ARK && typeof data.BTC.ARK === "object" && data.BTC.ARK.hash && typeof data.BTC.ARK.hash === "string" && typeof data.BTC.ARK.rate === "number" && data.BTC.ARK.limits && typeof data.BTC.ARK.limits === "object" && typeof data.BTC.ARK.limits.maximal === "number" && typeof data.BTC.ARK.limits.minimal === "number" && data.BTC.ARK.fees && typeof data.BTC.ARK.fees === "object" && typeof data.BTC.ARK.fees.percentage === "number" && typeof data.BTC.ARK.fees.minerFees === "object" && typeof data.BTC.ARK.fees.minerFees.claim === "number" && typeof data.BTC.ARK.fees.minerFees.lockup === "number";
};
var isCreateSubmarineSwapResponse = (data) => {
  return data && typeof data === "object" && typeof data.id === "string" && typeof data.address === "string" && typeof data.expectedAmount === "number" && typeof data.claimPublicKey === "string" && typeof data.acceptZeroConf === "boolean" && isTimeoutBlockHeights(data.timeoutBlockHeights);
};
var isGetSwapPreimageResponse = (data) => {
  return data && typeof data === "object" && typeof data.preimage === "string";
};
var isCreateReverseSwapResponse = (data) => {
  return data && typeof data === "object" && typeof data.id === "string" && typeof data.invoice === "string" && typeof data.onchainAmount === "number" && typeof data.lockupAddress === "string" && typeof data.refundPublicKey === "string" && isTimeoutBlockHeights(data.timeoutBlockHeights);
};
var isRefundSubmarineSwapResponse = (data) => {
  return data && typeof data === "object" && typeof data.transaction === "string" && typeof data.checkpoint === "string";
};
var isRefundChainSwapResponse = (data) => {
  return data && typeof data === "object" && typeof data.transaction === "string" && typeof data.checkpoint === "string";
};
var isGetChainPairsResponse = (data) => {
  return data && typeof data === "object" && data.ARK && data.BTC && typeof data.ARK === "object" && typeof data.BTC === "object" && data.ARK.BTC && data.BTC.ARK && typeof data.ARK.BTC === "object" && typeof data.BTC.ARK === "object" && typeof data.ARK.BTC.hash === "string" && typeof data.BTC.ARK.hash === "string" && typeof data.ARK.BTC.rate === "number" && typeof data.BTC.ARK.rate === "number" && data.ARK.BTC.limits && data.BTC.ARK.limits && typeof data.ARK.BTC.limits === "object" && typeof data.BTC.ARK.limits === "object" && typeof data.ARK.BTC.limits.maximal === "number" && typeof data.BTC.ARK.limits.maximal === "number" && typeof data.ARK.BTC.limits.minimal === "number" && typeof data.BTC.ARK.limits.minimal === "number" && typeof data.ARK.BTC.limits.maximalZeroConf === "number" && typeof data.BTC.ARK.limits.maximalZeroConf === "number" && data.ARK.BTC.fees && data.BTC.ARK.fees && typeof data.ARK.BTC.fees === "object" && typeof data.BTC.ARK.fees === "object" && typeof data.ARK.BTC.fees.percentage === "number" && typeof data.BTC.ARK.fees.percentage === "number" && typeof data.ARK.BTC.fees.minerFees === "object" && typeof data.BTC.ARK.fees.minerFees === "object" && typeof data.ARK.BTC.fees.minerFees.server === "number" && typeof data.BTC.ARK.fees.minerFees.server === "number" && data.ARK.BTC.fees.minerFees.user && data.BTC.ARK.fees.minerFees.user && typeof data.ARK.BTC.fees.minerFees.user === "object" && typeof data.BTC.ARK.fees.minerFees.user === "object" && typeof data.ARK.BTC.fees.minerFees.user.claim === "number" && typeof data.BTC.ARK.fees.minerFees.user.claim === "number" && typeof data.ARK.BTC.fees.minerFees.user.lockup === "number" && typeof data.BTC.ARK.fees.minerFees.user.lockup === "number";
};
var isSwapTree = (data) => {
  return data && typeof data === "object" && typeof data.claimLeaf === "object" && typeof data.claimLeaf.version === "number" && typeof data.claimLeaf.output === "string" && typeof data.refundLeaf === "object" && typeof data.refundLeaf.version === "number" && typeof data.refundLeaf.output === "string";
};
var isChainSwapDetailsResponse = (data) => {
  return data && typeof data === "object" && typeof data.amount === "number" && typeof data.lockupAddress === "string" && typeof data.serverPublicKey === "string" && typeof data.timeoutBlockHeight === "number" && (data.swapTree === void 0 || isSwapTree(data.swapTree)) && (data.timeouts === void 0 || isTimeoutBlockHeights(data.timeouts)) && (data.bip21 === void 0 || typeof data.bip21 === "string");
};
var isCreateChainSwapResponse = (data) => {
  return data && typeof data === "object" && typeof data.id === "string" && isChainSwapDetailsResponse(data.claimDetails) && isChainSwapDetailsResponse(data.lockupDetails);
};
var isGetChainClaimDetailsResponse = (data) => {
  return data && typeof data === "object" && typeof data.pubNonce === "string" && typeof data.publicKey === "string" && typeof data.transactionHash === "string";
};
var isPostChainClaimDetailsResponse = (data) => {
  return data && typeof data === "object" && (typeof data.pubNonce === "string" && typeof data.partialSignature === "string" || typeof data.pubNonce === "undefined" && typeof data.partialSignature === "undefined");
};
var isGetChainQuoteResponse = (data) => {
  return data && typeof data === "object" && typeof data.amount === "number";
};
var isPostChainQuoteResponse = (data) => {
  return data && typeof data === "object" && Object.keys(data).length === 0 && data.constructor === Object;
};
var isPostBtcTransactionResponse = (data) => {
  return data && typeof data === "object" && typeof data.id === "string";
};
var isLeaf = (data) => {
  return data && typeof data === "object" && typeof data.version === "number" && typeof data.output === "string";
};
var isTree = (data) => {
  return data && typeof data === "object" && isLeaf(data.claimLeaf) && isLeaf(data.refundLeaf) && isLeaf(data.refundWithoutBoltzLeaf) && isLeaf(data.unilateralClaimLeaf) && isLeaf(data.unilateralRefundLeaf) && isLeaf(data.unilateralRefundWithoutBoltzLeaf);
};
var isDetails = (data) => {
  return data && typeof data === "object" && isTree(data.tree) && (data.amount === void 0 || typeof data.amount === "number") && typeof data.keyIndex === "number" && (data.transaction === void 0 || data.transaction && typeof data.transaction === "object" && typeof data.transaction.id === "string" && typeof data.transaction.vout === "number") && typeof data.lockupAddress === "string" && typeof data.serverPublicKey === "string" && typeof data.timeoutBlockHeight === "number" && isTimeoutBlockHeights(data.timeoutBlockHeights) && (data.preimageHash === void 0 || typeof data.preimageHash === "string");
};
var isRestoredSubmarineSwap = (data) => {
  return data && typeof data === "object" && data.to === "BTC" && typeof data.id === "string" && data.from === "ARK" && data.type === "submarine" && typeof data.createdAt === "number" && typeof data.preimageHash === "string" && typeof data.status === "string" && isDetails(data.refundDetails);
};
var isRestoredReverseSwap = (data) => {
  return data && typeof data === "object" && data.to === "ARK" && typeof data.id === "string" && data.from === "BTC" && data.type === "reverse" && typeof data.createdAt === "number" && typeof data.preimageHash === "string" && typeof data.status === "string" && isDetails(data.claimDetails);
};
var isCreateSwapsRestoreResponse = (data) => {
  return Array.isArray(data) && data.every(
    (item) => isRestoredReverseSwap(item) || isRestoredSubmarineSwap(item)
  );
};
var BASE_URLS = {
  mutinynet: "https://api.boltz.mutinynet.arkade.sh",
  regtest: "http://localhost:9069"
};
var BoltzSwapProvider = class {
  wsUrl;
  apiUrl;
  network;
  referralId;
  constructor(config) {
    this.network = config.network;
    this.referralId = config.referralId;
    const apiUrl = config.apiUrl || BASE_URLS[config.network];
    if (!apiUrl)
      throw new Error(
        `API URL is required for network: ${config.network}`
      );
    this.apiUrl = apiUrl;
    this.wsUrl = this.apiUrl.replace(/^http(s)?:\/\//, "ws$1://").replace("9069", "9004") + "/v2/ws";
  }
  getApiUrl() {
    return this.apiUrl;
  }
  getWsUrl() {
    return this.wsUrl;
  }
  getNetwork() {
    return this.network;
  }
  async getFees() {
    const [submarine, reverse] = await Promise.all([
      this.request(
        "/v2/swap/submarine",
        "GET"
      ),
      this.request("/v2/swap/reverse", "GET")
    ]);
    if (!isGetSubmarinePairsResponse(submarine))
      throw new SchemaError({ message: "error fetching submarine fees" });
    if (!isGetReversePairsResponse(reverse))
      throw new SchemaError({ message: "error fetching reverse fees" });
    return {
      submarine: {
        percentage: submarine.ARK.BTC.fees.percentage,
        minerFees: submarine.ARK.BTC.fees.minerFees
      },
      reverse: {
        percentage: reverse.BTC.ARK.fees.percentage,
        minerFees: reverse.BTC.ARK.fees.minerFees
      }
    };
  }
  async getLimits() {
    const response = await this.request(
      "/v2/swap/submarine",
      "GET"
    );
    if (!isGetSubmarinePairsResponse(response))
      throw new SchemaError({ message: "error fetching limits" });
    return {
      min: response.ARK.BTC.limits.minimal,
      max: response.ARK.BTC.limits.maximal
    };
  }
  async getReverseSwapTxId(id) {
    const res = await this.request(
      `/v2/swap/reverse/${id}/transaction`,
      "GET"
    );
    if (!isGetReverseSwapTxIdResponse(res))
      throw new SchemaError({
        message: `error fetching txid for swap: ${id}`
      });
    return res;
  }
  async getSwapStatus(id) {
    const response = await this.request(
      `/v2/swap/${id}`,
      "GET"
    );
    if (!isGetSwapStatusResponse(response))
      throw new SchemaError({
        message: `error fetching status for swap: ${id}`
      });
    return response;
  }
  async getSwapPreimage(id) {
    const res = await this.request(
      `/v2/swap/submarine/${id}/preimage`,
      "GET"
    );
    if (!isGetSwapPreimageResponse(res))
      throw new SchemaError({
        message: `error fetching preimage for swap: ${id}`
      });
    return res;
  }
  async createSubmarineSwap({
    invoice,
    refundPublicKey
  }) {
    if (refundPublicKey.length != 66) {
      throw new SwapError({
        message: "refundPublicKey must be a compressed public key"
      });
    }
    const requestBody = {
      from: "ARK",
      to: "BTC",
      invoice,
      refundPublicKey,
      ...this.referralId ? { referralId: this.referralId } : {}
    };
    const response = await this.request(
      "/v2/swap/submarine",
      "POST",
      requestBody
    );
    if (!isCreateSubmarineSwapResponse(response))
      throw new SchemaError({ message: "Error creating submarine swap" });
    return response;
  }
  async createReverseSwap({
    invoiceAmount,
    claimPublicKey,
    preimageHash,
    description
  }) {
    if (claimPublicKey.length != 66) {
      throw new SwapError({
        message: "claimPublicKey must be a compressed public key"
      });
    }
    const requestBody = {
      from: "BTC",
      to: "ARK",
      invoiceAmount,
      claimPublicKey,
      preimageHash,
      ...description?.trim() ? { description: description.trim() } : {},
      ...this.referralId ? { referralId: this.referralId } : {}
    };
    const response = await this.request(
      "/v2/swap/reverse",
      "POST",
      requestBody
    );
    if (!isCreateReverseSwapResponse(response))
      throw new SchemaError({ message: "Error creating reverse swap" });
    return response;
  }
  async createChainSwap({
    to,
    from,
    preimageHash,
    feeSatsPerByte,
    claimPublicKey,
    refundPublicKey,
    serverLockAmount,
    userLockAmount
  }) {
    if (["BTC", "ARK"].indexOf(to) === -1)
      throw new SwapError({ message: "Invalid 'to' chain" });
    if (["BTC", "ARK"].indexOf(from) === -1)
      throw new SwapError({ message: "Invalid 'from' chain" });
    if (to === from)
      throw new SwapError({ message: "Invalid swap direction" });
    if (!preimageHash || preimageHash.length != 64)
      throw new SwapError({ message: "Invalid preimageHash" });
    if (feeSatsPerByte <= 0)
      throw new SwapError({ message: "Invalid feeSatsPerByte" });
    if (serverLockAmount !== void 0 && userLockAmount !== void 0 || serverLockAmount === void 0 && userLockAmount === void 0)
      throw new SwapError({
        message: "Either serverLockAmount or userLockAmount must be provided"
      });
    if (userLockAmount !== void 0 && userLockAmount <= 0)
      throw new SwapError({ message: "Invalid userLockAmount" });
    if (serverLockAmount !== void 0 && serverLockAmount <= 0)
      throw new SwapError({ message: "Invalid serverLockAmount" });
    if (claimPublicKey.length != 66) {
      throw new SwapError({
        message: "claimPublicKey must be a compressed public key"
      });
    }
    if (refundPublicKey.length != 66) {
      throw new SwapError({
        message: "refundPublicKey must be a compressed public key"
      });
    }
    const requestBody = {
      to,
      from,
      preimageHash,
      feeSatsPerByte,
      claimPublicKey,
      refundPublicKey,
      serverLockAmount,
      userLockAmount,
      ...this.referralId ? { referralId: this.referralId } : {}
    };
    const response = await this.request(
      "/v2/swap/chain",
      "POST",
      requestBody
    );
    if (!isCreateChainSwapResponse(response))
      throw new SchemaError({ message: "Error creating chain swap" });
    return response;
  }
  async refundSubmarineSwap(swapId, transaction, checkpoint) {
    const requestBody = {
      checkpoint: base64.encode(checkpoint.toPSBT()),
      transaction: base64.encode(transaction.toPSBT())
    };
    const response = await this.request(
      `/v2/swap/submarine/${swapId}/refund/ark`,
      "POST",
      requestBody
    );
    if (!isRefundSubmarineSwapResponse(response))
      throw new SchemaError({
        message: "Error refunding submarine swap"
      });
    return {
      transaction: Transaction.fromPSBT(
        base64.decode(response.transaction)
      ),
      checkpoint: Transaction.fromPSBT(
        base64.decode(response.checkpoint)
      )
    };
  }
  async refundChainSwap(swapId, transaction, checkpoint) {
    const requestBody = {
      checkpoint: base64.encode(checkpoint.toPSBT()),
      transaction: base64.encode(transaction.toPSBT())
    };
    const response = await this.request(
      `/v2/swap/chain/${swapId}/refund/ark`,
      "POST",
      requestBody
    );
    if (!isRefundChainSwapResponse(response))
      throw new SchemaError({
        message: "Error refunding chain swap"
      });
    return {
      transaction: Transaction.fromPSBT(
        base64.decode(response.transaction)
      ),
      checkpoint: Transaction.fromPSBT(
        base64.decode(response.checkpoint)
      )
    };
  }
  async monitorSwap(swapId, update) {
    return new Promise((resolve, reject) => {
      const webSocket = new globalThis.WebSocket(this.wsUrl);
      const connectionTimeout = setTimeout(() => {
        webSocket.close();
        reject(new NetworkError("WebSocket connection timeout"));
      }, 3e4);
      webSocket.onerror = (error) => {
        clearTimeout(connectionTimeout);
        reject(
          new NetworkError(
            `WebSocket error: ${error.message}`
          )
        );
      };
      webSocket.onopen = () => {
        clearTimeout(connectionTimeout);
        webSocket.send(
          JSON.stringify({
            op: "subscribe",
            channel: "swap.update",
            args: [swapId]
          })
        );
      };
      webSocket.onclose = () => {
        clearTimeout(connectionTimeout);
        resolve();
      };
      webSocket.onmessage = async (rawMsg) => {
        const msg = JSON.parse(rawMsg.data);
        if (msg.event !== "update" || msg.args[0].id !== swapId) return;
        if (msg.args[0].error) {
          webSocket.close();
          reject(new SwapError({ message: msg.args[0].error }));
        }
        const status = msg.args[0].status;
        const negotiable = status === "transaction.lockupFailed" && msg.args[0].failureDetails?.actual !== void 0 && msg.args[0].failureDetails?.expected !== void 0;
        switch (status) {
          case "invoice.settled":
          case "transaction.claimed":
          case "transaction.refunded":
          case "invoice.expired":
          case "invoice.failedToPay":
          case "transaction.failed":
          case "swap.expired":
            webSocket.close();
            update(status, msg.args[0]);
            break;
          case "transaction.lockupFailed":
            if (!negotiable) webSocket.close();
            update(status, msg.args[0]);
            break;
          case "invoice.paid":
          case "invoice.pending":
          case "invoice.set":
          case "swap.created":
          case "transaction.mempool":
          case "transaction.confirmed":
          case "transaction.claim.pending":
          case "transaction.server.mempool":
          case "transaction.server.confirmed":
            update(status, msg.args[0]);
        }
      };
    });
  }
  async getChainFees(from, to) {
    if (from === to) {
      throw new SwapError({ message: "Invalid chain pair" });
    }
    const response = await this.request(
      "/v2/swap/chain",
      "GET"
    );
    if (!isGetChainPairsResponse(response))
      throw new SchemaError({ message: "error fetching fees" });
    if (!response[from]?.[to]) {
      throw new SchemaError({
        message: `unsupported chain pair: ${from} -> ${to}`
      });
    }
    return response[from][to].fees;
  }
  async getChainLimits(from, to) {
    if (from === to) {
      throw new SwapError({ message: "Invalid chain pair" });
    }
    const response = await this.request(
      "/v2/swap/chain",
      "GET"
    );
    if (!isGetChainPairsResponse(response))
      throw new SchemaError({ message: "error fetching limits" });
    if (!response[from]?.[to]) {
      throw new SchemaError({
        message: `unsupported chain pair: ${from} -> ${to}`
      });
    }
    return {
      min: response[from][to].limits.minimal,
      max: response[from][to].limits.maximal
    };
  }
  async getChainClaimDetails(swapId) {
    const response = await this.request(
      `/v2/swap/chain/${swapId}/claim`,
      "GET"
    );
    if (!isGetChainClaimDetailsResponse(response))
      throw new SchemaError({
        message: `error fetching claim details for swap: ${swapId}`
      });
    return response;
  }
  async getChainQuote(swapId) {
    const response = await this.request(
      `/v2/swap/chain/${swapId}/quote`,
      "GET"
    );
    if (!isGetChainQuoteResponse(response))
      throw new SchemaError({
        message: `error fetching quote for swap: ${swapId}`
      });
    return response;
  }
  async postChainQuote(swapId, request) {
    const response = await this.request(
      `/v2/swap/chain/${swapId}/quote`,
      "POST",
      request
    );
    if (!isPostChainQuoteResponse(response))
      throw new SchemaError({
        message: `error posting quote for swap: ${swapId}`
      });
    return response;
  }
  async postBtcTransaction(hex9) {
    const requestBody = { hex: hex9 };
    const response = await this.request(
      "/v2/chain/BTC/transaction",
      "POST",
      requestBody
    );
    if (!isPostBtcTransactionResponse(response))
      throw new SchemaError({
        message: "error posting BTC transaction"
      });
    return response;
  }
  async postChainClaimDetails(swapId, request) {
    const response = await this.request(
      `/v2/swap/chain/${swapId}/claim`,
      "POST",
      request
    );
    if (!isPostChainClaimDetailsResponse(response))
      throw new SchemaError({
        message: `error posting claim details for swap: ${swapId}`
      });
    return response;
  }
  async restoreSwaps(publicKey) {
    const requestBody = {
      publicKey
    };
    const response = await this.request(
      "/v2/swap/restore",
      "POST",
      requestBody
    );
    if (!isCreateSwapsRestoreResponse(response))
      throw new SchemaError({
        message: "Invalid schema in response for swap restoration"
      });
    return response;
  }
  async request(path, method, body) {
    const url = `${this.apiUrl}${path}`;
    try {
      const response = await globalThis.fetch(url, {
        method,
        headers: { "Content-Type": "application/json" },
        body: body ? JSON.stringify(body) : void 0
      });
      if (!response.ok) {
        const errorBody = await response.text();
        let errorData;
        try {
          errorData = JSON.parse(errorBody);
        } catch {
        }
        const message = `Boltz API error: ${response.status} ${errorBody}`;
        throw new NetworkError(message, response.status, errorData);
      }
      if (response.headers.get("content-length") === "0") {
        throw new NetworkError("Empty response from Boltz API");
      }
      return await response.json();
    } catch (error) {
      if (error instanceof NetworkError) throw error;
      throw new NetworkError(
        `Request to ${url} failed: ${error.message}`
      );
    }
  }
};

// src/utils/decoding.ts
import bolt11 from "light-bolt11-decoder";
import { ArkAddress } from "@arkade-os/sdk";
var decodeInvoice = (invoice) => {
  const decoded = bolt11.decode(invoice);
  const millisats = Number(
    decoded.sections.find((s) => s.name === "amount")?.value ?? "0"
  );
  return {
    expiry: decoded.expiry ?? 3600,
    amountSats: Math.floor(millisats / 1e3),
    description: decoded.sections.find((s) => s.name === "description")?.value ?? "",
    paymentHash: decoded.sections.find((s) => s.name === "payment_hash")?.value ?? ""
  };
};
var getInvoiceSatoshis = (invoice) => {
  return decodeInvoice(invoice).amountSats;
};
var getInvoicePaymentHash = (invoice) => {
  return decodeInvoice(invoice).paymentHash;
};
var isValidArkAddress = (address) => {
  try {
    ArkAddress.decode(address);
    return true;
  } catch (e) {
    return false;
  }
};

// src/utils/signatures.ts
import { verifyTapscriptSignatures } from "@arkade-os/sdk";
import { hex } from "@scure/base";
var verifySignatures = (tx, inputIndex, requiredSigners) => {
  try {
    verifyTapscriptSignatures(tx, inputIndex, requiredSigners);
    return true;
  } catch (_) {
    return false;
  }
};
var normalizeToXOnlyKey = (someKey, keyName = "", swapId = "") => {
  const keyBytes = typeof someKey === "string" ? hex.decode(someKey) : someKey;
  if (keyBytes.length === 33) {
    return keyBytes.slice(1);
  }
  if (keyBytes.length !== 32) {
    throw new Error(
      `Invalid ${keyName} key length: ${keyBytes.length} ${swapId ? "for swap " + swapId : ""}`
    );
  }
  return keyBytes;
};

// src/logger.ts
var logger = console;
function setLogger(customLogger) {
  logger = customLogger;
}

// src/swap-manager.ts
var SwapManager = class {
  swapProvider;
  config;
  // Event listeners storage (supports multiple listeners per event)
  swapUpdateListeners = /* @__PURE__ */ new Set();
  swapCompletedListeners = /* @__PURE__ */ new Set();
  swapFailedListeners = /* @__PURE__ */ new Set();
  actionExecutedListeners = /* @__PURE__ */ new Set();
  wsConnectedListeners = /* @__PURE__ */ new Set();
  wsDisconnectedListeners = /* @__PURE__ */ new Set();
  // State
  websocket = null;
  monitoredSwaps = /* @__PURE__ */ new Map();
  initialSwaps = /* @__PURE__ */ new Map();
  // All swaps passed to start(), including completed ones
  pollTimer = null;
  reconnectTimer = null;
  isRunning = false;
  currentReconnectDelay;
  currentPollRetryDelay;
  usePollingFallback = false;
  isReconnecting = false;
  webSocketUnavailable = false;
  // Race condition prevention
  swapsInProgress = /* @__PURE__ */ new Set();
  // Per-swap subscriptions for UI hooks
  swapSubscriptions = /* @__PURE__ */ new Map();
  // Action callbacks (injected via setCallbacks)
  claimCallback = null;
  refundCallback = null;
  claimArkCallback = null;
  claimBtcCallback = null;
  refundArkCallback = null;
  signServerClaimCallback = null;
  saveSwapCallback = null;
  constructor(swapProvider, config = {}) {
    this.swapProvider = swapProvider;
    this.config = {
      enableAutoActions: config.enableAutoActions ?? true,
      pollInterval: config.pollInterval ?? 3e4,
      reconnectDelayMs: config.reconnectDelayMs ?? 1e3,
      maxReconnectDelayMs: config.maxReconnectDelayMs ?? 6e4,
      pollRetryDelayMs: config.pollRetryDelayMs ?? 5e3,
      maxPollRetryDelayMs: config.maxPollRetryDelayMs ?? 3e5,
      maxPollIntervalMs: config.maxPollIntervalMs ?? 3e5,
      events: config.events ?? {}
    };
    if (config.events?.onSwapUpdate) {
      this.swapUpdateListeners.add(config.events.onSwapUpdate);
    }
    if (config.events?.onSwapCompleted) {
      this.swapCompletedListeners.add(config.events.onSwapCompleted);
    }
    if (config.events?.onSwapFailed) {
      this.swapFailedListeners.add(config.events.onSwapFailed);
    }
    if (config.events?.onActionExecuted) {
      this.actionExecutedListeners.add(config.events.onActionExecuted);
    }
    if (config.events?.onWebSocketConnected) {
      this.wsConnectedListeners.add(config.events.onWebSocketConnected);
    }
    if (config.events?.onWebSocketDisconnected) {
      this.wsDisconnectedListeners.add(
        config.events.onWebSocketDisconnected
      );
    }
    this.currentReconnectDelay = this.config.reconnectDelayMs;
    this.currentPollRetryDelay = this.config.pollRetryDelayMs;
  }
  /**
   * Set callbacks for claim, refund, and save operations.
   * These are called by the manager when autonomous actions are needed.
   */
  setCallbacks(callbacks) {
    this.claimCallback = callbacks.claim;
    this.refundCallback = callbacks.refund;
    this.claimArkCallback = callbacks.claimArk;
    this.claimBtcCallback = callbacks.claimBtc;
    this.refundArkCallback = callbacks.refundArk;
    this.signServerClaimCallback = callbacks.signServerClaim ?? null;
    this.saveSwapCallback = callbacks.saveSwap;
  }
  /**
   * Add an event listener for swap updates
   * @returns Unsubscribe function
   */
  async onSwapUpdate(listener) {
    this.swapUpdateListeners.add(listener);
    return () => this.swapUpdateListeners.delete(listener);
  }
  /**
   * Add an event listener for swap completion
   * @returns Unsubscribe function
   */
  async onSwapCompleted(listener) {
    this.swapCompletedListeners.add(listener);
    return () => this.swapCompletedListeners.delete(listener);
  }
  /**
   * Add an event listener for swap failures
   * @returns Unsubscribe function
   */
  async onSwapFailed(listener) {
    this.swapFailedListeners.add(listener);
    return () => this.swapFailedListeners.delete(listener);
  }
  /**
   * Add an event listener for executed actions (claim/refund)
   * @returns Unsubscribe function
   */
  async onActionExecuted(listener) {
    this.actionExecutedListeners.add(listener);
    return () => this.actionExecutedListeners.delete(listener);
  }
  /**
   * Add an event listener for WebSocket connection
   * @returns Unsubscribe function
   */
  async onWebSocketConnected(listener) {
    this.wsConnectedListeners.add(listener);
    return () => this.wsConnectedListeners.delete(listener);
  }
  /**
   * Add an event listener for WebSocket disconnection
   * @returns Unsubscribe function
   */
  async onWebSocketDisconnected(listener) {
    this.wsDisconnectedListeners.add(listener);
    return () => this.wsDisconnectedListeners.delete(listener);
  }
  /** Remove a swap update listener */
  offSwapUpdate(listener) {
    this.swapUpdateListeners.delete(listener);
  }
  /** Remove a swap completed listener */
  offSwapCompleted(listener) {
    this.swapCompletedListeners.delete(listener);
  }
  /** Remove a swap failed listener */
  offSwapFailed(listener) {
    this.swapFailedListeners.delete(listener);
  }
  /** Remove an action executed listener */
  offActionExecuted(listener) {
    this.actionExecutedListeners.delete(listener);
  }
  /** Remove a WebSocket connected listener */
  offWebSocketConnected(listener) {
    this.wsConnectedListeners.delete(listener);
  }
  /** Remove a WebSocket disconnected listener */
  offWebSocketDisconnected(listener) {
    this.wsDisconnectedListeners.delete(listener);
  }
  /**
   * Start the swap manager
   * This will:
   * 1. Load pending swaps
   * 2. Connect WebSocket (with fallback to polling)
   * 3. Poll all swaps after connection
   * 4. Resume any actionable swaps
   */
  async start(pendingSwaps) {
    if (this.isRunning) {
      logger.warn("SwapManager is already running");
      return;
    }
    this.isRunning = true;
    this.initialSwaps.clear();
    for (const swap of pendingSwaps) {
      this.initialSwaps.set(swap.id, swap);
    }
    for (const swap of pendingSwaps) {
      if (!this.isFinalStatus(swap)) {
        this.monitoredSwaps.set(swap.id, swap);
      }
    }
    await this.tryConnectWebSocket();
    await this.resumeActionableSwaps();
  }
  /**
   * Stop the swap manager
   * Cleanup: close WebSocket, stop all timers
   */
  async stop() {
    if (!this.isRunning) return;
    this.isRunning = false;
    if (this.websocket) {
      this.websocket.close();
      this.websocket = null;
    }
    if (this.pollTimer) {
      clearTimeout(this.pollTimer);
      this.pollTimer = null;
    }
    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }
  }
  /**
   * Set the polling interval (ms).
   * Restarts any running timer so the new interval takes effect immediately.
   */
  setPollInterval(ms) {
    if (ms <= 0) {
      throw new RangeError(
        `setPollInterval: ms must be a positive number, got ${ms}`
      );
    }
    const cappedInterval = Math.min(ms, this.config.maxPollIntervalMs);
    if (cappedInterval !== ms) {
      logger.warn(
        `setPollInterval: requested ${ms}ms exceeds maxPollIntervalMs ${this.config.maxPollIntervalMs}ms, clamping to ${cappedInterval}ms`
      );
    }
    this.config.pollInterval = cappedInterval;
    this.currentPollRetryDelay = Math.min(
      cappedInterval,
      this.config.pollRetryDelayMs
    );
    if (this.isRunning) {
      if (this.usePollingFallback) {
        this.startPollingFallback();
      } else if (this.pollTimer) {
        this.startPolling();
      }
    }
  }
  /**
   * Add a new swap to monitoring.
   * If fallback polling is active, this triggers an immediate poll and resets
   * fallback delay so newly-added swaps are checked without waiting.
   */
  async addSwap(swap) {
    this.monitoredSwaps.set(swap.id, swap);
    if (this.websocket && this.websocket.readyState === WebSocket.OPEN) {
      this.subscribeToSwap(swap.id);
    }
    if (this.usePollingFallback && this.isRunning) {
      this.currentPollRetryDelay = Math.min(
        this.config.pollInterval,
        this.config.pollRetryDelayMs
      );
      this.pollAllSwaps();
      this.startPollingFallback();
    }
  }
  /**
   * Remove a swap from monitoring
   */
  async removeSwap(swapId) {
    this.monitoredSwaps.delete(swapId);
    this.swapSubscriptions.delete(swapId);
    logger.log(`Removed swap ${swapId} from monitoring`);
  }
  /**
   * Get all currently monitored swaps
   */
  async getPendingSwaps() {
    return Array.from(this.monitoredSwaps.values());
  }
  /**
   * Subscribe to updates for a specific swap
   * Returns an unsubscribe function
   * Useful for UI components that need to track specific swap progress
   */
  async subscribeToSwapUpdates(swapId, callback) {
    if (!this.swapSubscriptions.has(swapId)) {
      this.swapSubscriptions.set(swapId, /* @__PURE__ */ new Set());
    }
    const subscribers = this.swapSubscriptions.get(swapId);
    subscribers.add(callback);
    return () => {
      subscribers.delete(callback);
      if (subscribers.size === 0) {
        this.swapSubscriptions.delete(swapId);
      }
    };
  }
  /**
   * Wait for a specific swap to complete.
   * Blocks until the swap reaches a final status or fails.
   * Useful when you want blocking behavior even with SwapManager enabled.
   *
   * @throws If the swap is already in a final status (submarine/chain swaps throw immediately;
   *         reverse swaps return the existing txid).
   * @throws If the swap is not found in the manager.
   */
  async waitForSwapCompletion(swapId) {
    let swap = this.monitoredSwaps.get(swapId);
    if (!swap) {
      swap = this.initialSwaps.get(swapId);
      if (!swap) {
        throw new Error(`Swap ${swapId} not found in manager`);
      }
    }
    if (this.isFinalStatus(swap)) {
      if (isPendingReverseSwap(swap)) {
        const response = await this.swapProvider.getReverseSwapTxId(
          swap.id
        );
        return { txid: response.id };
      }
      if (isPendingSubmarineSwap(swap)) {
        throw new Error("Submarine swap already completed");
      }
      throw new Error("Chain swap already completed");
    }
    return new Promise((resolve, reject) => {
      let unsubscribe = null;
      const handleUpdate = (updatedSwap, _oldStatus) => {
        if (!this.isFinalStatus(updatedSwap)) return;
        unsubscribe?.();
        if (isPendingReverseSwap(updatedSwap)) {
          if (updatedSwap.status === "invoice.settled") {
            this.swapProvider.getReverseSwapTxId(updatedSwap.id).then((response) => resolve({ txid: response.id })).catch((error) => reject(error));
          } else {
            reject(
              new Error(
                `Swap failed with status: ${updatedSwap.status}`
              )
            );
          }
        } else if (isPendingSubmarineSwap(updatedSwap)) {
          if (updatedSwap.status === "transaction.claimed") {
            resolve({ txid: updatedSwap.id });
          } else {
            reject(
              new Error(
                `Swap failed with status: ${updatedSwap.status}`
              )
            );
          }
        } else if (isPendingChainSwap(updatedSwap)) {
          if (updatedSwap.status === "transaction.claimed") {
            resolve({ txid: updatedSwap.id });
          } else {
            reject(
              new Error(
                `Swap failed with status: ${updatedSwap.status}`
              )
            );
          }
        }
      };
      this.subscribeToSwapUpdates(swapId, handleUpdate).then((unsub) => {
        unsubscribe = unsub;
      }).catch(reject);
    });
  }
  /**
   * Check if a swap is currently being processed
   * Useful for preventing race conditions
   */
  async isProcessing(swapId) {
    return this.swapsInProgress.has(swapId);
  }
  /**
   * Check if manager has a specific swap
   */
  async hasSwap(swapId) {
    return this.monitoredSwaps.has(swapId);
  }
  /**
   * Try connecting to WebSocket for real-time swap updates.
   * If WebSocket is unavailable in the current runtime, switch directly to polling fallback.
   * If connection setup fails, also switches to polling fallback.
   */
  async tryConnectWebSocket() {
    if (this.isReconnecting || this.webSocketUnavailable) return;
    if (!this.hasWebSocketSupport()) {
      this.webSocketUnavailable = true;
      this.enterPollingFallback(
        new NetworkError("WebSocket is not available in this runtime")
      );
      return;
    }
    this.isReconnecting = true;
    try {
      const wsUrl = this.swapProvider.getWsUrl();
      this.websocket = new globalThis.WebSocket(wsUrl);
      const connectionTimeout = setTimeout(() => {
        logger.error("WebSocket connection timeout");
        this.websocket?.close();
        this.enterPollingFallback(
          new NetworkError("WebSocket connection failed")
        );
      }, 1e4);
      this.websocket.onerror = (error) => {
        clearTimeout(connectionTimeout);
        logger.error("WebSocket error:", error);
        this.enterPollingFallback(
          new NetworkError("WebSocket connection failed")
        );
      };
      this.websocket.onopen = () => {
        clearTimeout(connectionTimeout);
        this.currentReconnectDelay = this.config.reconnectDelayMs;
        this.usePollingFallback = false;
        this.isReconnecting = false;
        for (const swapId of this.monitoredSwaps.keys()) {
          this.subscribeToSwap(swapId);
        }
        this.pollAllSwaps();
        this.startPolling();
        this.wsConnectedListeners.forEach((listener) => listener());
      };
      this.websocket.onclose = () => {
        clearTimeout(connectionTimeout);
        this.websocket = null;
        if (this.isRunning) {
          this.scheduleReconnect();
        }
        this.wsDisconnectedListeners.forEach((listener) => listener());
      };
      this.websocket.onmessage = async (rawMsg) => {
        await this.handleWebSocketMessage(rawMsg);
      };
    } catch (error) {
      logger.error("Failed to create WebSocket:", error);
      this.enterPollingFallback(
        new NetworkError("WebSocket connection failed")
      );
    }
  }
  /**
   * Runtime feature detection for WebSocket API.
   */
  hasWebSocketSupport() {
    return typeof globalThis.WebSocket === "function";
  }
  /**
   * Enter polling fallback mode and notify listeners about WebSocket unavailability/failure.
   * Resets fallback delay to its configured initial value (capped by max poll interval).
   */
  enterPollingFallback(error) {
    if (!this.isRunning) return;
    this.isReconnecting = false;
    this.websocket = null;
    this.usePollingFallback = true;
    this.currentPollRetryDelay = Math.min(
      this.config.pollRetryDelayMs,
      this.config.maxPollIntervalMs
    );
    this.startPollingFallback();
    this.wsDisconnectedListeners.forEach((listener) => {
      listener(error);
    });
  }
  /**
   * Schedule WebSocket reconnection with exponential backoff
   */
  scheduleReconnect() {
    if (this.reconnectTimer || this.webSocketUnavailable || !this.hasWebSocketSupport())
      return;
    logger.log(
      `Scheduling WebSocket reconnect in ${this.currentReconnectDelay}ms`
    );
    this.reconnectTimer = setTimeout(() => {
      this.reconnectTimer = null;
      this.isReconnecting = false;
      this.tryConnectWebSocket();
    }, this.currentReconnectDelay);
    this.currentReconnectDelay = Math.min(
      this.currentReconnectDelay * 2,
      this.config.maxReconnectDelayMs
    );
  }
  /**
   * Subscribe to a specific swap ID on the WebSocket
   */
  subscribeToSwap(swapId) {
    if (!this.websocket || this.websocket.readyState !== WebSocket.OPEN)
      return;
    this.websocket.send(
      JSON.stringify({
        op: "subscribe",
        channel: "swap.update",
        args: [swapId]
      })
    );
  }
  /**
   * Handle incoming WebSocket message
   */
  async handleWebSocketMessage(rawMsg) {
    try {
      const msg = JSON.parse(rawMsg.data);
      if (msg.event !== "update") return;
      const swapId = msg.args[0]?.id;
      if (!swapId) return;
      const swap = this.monitoredSwaps.get(swapId);
      if (!swap) return;
      if (msg.args[0].error) {
        logger.error(`Swap ${swapId} error:`, msg.args[0].error);
        const error = new Error(msg.args[0].error);
        this.swapFailedListeners.forEach(
          (listener) => listener(swap, error)
        );
        return;
      }
      const newStatus = msg.args[0].status;
      await this.handleSwapStatusUpdate(swap, newStatus);
    } catch (error) {
      logger.error("Error handling WebSocket message:", error);
    }
  }
  /**
   * Handle status update for a swap
   * This is the core logic that determines what actions to take
   */
  async handleSwapStatusUpdate(swap, newStatus) {
    const oldStatus = swap.status;
    if (oldStatus === newStatus) return;
    swap.status = newStatus;
    this.swapUpdateListeners.forEach(
      (listener) => listener(swap, oldStatus)
    );
    const subscribers = this.swapSubscriptions.get(swap.id);
    if (subscribers) {
      subscribers.forEach((callback) => {
        try {
          callback(swap, oldStatus);
        } catch (error) {
          logger.error(
            `Error in swap subscription callback for ${swap.id}:`,
            error
          );
        }
      });
    }
    await this.saveSwap(swap);
    if (this.config.enableAutoActions) {
      await this.executeAutonomousAction(swap);
    }
    if (this.isFinalStatus(swap)) {
      this.monitoredSwaps.delete(swap.id);
      this.swapSubscriptions.delete(swap.id);
      this.swapCompletedListeners.forEach((listener) => listener(swap));
    }
  }
  /**
   * Execute autonomous action based on swap status
   * Uses locking to prevent race conditions with manual operations
   */
  async executeAutonomousAction(swap) {
    if (this.swapsInProgress.has(swap.id)) {
      logger.log(
        `Swap ${swap.id} is already being processed, skipping autonomous action`
      );
      return;
    }
    try {
      this.swapsInProgress.add(swap.id);
      if (isPendingReverseSwap(swap)) {
        if (!swap.preimage || swap.preimage.length === 0) {
          logger.log(
            `Skipping claim for swap ${swap.id}: missing preimage (restored swap)`
          );
          return;
        }
        if (isReverseClaimableStatus(swap.status)) {
          logger.log(`Auto-claiming reverse swap ${swap.id}`);
          await this.executeClaimAction(swap);
          this.actionExecutedListeners.forEach(
            (listener) => listener(swap, "claim")
          );
        }
      } else if (isPendingSubmarineSwap(swap)) {
        if (!swap.request?.invoice || swap.request.invoice.length === 0) {
          logger.log(
            `Skipping refund for swap ${swap.id}: missing invoice (restored swap)`
          );
          return;
        }
        if (isSubmarineRefundableStatus(swap.status)) {
          logger.log(`Auto-refunding submarine swap ${swap.id}`);
          await this.executeRefundAction(swap);
          this.actionExecutedListeners.forEach(
            (listener) => listener(swap, "refund")
          );
        }
      } else if (isPendingChainSwap(swap)) {
        if (isChainClaimableStatus(swap.status)) {
          if (swap.request.to === "ARK") {
            logger.log(`Auto-claiming ARK chain swap ${swap.id}`);
            await this.executeClaimArkAction(swap);
            this.actionExecutedListeners.forEach(
              (listener) => listener(swap, "claimArk")
            );
          } else if (swap.request.to === "BTC") {
            logger.log(`Auto-claiming BTC chain swap ${swap.id}`);
            await this.executeClaimBtcAction(swap);
            this.actionExecutedListeners.forEach(
              (listener) => listener(swap, "claimBtc")
            );
          }
        } else if (isChainRefundableStatus(swap.status)) {
          if (swap.request.from === "ARK") {
            logger.log(`Auto-refunding ARK chain swap ${swap.id}`);
            await this.executeRefundArkAction(swap);
            this.actionExecutedListeners.forEach(
              (listener) => listener(swap, "refundArk")
            );
          }
          if (swap.request.from === "BTC") {
            logger.warn(
              `Chain swap ${swap.id} expired: BTC lockup will be refunded by Boltz after timelock`
            );
          }
        } else if (swap.request.to === "ARK" && isChainSignableStatus(swap.status)) {
          logger.log(
            `Auto-signing server's cooperative claim for ARK chain swap ${swap.id}`
          );
          try {
            const signed = await this.executeSignServerClaimAction(swap);
            if (signed) {
              this.actionExecutedListeners.forEach(
                (listener) => listener(swap, "signServerClaim")
              );
            }
          } catch (error) {
            logger.error(
              `Non-fatal: failed to sign server claim for swap ${swap.id}:`,
              error
            );
          }
        }
      }
    } catch (error) {
      logger.error(
        `Failed to execute autonomous action for swap ${swap.id}:`,
        error
      );
      this.swapFailedListeners.forEach(
        (listener) => listener(swap, error)
      );
    } finally {
      this.swapsInProgress.delete(swap.id);
    }
  }
  /**
   * Execute claim action for reverse swap
   */
  async executeClaimAction(swap) {
    if (!this.claimCallback) {
      logger.error("Claim callback not set");
      return;
    }
    await this.claimCallback(swap);
  }
  /**
   * Execute refund action for submarine swap
   */
  async executeRefundAction(swap) {
    if (!this.refundCallback) {
      logger.error("Refund callback not set");
      return;
    }
    await this.refundCallback(swap);
  }
  /**
   * Execute claim action for chain swap Btc to Ark
   */
  async executeClaimArkAction(swap) {
    if (!this.claimArkCallback) {
      logger.error("claimArk callback not set");
      return;
    }
    await this.claimArkCallback(swap);
  }
  /**
   * Execute claim action for chain swap Ark to Btc
   */
  async executeClaimBtcAction(swap) {
    if (!this.claimBtcCallback) {
      logger.error("claimBtc callback not set");
      return;
    }
    await this.claimBtcCallback(swap);
  }
  /**
   * Execute refund action for chain swap Ark to Btc
   */
  async executeRefundArkAction(swap) {
    if (!this.refundArkCallback) {
      logger.error("refundArk callback not set");
      return;
    }
    await this.refundArkCallback(swap);
  }
  /**
   * Execute sign server claim action for chain swap.
   * Returns true on success, false if no callback is set.
   * Throws if the callback itself throws.
   */
  async executeSignServerClaimAction(swap) {
    if (!this.signServerClaimCallback) {
      logger.error("signServerClaim callback not set");
      return false;
    }
    await this.signServerClaimCallback(swap);
    return true;
  }
  /**
   * Save swap to storage
   */
  async saveSwap(swap) {
    if (!this.saveSwapCallback) {
      logger.error("Save swap callback not set");
      return;
    }
    await this.saveSwapCallback(swap);
  }
  /**
   * Resume actionable swaps on startup
   * This checks all pending swaps and executes actions if needed
   */
  async resumeActionableSwaps() {
    if (!this.config.enableAutoActions) {
      return;
    }
    for (const swap of this.monitoredSwaps.values()) {
      try {
        if (isPendingReverseSwap(swap) && isReverseClaimableStatus(swap.status)) {
          logger.log(`Resuming claim for swap ${swap.id}`);
          await this.executeAutonomousAction(swap);
        } else if (isPendingSubmarineSwap(swap) && isSubmarineRefundableStatus(swap.status)) {
          logger.log(`Resuming refund for swap ${swap.id}`);
          await this.executeAutonomousAction(swap);
        } else if (isPendingChainSwap(swap) && isChainClaimableStatus(swap.status)) {
          logger.log(`Resuming chain claim for swap ${swap.id}`);
          await this.executeAutonomousAction(swap);
        } else if (isPendingChainSwap(swap) && isChainRefundableStatus(swap.status)) {
          logger.log(`Resuming chain refund for swap ${swap.id}`);
          await this.executeAutonomousAction(swap);
        } else if (isPendingChainSwap(swap) && swap.request.to === "ARK" && isChainSignableStatus(swap.status)) {
          logger.log(
            `Resuming server claim signing for swap ${swap.id}`
          );
          await this.executeAutonomousAction(swap);
        }
      } catch (error) {
        logger.error(`Failed to resume swap ${swap.id}:`, error);
      }
    }
  }
  /**
   * Start regular polling
   * Polls all swaps at configured interval when WebSocket is active
   */
  startPolling() {
    if (this.pollTimer) {
      clearTimeout(this.pollTimer);
    }
    this.pollTimer = setTimeout(async () => {
      await this.pollAllSwaps();
      if (this.isRunning) {
        this.startPolling();
      }
    }, this.config.pollInterval);
  }
  /**
   * Start polling fallback when WebSocket is unavailable
   * Uses exponential backoff for retry delay
   */
  startPollingFallback() {
    if (this.pollTimer) {
      clearTimeout(this.pollTimer);
    }
    this.pollTimer = setTimeout(async () => {
      await this.pollAllSwaps();
      this.currentPollRetryDelay = Math.min(
        this.currentPollRetryDelay * 2,
        this.config.maxPollRetryDelayMs
      );
      if (this.isRunning && this.usePollingFallback) {
        this.startPollingFallback();
      }
    }, this.currentPollRetryDelay);
  }
  /**
   * Poll all monitored swaps for status updates
   * This is called:
   * 1. After WebSocket connects
   * 2. After WebSocket reconnects
   * 3. Periodically while WebSocket is active
   * 4. As fallback when WebSocket is unavailable
   */
  async pollAllSwaps() {
    if (this.monitoredSwaps.size === 0) return;
    const pollPromises = Array.from(this.monitoredSwaps.values()).map(
      async (swap) => {
        try {
          const statusResponse = await this.swapProvider.getSwapStatus(swap.id);
          if (statusResponse.status !== swap.status) {
            await this.handleSwapStatusUpdate(
              swap,
              statusResponse.status
            );
          }
        } catch (error) {
          logger.error(`Failed to poll swap ${swap.id}:`, error);
        }
      }
    );
    await Promise.allSettled(pollPromises);
  }
  /**
   * Check if a status is final (no more updates expected)
   */
  isFinalStatus(pendingSwap) {
    const status = pendingSwap.status;
    return isPendingReverseSwap(pendingSwap) && isReverseFinalStatus(status) || isPendingSubmarineSwap(pendingSwap) && isSubmarineFinalStatus(status) || isPendingChainSwap(pendingSwap) && isChainFinalStatus(status);
  }
  /**
   * Get current manager statistics (for debugging/monitoring)
   */
  async getStats() {
    return {
      isRunning: this.isRunning,
      monitoredSwaps: this.monitoredSwaps.size,
      websocketConnected: this.websocket !== null && this.websocket.readyState === WebSocket.OPEN,
      usePollingFallback: this.usePollingFallback,
      currentReconnectDelay: this.currentReconnectDelay,
      currentPollRetryDelay: this.currentPollRetryDelay
    };
  }
};

// src/utils/swap-helpers.ts
import { sha256 } from "@noble/hashes/sha2.js";
import { hex as hex2 } from "@scure/base";
async function saveSwap(swap, saver) {
  if (isPendingReverseSwap(swap)) {
    if (saver.saveReverseSwap) {
      await saver.saveReverseSwap(swap);
    } else {
      console.warn("No saveReverseSwap handler provided, swap not saved");
    }
  } else if (isPendingSubmarineSwap(swap)) {
    if (saver.saveSubmarineSwap) {
      await saver.saveSubmarineSwap(swap);
    } else {
      console.warn(
        "No saveSubmarineSwap handler provided, swap not saved"
      );
    }
  } else if (isPendingChainSwap(swap)) {
    if (saver.saveChainSwap) {
      await saver.saveChainSwap(swap);
    } else {
      console.warn("No saveChainSwap handler provided, swap not saved");
    }
  }
}
async function updateReverseSwapStatus(swap, status, saveFunc, additionalFields) {
  await saveFunc({
    ...swap,
    status,
    ...additionalFields
  });
}
async function updateSubmarineSwapStatus(swap, status, saveFunc, additionalFields) {
  await saveFunc({
    ...swap,
    status,
    ...additionalFields
  });
}
async function updateChainSwapStatus(swap, status, saveFunc, additionalFields) {
  await saveFunc({
    ...swap,
    status,
    ...additionalFields
  });
}
function enrichReverseSwapPreimage(swap, preimage) {
  const computedHash = hex2.encode(sha256(hex2.decode(preimage)));
  if (computedHash !== swap.request.preimageHash) {
    throw new Error(
      `Preimage does not match swap: expected hash ${swap.request.preimageHash}, got ${computedHash}`
    );
  }
  swap.preimage = preimage;
  return swap;
}
function enrichSubmarineSwapInvoice(swap, invoice) {
  let paymentHash;
  try {
    const decoded = decodeInvoice(invoice);
    if (!decoded.paymentHash) {
      throw new Error("Invoice missing payment hash");
    }
    paymentHash = decoded.paymentHash;
  } catch (error) {
    if (error instanceof Error) {
      throw new Error(`Invalid Lightning invoice: ${error.message}`);
    }
    throw new Error(`Invalid Lightning invoice format`);
  }
  if (swap.preimageHash && paymentHash !== swap.preimageHash) {
    throw new Error(
      `Invoice payment hash does not match swap: expected ${swap.preimageHash}, got ${paymentHash}`
    );
  }
  swap.request.invoice = invoice;
  return swap;
}

// src/repositories/IndexedDb/swap-repository.ts
import { closeDatabase, openDatabase } from "@arkade-os/sdk";
var DEFAULT_DB_NAME = "arkade-boltz-swap";
var DB_VERSION = 2;
var STORE_SWAPS_STATE = "swaps";
function initDatabase(db) {
  if (!db.objectStoreNames.contains(STORE_SWAPS_STATE)) {
    const swapStore = db.createObjectStore(STORE_SWAPS_STATE, {
      keyPath: "id"
    });
    swapStore.createIndex("status", "status", { unique: false });
    swapStore.createIndex("type", "type", { unique: false });
    swapStore.createIndex("createdAt", "createdAt", { unique: false });
  }
}
var IndexedDbSwapRepository = class {
  constructor(dbName = DEFAULT_DB_NAME) {
    this.dbName = dbName;
  }
  version = 1;
  db = null;
  async getDB() {
    if (this.db) return this.db;
    this.db = await openDatabase(this.dbName, DB_VERSION, initDatabase);
    return this.db;
  }
  async saveSwap(swap) {
    const db = await this.getDB();
    return new Promise((resolve, reject) => {
      const transaction = db.transaction(
        [STORE_SWAPS_STATE],
        "readwrite"
      );
      const store = transaction.objectStore(STORE_SWAPS_STATE);
      const request = store.put(swap);
      request.onsuccess = () => resolve();
      request.onerror = () => reject(request.error);
    });
  }
  async deleteSwap(id) {
    const db = await this.getDB();
    return new Promise((resolve, reject) => {
      const transaction = db.transaction(
        [STORE_SWAPS_STATE],
        "readwrite"
      );
      const store = transaction.objectStore(STORE_SWAPS_STATE);
      const request = store.delete(id);
      request.onsuccess = () => resolve();
      request.onerror = () => reject(request.error);
    });
  }
  async getAllSwaps(filter) {
    return this.getAllSwapsFromStore(filter);
  }
  async clear() {
    const db = await this.getDB();
    return new Promise((resolve, reject) => {
      const transaction = db.transaction(
        [STORE_SWAPS_STATE],
        "readwrite"
      );
      const store = transaction.objectStore(STORE_SWAPS_STATE);
      const request = store.clear();
      request.onsuccess = () => resolve();
      request.onerror = () => reject(request.error);
    });
  }
  getSwapsByIndexValues(store, indexName, values) {
    if (values.length === 0) return Promise.resolve([]);
    const index = store.index(indexName);
    const requests = values.map(
      (value) => new Promise((resolve, reject) => {
        const request = index.getAll(value);
        request.onerror = () => reject(request.error);
        request.onsuccess = () => resolve(request.result ?? []);
      })
    );
    return Promise.all(requests).then(
      (results) => results.flatMap((result) => result)
    );
  }
  async getAllSwapsFromStore(filter) {
    const db = await this.getDB();
    const store = db.transaction([STORE_SWAPS_STATE], "readonly").objectStore(STORE_SWAPS_STATE);
    if (!filter || Object.keys(filter).length === 0) {
      return new Promise((resolve, reject) => {
        const request = store.getAll();
        request.onsuccess = () => resolve(request.result ?? []);
        request.onerror = () => reject(request.error);
      });
    }
    const normalizedFilter = normalizeFilter(filter);
    if (normalizedFilter.has("id")) {
      const ids = normalizedFilter.get("id");
      const swaps = await Promise.all(
        ids.map(
          (id) => new Promise((resolve, reject) => {
            const request = store.get(id);
            request.onsuccess = () => resolve(request.result);
            request.onerror = () => reject(request.error);
          })
        )
      );
      return this.sortIfNeeded(
        this.applySwapsFilter(swaps, normalizedFilter),
        filter
      );
    }
    if (normalizedFilter.has("type")) {
      const types = normalizedFilter.get("type");
      const swaps = await this.getSwapsByIndexValues(
        store,
        "type",
        types
      );
      return this.sortIfNeeded(
        this.applySwapsFilter(swaps, normalizedFilter),
        filter
      );
    }
    if (normalizedFilter.has("status")) {
      const ids = normalizedFilter.get("status");
      const swaps = await this.getSwapsByIndexValues(
        store,
        "status",
        ids
      );
      return this.sortIfNeeded(
        this.applySwapsFilter(swaps, normalizedFilter),
        filter
      );
    }
    if (filter.orderBy === "createdAt") {
      return this.getAllSwapsByCreatedAt(store, filter.orderDirection);
    }
    const allSwaps = await new Promise((resolve, reject) => {
      const request = store.getAll();
      request.onsuccess = () => resolve(request.result ?? []);
      request.onerror = () => reject(request.error);
    });
    return this.sortIfNeeded(
      this.applySwapsFilter(allSwaps, normalizedFilter),
      filter
    );
  }
  applySwapsFilter(swaps, filter) {
    return swaps.filter((swap) => {
      if (swap === void 0) return false;
      if (filter.has("id") && !filter.get("id")?.includes(swap.id))
        return false;
      if (filter.has("status") && !filter.get("status")?.includes(swap.status))
        return false;
      if (filter.has("type") && !filter.get("type")?.includes(swap.type))
        return false;
      return true;
    });
  }
  async getAllSwapsByCreatedAt(store, orderDirection) {
    const index = store.index("createdAt");
    const direction = orderDirection === "desc" ? "prev" : "next";
    return new Promise((resolve, reject) => {
      const results = [];
      const request = index.openCursor(null, direction);
      request.onerror = () => reject(request.error);
      request.onsuccess = () => {
        const cursor = request.result;
        if (!cursor) {
          resolve(results);
          return;
        }
        results.push(cursor.value);
        cursor.continue();
      };
    });
  }
  sortIfNeeded(swaps, filter) {
    if (filter?.orderBy !== "createdAt") return swaps;
    const direction = filter.orderDirection === "asc" ? 1 : -1;
    return swaps.slice().sort((a, b) => (a.createdAt - b.createdAt) * direction);
  }
  async [Symbol.asyncDispose]() {
    if (!this.db) return;
    await closeDatabase(this.dbName);
    this.db = null;
  }
};
var FILTER_FIELDS = ["id", "status", "type"];
function normalizeFilter(filter) {
  const res = /* @__PURE__ */ new Map();
  FILTER_FIELDS.forEach((current) => {
    if (!filter?.[current]) return;
    if (Array.isArray(filter[current])) {
      res.set(current, filter[current]);
    } else {
      res.set(current, [filter[current]]);
    }
  });
  return res;
}

// src/arkade-swaps.ts
import {
  ArkAddress as ArkAddress2,
  isRecoverable
} from "@arkade-os/sdk";
import { sha256 as sha2563 } from "@noble/hashes/sha2.js";
import { hex as hex8 } from "@scure/base";
import { secp256k1 as secp256k12 } from "@noble/curves/secp256k1.js";
import { randomBytes } from "@noble/hashes/utils.js";
import { Address as Address2, OutScript as OutScript2, SigHash as SigHash2, Transaction as Transaction6 } from "@scure/btc-signer";
import { NETWORK } from "@scure/btc-signer/utils.js";

// src/utils/musig.ts
import { secp256k1 } from "@noble/curves/secp256k1.js";
import { hex as hex3 } from "@scure/base";
import {
  Session,
  keyAggregate,
  keyAggExport,
  nonceAggregate,
  nonceGen
} from "@scure/btc-signer/musig2.js";
import { equalBytes } from "@scure/btc-signer/utils.js";
var findKeyIndex = (keys, target) => keys.findIndex((k) => equalBytes(target, k));
var assertPublicKeys = (keys) => {
  const seen = /* @__PURE__ */ new Set();
  for (const key of keys) {
    if (key.length !== 33)
      throw new Error(`public key must be 33 bytes, got ${key.length}`);
    const enc = hex3.encode(key);
    if (seen.has(enc)) throw new Error(`duplicate public key ${enc}`);
    seen.add(enc);
  }
};
var aggregateKeys = (publicKeys, tweak) => {
  assertPublicKeys([...publicKeys]);
  return keyAggExport(
    keyAggregate([...publicKeys], tweak ? [tweak] : [], tweak ? [true] : [])
  );
};
var MusigKeyAgg = class _MusigKeyAgg {
  constructor(privateKey, myPublicKey, publicKeys, myIndex, aggPubkey, internalKey, _tweak) {
    this.privateKey = privateKey;
    this.myPublicKey = myPublicKey;
    this.publicKeys = publicKeys;
    this.myIndex = myIndex;
    this.aggPubkey = aggPubkey;
    this.internalKey = internalKey;
    this._tweak = _tweak;
  }
  xonlyTweakAdd(tweak) {
    if (this._tweak) throw new Error("musig key already tweaked");
    return new _MusigKeyAgg(
      this.privateKey,
      this.myPublicKey,
      this.publicKeys,
      this.myIndex,
      aggregateKeys(this.publicKeys, tweak),
      this.internalKey,
      tweak
    );
  }
  message(msg) {
    return new MusigWithMessage(
      this.privateKey,
      this.myPublicKey,
      this.publicKeys,
      this.myIndex,
      this.aggPubkey,
      this._tweak,
      msg
    );
  }
};
var MusigWithMessage = class {
  constructor(privateKey, myPublicKey, publicKeys, myIndex, aggPubkey, tweak, msg) {
    this.privateKey = privateKey;
    this.myPublicKey = myPublicKey;
    this.publicKeys = publicKeys;
    this.myIndex = myIndex;
    this.aggPubkey = aggPubkey;
    this.tweak = tweak;
    this.msg = msg;
  }
  generateNonce() {
    const nonce = nonceGen(
      this.myPublicKey,
      this.privateKey,
      this.aggPubkey,
      this.msg
    );
    return new MusigWithNonce(
      this.privateKey,
      this.myPublicKey,
      this.publicKeys,
      this.myIndex,
      this.aggPubkey,
      this.tweak,
      this.msg,
      nonce
    );
  }
};
var MusigWithNonce = class {
  constructor(privateKey, myPublicKey, publicKeys, myIndex, aggPubkey, tweak, msg, nonce) {
    this.privateKey = privateKey;
    this.myPublicKey = myPublicKey;
    this.publicKeys = publicKeys;
    this.myIndex = myIndex;
    this.aggPubkey = aggPubkey;
    this.tweak = tweak;
    this.msg = msg;
    this.nonce = nonce;
  }
  get publicNonce() {
    return this.nonce.public;
  }
  aggregateNonces(nonces) {
    const pairs = Array.from(nonces);
    const ours = pairs.find(([k]) => equalBytes(this.myPublicKey, k));
    if (!ours) {
      pairs.push([this.myPublicKey, this.publicNonce]);
    } else if (!equalBytes(ours[1], this.publicNonce)) {
      throw new Error("nonce for our public key does not match");
    }
    if (this.publicKeys.length !== pairs.length) {
      throw new Error("number of nonces != number of public keys");
    }
    const nonceByKey = /* @__PURE__ */ new Map();
    for (const [key, nonce] of pairs) {
      nonceByKey.set(hex3.encode(key), nonce);
    }
    const ordered = [];
    for (const key of this.publicKeys) {
      const n = nonceByKey.get(hex3.encode(key));
      if (!n) throw new Error("missing nonce for public key");
      ordered.push(n);
    }
    const aggregatedNonce = nonceAggregate([...ordered]);
    return new MusigNoncesAggregated(
      this.privateKey,
      this.myPublicKey,
      this.publicKeys,
      this.myIndex,
      this.aggPubkey,
      this.tweak,
      this.msg,
      this.nonce,
      Object.freeze(ordered),
      aggregatedNonce
    );
  }
};
var MusigNoncesAggregated = class {
  constructor(privateKey, myPublicKey, publicKeys, myIndex, aggPubkey, tweak, msg, nonce, pubNonces, aggregatedNonce) {
    this.privateKey = privateKey;
    this.myPublicKey = myPublicKey;
    this.publicKeys = publicKeys;
    this.myIndex = myIndex;
    this.aggPubkey = aggPubkey;
    this.tweak = tweak;
    this.msg = msg;
    this.nonce = nonce;
    this.pubNonces = pubNonces;
    this.aggregatedNonce = aggregatedNonce;
  }
  get publicNonce() {
    return this.nonce.public;
  }
  initializeSession() {
    const session = new Session(
      this.aggregatedNonce,
      [...this.publicKeys],
      this.msg,
      this.tweak ? [this.tweak] : [],
      this.tweak ? [true] : []
    );
    return new MusigSession(
      this.privateKey,
      this.publicKeys,
      this.myIndex,
      this.nonce,
      this.pubNonces,
      session
    );
  }
};
var MusigSession = class {
  constructor(privateKey, publicKeys, myIndex, nonce, pubNonces, session) {
    this.privateKey = privateKey;
    this.publicKeys = publicKeys;
    this.myIndex = myIndex;
    this.nonce = nonce;
    this.pubNonces = pubNonces;
    this.session = session;
    this.partialSignatures = Array(publicKeys.length).fill(null);
  }
  partialSignatures;
  get publicNonce() {
    return this.nonce.public;
  }
  addPartial(publicKeyOrIndex, signature) {
    const index = typeof publicKeyOrIndex === "number" ? publicKeyOrIndex : findKeyIndex(this.publicKeys, publicKeyOrIndex);
    if (index < 0 || index >= this.publicKeys.length)
      throw new Error("public key not found or index out of range");
    if (!this.session.partialSigVerify(
      signature,
      [...this.pubNonces],
      index
    )) {
      throw new Error("invalid partial signature");
    }
    this.partialSignatures[index] = signature;
    return this;
  }
  signPartial() {
    const sig = this.session.sign(this.nonce.secret, this.privateKey, true);
    this.partialSignatures[this.myIndex] = sig;
    return new MusigSigned(
      this.session,
      [...this.partialSignatures],
      sig,
      this.nonce.public
    );
  }
};
var MusigSigned = class {
  constructor(session, partialSignatures, ourPartialSignature, publicNonce) {
    this.session = session;
    this.partialSignatures = partialSignatures;
    this.ourPartialSignature = ourPartialSignature;
    this.publicNonce = publicNonce;
  }
  aggregatePartials() {
    if (this.partialSignatures.some((s) => s === null)) {
      throw new Error("not all partial signatures are set");
    }
    return this.session.partialSigAgg(
      this.partialSignatures
    );
  }
};
var create = (privateKey, publicKeys) => {
  if (publicKeys.length < 2)
    throw new Error("need at least 2 keys to aggregate");
  const keys = [...publicKeys];
  assertPublicKeys(keys);
  Object.freeze(keys);
  const myPublicKey = secp256k1.getPublicKey(privateKey);
  const myIndex = findKeyIndex(keys, myPublicKey);
  if (myIndex === -1) throw new Error("our key is not in publicKeys");
  const aggPubkey = aggregateKeys(keys);
  return new MusigKeyAgg(
    privateKey,
    myPublicKey,
    keys,
    myIndex,
    aggPubkey,
    aggPubkey
  );
};

// src/utils/boltz-swap-tx.ts
import { schnorr } from "@noble/curves/secp256k1.js";
import { hex as hex4 } from "@scure/base";
import { Script, Transaction as Transaction2 } from "@scure/btc-signer";
import { tapLeafHash } from "@scure/btc-signer/payment.js";
import { compareBytes, equalBytes as equalBytes2 } from "@scure/btc-signer/utils.js";
var REGTEST_NETWORK = {
  bech32: "bcrt",
  pubKeyHash: 111,
  scriptHash: 196,
  wif: 239
};
var MUTINYNET_NETWORK = {
  bech32: "tb",
  pubKeyHash: 111,
  scriptHash: 196,
  wif: 239
};
var deserializeLeaf = (leaf) => ({
  version: leaf.version,
  output: hex4.decode(leaf.output)
});
var sortTree = (nodes) => {
  const sorted = [...nodes].sort((a, b) => b.probability - a.probability);
  const sub = (items) => {
    if (items.length === 1) return items[0].value;
    if (items.length === 2) return [items[0].value, items[1].value];
    const sum = items.reduce((s, n) => s + n.probability, 0);
    let mid = 0;
    let midSum = 0;
    while (midSum < sum / 2) {
      midSum += items[mid].probability;
      mid++;
    }
    return [sub(items.slice(0, mid)), sub(items.slice(mid))];
  };
  return sub(sorted);
};
var deserializeSwapTree = (tree) => {
  const parsed = typeof tree === "string" ? JSON.parse(tree) : tree;
  const claimLeaf = deserializeLeaf(parsed.claimLeaf);
  const refundLeaf = deserializeLeaf(parsed.refundLeaf);
  return {
    claimLeaf,
    refundLeaf,
    tree: sortTree([
      { probability: 51, value: claimLeaf },
      { probability: 49, value: refundLeaf }
    ])
  };
};
var taprootHashTree = (tree) => {
  if (!Array.isArray(tree)) {
    return {
      type: "leaf",
      version: tree.version,
      script: tree.output,
      hash: tapLeafHash(tree.output, tree.version)
    };
  }
  const left = taprootHashTree(tree[0]);
  const right = taprootHashTree(tree[1]);
  let [lH, rH] = [left.hash, right.hash];
  if (compareBytes(rH, lH) === -1) [lH, rH] = [rH, lH];
  return {
    type: "branch",
    left,
    right,
    hash: schnorr.utils.taggedHash("TapBranch", lH, rH)
  };
};
var tweakMusig = (musig, tree) => {
  const tweak = taprootHashTree(tree).hash;
  return musig.xonlyTweakAdd(
    schnorr.utils.taggedHash("TapTweak", musig.aggPubkey, tweak)
  );
};
var toXOnly = (pubKey) => {
  if (pubKey.length === 32) return pubKey;
  if (pubKey.length === 33) {
    if (pubKey[0] !== 2 && pubKey[0] !== 3) {
      throw new Error(
        `Invalid compressed public key prefix: 0x${pubKey[0].toString(16).padStart(2, "0")}`
      );
    }
    return pubKey.subarray(1, 33);
  }
  throw new Error(
    `Invalid public key length: expected 32 or 33 bytes, got ${pubKey.length}`
  );
};
var p2trScript = (publicKey) => Script.encode(["OP_1", toXOnly(publicKey)]);
var detectSwapOutput = (tweakedKey, transaction) => {
  const target = p2trScript(tweakedKey);
  for (let vout = 0; vout < transaction.outputsLength; vout++) {
    const output = transaction.getOutput(vout);
    if (output.script !== void 0 && output.amount !== void 0 && equalBytes2(target, output.script)) {
      return { ...output, vout };
    }
  }
  throw new Error("Swap output not found in transaction");
};
var DUMMY_TAPROOT_SIGNATURE = new Uint8Array(64);
var constructClaimTransaction = (utxo, destinationScript, fee) => {
  if (fee < BigInt(0) || fee >= utxo.amount)
    throw new Error("fee exceeds utxo amount");
  const tx = new Transaction2({ version: 2 });
  tx.addOutput({
    amount: utxo.amount - fee,
    script: destinationScript
  });
  tx.addInput({
    txid: utxo.transactionId,
    index: utxo.vout,
    sequence: 4294967293
    // RBF enabled
  });
  tx.updateInput(0, {
    finalScriptWitness: [DUMMY_TAPROOT_SIGNATURE]
  });
  return tx;
};
var targetFee = (satPerVbyte, constructTx) => {
  const tx = constructTx(BigInt(1));
  return constructTx(
    BigInt(Math.ceil((tx.vsize + tx.inputsLength) * satPerVbyte))
  );
};

// src/utils/restoration.ts
import { hex as hex5 } from "@scure/base";
import { Script as Script2 } from "@scure/btc-signer";
import bip68 from "bip68";
function extractTimeLockFromLeafOutput(scriptHex) {
  if (!scriptHex) return 0;
  try {
    const opcodes = Script2.decode(hex5.decode(scriptHex));
    const hasCLTV = opcodes.findIndex((op) => op === "CHECKLOCKTIMEVERIFY");
    if (hasCLTV > 0) {
      const data = opcodes[hasCLTV - 1];
      if (data instanceof Uint8Array) {
        const dataBytes = new Uint8Array(data).reverse();
        return parseInt(hex5.encode(dataBytes), 16);
      }
    }
    const hasCSV = opcodes.findIndex((op) => op === "CHECKSEQUENCEVERIFY");
    if (hasCSV > 0) {
      const data = opcodes[hasCSV - 1];
      if (data instanceof Uint8Array) {
        const dataBytes = new Uint8Array(data).reverse();
        const {
          blocks,
          seconds
        } = bip68.decode(
          parseInt(hex5.encode(dataBytes), 16)
        );
        return blocks ?? seconds ?? 0;
      }
    }
  } catch (error) {
    return 0;
  }
  return 0;
}
function extractInvoiceAmount(amountSats, fees) {
  if (!amountSats) return 0;
  const { percentage, minerFees } = fees.reverse;
  const miner = minerFees.lockup + minerFees.claim;
  if (percentage >= 100 || percentage < 0) return 0;
  if (miner >= amountSats) return 0;
  return Math.ceil((amountSats - miner) / (1 - percentage / 100));
}

// src/utils/identity.ts
import {
  ConditionWitness,
  setArkPsbtField,
  Transaction as Transaction3
} from "@arkade-os/sdk";
function claimVHTLCIdentity(identity, preimage) {
  return {
    ...identity,
    sign: async (tx, inputIndexes) => {
      const cpy = tx.clone();
      let signedTx = await identity.sign(cpy, inputIndexes);
      signedTx = Transaction3.fromPSBT(signedTx.toPSBT());
      if (preimage) {
        for (const inputIndex of inputIndexes || Array.from(
          { length: signedTx.inputsLength },
          (_, i) => i
        )) {
          setArkPsbtField(signedTx, inputIndex, ConditionWitness, [
            preimage
          ]);
        }
      }
      return signedTx;
    }
  };
}

// src/utils/vhtlc.ts
import {
  Batch,
  buildOffchainTx,
  getSequence as getSequence2,
  Intent,
  networks,
  VHTLC,
  VtxoScript as VtxoScript2,
  VtxoTaprootTree,
  CSVMultisigTapscript as CSVMultisigTapscript2,
  combineTapscriptSigs
} from "@arkade-os/sdk";
import { hex as hex7, base64 as base643 } from "@scure/base";

// src/batch.ts
import {
  CSVMultisigTapscript,
  Transaction as Transaction4,
  validateVtxoTxGraph,
  validateConnectorsTxGraph,
  VtxoScript,
  buildForfeitTx,
  getSequence
} from "@arkade-os/sdk";
import { sha256 as sha2562 } from "@noble/hashes/sha2.js";
import { base64 as base642, hex as hex6 } from "@scure/base";
import { SigHash } from "@scure/btc-signer";
import { tapLeafHash as tapLeafHash2 } from "@scure/btc-signer/payment.js";
function createVHTLCBatchHandler(intentId, vhtlc, arkProvider, identity, session, sweepPublicKey, forfeitOutputScript, connectorIndex = 0) {
  const utf8IntentId = new TextEncoder().encode(intentId);
  const intentIdHash = sha2562(utf8IntentId);
  const intentIdHashStr = hex6.encode(intentIdHash);
  let sweepTapTreeRoot;
  return {
    onBatchStarted: async (event) => {
      let skip = true;
      for (const idHash of event.intentIdHashes) {
        if (idHash === intentIdHashStr) {
          if (!arkProvider) {
            throw new Error("Ark provider not configured");
          }
          await arkProvider.confirmRegistration(intentId);
          skip = false;
        }
      }
      if (skip) {
        return { skip };
      }
      const sweepTapscript = CSVMultisigTapscript.encode({
        timelock: {
          value: event.batchExpiry,
          type: event.batchExpiry >= 512n ? "seconds" : "blocks"
        },
        pubkeys: [sweepPublicKey]
      }).script;
      sweepTapTreeRoot = tapLeafHash2(sweepTapscript);
      return { skip: false };
    },
    onTreeSigningStarted: async (event, vtxoTree) => {
      if (!session) {
        return { skip: true };
      }
      if (!sweepTapTreeRoot) {
        throw new Error("Sweep tap tree root not set");
      }
      const xOnlyPublicKeys = event.cosignersPublicKeys.map(
        (k) => k.slice(2)
      );
      const signerPublicKey = await session.getPublicKey();
      const xonlySignerPublicKey = signerPublicKey.subarray(1);
      if (!xOnlyPublicKeys.includes(hex6.encode(xonlySignerPublicKey))) {
        return { skip: true };
      }
      const commitmentTx = Transaction4.fromPSBT(
        base642.decode(event.unsignedCommitmentTx)
      );
      validateVtxoTxGraph(vtxoTree, commitmentTx, sweepTapTreeRoot);
      const sharedOutput = commitmentTx.getOutput(0);
      if (!sharedOutput?.amount) {
        throw new Error("Shared output not found");
      }
      await session.init(vtxoTree, sweepTapTreeRoot, sharedOutput.amount);
      const pubkey = hex6.encode(await session.getPublicKey());
      const nonces = await session.getNonces();
      await arkProvider.submitTreeNonces(event.id, pubkey, nonces);
      return { skip: false };
    },
    onTreeNonces: async (event) => {
      if (!session) {
        return { fullySigned: true };
      }
      const { hasAllNonces } = await session.aggregatedNonces(
        event.txid,
        event.nonces
      );
      if (!hasAllNonces) return { fullySigned: false };
      const signatures = await session.sign();
      const pubkey = hex6.encode(await session.getPublicKey());
      await arkProvider.submitTreeSignatures(
        event.id,
        pubkey,
        signatures
      );
      return { fullySigned: true };
    },
    onBatchFinalization: async (event, _, connectorTree) => {
      if (!forfeitOutputScript) {
        return;
      }
      if (!connectorTree) {
        throw new Error(
          "BatchFinalizationEvent: expected connector tree to be defined"
        );
      }
      validateConnectorsTxGraph(event.commitmentTx, connectorTree);
      const connectors = connectorTree.leaves();
      if (connectors.length <= connectorIndex) {
        throw new Error(
          `BatchFinalizationEvent: expected connector tree has ${connectors.length} leaves, expected at least ${connectorIndex + 1}`
        );
      }
      const forfeitTx = createForfeitTx(
        vhtlc,
        forfeitOutputScript,
        connectors[connectorIndex]
      );
      const signedForfeitTx = await identity.sign(forfeitTx);
      await arkProvider.submitSignedForfeitTxs([
        base642.encode(signedForfeitTx.toPSBT())
      ]);
    }
  };
}
function createForfeitTx(input, forfeitOutputScript, connector) {
  const connectorTxId = connector.id;
  const connectorOutput = connector.getOutput(0);
  if (!connectorOutput) {
    throw new Error("connector output not found");
  }
  const connectorAmount = connectorOutput.amount;
  const connectorPkScript = connectorOutput.script;
  if (!connectorAmount || !connectorPkScript) {
    throw new Error("invalid connector output");
  }
  const sequence = getSequence(input.tapLeafScript);
  return buildForfeitTx(
    [
      {
        txid: input.txid,
        index: input.vout,
        witnessUtxo: {
          amount: BigInt(input.value),
          script: VtxoScript.decode(input.tapTree).pkScript
        },
        sighashType: SigHash.DEFAULT,
        tapLeafScript: [input.tapLeafScript],
        sequence
      },
      {
        txid: connectorTxId,
        index: 0,
        witnessUtxo: {
          amount: connectorAmount,
          script: connectorPkScript
        }
      }
    ],
    forfeitOutputScript,
    sequence
  );
}

// src/utils/vhtlc.ts
import { ripemd160 } from "@noble/hashes/legacy.js";
import { Address, OutScript, Transaction as Transaction5 } from "@scure/btc-signer";
var createVHTLCScript = (args) => {
  const {
    network,
    preimageHash,
    receiverPubkey,
    senderPubkey,
    serverPubkey,
    timeoutBlockHeights
  } = args;
  const receiverXOnlyPublicKey = normalizeToXOnlyKey(
    hex7.decode(receiverPubkey),
    "receiver"
  );
  const senderXOnlyPublicKey = normalizeToXOnlyKey(
    hex7.decode(senderPubkey),
    "sender"
  );
  const serverXOnlyPublicKey = normalizeToXOnlyKey(
    hex7.decode(serverPubkey),
    "server"
  );
  const delayType = (num) => num < 512 ? "blocks" : "seconds";
  const vhtlcScript = new VHTLC.Script({
    preimageHash: ripemd160(preimageHash),
    sender: senderXOnlyPublicKey,
    receiver: receiverXOnlyPublicKey,
    server: serverXOnlyPublicKey,
    refundLocktime: BigInt(timeoutBlockHeights.refund),
    unilateralClaimDelay: {
      type: delayType(timeoutBlockHeights.unilateralClaim),
      value: BigInt(timeoutBlockHeights.unilateralClaim)
    },
    unilateralRefundDelay: {
      type: delayType(timeoutBlockHeights.unilateralRefund),
      value: BigInt(timeoutBlockHeights.unilateralRefund)
    },
    unilateralRefundWithoutReceiverDelay: {
      type: delayType(
        timeoutBlockHeights.unilateralRefundWithoutReceiver
      ),
      value: BigInt(timeoutBlockHeights.unilateralRefundWithoutReceiver)
    }
  });
  if (!vhtlcScript.claimScript)
    throw new Error("Failed to create VHTLC script");
  const hrp = network === "bitcoin" ? "ark" : "tark";
  const vhtlcAddress = vhtlcScript.address(hrp, serverXOnlyPublicKey).encode();
  return { vhtlcScript, vhtlcAddress };
};
var joinBatch = async (arkProvider, identity, input, output, {
  forfeitPubkey,
  forfeitAddress,
  network
}, isRecoverable2 = true) => {
  const signerSession = identity.signerSession();
  const signerPublicKey = await signerSession.getPublicKey();
  const intentMessage = {
    type: "register",
    onchain_output_indexes: [],
    valid_at: 0,
    expire_at: 0,
    cosigners_public_keys: [hex7.encode(signerPublicKey)]
  };
  const deleteMessage = {
    type: "delete",
    expire_at: 0
  };
  const intentInput = {
    txid: hex7.decode(input.txid),
    index: input.vout,
    witnessUtxo: {
      amount: BigInt(input.value),
      script: VtxoScript2.decode(input.tapTree).pkScript
    },
    tapLeafScript: [input.tapLeafScript],
    unknown: [VtxoTaprootTree.encode(input.tapTree)],
    sequence: getSequence2(input.tapLeafScript)
  };
  const registerIntent = Intent.create(
    intentMessage,
    [intentInput],
    [output]
  );
  const deleteIntent = Intent.create(deleteMessage, [intentInput]);
  const [signedRegisterIntent, signedDeleteIntent] = await Promise.all([
    identity.sign(registerIntent),
    identity.sign(deleteIntent)
  ]);
  const abortController = new AbortController();
  const intentId = await arkProvider.registerIntent({
    message: intentMessage,
    proof: base643.encode(signedRegisterIntent.toPSBT())
  });
  const decodedAddress = Address(
    network in networks ? networks[network] : networks.bitcoin
  ).decode(forfeitAddress);
  try {
    const handler = createVHTLCBatchHandler(
      intentId,
      input,
      arkProvider,
      identity,
      signerSession,
      normalizeToXOnlyKey(forfeitPubkey, "forfeit"),
      isRecoverable2 ? void 0 : OutScript.encode(decodedAddress)
    );
    const topics = [
      hex7.encode(signerPublicKey),
      `${input.txid}:${input.vout}`
    ];
    const eventStream = arkProvider.getEventStream(
      abortController.signal,
      topics
    );
    const commitmentTxid = await Batch.join(eventStream, handler, {
      abortController
    });
    return commitmentTxid;
  } catch (error) {
    abortController.abort();
    logger.error("Failed to join batch:", error);
    try {
      await arkProvider.deleteIntent({
        message: deleteMessage,
        proof: base643.encode(signedDeleteIntent.toPSBT())
      });
    } catch (error2) {
      logger.error("Failed to delete intent:", error2);
    }
    throw error;
  }
};
var claimVHTLCwithOffchainTx = async (identity, vhtlcScript, serverXOnlyPublicKey, input, output, arkInfo, arkProvider) => {
  const rawCheckpointTapscript = hex7.decode(arkInfo.checkpointTapscript);
  const serverUnrollScript = CSVMultisigTapscript2.decode(
    rawCheckpointTapscript
  );
  const { arkTx, checkpoints } = buildOffchainTx(
    [input],
    [output],
    serverUnrollScript
  );
  const signedArkTx = await identity.sign(arkTx);
  const { arkTxid, finalArkTx, signedCheckpointTxs } = await arkProvider.submitTx(
    base643.encode(signedArkTx.toPSBT()),
    checkpoints.map((c) => base643.encode(c.toPSBT()))
  );
  if (!validFinalArkTx(finalArkTx, serverXOnlyPublicKey, vhtlcScript.leaves)) {
    throw new Error("Invalid final Ark transaction");
  }
  const finalCheckpoints = await Promise.all(
    signedCheckpointTxs.map(async (c) => {
      const tx = Transaction5.fromPSBT(base643.decode(c), {
        allowUnknown: true
      });
      const signedCheckpoint = await identity.sign(tx, [0]);
      return base643.encode(signedCheckpoint.toPSBT());
    })
  );
  await arkProvider.finalizeTx(arkTxid, finalCheckpoints);
};
var refundVHTLCwithOffchainTx = async (swapId, identity, arkProvider, boltzXOnlyPublicKey, ourXOnlyPublicKey, serverXOnlyPublicKey, input, output, arkInfo, refundFunc) => {
  const rawCheckpointTapscript = hex7.decode(arkInfo.checkpointTapscript);
  const serverUnrollScript = CSVMultisigTapscript2.decode(
    rawCheckpointTapscript
  );
  const { arkTx: unsignedRefundTx, checkpoints: checkpointPtxs } = buildOffchainTx([input], [output], serverUnrollScript);
  if (checkpointPtxs.length !== 1)
    throw new Error(
      `Expected one checkpoint transaction, got ${checkpointPtxs.length}`
    );
  const unsignedCheckpointTx = checkpointPtxs[0];
  const {
    transaction: boltzSignedRefundTx,
    checkpoint: boltzSignedCheckpointTx
  } = await refundFunc(swapId, unsignedRefundTx, unsignedCheckpointTx);
  const boltzXOnlyPublicKeyHex = hex7.encode(boltzXOnlyPublicKey);
  if (!verifySignatures(boltzSignedRefundTx, 0, [boltzXOnlyPublicKeyHex])) {
    throw new Error("Invalid Boltz signature in refund transaction");
  }
  if (!verifySignatures(boltzSignedCheckpointTx, 0, [boltzXOnlyPublicKeyHex])) {
    throw new Error("Invalid Boltz signature in checkpoint transaction");
  }
  const signedRefundTx = await identity.sign(unsignedRefundTx);
  const signedCheckpointTx = await identity.sign(unsignedCheckpointTx);
  const combinedSignedRefundTx = combineTapscriptSigs(
    boltzSignedRefundTx,
    signedRefundTx
  );
  const combinedSignedCheckpointTx = combineTapscriptSigs(
    boltzSignedCheckpointTx,
    signedCheckpointTx
  );
  const { arkTxid, finalArkTx, signedCheckpointTxs } = await arkProvider.submitTx(
    base643.encode(combinedSignedRefundTx.toPSBT()),
    [base643.encode(unsignedCheckpointTx.toPSBT())]
  );
  const tx = Transaction5.fromPSBT(base643.decode(finalArkTx));
  const inputIndex = 0;
  const requiredSigners = [
    hex7.encode(ourXOnlyPublicKey),
    hex7.encode(boltzXOnlyPublicKey),
    hex7.encode(serverXOnlyPublicKey)
  ];
  if (!verifySignatures(tx, inputIndex, requiredSigners)) {
    throw new Error("Invalid refund transaction");
  }
  if (signedCheckpointTxs.length !== 1) {
    throw new Error(
      `Expected one signed checkpoint transaction, got ${signedCheckpointTxs.length}`
    );
  }
  const serverSignedCheckpointTx = Transaction5.fromPSBT(
    base643.decode(signedCheckpointTxs[0])
  );
  const finalCheckpointTx = combineTapscriptSigs(
    combinedSignedCheckpointTx,
    serverSignedCheckpointTx
  );
  await arkProvider.finalizeTx(arkTxid, [
    base643.encode(finalCheckpointTx.toPSBT())
  ]);
};
var validFinalArkTx = (finalArkTx, _pubkey, _tapLeaves) => {
  const tx = Transaction5.fromPSBT(base643.decode(finalArkTx), {
    allowUnknown: true
  });
  if (!tx) return false;
  const inputs = [];
  for (let i = 0; i < tx.inputsLength; i++) {
    inputs.push(tx.getInput(i));
  }
  return inputs.every((input) => input.witnessUtxo);
};

// src/arkade-swaps.ts
var ArkadeSwaps = class {
  wallet;
  arkProvider;
  swapProvider;
  indexerProvider;
  swapManager = null;
  swapRepository;
  constructor(config) {
    if (!config.wallet) throw new Error("Wallet is required.");
    if (!config.swapProvider) throw new Error("Swap provider is required.");
    this.wallet = config.wallet;
    const arkProvider = config.wallet.arkProvider ?? config.arkProvider;
    if (!arkProvider)
      throw new Error(
        "Ark provider is required either in wallet or config."
      );
    this.arkProvider = arkProvider;
    const indexerProvider = config.wallet.indexerProvider ?? config.indexerProvider;
    if (!indexerProvider)
      throw new Error(
        "Indexer provider is required either in wallet or config."
      );
    this.indexerProvider = indexerProvider;
    this.swapProvider = config.swapProvider;
    if (config.swapRepository) {
      this.swapRepository = config.swapRepository;
    } else {
      this.swapRepository = new IndexedDbSwapRepository();
    }
    if (config.swapManager) {
      const swapManagerConfig = config.swapManager === true ? {} : config.swapManager;
      const shouldAutostart = swapManagerConfig.autoStart ?? true;
      this.swapManager = new SwapManager(
        this.swapProvider,
        swapManagerConfig
      );
      this.swapManager.setCallbacks({
        claim: async (swap) => {
          await this.claimVHTLC(swap);
        },
        refund: async (swap) => {
          await this.refundVHTLC(swap);
        },
        claimArk: async (swap) => {
          await this.claimArk(swap);
        },
        claimBtc: async (swap) => {
          await this.claimBtc(swap);
        },
        refundArk: async (swap) => {
          await this.refundArk(swap);
        },
        signServerClaim: async (swap) => {
          await this.signCooperativeClaimForServer(swap);
        },
        saveSwap: async (swap) => {
          await saveSwap(swap, {
            saveReverseSwap: this.savePendingReverseSwap.bind(this),
            saveSubmarineSwap: this.savePendingSubmarineSwap.bind(this),
            saveChainSwap: this.savePendingChainSwap.bind(this)
          });
        }
      });
      if (shouldAutostart) {
        this.startSwapManager().catch((error) => {
          logger.error("Failed to autostart SwapManager:", error);
        });
      }
    }
  }
  // =========================================================================
  // Storage helpers
  // =========================================================================
  async savePendingReverseSwap(swap) {
    await this.swapRepository.saveSwap(swap);
  }
  async savePendingSubmarineSwap(swap) {
    await this.swapRepository.saveSwap(swap);
  }
  async savePendingChainSwap(swap) {
    await this.swapRepository.saveSwap(swap);
  }
  async getPendingReverseSwapsFromStorage() {
    return this.swapRepository.getAllSwaps({
      type: "reverse"
    });
  }
  async getPendingSubmarineSwapsFromStorage() {
    return this.swapRepository.getAllSwaps({
      type: "submarine"
    });
  }
  async getPendingChainSwapsFromStorage() {
    return this.swapRepository.getAllSwaps({
      type: "chain"
    });
  }
  // =========================================================================
  // SwapManager lifecycle
  // =========================================================================
  /**
   * Start the background swap manager.
   * This will load all pending swaps and begin monitoring them.
   * Automatically called when SwapManager is enabled.
   */
  async startSwapManager() {
    if (!this.swapManager) {
      throw new Error(
        "SwapManager is not enabled. Provide 'swapManager' config in ArkadeSwapsConfig."
      );
    }
    const allSwaps = await this.swapRepository.getAllSwaps();
    await this.swapManager.start(allSwaps);
  }
  /**
   * Stop the background swap manager.
   */
  async stopSwapManager() {
    if (!this.swapManager) return;
    await this.swapManager.stop();
  }
  /**
   * Get the SwapManager instance.
   * Useful for accessing manager stats or manually controlling swaps.
   */
  getSwapManager() {
    return this.swapManager;
  }
  /**
   * Dispose of resources (stops SwapManager and cleans up).
   * Can be called manually or automatically with `await using` syntax (TypeScript 5.2+).
   */
  async dispose() {
    if (this.swapManager) {
      await this.stopSwapManager();
    }
  }
  /**
   * Symbol.asyncDispose for automatic cleanup with `await using` syntax.
   */
  async [Symbol.asyncDispose]() {
    await this.dispose();
  }
  // =========================================================================
  // Lightning: Reverse swaps (receive Lightning -> Arkade)
  // =========================================================================
  /**
   * Creates a Lightning invoice.
   * @param args - The arguments for creating a Lightning invoice.
   * @returns The response containing the created Lightning invoice.
   */
  async createLightningInvoice(args) {
    const pendingSwap = await this.createReverseSwap(args);
    const decodedInvoice = decodeInvoice(pendingSwap.response.invoice);
    return {
      amount: pendingSwap.response.onchainAmount,
      expiry: decodedInvoice.expiry,
      invoice: pendingSwap.response.invoice,
      paymentHash: decodedInvoice.paymentHash,
      pendingSwap,
      preimage: pendingSwap.preimage
    };
  }
  /**
   * Creates a reverse swap.
   * @param args - The arguments for creating a reverse swap.
   * @returns The created pending reverse swap.
   */
  async createReverseSwap(args) {
    if (args.amount <= 0)
      throw new SwapError({ message: "Amount must be greater than 0" });
    const claimPublicKey = hex8.encode(
      await this.wallet.identity.compressedPublicKey()
    );
    if (!claimPublicKey)
      throw new SwapError({
        message: "Failed to get claim public key from wallet"
      });
    const preimage = randomBytes(32);
    const preimageHash = hex8.encode(sha2563(preimage));
    if (!preimageHash)
      throw new SwapError({ message: "Failed to get preimage hash" });
    const swapRequest = {
      invoiceAmount: args.amount,
      claimPublicKey,
      preimageHash,
      ...args.description?.trim() ? { description: args.description.trim() } : {}
    };
    const swapResponse = await this.swapProvider.createReverseSwap(swapRequest);
    const pendingSwap = {
      id: swapResponse.id,
      type: "reverse",
      createdAt: Math.floor(Date.now() / 1e3),
      preimage: hex8.encode(preimage),
      request: swapRequest,
      response: swapResponse,
      status: "swap.created"
    };
    await this.savePendingReverseSwap(pendingSwap);
    if (this.swapManager) {
      this.swapManager.addSwap(pendingSwap);
    }
    return pendingSwap;
  }
  /**
   * Claims the VHTLC for a pending reverse swap.
   * @param pendingSwap - The pending reverse swap to claim the VHTLC.
   */
  async claimVHTLC(pendingSwap) {
    if (!pendingSwap.preimage)
      throw new Error("Preimage is required to claim VHTLC");
    const preimage = hex8.decode(pendingSwap.preimage);
    const arkInfo = await this.arkProvider.getInfo();
    const address = await this.wallet.getAddress();
    const receiverXOnly = normalizeToXOnlyKey(
      await this.wallet.identity.xOnlyPublicKey(),
      "our",
      pendingSwap.id
    );
    const senderXOnly = normalizeToXOnlyKey(
      hex8.decode(pendingSwap.response.refundPublicKey),
      "boltz",
      pendingSwap.id
    );
    const serverXOnly = normalizeToXOnlyKey(
      hex8.decode(arkInfo.signerPubkey),
      "server",
      pendingSwap.id
    );
    const { vhtlcScript, vhtlcAddress } = this.createVHTLCScript({
      network: arkInfo.network,
      preimageHash: sha2563(preimage),
      receiverPubkey: hex8.encode(receiverXOnly),
      senderPubkey: hex8.encode(senderXOnly),
      serverPubkey: hex8.encode(serverXOnly),
      timeoutBlockHeights: pendingSwap.response.timeoutBlockHeights
    });
    if (!vhtlcScript.claimScript)
      throw new Error("Failed to create VHTLC script for reverse swap");
    if (vhtlcAddress !== pendingSwap.response.lockupAddress)
      throw new Error("Boltz is trying to scam us");
    const { vtxos } = await this.indexerProvider.getVtxos({
      scripts: [hex8.encode(vhtlcScript.pkScript)]
    });
    if (vtxos.length === 0)
      throw new Error("No spendable virtual coins found");
    const vtxo = vtxos[0];
    if (vtxo.isSpent) {
      throw new Error("VHTLC is already spent");
    }
    const input = {
      ...vtxo,
      tapLeafScript: vhtlcScript.claim(),
      tapTree: vhtlcScript.encode()
    };
    const output = {
      amount: BigInt(vtxo.value),
      script: ArkAddress2.decode(address).pkScript
    };
    const vhtlcIdentity = claimVHTLCIdentity(
      this.wallet.identity,
      preimage
    );
    let finalStatus;
    if (isRecoverable(vtxo)) {
      await this.joinBatch(vhtlcIdentity, input, output, arkInfo);
      finalStatus = "transaction.claimed";
    } else {
      await claimVHTLCwithOffchainTx(
        vhtlcIdentity,
        vhtlcScript,
        serverXOnly,
        input,
        output,
        arkInfo,
        this.arkProvider
      );
      finalStatus = (await this.getSwapStatus(pendingSwap.id)).status;
    }
    await updateReverseSwapStatus(
      pendingSwap,
      finalStatus,
      this.savePendingReverseSwap.bind(this)
    );
  }
  /**
   * Waits for the swap to be confirmed and claims the VHTLC.
   * @param pendingSwap - The pending reverse swap.
   * @returns The transaction ID of the claimed VHTLC.
   */
  async waitAndClaim(pendingSwap) {
    if (this.swapManager && await this.swapManager.hasSwap(pendingSwap.id)) {
      return this.swapManager.waitForSwapCompletion(pendingSwap.id);
    }
    return new Promise((resolve, reject) => {
      const onStatusUpdate = async (status, data) => {
        const saveStatus = (additionalFields) => updateReverseSwapStatus(
          pendingSwap,
          status,
          this.savePendingReverseSwap.bind(this),
          additionalFields
        );
        switch (status) {
          case "transaction.mempool":
          case "transaction.confirmed":
            await saveStatus();
            this.claimVHTLC(pendingSwap).catch(reject);
            break;
          case "invoice.settled": {
            await saveStatus();
            const swapStatus = await this.swapProvider.getReverseSwapTxId(
              pendingSwap.id
            );
            const txid = swapStatus.id;
            if (!txid || txid.trim() === "") {
              reject(
                new SwapError({
                  message: `Transaction ID not available for settled swap ${pendingSwap.id}.`
                })
              );
              break;
            }
            resolve({ txid });
            break;
          }
          case "invoice.expired":
            await saveStatus();
            reject(
              new InvoiceExpiredError({
                isRefundable: true,
                pendingSwap
              })
            );
            break;
          case "swap.expired":
            await saveStatus();
            reject(
              new SwapExpiredError({
                isRefundable: true,
                pendingSwap
              })
            );
            break;
          case "transaction.failed":
            await saveStatus();
            reject(
              new TransactionFailedError({
                message: data?.failureReason ?? "Transaction failed",
                isRefundable: true
              })
            );
            break;
          case "transaction.refunded":
            await saveStatus();
            reject(new TransactionRefundedError());
            break;
          default:
            await saveStatus();
            break;
        }
      };
      this.swapProvider.monitorSwap(pendingSwap.id, onStatusUpdate).catch(reject);
    });
  }
  // =========================================================================
  // Lightning: Submarine swaps (send Arkade -> Lightning)
  // =========================================================================
  /**
   * Sends a Lightning payment.
   * @param args - The arguments for sending a Lightning payment.
   * @returns The result of the payment.
   */
  async sendLightningPayment(args) {
    const pendingSwap = await this.createSubmarineSwap(args);
    await this.savePendingSubmarineSwap(pendingSwap);
    const txid = await this.wallet.sendBitcoin({
      address: pendingSwap.response.address,
      amount: pendingSwap.response.expectedAmount
    });
    try {
      const { preimage } = await this.waitForSwapSettlement(pendingSwap);
      return {
        amount: pendingSwap.response.expectedAmount,
        preimage,
        txid
      };
    } catch (error) {
      if (error.isRefundable) {
        await this.refundVHTLC(pendingSwap);
        const finalStatus = await this.getSwapStatus(pendingSwap.id);
        await updateSubmarineSwapStatus(
          pendingSwap,
          finalStatus.status,
          this.savePendingSubmarineSwap.bind(this)
        );
      }
      throw new TransactionFailedError();
    }
  }
  /**
   * Creates a submarine swap.
   * @param args - The arguments for creating a submarine swap.
   * @returns The created pending submarine swap.
   */
  async createSubmarineSwap(args) {
    const refundPublicKey = hex8.encode(
      await this.wallet.identity.compressedPublicKey()
    );
    if (!refundPublicKey)
      throw new SwapError({
        message: "Failed to get refund public key from wallet"
      });
    const invoice = args.invoice;
    if (!invoice) throw new SwapError({ message: "Invoice is required" });
    const swapRequest = {
      invoice,
      refundPublicKey
    };
    const swapResponse = await this.swapProvider.createSubmarineSwap(swapRequest);
    const pendingSwap = {
      id: swapResponse.id,
      type: "submarine",
      createdAt: Math.floor(Date.now() / 1e3),
      request: swapRequest,
      response: swapResponse,
      status: "invoice.set"
    };
    await this.savePendingSubmarineSwap(pendingSwap);
    if (this.swapManager) {
      this.swapManager.addSwap(pendingSwap);
    }
    return pendingSwap;
  }
  /**
   * Claims the VHTLC for a pending submarine swap (aka refund).
   * @param pendingSwap - The pending submarine swap to refund the VHTLC.
   */
  async refundVHTLC(pendingSwap) {
    const preimageHash = pendingSwap.request.invoice ? getInvoicePaymentHash(pendingSwap.request.invoice) : pendingSwap.preimageHash;
    if (!preimageHash)
      throw new Error("Preimage hash is required to refund VHTLC");
    const vhtlcPkScript = ArkAddress2.decode(
      pendingSwap.response.address
    ).pkScript;
    const { vtxos } = await this.indexerProvider.getVtxos({
      scripts: [hex8.encode(vhtlcPkScript)]
    });
    if (vtxos.length === 0) {
      throw new Error(
        `VHTLC not found for address ${pendingSwap.response.address}`
      );
    }
    const vtxo = vtxos[0];
    if (vtxo.isSpent) {
      throw new Error("VHTLC is already spent");
    }
    const arkInfo = await this.arkProvider.getInfo();
    const address = await this.wallet.getAddress();
    if (!address) throw new Error("Failed to get ark address from wallet");
    const ourXOnlyPublicKey = normalizeToXOnlyKey(
      await this.wallet.identity.xOnlyPublicKey(),
      "our",
      pendingSwap.id
    );
    const serverXOnlyPublicKey = normalizeToXOnlyKey(
      hex8.decode(arkInfo.signerPubkey),
      "server",
      pendingSwap.id
    );
    const boltzXOnlyPublicKey = normalizeToXOnlyKey(
      hex8.decode(pendingSwap.response.claimPublicKey),
      "boltz",
      pendingSwap.id
    );
    const { vhtlcScript } = this.createVHTLCScript({
      network: arkInfo.network,
      preimageHash: hex8.decode(preimageHash),
      receiverPubkey: hex8.encode(boltzXOnlyPublicKey),
      senderPubkey: hex8.encode(ourXOnlyPublicKey),
      serverPubkey: hex8.encode(serverXOnlyPublicKey),
      timeoutBlockHeights: pendingSwap.response.timeoutBlockHeights
    });
    if (!vhtlcScript.claimScript)
      throw new Error("Failed to create VHTLC script for submarine swap");
    const isRecoverableVtxo = isRecoverable(vtxo);
    const input = {
      ...vtxo,
      tapLeafScript: isRecoverableVtxo ? vhtlcScript.refundWithoutReceiver() : vhtlcScript.refund(),
      tapTree: vhtlcScript.encode()
    };
    const output = {
      amount: BigInt(vtxo.value),
      script: ArkAddress2.decode(address).pkScript
    };
    if (isRecoverableVtxo) {
      await this.joinBatch(this.wallet.identity, input, output, arkInfo);
    } else {
      await refundVHTLCwithOffchainTx(
        pendingSwap.id,
        this.wallet.identity,
        this.arkProvider,
        boltzXOnlyPublicKey,
        ourXOnlyPublicKey,
        serverXOnlyPublicKey,
        input,
        output,
        arkInfo,
        this.swapProvider.refundSubmarineSwap.bind(this.swapProvider)
      );
    }
    await updateSubmarineSwapStatus(
      pendingSwap,
      pendingSwap.status,
      // Keep current status
      this.savePendingSubmarineSwap.bind(this),
      { refundable: true, refunded: true }
    );
  }
  /**
   * Waits for the swap settlement.
   * @param pendingSwap - The pending submarine swap.
   * @returns The status of the swap settlement.
   */
  async waitForSwapSettlement(pendingSwap) {
    return new Promise((resolve, reject) => {
      let isResolved = false;
      const onStatusUpdate = async (status) => {
        if (isResolved) return;
        const saveStatus = (additionalFields) => updateSubmarineSwapStatus(
          pendingSwap,
          status,
          this.savePendingSubmarineSwap.bind(this),
          additionalFields
        );
        switch (status) {
          case "swap.expired":
            isResolved = true;
            await saveStatus({ refundable: true });
            reject(
              new SwapExpiredError({
                isRefundable: true,
                pendingSwap
              })
            );
            break;
          case "invoice.failedToPay":
            isResolved = true;
            await saveStatus({ refundable: true });
            reject(
              new InvoiceFailedToPayError({
                isRefundable: true,
                pendingSwap
              })
            );
            break;
          case "transaction.lockupFailed":
            isResolved = true;
            await saveStatus({ refundable: true });
            reject(
              new TransactionLockupFailedError({
                isRefundable: true,
                pendingSwap
              })
            );
            break;
          case "transaction.claimed": {
            isResolved = true;
            const { preimage } = await this.swapProvider.getSwapPreimage(
              pendingSwap.id
            );
            await saveStatus({ preimage });
            resolve({ preimage });
            break;
          }
          default:
            await saveStatus();
            break;
        }
      };
      this.swapProvider.monitorSwap(pendingSwap.id, onStatusUpdate).catch((error) => {
        if (!isResolved) {
          isResolved = true;
          reject(error);
        }
      });
    });
  }
  // =========================================================================
  // Chain swaps: ARK -> BTC
  // =========================================================================
  /**
   * Creates a chain swap from ARK to BTC.
   * @param args - Swap arguments.
   * @returns The payment details and pending chain swap.
   */
  async arkToBtc(args) {
    const pendingSwap = await this.createChainSwap({
      to: "BTC",
      from: "ARK",
      feeSatsPerByte: args.feeSatsPerByte,
      senderLockAmount: args.senderLockAmount,
      receiverLockAmount: args.receiverLockAmount,
      toAddress: args.btcAddress
    });
    await this.verifyChainSwap({
      to: "BTC",
      from: "ARK",
      swap: pendingSwap,
      arkInfo: await this.arkProvider.getInfo()
    }).catch((err) => {
      throw new SwapError({
        message: `Chain swap verification failed: ${err.message}`
      });
    });
    return {
      amountToPay: pendingSwap.response.lockupDetails.amount,
      arkAddress: pendingSwap.response.lockupDetails.lockupAddress,
      pendingSwap
    };
  }
  /**
   * Waits for the swap to be confirmed and claims BTC.
   * @param pendingSwap - The pending chain swap to monitor.
   * @returns The transaction ID of the claimed HTLC.
   */
  async waitAndClaimBtc(pendingSwap) {
    if (this.swapManager && await this.swapManager.hasSwap(pendingSwap.id)) {
      const { txid } = await this.swapManager.waitForSwapCompletion(
        pendingSwap.id
      );
      return { txid };
    }
    return new Promise((resolve, reject) => {
      let claimStarted = false;
      const swap = { ...pendingSwap };
      const onStatusUpdate = async (status, data) => {
        const updateSwapStatus = async () => {
          swap.status = status;
          await this.savePendingChainSwap(swap);
          return swap;
        };
        switch (status) {
          case "transaction.mempool":
          case "transaction.confirmed":
            await updateSwapStatus();
            break;
          case "transaction.server.mempool":
          case "transaction.server.confirmed": {
            const updatedSwap = await updateSwapStatus();
            if (claimStarted) return;
            claimStarted = true;
            this.claimBtc(updatedSwap).catch(reject);
            break;
          }
          case "transaction.claimed":
            await updateSwapStatus();
            const claimedStatus = await this.getSwapStatus(
              pendingSwap.id
            );
            resolve({
              txid: claimedStatus.transaction?.id ?? ""
            });
            break;
          case "transaction.lockupFailed":
            await updateSwapStatus();
            await this.quoteSwap(swap.response.id).catch((err) => {
              reject(
                new SwapError({
                  message: `Failed to renegotiate quote: ${err.message}`,
                  isRefundable: true,
                  pendingSwap: swap
                })
              );
            });
            break;
          case "swap.expired":
            await updateSwapStatus();
            reject(
              new SwapExpiredError({
                isRefundable: true,
                pendingSwap: swap
              })
            );
            break;
          case "transaction.failed":
            await updateSwapStatus();
            reject(
              new TransactionFailedError({
                message: data.failureReason,
                isRefundable: true
              })
            );
            break;
          case "transaction.refunded":
            await updateSwapStatus();
            reject(new TransactionRefundedError());
            break;
          default:
            await updateSwapStatus();
            break;
        }
      };
      this.swapProvider.monitorSwap(swap.id, onStatusUpdate).catch(reject);
    });
  }
  /**
   * Claim sats on BTC chain by claiming the HTLC.
   * @param pendingSwap - The pending chain swap with BTC transaction hex.
   */
  async claimBtc(pendingSwap) {
    if (!pendingSwap.toAddress)
      throw new Error("Destination address is required");
    if (!pendingSwap.response.claimDetails.swapTree)
      throw new Error("Missing swap tree in claim details");
    if (!pendingSwap.response.claimDetails.serverPublicKey)
      throw new Error("Missing server public key in claim details");
    const swapStatus = await this.getSwapStatus(pendingSwap.id);
    if (!swapStatus.transaction?.hex)
      throw new Error("BTC transaction hex is required");
    const lockupTx = Transaction6.fromRaw(
      hex8.decode(swapStatus.transaction.hex)
    );
    const arkInfo = await this.arkProvider.getInfo();
    const network = arkInfo.network === "bitcoin" ? NETWORK : arkInfo.network === "mutinynet" ? MUTINYNET_NETWORK : REGTEST_NETWORK;
    const swapTree = deserializeSwapTree(
      pendingSwap.response.claimDetails.swapTree
    );
    const musig = tweakMusig(
      create(hex8.decode(pendingSwap.ephemeralKey), [
        hex8.decode(pendingSwap.response.claimDetails.serverPublicKey),
        secp256k12.getPublicKey(hex8.decode(pendingSwap.ephemeralKey))
      ]),
      swapTree.tree
    );
    const swapOutput = detectSwapOutput(musig.aggPubkey, lockupTx);
    const feeToDeliverExactAmount = BigInt(
      pendingSwap.request.serverLockAmount ? pendingSwap.request.serverLockAmount - pendingSwap.amount : 0
    );
    const claimTx = targetFee(
      pendingSwap.feeSatsPerByte,
      (fee) => constructClaimTransaction(
        {
          script: swapOutput.script,
          amount: swapOutput.amount,
          vout: swapOutput.vout,
          transactionId: lockupTx.id
        },
        OutScript2.encode(
          Address2(network).decode(pendingSwap.toAddress)
        ),
        feeToDeliverExactAmount > fee ? feeToDeliverExactAmount : fee
      )
    );
    const musigMessage = musig.message(
      claimTx.preimageWitnessV1(
        0,
        [swapOutput.script],
        SigHash2.DEFAULT,
        [swapOutput.amount]
      )
    ).generateNonce();
    const signedTxData = await this.swapProvider.postChainClaimDetails(
      pendingSwap.response.id,
      {
        preimage: pendingSwap.preimage,
        toSign: {
          pubNonce: hex8.encode(musigMessage.publicNonce),
          transaction: claimTx.hex,
          index: 0
        }
      }
    );
    if (!signedTxData.pubNonce || !signedTxData.partialSignature)
      throw new Error("Invalid signature data from server");
    const musigSession = musigMessage.aggregateNonces([
      [
        hex8.decode(
          pendingSwap.response.claimDetails.serverPublicKey
        ),
        hex8.decode(signedTxData.pubNonce)
      ]
    ]).initializeSession();
    musigSession.addPartial(
      hex8.decode(pendingSwap.response.claimDetails.serverPublicKey),
      hex8.decode(signedTxData.partialSignature)
    );
    const musigSigned = musigSession.signPartial();
    claimTx.updateInput(0, {
      finalScriptWitness: [musigSigned.aggregatePartials()]
    });
    await this.swapProvider.postBtcTransaction(claimTx.hex);
  }
  /**
   * When an ARK to BTC swap fails, refund sats on ARK chain by claiming the VHTLC.
   * @param pendingSwap - The pending chain swap to refund.
   */
  async refundArk(pendingSwap) {
    if (!pendingSwap.response.lockupDetails.serverPublicKey)
      throw new Error("Missing server public key in lockup details");
    if (!pendingSwap.response.lockupDetails.timeouts)
      throw new Error("Missing timeouts in lockup details");
    const arkInfo = await this.arkProvider.getInfo();
    const address = await this.wallet.getAddress();
    const ourXOnlyPublicKey = normalizeToXOnlyKey(
      await this.wallet.identity.xOnlyPublicKey(),
      "user",
      pendingSwap.id
    );
    const serverXOnlyPublicKey = normalizeToXOnlyKey(
      hex8.decode(arkInfo.signerPubkey),
      "server",
      pendingSwap.id
    );
    const boltzXOnlyPublicKey = normalizeToXOnlyKey(
      hex8.decode(pendingSwap.response.lockupDetails.serverPublicKey),
      "boltz",
      pendingSwap.id
    );
    const vhtlcPkScript = ArkAddress2.decode(
      pendingSwap.response.lockupDetails.lockupAddress
    ).pkScript;
    const { vtxos } = await this.indexerProvider.getVtxos({
      scripts: [hex8.encode(vhtlcPkScript)]
    });
    if (vtxos.length === 0) {
      throw new Error(
        `VHTLC not found for address ${pendingSwap.response.lockupDetails.lockupAddress}`
      );
    }
    const vtxo = vtxos[0];
    if (vtxo.isSpent) {
      throw new Error("VHTLC is already spent");
    }
    const { vhtlcAddress, vhtlcScript } = this.createVHTLCScript({
      network: arkInfo.network,
      preimageHash: hex8.decode(pendingSwap.request.preimageHash),
      serverPubkey: hex8.encode(serverXOnlyPublicKey),
      senderPubkey: hex8.encode(ourXOnlyPublicKey),
      receiverPubkey: hex8.encode(boltzXOnlyPublicKey),
      timeoutBlockHeights: pendingSwap.response.lockupDetails.timeouts
    });
    if (!vhtlcScript.refundScript)
      throw new Error("Failed to create VHTLC script for chain swap");
    if (pendingSwap.response.lockupDetails.lockupAddress !== vhtlcAddress) {
      throw new SwapError({
        message: "Unable to claim: invalid VHTLC address"
      });
    }
    const isRecoverableVtxo = isRecoverable(vtxo);
    const input = {
      ...vtxo,
      tapLeafScript: isRecoverableVtxo ? vhtlcScript.refundWithoutReceiver() : vhtlcScript.refund(),
      tapTree: vhtlcScript.encode()
    };
    const output = {
      amount: BigInt(vtxo.value),
      script: ArkAddress2.decode(address).pkScript
    };
    if (isRecoverableVtxo) {
      await this.joinBatch(this.wallet.identity, input, output, arkInfo);
    } else {
      await refundVHTLCwithOffchainTx(
        pendingSwap.id,
        this.wallet.identity,
        this.arkProvider,
        boltzXOnlyPublicKey,
        ourXOnlyPublicKey,
        serverXOnlyPublicKey,
        input,
        output,
        arkInfo,
        this.swapProvider.refundChainSwap.bind(this.swapProvider)
      );
    }
    const finalStatus = await this.getSwapStatus(pendingSwap.id);
    await this.savePendingChainSwap({
      ...pendingSwap,
      status: finalStatus.status
    });
  }
  // =========================================================================
  // Chain swaps: BTC -> ARK
  // =========================================================================
  /**
   * Creates a chain swap from BTC to ARK.
   * @param args - Swap arguments.
   * @returns The pending chain swap and payment details.
   */
  async btcToArk(args) {
    const pendingSwap = await this.createChainSwap({
      to: "ARK",
      from: "BTC",
      feeSatsPerByte: args.feeSatsPerByte,
      senderLockAmount: args.senderLockAmount,
      receiverLockAmount: args.receiverLockAmount,
      toAddress: await this.wallet.getAddress()
    });
    await this.verifyChainSwap({
      to: "ARK",
      from: "BTC",
      swap: pendingSwap,
      arkInfo: await this.arkProvider.getInfo()
    }).catch((err) => {
      throw new SwapError({
        message: `Chain swap verification failed: ${err.message}`
      });
    });
    return {
      amountToPay: pendingSwap.response.lockupDetails.amount,
      btcAddress: pendingSwap.response.lockupDetails.lockupAddress,
      pendingSwap
    };
  }
  /**
   * Waits for the swap to be confirmed and claims ARK.
   * @param pendingSwap - The pending chain swap to monitor.
   * @returns The transaction ID of the claimed VHTLC.
   */
  async waitAndClaimArk(pendingSwap) {
    if (this.swapManager && await this.swapManager.hasSwap(pendingSwap.id)) {
      const { txid } = await this.swapManager.waitForSwapCompletion(
        pendingSwap.id
      );
      return { txid };
    }
    return new Promise((resolve, reject) => {
      let claimStarted = false;
      const swap = { ...pendingSwap };
      const onStatusUpdate = async (status, data) => {
        const updateSwapStatus = () => {
          swap.status = status;
          return this.savePendingChainSwap(swap);
        };
        switch (status) {
          case "transaction.server.mempool":
          case "transaction.server.confirmed":
            await updateSwapStatus();
            if (claimStarted) return;
            claimStarted = true;
            this.claimArk(swap).catch(reject);
            break;
          case "transaction.claimed":
            await updateSwapStatus();
            const claimedStatus = await this.getSwapStatus(
              pendingSwap.id
            );
            resolve({
              txid: claimedStatus.transaction?.id ?? ""
            });
            break;
          case "transaction.claim.pending":
            await updateSwapStatus();
            await this.signCooperativeClaimForServer(swap).catch(
              (err) => {
                logger.error(
                  `Failed to sign cooperative claim for ${swap.id}:`,
                  err
                );
              }
            );
            break;
          case "transaction.lockupFailed":
            await updateSwapStatus();
            await this.quoteSwap(swap.response.id).catch((err) => {
              reject(
                new SwapError({
                  message: `Failed to renegotiate quote: ${err.message}`,
                  isRefundable: false,
                  // TODO btc refund not implemented yet
                  pendingSwap: swap
                })
              );
            });
            break;
          case "swap.expired":
            await updateSwapStatus();
            reject(
              new SwapExpiredError({
                isRefundable: false,
                // TODO btc refund not implemented yet
                pendingSwap: swap
              })
            );
            break;
          case "transaction.failed":
            await updateSwapStatus();
            reject(
              new TransactionFailedError({
                message: data.failureReason,
                isRefundable: false
                // TODO btc refund not implemented yet
              })
            );
            break;
          case "transaction.refunded":
            await updateSwapStatus();
            reject(new TransactionRefundedError());
            break;
          default:
            await updateSwapStatus();
            break;
        }
      };
      this.swapProvider.monitorSwap(swap.id, onStatusUpdate).catch(reject);
    });
  }
  /**
   * Claim sats on ARK chain by claiming the VHTLC.
   * Refactored to use claimVHTLCIdentity + claimVHTLCwithOffchainTx utilities.
   * @param pendingSwap - The pending chain swap.
   */
  async claimArk(pendingSwap) {
    if (!pendingSwap.toAddress)
      throw new Error("Destination address is required");
    if (!pendingSwap.response.claimDetails.serverPublicKey)
      throw new Error("Missing server public key in claim details");
    if (!pendingSwap.response.claimDetails.timeouts)
      throw new Error("Missing timeouts in claim details");
    const arkInfo = await this.arkProvider.getInfo();
    const preimage = hex8.decode(pendingSwap.preimage);
    const address = await this.wallet.getAddress();
    const receiverXOnlyPublicKey = normalizeToXOnlyKey(
      pendingSwap.request.claimPublicKey,
      "receiver"
    );
    const senderXOnlyPublicKey = normalizeToXOnlyKey(
      pendingSwap.response.claimDetails.serverPublicKey,
      "sender"
    );
    const serverXOnlyPublicKey = normalizeToXOnlyKey(
      arkInfo.signerPubkey,
      "server"
    );
    const { vhtlcAddress, vhtlcScript } = this.createVHTLCScript({
      network: arkInfo.network,
      preimageHash: hex8.decode(pendingSwap.request.preimageHash),
      serverPubkey: hex8.encode(serverXOnlyPublicKey),
      senderPubkey: hex8.encode(senderXOnlyPublicKey),
      receiverPubkey: hex8.encode(receiverXOnlyPublicKey),
      timeoutBlockHeights: pendingSwap.response.claimDetails.timeouts
    });
    if (!vhtlcScript.claimScript)
      throw new Error("Failed to create VHTLC script for chain swap");
    if (pendingSwap.response.claimDetails.lockupAddress !== vhtlcAddress) {
      throw new SwapError({
        message: "Unable to claim: invalid VHTLC address"
      });
    }
    const spendableVtxos = await this.indexerProvider.getVtxos({
      scripts: [hex8.encode(vhtlcScript.pkScript)],
      spendableOnly: true
    });
    if (spendableVtxos.vtxos.length === 0)
      throw new Error("No spendable virtual coins found");
    const vtxo = spendableVtxos.vtxos[0];
    const input = {
      ...vtxo,
      tapLeafScript: vhtlcScript.claim(),
      tapTree: vhtlcScript.encode()
    };
    const output = {
      amount: BigInt(vtxo.value),
      script: ArkAddress2.decode(address).pkScript
    };
    const vhtlcIdentity = claimVHTLCIdentity(
      this.wallet.identity,
      preimage
    );
    if (isRecoverable(vtxo)) {
      await this.joinBatch(vhtlcIdentity, input, output, arkInfo);
    } else {
      await claimVHTLCwithOffchainTx(
        vhtlcIdentity,
        vhtlcScript,
        serverXOnlyPublicKey,
        input,
        output,
        arkInfo,
        this.arkProvider
      );
    }
    const finalStatus = await this.getSwapStatus(pendingSwap.id);
    await this.savePendingChainSwap({
      ...pendingSwap,
      status: finalStatus.status
    });
  }
  /**
   * Sign a cooperative claim for the server in BTC => ARK swaps.
   * @param pendingSwap - The pending chain swap.
   */
  async signCooperativeClaimForServer(pendingSwap) {
    if (!pendingSwap.response.lockupDetails.swapTree)
      throw new Error("Missing swap tree in lockup details");
    if (!pendingSwap.response.lockupDetails.serverPublicKey)
      throw new Error("Missing server public key in lockup details");
    const claimDetails = await this.swapProvider.getChainClaimDetails(
      pendingSwap.id
    );
    const serverPubKey = pendingSwap.response.lockupDetails.serverPublicKey;
    if (claimDetails.publicKey !== serverPubKey) {
      throw new Error(
        `Server public key mismatch: claim response has ${claimDetails.publicKey}, expected ${serverPubKey}`
      );
    }
    const musig = tweakMusig(
      create(hex8.decode(pendingSwap.ephemeralKey), [
        hex8.decode(serverPubKey),
        secp256k12.getPublicKey(hex8.decode(pendingSwap.ephemeralKey))
      ]),
      deserializeSwapTree(pendingSwap.response.lockupDetails.swapTree).tree
    );
    const musigNonces = musig.message(hex8.decode(claimDetails.transactionHash)).generateNonce().aggregateNonces([
      [
        hex8.decode(
          pendingSwap.response.lockupDetails.serverPublicKey
        ),
        hex8.decode(claimDetails.pubNonce)
      ]
    ]).initializeSession();
    const partialSig = musigNonces.signPartial();
    await this.swapProvider.postChainClaimDetails(pendingSwap.response.id, {
      signature: {
        partialSignature: hex8.encode(partialSig.ourPartialSignature),
        pubNonce: hex8.encode(partialSig.publicNonce)
      }
    });
  }
  /**
   * Waits for a chain swap to be claimable and then claims it.
   * Dispatches to waitAndClaimArk or waitAndClaimBtc based on swap direction.
   * @param pendingSwap - The pending swap to wait for and claim.
   * @returns The transaction ID of the claim.
   */
  async waitAndClaimChain(pendingSwap) {
    if (pendingSwap.request.to === "ARK")
      return this.waitAndClaimArk(pendingSwap);
    if (pendingSwap.request.to === "BTC")
      return this.waitAndClaimBtc(pendingSwap);
    throw new SwapError({
      message: `Unsupported swap destination: ${pendingSwap.request.to}`
    });
  }
  // =========================================================================
  // Chain swap creation and verification
  // =========================================================================
  /**
   * Creates a chain swap.
   * @param args - The arguments for creating a chain swap.
   * @returns The created pending chain swap.
   */
  async createChainSwap(args) {
    const { to, from, receiverLockAmount, senderLockAmount, toAddress } = args;
    if (!toAddress)
      throw new SwapError({ message: "Destination address is required" });
    const feeSatsPerByte = args.feeSatsPerByte ?? 1;
    if (feeSatsPerByte <= 0)
      throw new SwapError({ message: "Invalid feeSatsPerByte" });
    let amount, serverLockAmount, userLockAmount;
    if (receiverLockAmount) {
      amount = receiverLockAmount;
      const fees = await this.getFees(from, to);
      serverLockAmount = receiverLockAmount + fees.minerFees.user.claim;
    } else if (senderLockAmount) {
      amount = senderLockAmount;
      userLockAmount = senderLockAmount;
    }
    if (!amount || amount <= 0) {
      throw new SwapError({ message: "Invalid lock amount" });
    }
    const preimage = randomBytes(32);
    const preimageHash = hex8.encode(sha2563(preimage));
    if (!preimageHash)
      throw new SwapError({ message: "Failed to get preimage hash" });
    const ephemeralKey = secp256k12.utils.randomSecretKey();
    const refundPublicKey = to === "ARK" ? hex8.encode(secp256k12.getPublicKey(ephemeralKey)) : hex8.encode(await this.wallet.identity.compressedPublicKey());
    if (!refundPublicKey)
      throw new SwapError({
        message: "Failed to get refund public key"
      });
    const claimPublicKey = to === "ARK" ? hex8.encode(await this.wallet.identity.compressedPublicKey()) : hex8.encode(secp256k12.getPublicKey(ephemeralKey));
    if (!claimPublicKey)
      throw new SwapError({
        message: "Failed to get claim public key"
      });
    const swapRequest = {
      to,
      from,
      preimageHash,
      feeSatsPerByte,
      claimPublicKey,
      refundPublicKey,
      serverLockAmount,
      userLockAmount
    };
    const swapResponse = await this.swapProvider.createChainSwap(swapRequest);
    const pendingSwap = {
      amount,
      createdAt: Math.floor(Date.now() / 1e3),
      ephemeralKey: hex8.encode(ephemeralKey),
      feeSatsPerByte,
      id: swapResponse.id,
      preimage: hex8.encode(preimage),
      request: swapRequest,
      response: swapResponse,
      status: "swap.created",
      toAddress: args.toAddress,
      type: "chain"
    };
    await this.savePendingChainSwap(pendingSwap);
    this.swapManager?.addSwap(pendingSwap);
    return pendingSwap;
  }
  /**
   * Validates the lockup and claim addresses match the expected scripts.
   * @param args - The arguments for verifying a chain swap.
   * @returns True if the addresses match.
   */
  async verifyChainSwap(args) {
    const { to, from, swap, arkInfo } = args;
    if (from === "ARK") {
      if (!swap.response.lockupDetails.serverPublicKey)
        throw new Error("Missing serverPublicKey in lockup details");
      if (!swap.response.lockupDetails.timeouts)
        throw new Error("Missing timeouts in lockup details");
    }
    if (to === "ARK") {
      if (!swap.response.claimDetails.serverPublicKey)
        throw new Error("Missing serverPublicKey in claim details");
      if (!swap.response.claimDetails.timeouts)
        throw new Error("Missing timeouts in claim details");
    }
    const lockupAddress = to === "ARK" ? swap.response.claimDetails.lockupAddress : swap.response.lockupDetails.lockupAddress;
    const receiverPubkey = to === "ARK" ? swap.request.claimPublicKey : swap.response.lockupDetails.serverPublicKey;
    const senderPubkey = to === "ARK" ? swap.response.claimDetails.serverPublicKey : swap.request.refundPublicKey;
    const serverPubkey = hex8.encode(
      normalizeToXOnlyKey(arkInfo.signerPubkey, "server")
    );
    const timeoutBlockHeights = to === "ARK" ? swap.response.claimDetails.timeouts : swap.response.lockupDetails.timeouts;
    const { vhtlcAddress } = this.createVHTLCScript({
      network: arkInfo.network,
      preimageHash: hex8.decode(swap.request.preimageHash),
      receiverPubkey,
      senderPubkey,
      serverPubkey,
      timeoutBlockHeights
    });
    if (lockupAddress !== vhtlcAddress) {
      throw new SwapError({
        message: "Boltz is trying to scam us (invalid address)"
      });
    }
    return true;
  }
  /**
   * Renegotiates the quote for an existing swap.
   * @param swapId - The ID of the swap.
   * @returns The accepted quote amount.
   */
  async quoteSwap(swapId) {
    const { amount } = await this.swapProvider.getChainQuote(swapId);
    await this.swapProvider.postChainQuote(swapId, { amount });
    return amount;
  }
  // =========================================================================
  // Shared utilities
  // =========================================================================
  /**
   * Joins a batch to spend the vtxo via commitment transaction.
   * @param identity - The identity to use for signing.
   * @param input - The input vtxo.
   * @param output - The output script.
   * @param arkInfo - Chain information used for building transactions.
   * @param isRecoverable - Whether the input is recoverable.
   * @returns The commitment transaction ID.
   */
  async joinBatch(identity, input, output, arkInfo, isRecoverable2 = true) {
    return joinBatch(
      this.arkProvider,
      identity,
      input,
      output,
      arkInfo,
      isRecoverable2
    );
  }
  /**
   * Creates a VHTLC script for the swap.
   * Works for submarine, reverse, and chain swaps.
   */
  createVHTLCScript(args) {
    return createVHTLCScript(args);
  }
  async getFees(from, to) {
    if (from && to) {
      return this.swapProvider.getChainFees(from, to);
    }
    return this.swapProvider.getFees();
  }
  async getLimits(from, to) {
    if (from && to) {
      return this.swapProvider.getChainLimits(from, to);
    }
    return this.swapProvider.getLimits();
  }
  /**
   * Retrieves swap status by ID.
   * @param swapId - The ID of the swap.
   * @returns The status of the swap.
   */
  async getSwapStatus(swapId) {
    return this.swapProvider.getSwapStatus(swapId);
  }
  // =========================================================================
  // Storage queries
  // =========================================================================
  /**
   * Returns pending submarine swaps (those with status `invoice.set`).
   */
  async getPendingSubmarineSwaps() {
    const swaps = await this.getPendingSubmarineSwapsFromStorage();
    if (!swaps) return [];
    return swaps.filter(
      (swap) => swap.status === "invoice.set"
    );
  }
  /**
   * Returns pending reverse swaps (those with status `swap.created`).
   */
  async getPendingReverseSwaps() {
    const swaps = await this.getPendingReverseSwapsFromStorage();
    if (!swaps) return [];
    return swaps.filter(
      (swap) => swap.status === "swap.created"
    );
  }
  /**
   * Returns pending chain swaps (those with status `swap.created`).
   */
  async getPendingChainSwaps() {
    const swaps = await this.getPendingChainSwapsFromStorage();
    return swaps.filter((swap) => swap.status === "swap.created");
  }
  /**
   * Retrieves swap history from storage.
   * @returns Array of all swaps (reverse + submarine + chain) sorted by creation date (newest first).
   */
  async getSwapHistory() {
    const reverseSwaps = await this.getPendingReverseSwapsFromStorage();
    const submarineSwaps = await this.getPendingSubmarineSwapsFromStorage();
    const chainSwaps = await this.getPendingChainSwapsFromStorage();
    const allSwaps = [
      ...reverseSwaps || [],
      ...submarineSwaps || [],
      ...chainSwaps || []
    ];
    return allSwaps.sort((a, b) => b.createdAt - a.createdAt);
  }
  /**
   * Refreshes the status of all pending swaps in the storage provider.
   */
  async refreshSwapsStatus() {
    const promises = [];
    for (const swap of await this.getPendingReverseSwapsFromStorage()) {
      if (isReverseFinalStatus(swap.status)) continue;
      promises.push(
        this.getSwapStatus(swap.id).then(
          ({ status }) => updateReverseSwapStatus(
            swap,
            status,
            this.savePendingReverseSwap.bind(this)
          )
        ).catch((error) => {
          logger.error(
            `Failed to refresh swap status for ${swap.id}:`,
            error
          );
        })
      );
    }
    for (const swap of await this.getPendingSubmarineSwapsFromStorage()) {
      if (isSubmarineFinalStatus(swap.status)) continue;
      promises.push(
        this.getSwapStatus(swap.id).then(
          ({ status }) => updateSubmarineSwapStatus(
            swap,
            status,
            this.savePendingSubmarineSwap.bind(this)
          )
        ).catch((error) => {
          logger.error(
            `Failed to refresh swap status for ${swap.id}:`,
            error
          );
        })
      );
    }
    for (const swap of await this.getPendingChainSwapsFromStorage()) {
      if (isChainFinalStatus(swap.status)) continue;
      promises.push(
        this.getSwapStatus(swap.id).then(
          ({ status }) => this.savePendingChainSwap({ ...swap, status })
        ).catch((error) => {
          logger.error(
            `Failed to refresh swap status for ${swap.id}:`,
            error
          );
        })
      );
    }
    await Promise.all(promises);
  }
  // =========================================================================
  // Swap restoration and enrichment
  // =========================================================================
  /**
   * Restore swaps from Boltz API.
   *
   * Note: restored swaps may lack local-only data such as the original
   * Lightning invoice or preimage. They are intended primarily for
   * display/monitoring and are not automatically wired into the SwapManager.
   */
  async restoreSwaps(boltzFees) {
    const publicKey = hex8.encode(
      await this.wallet.identity.compressedPublicKey()
    );
    if (!publicKey) throw new Error("Failed to get public key from wallet");
    const fees = boltzFees ?? await this.swapProvider.getFees();
    const reverseSwaps = [];
    const submarineSwaps = [];
    const restoredSwaps = await this.swapProvider.restoreSwaps(publicKey);
    for (const swap of restoredSwaps) {
      const { id, createdAt, status } = swap;
      if (isRestoredReverseSwap(swap)) {
        const {
          amount,
          lockupAddress,
          preimageHash,
          serverPublicKey,
          tree
        } = swap.claimDetails;
        reverseSwaps.push({
          id,
          createdAt,
          request: {
            invoiceAmount: extractInvoiceAmount(amount, fees),
            claimPublicKey: publicKey,
            preimageHash
          },
          response: {
            id,
            invoice: "",
            onchainAmount: amount,
            lockupAddress,
            refundPublicKey: serverPublicKey,
            timeoutBlockHeights: {
              refund: extractTimeLockFromLeafOutput(
                tree.refundWithoutBoltzLeaf.output
              ),
              unilateralClaim: extractTimeLockFromLeafOutput(
                tree.unilateralClaimLeaf.output
              ),
              unilateralRefund: extractTimeLockFromLeafOutput(
                tree.unilateralRefundLeaf.output
              ),
              unilateralRefundWithoutReceiver: extractTimeLockFromLeafOutput(
                tree.unilateralRefundWithoutBoltzLeaf.output
              )
            }
          },
          status,
          type: "reverse",
          preimage: ""
        });
      } else if (isRestoredSubmarineSwap(swap)) {
        const { amount, lockupAddress, serverPublicKey, tree } = swap.refundDetails;
        let preimage = "";
        try {
          const data = await this.swapProvider.getSwapPreimage(
            swap.id
          );
          preimage = data.preimage;
        } catch (error) {
          logger.warn(
            `Failed to restore preimage for submarine swap ${id}`,
            error
          );
        }
        submarineSwaps.push({
          id,
          type: "submarine",
          createdAt,
          preimage,
          preimageHash: swap.preimageHash,
          status,
          request: {
            invoice: "",
            refundPublicKey: publicKey
          },
          response: {
            id,
            address: lockupAddress,
            expectedAmount: amount,
            claimPublicKey: serverPublicKey,
            timeoutBlockHeights: {
              refund: extractTimeLockFromLeafOutput(
                tree.refundWithoutBoltzLeaf.output
              ),
              unilateralClaim: extractTimeLockFromLeafOutput(
                tree.unilateralClaimLeaf.output
              ),
              unilateralRefund: extractTimeLockFromLeafOutput(
                tree.unilateralRefundLeaf.output
              ),
              unilateralRefundWithoutReceiver: extractTimeLockFromLeafOutput(
                tree.unilateralRefundWithoutBoltzLeaf.output
              )
            }
          }
        });
      }
    }
    return { reverseSwaps, submarineSwaps };
  }
  /**
   * Enrich a restored reverse swap with its preimage.
   */
  enrichReverseSwapPreimage(swap, preimage) {
    return enrichReverseSwapPreimage(swap, preimage);
  }
  /**
   * Enrich a restored submarine swap with its invoice.
   */
  enrichSubmarineSwapInvoice(swap, invoice) {
    return enrichSubmarineSwapInvoice(swap, invoice);
  }
};

export {
  SwapError,
  InvoiceExpiredError,
  InvoiceFailedToPayError,
  InsufficientFundsError,
  NetworkError,
  SchemaError,
  SwapExpiredError,
  TransactionFailedError,
  PreimageFetchError,
  isSubmarineFailedStatus,
  isSubmarineFinalStatus,
  isSubmarinePendingStatus,
  isSubmarineRefundableStatus,
  isSubmarineSuccessStatus,
  isReverseFailedStatus,
  isReverseFinalStatus,
  isReversePendingStatus,
  isReverseClaimableStatus,
  isReverseSuccessStatus,
  isChainFailedStatus,
  isChainClaimableStatus,
  isChainFinalStatus,
  isChainPendingStatus,
  isChainRefundableStatus,
  isChainSignableStatus,
  isChainSuccessStatus,
  isPendingReverseSwap,
  isPendingSubmarineSwap,
  isPendingChainSwap,
  isSubmarineSwapRefundable,
  isChainSwapRefundable,
  isReverseSwapClaimable,
  isChainSwapClaimable,
  BoltzSwapProvider,
  decodeInvoice,
  getInvoiceSatoshis,
  getInvoicePaymentHash,
  isValidArkAddress,
  verifySignatures,
  logger,
  setLogger,
  SwapManager,
  saveSwap,
  updateReverseSwapStatus,
  updateSubmarineSwapStatus,
  updateChainSwapStatus,
  enrichReverseSwapPreimage,
  enrichSubmarineSwapInvoice,
  IndexedDbSwapRepository,
  ArkadeSwaps
};
