// @arkade-os/boltz-swap/dist/chunk-WNTQXNGX.js
import {
  ArkadeSwaps,
  BoltzSwapProvider,
  isPendingReverseSwap,
  isPendingSubmarineSwap,
  isReverseClaimableStatus,
  isReverseFinalStatus,
  isSubmarineFinalStatus,
  isSubmarineSwapRefundable,
  logger
} from "./chunk-XVJ47GNJ.js";
import {
  __require
} from "./chunk-3RG5ZIWI.js";

// src/expo/background.ts
import { runTasks } from "@arkade-os/sdk/worker/expo";
import {
  ExpoArkProvider,
  ExpoIndexerProvider
} from "@arkade-os/sdk/adapters/expo";

// src/expo/swapsPollProcessor.ts
var SWAP_POLL_TASK_TYPE = "swap-poll";
var swapsPollProcessor = {
  taskType: SWAP_POLL_TASK_TYPE,
  async execute(item, deps) {
    const {
      swapRepository,
      swapProvider,
      wallet,
      arkProvider,
      indexerProvider
    } = deps;
    const allSwaps = await swapRepository.getAllSwaps();
    const pendingSwaps = allSwaps.filter((swap) => {
      if (isPendingReverseSwap(swap))
        return !isReverseFinalStatus(swap.status);
      if (isPendingSubmarineSwap(swap))
        return !isSubmarineFinalStatus(swap.status);
      return false;
    });
    let polled = 0;
    let updated = 0;
    let claimed = 0;
    let refunded = 0;
    let errors = 0;
    const tempSwaps = new ArkadeSwaps({
      wallet,
      arkProvider,
      indexerProvider,
      swapProvider,
      swapManager: false,
      swapRepository
    });
    try {
      for (const swap of pendingSwaps) {
        try {
          const { status: currentStatus } = await swapProvider.getSwapStatus(swap.id);
          polled++;
          if (currentStatus !== swap.status) {
            await swapRepository.saveSwap({
              ...swap,
              status: currentStatus
            });
            updated++;
          }
          if (isPendingReverseSwap(swap) && isReverseClaimableStatus(currentStatus)) {
            if (!swap.preimage) {
              logger.warn(
                `[swap-poll] Skipping claim for ${swap.id}: no preimage`
              );
              continue;
            }
            try {
              await tempSwaps.claimVHTLC(swap);
              claimed++;
            } catch (claimError) {
              logger.error(
                `[swap-poll] Claim failed for ${swap.id}:`,
                claimError
              );
              errors++;
            }
          }
          const swapWithStatus = isPendingSubmarineSwap(swap) ? { ...swap, status: currentStatus } : null;
          if (isPendingSubmarineSwap(swap) && isSubmarineSwapRefundable(swapWithStatus)) {
            if (!swap.request.invoice && !swap.preimageHash) {
              logger.warn(
                `[swap-poll] Skipping refund for ${swap.id}: no invoice or preimageHash`
              );
              continue;
            }
            try {
              await tempSwaps.refundVHTLC(swapWithStatus);
              refunded++;
            } catch (refundError) {
              logger.error(
                `[swap-poll] Refund failed for ${swap.id}:`,
                refundError
              );
              errors++;
            }
          }
        } catch (swapError) {
          logger.error(
            `[swap-poll] Error processing swap ${swap.id}:`,
            swapError
          );
          errors++;
        }
      }
    } finally {
      await tempSwaps.dispose();
    }
    return {
      taskItemId: item.id,
      type: SWAP_POLL_TASK_TYPE,
      status: errors > 0 && polled === 0 ? "failed" : "success",
      data: { polled, updated, claimed, refunded, errors }
    };
  }
};

// src/expo/background.ts
function requireTaskManager() {
  try {
    return __require("expo-task-manager");
  } catch {
    throw new Error(
      "expo-task-manager is required for background tasks. Install it with: npx expo install expo-task-manager"
    );
  }
}
function requireBackgroundTask() {
  try {
    return __require("expo-background-task");
  } catch {
    throw new Error(
      "expo-background-task is required for background tasks. Install it with: npx expo install expo-background-task"
    );
  }
}
function getRandomId() {
  return Math.random().toString(36).slice(2) + Date.now().toString(36);
}
function createBackgroundWalletShim(args) {
  const notImplemented = (method) => {
    throw new Error(
      `[boltz-swap] Background wallet shim: "${String(method)}" is not implemented`
    );
  };
  return {
    identity: args.identity,
    getAddress: args.getAddress,
    getBoardingAddress: async () => notImplemented("getBoardingAddress"),
    getBalance: async () => notImplemented("getBalance"),
    getVtxos: async () => notImplemented("getVtxos"),
    getBoardingUtxos: async () => notImplemented("getBoardingUtxos"),
    getTransactionHistory: async () => notImplemented("getTransactionHistory"),
    getContractManager: async () => notImplemented("getContractManager"),
    getDelegatorManager: async () => notImplemented("getDelegatorManager"),
    sendBitcoin: async () => notImplemented("sendBitcoin"),
    send: async () => notImplemented("send"),
    settle: async () => notImplemented("settle"),
    assetManager: new Proxy({}, {
      get: () => notImplemented("assetManager")
    })
  };
}
function defineExpoSwapBackgroundTask(taskName, options) {
  const TaskManager = requireTaskManager();
  const BackgroundTask = requireBackgroundTask();
  const { taskQueue, swapRepository, identityFactory } = options;
  TaskManager.defineTask(taskName, async () => {
    try {
      const config = await taskQueue.loadConfig();
      if (!config) {
        return BackgroundTask.BackgroundTaskResult.Success;
      }
      const identity = await identityFactory();
      const arkProvider = new ExpoArkProvider(config.arkServerUrl);
      const indexerProvider = new ExpoIndexerProvider(
        config.arkServerUrl
      );
      const swapProvider = new BoltzSwapProvider({
        network: config.network,
        apiUrl: config.boltzApiUrl
      });
      const wallet = createBackgroundWalletShim({
        identity,
        getAddress: async () => {
          const { ArkAddress } = await import("@arkade-os/sdk");
          const { hex } = await import("@scure/base");
          const info = await arkProvider.getInfo();
          const pubkey = await identity.xOnlyPublicKey();
          const serverPubKey = hex.decode(info.signerPubkey);
          const xOnlyServerPubKey = serverPubKey.length === 33 ? serverPubKey.slice(1) : serverPubKey;
          const hrp = info.network === "bitcoin" ? "ark" : "tark";
          return new ArkAddress(
            xOnlyServerPubKey,
            pubkey,
            hrp
          ).encode();
        }
      });
      const deps = {
        swapRepository,
        swapProvider,
        arkProvider,
        indexerProvider,
        identity,
        wallet
      };
      await runTasks(taskQueue, [swapsPollProcessor], deps);
      const results = await taskQueue.getResults();
      if (results.length > 0) {
        await taskQueue.acknowledgeResults(
          results.map((r) => r.id)
        );
      }
      const existing = await taskQueue.getTasks(SWAP_POLL_TASK_TYPE);
      if (existing.length === 0) {
        const task = {
          id: getRandomId(),
          type: SWAP_POLL_TASK_TYPE,
          data: {},
          createdAt: Date.now()
        };
        await taskQueue.addTask(task);
      }
      return BackgroundTask.BackgroundTaskResult.Success;
    } catch (error) {
      console.error(
        "[boltz-swap] Background task failed:",
        error instanceof Error ? error.message : error
      );
      return BackgroundTask.BackgroundTaskResult.Failed;
    }
  });
}
async function registerExpoSwapBackgroundTask(taskName, options) {
  const BackgroundTask = requireBackgroundTask();
  await BackgroundTask.registerTaskAsync(taskName, {
    // expo-background-task expects minutes:
    // https://docs.expo.dev/versions/latest/sdk/background-task/#backgroundtaskoptions
    minimumInterval: options?.minimumInterval ?? 15
  });
}
async function unregisterExpoSwapBackgroundTask(taskName) {
  const BackgroundTask = requireBackgroundTask();
  await BackgroundTask.unregisterTaskAsync(taskName);
}

export {
  SWAP_POLL_TASK_TYPE,
  swapsPollProcessor,
  defineExpoSwapBackgroundTask,
  registerExpoSwapBackgroundTask,
  unregisterExpoSwapBackgroundTask
};
