// @arkade-os/boltz-swap/dist/index.js
import {
  ArkadeSwaps,
  BoltzSwapProvider,
  IndexedDbSwapRepository,
  InsufficientFundsError,
  InvoiceExpiredError,
  InvoiceFailedToPayError,
  NetworkError,
  PreimageFetchError,
  SchemaError,
  SwapError,
  SwapExpiredError,
  SwapManager,
  TransactionFailedError,
  decodeInvoice,
  enrichReverseSwapPreimage,
  enrichSubmarineSwapInvoice,
  getInvoicePaymentHash,
  getInvoiceSatoshis,
  isChainClaimableStatus,
  isChainFailedStatus,
  isChainFinalStatus,
  isChainPendingStatus,
  isChainRefundableStatus,
  isChainSignableStatus,
  isChainSuccessStatus,
  isChainSwapClaimable,
  isChainSwapRefundable,
  isPendingChainSwap,
  isPendingReverseSwap,
  isPendingSubmarineSwap,
  isReverseClaimableStatus,
  isReverseFailedStatus,
  isReverseFinalStatus,
  isReversePendingStatus,
  isReverseSuccessStatus,
  isReverseSwapClaimable,
  isSubmarineFailedStatus,
  isSubmarineFinalStatus,
  isSubmarinePendingStatus,
  isSubmarineRefundableStatus,
  isSubmarineSuccessStatus,
  isSubmarineSwapRefundable,
  isValidArkAddress,
  logger,
  saveSwap,
  setLogger,
  updateChainSwapStatus,
  updateReverseSwapStatus,
  updateSubmarineSwapStatus,
  verifySignatures
} from "./chunk-XVJ47GNJ.js";
import "./chunk-3RG5ZIWI.js";

// src/serviceWorker/arkade-swaps-message-handler.ts
import {
  RestArkProvider,
  RestIndexerProvider
} from "@arkade-os/sdk";
var DEFAULT_MESSAGE_TAG = "ARKADE_SWAPS_UPDATER";
var ArkadeSwapsMessageHandler = class _ArkadeSwapsMessageHandler {
  constructor(swapRepository) {
    this.swapRepository = swapRepository;
  }
  static messageTag = DEFAULT_MESSAGE_TAG;
  messageTag = _ArkadeSwapsMessageHandler.messageTag;
  arkProvider;
  indexerProvider;
  swapProvider;
  wallet;
  handler;
  swapManager;
  getSwapManagerOrThrow() {
    const sm = this.handler?.getSwapManager();
    if (!sm) throw new Error("SwapManager is not enabled");
    return sm;
  }
  async start(opts) {
    if (!opts.wallet) throw new Error("Wallet is required");
    this.wallet = opts.wallet;
  }
  async stop() {
    const handler = this.handler;
    if (!handler) return;
    const swapManager = this.swapManager ?? handler.getSwapManager();
    if (swapManager) {
      await swapManager.stop();
    }
    if (typeof handler.dispose === "function") {
      await handler.dispose();
    }
    this.swapManager = null;
    this.handler = void 0;
    this.wallet = void 0;
    this.arkProvider = void 0;
    this.indexerProvider = void 0;
    this.swapProvider = void 0;
  }
  async tick(_now) {
    return [];
  }
  tagged(res) {
    return {
      ...res,
      tag: this.messageTag
    };
  }
  async broadcastEvent(event) {
    const sw = self;
    if (!sw?.clients?.matchAll) return;
    const clients = await sw.clients.matchAll();
    for (const client of clients) {
      try {
        client.postMessage(event);
      } catch {
      }
    }
  }
  async handleMessage(message) {
    const id = message.id;
    if (message.type === "INIT_ARKADE_SWAPS") {
      try {
        await this.handleInit(message);
        return this.tagged({
          id,
          type: "ARKADE_SWAPS_INITIALIZED"
        });
      } catch (error) {
        return this.tagged({ id, error });
      }
    }
    if (!this.handler || !this.wallet) {
      return this.tagged({
        id,
        error: new Error("handler not initialized")
      });
    }
    try {
      switch (message.type) {
        case "CREATE_LIGHTNING_INVOICE": {
          const res = await this.handler.createLightningInvoice(
            message.payload
          );
          return this.tagged({
            id,
            type: "LIGHTNING_INVOICE_CREATED",
            payload: res
          });
        }
        case "SEND_LIGHTNING_PAYMENT": {
          const res = await this.handler.sendLightningPayment(
            message.payload
          );
          return this.tagged({
            id,
            type: "LIGHTNING_PAYMENT_SENT",
            payload: res
          });
        }
        case "CREATE_SUBMARINE_SWAP": {
          const res = await this.handler.createSubmarineSwap(
            message.payload
          );
          return this.tagged({
            id,
            type: "SUBMARINE_SWAP_CREATED",
            payload: res
          });
        }
        case "CREATE_REVERSE_SWAP": {
          const res = await this.handler.createReverseSwap(
            message.payload
          );
          return this.tagged({
            id,
            type: "REVERSE_SWAP_CREATED",
            payload: res
          });
        }
        case "CLAIM_VHTLC":
          await this.handler.claimVHTLC(message.payload);
          return this.tagged({ id, type: "VHTLC_CLAIMED" });
        case "REFUND_VHTLC":
          await this.handler.refundVHTLC(message.payload);
          return this.tagged({ id, type: "VHTLC_REFUNDED" });
        case "WAIT_AND_CLAIM": {
          const res = await this.handler.waitAndClaim(
            message.payload
          );
          return this.tagged({
            id,
            type: "WAIT_AND_CLAIMED",
            payload: res
          });
        }
        case "WAIT_FOR_SWAP_SETTLEMENT": {
          const res = await this.handler.waitForSwapSettlement(
            message.payload
          );
          return this.tagged({
            id,
            type: "SWAP_SETTLED",
            payload: res
          });
        }
        case "RESTORE_SWAPS": {
          const res = await this.handler.restoreSwaps(
            message.payload
          );
          return this.tagged({
            id,
            type: "SWAPS_RESTORED",
            payload: res
          });
        }
        case "ENRICH_REVERSE_SWAP_PREIMAGE": {
          const res = this.handler.enrichReverseSwapPreimage(
            message.payload.swap,
            message.payload.preimage
          );
          return this.tagged({
            id,
            type: "REVERSE_SWAP_PREIMAGE_ENRICHED",
            payload: res
          });
        }
        case "ENRICH_SUBMARINE_SWAP_INVOICE": {
          const res = this.handler.enrichSubmarineSwapInvoice(
            message.payload.swap,
            message.payload.invoice
          );
          return this.tagged({
            id,
            type: "SUBMARINE_SWAP_INVOICE_ENRICHED",
            payload: res
          });
        }
        case "GET_FEES": {
          const res = message.payload ? await this.handler.getFees(
            message.payload.from,
            message.payload.to
          ) : await this.handler.getFees();
          return this.tagged({ id, type: "FEES", payload: res });
        }
        case "GET_LIMITS": {
          const res = message.payload ? await this.handler.getLimits(
            message.payload.from,
            message.payload.to
          ) : await this.handler.getLimits();
          return this.tagged({ id, type: "LIMITS", payload: res });
        }
        case "GET_SWAP_STATUS": {
          const res = await this.handler.getSwapStatus(
            message.payload.swapId
          );
          return this.tagged({
            id,
            type: "SWAP_STATUS",
            payload: res
          });
        }
        case "GET_PENDING_SUBMARINE_SWAPS": {
          const res = await this.handler.getPendingSubmarineSwaps();
          return this.tagged({
            id,
            type: "PENDING_SUBMARINE_SWAPS",
            payload: res
          });
        }
        case "GET_PENDING_REVERSE_SWAPS": {
          const res = await this.handler.getPendingReverseSwaps();
          return this.tagged({
            id,
            type: "PENDING_REVERSE_SWAPS",
            payload: res
          });
        }
        case "GET_PENDING_CHAIN_SWAPS": {
          const res = await this.handler.getPendingChainSwaps();
          return this.tagged({
            id,
            type: "PENDING_CHAIN_SWAPS",
            payload: res
          });
        }
        case "GET_SWAP_HISTORY": {
          const res = await this.handler.getSwapHistory();
          return this.tagged({
            id,
            type: "SWAP_HISTORY",
            payload: res
          });
        }
        case "REFRESH_SWAPS_STATUS":
          await this.handler.refreshSwapsStatus();
          return this.tagged({ id, type: "SWAPS_STATUS_REFRESHED" });
        case "ARK_TO_BTC": {
          const res = await this.handler.arkToBtc(message.payload);
          return this.tagged({
            id,
            type: "ARK_TO_BTC_CREATED",
            payload: res
          });
        }
        case "BTC_TO_ARK": {
          const res = await this.handler.btcToArk(message.payload);
          return this.tagged({
            id,
            type: "BTC_TO_ARK_CREATED",
            payload: res
          });
        }
        case "CREATE_CHAIN_SWAP": {
          const res = await this.handler.createChainSwap(
            message.payload
          );
          return this.tagged({
            id,
            type: "CHAIN_SWAP_CREATED",
            payload: res
          });
        }
        case "WAIT_AND_CLAIM_CHAIN": {
          const res = await this.handler.waitAndClaimChain(
            message.payload
          );
          return this.tagged({
            id,
            type: "CHAIN_CLAIMED",
            payload: res
          });
        }
        case "WAIT_AND_CLAIM_ARK": {
          const res = await this.handler.waitAndClaimArk(
            message.payload
          );
          return this.tagged({
            id,
            type: "ARK_CLAIMED",
            payload: res
          });
        }
        case "WAIT_AND_CLAIM_BTC": {
          const res = await this.handler.waitAndClaimBtc(
            message.payload
          );
          return this.tagged({
            id,
            type: "BTC_CLAIMED",
            payload: res
          });
        }
        case "CLAIM_ARK":
          await this.handler.claimArk(message.payload);
          return this.tagged({ id, type: "ARK_CLAIM_EXECUTED" });
        case "CLAIM_BTC":
          await this.handler.claimBtc(message.payload);
          return this.tagged({ id, type: "BTC_CLAIM_EXECUTED" });
        case "REFUND_ARK":
          await this.handler.refundArk(message.payload);
          return this.tagged({ id, type: "ARK_REFUND_EXECUTED" });
        case "SIGN_SERVER_CLAIM":
          await this.handler.signCooperativeClaimForServer(
            message.payload
          );
          return this.tagged({ id, type: "SERVER_CLAIM_SIGNED" });
        case "VERIFY_CHAIN_SWAP": {
          const verified = await this.handler.verifyChainSwap({
            ...message.payload
          });
          return this.tagged({
            id,
            type: "CHAIN_SWAP_VERIFIED",
            payload: { verified }
          });
        }
        case "QUOTE_SWAP": {
          const amount = await this.handler.quoteSwap(
            message.payload.swapId
          );
          return this.tagged({
            id,
            type: "SWAP_QUOTED",
            payload: { amount }
          });
        }
        /* --- SwapManager methods --- */
        case "SM-START": {
          await this.handler.startSwapManager();
          return this.tagged({ id, type: "SM-STARTED" });
        }
        case "SM-STOP": {
          await this.handler.stopSwapManager();
          return this.tagged({ id, type: "SM-STOPPED" });
        }
        case "SM-ADD_SWAP": {
          await this.getSwapManagerOrThrow().addSwap(message.payload);
          return this.tagged({ id, type: "SM-SWAP_ADDED" });
        }
        case "SM-REMOVE_SWAP": {
          await this.getSwapManagerOrThrow().removeSwap(
            message.payload.swapId
          );
          return this.tagged({ id, type: "SM-SWAP_REMOVED" });
        }
        case "SM-GET_PENDING_SWAPS": {
          const res = await this.getSwapManagerOrThrow().getPendingSwaps();
          return this.tagged({
            id,
            type: "SM-PENDING_SWAPS",
            payload: res
          });
        }
        case "SM-HAS_SWAP": {
          const has = await this.getSwapManagerOrThrow().hasSwap(
            message.payload.swapId
          );
          return this.tagged({
            id,
            type: "SM-HAS_SWAP_RESULT",
            payload: { has }
          });
        }
        case "SM-IS_PROCESSING": {
          const processing = await this.getSwapManagerOrThrow().isProcessing(
            message.payload.swapId
          );
          return this.tagged({
            id,
            type: "SM-IS_PROCESSING_RESULT",
            payload: { processing }
          });
        }
        case "SM-GET_STATS": {
          const stats = await this.getSwapManagerOrThrow().getStats();
          return this.tagged({
            id,
            type: "SM-STATS",
            payload: stats
          });
        }
        case "SM-WAIT_FOR_COMPLETION": {
          const res = await this.getSwapManagerOrThrow().waitForSwapCompletion(
            message.payload.swapId
          );
          return this.tagged({
            id,
            type: "SM-COMPLETED",
            payload: res
          });
        }
        default:
          console.error("Unknown message type", message);
          throw new Error("Unknown message");
      }
    } catch (error) {
      return this.tagged({ id, error });
    }
  }
  async handleInit({ payload }) {
    if (!this.wallet) {
      throw new Error("Wallet is required");
    }
    const { arkServerUrl } = payload;
    this.arkProvider = new RestArkProvider(arkServerUrl);
    this.indexerProvider = new RestIndexerProvider(arkServerUrl);
    this.swapProvider = new BoltzSwapProvider({
      apiUrl: payload.swapProvider.baseUrl,
      network: payload.network
    });
    const handler = new ArkadeSwaps({
      wallet: this.wallet,
      arkProvider: this.arkProvider,
      swapProvider: this.swapProvider,
      indexerProvider: this.indexerProvider,
      swapRepository: this.swapRepository,
      swapManager: payload.swapManager
    });
    this.handler = handler;
    const sm = handler.getSwapManager();
    this.swapManager = sm;
    if (sm) {
      void sm.onSwapUpdate(async (swap, oldStatus) => {
        await this.broadcastEvent({
          tag: this.messageTag,
          type: "SM-EVENT-SWAP_UPDATE",
          payload: { swap, oldStatus }
        });
      });
      void sm.onSwapCompleted(async (swap) => {
        await this.broadcastEvent({
          tag: this.messageTag,
          type: "SM-EVENT-SWAP_COMPLETED",
          payload: { swap }
        });
      });
      void sm.onSwapFailed(async (swap, error) => {
        await this.broadcastEvent({
          tag: this.messageTag,
          type: "SM-EVENT-SWAP_FAILED",
          payload: { swap, error: { message: error.message } }
        });
      });
      void sm.onActionExecuted(async (swap, action) => {
        await this.broadcastEvent({
          tag: this.messageTag,
          type: "SM-EVENT-ACTION_EXECUTED",
          payload: { swap, action }
        });
      });
      void sm.onWebSocketConnected(async () => {
        await this.broadcastEvent({
          tag: this.messageTag,
          type: "SM-EVENT-WS_CONNECTED"
        });
      });
      void sm.onWebSocketDisconnected(async (error) => {
        await this.broadcastEvent({
          tag: this.messageTag,
          type: "SM-EVENT-WS_DISCONNECTED",
          payload: error ? { errorMessage: error.message } : void 0
        });
      });
    }
  }
};

// src/serviceWorker/arkade-swaps-runtime.ts
var ServiceWorkerArkadeSwaps = class _ServiceWorkerArkadeSwaps {
  constructor(messageTag, serviceWorker, swapRepository, withSwapManager) {
    this.messageTag = messageTag;
    this.serviceWorker = serviceWorker;
    this.swapRepository = swapRepository;
    this.withSwapManager = withSwapManager;
  }
  eventListenerInitialized = false;
  swapUpdateListeners = /* @__PURE__ */ new Set();
  swapCompletedListeners = /* @__PURE__ */ new Set();
  swapFailedListeners = /* @__PURE__ */ new Set();
  actionExecutedListeners = /* @__PURE__ */ new Set();
  wsConnectedListeners = /* @__PURE__ */ new Set();
  wsDisconnectedListeners = /* @__PURE__ */ new Set();
  static async create(config) {
    const messageTag = config.messageTag ?? DEFAULT_MESSAGE_TAG;
    const swapRepository = config.swapRepository ?? new IndexedDbSwapRepository();
    const svcArkadeSwaps = new _ServiceWorkerArkadeSwaps(
      messageTag,
      config.serviceWorker,
      swapRepository,
      Boolean(config.swapManager)
    );
    const initMessage = {
      tag: messageTag,
      id: getRandomId(),
      type: "INIT_ARKADE_SWAPS",
      payload: {
        network: config.network,
        arkServerUrl: config.arkServerUrl,
        swapProvider: { baseUrl: config.swapProvider.getApiUrl() },
        swapManager: config.swapManager
      }
    };
    await svcArkadeSwaps.sendMessage(initMessage);
    return svcArkadeSwaps;
  }
  async startSwapManager() {
    if (!this.withSwapManager) {
      throw new Error("SwapManager is not enabled.");
    }
    await this.sendMessage({
      id: getRandomId(),
      tag: this.messageTag,
      type: "SM-START"
    });
  }
  async stopSwapManager() {
    if (!this.withSwapManager) return;
    await this.sendMessage({
      id: getRandomId(),
      tag: this.messageTag,
      type: "SM-STOP"
    });
  }
  getSwapManager() {
    if (!this.withSwapManager) {
      return null;
    }
    this.initEventStream();
    const send = this.sendMessage.bind(this);
    const tag = this.messageTag;
    const proxy = {
      start: async () => {
        await send({
          id: getRandomId(),
          tag,
          type: "SM-START"
        });
      },
      stop: async () => {
        await send({
          id: getRandomId(),
          tag,
          type: "SM-STOP"
        });
      },
      addSwap: async (swap) => {
        await send({
          id: getRandomId(),
          tag,
          type: "SM-ADD_SWAP",
          payload: swap
        });
      },
      removeSwap: async (swapId) => {
        await send({
          id: getRandomId(),
          tag,
          type: "SM-REMOVE_SWAP",
          payload: { swapId }
        });
      },
      getPendingSwaps: async () => {
        const res = await send({
          id: getRandomId(),
          tag,
          type: "SM-GET_PENDING_SWAPS"
        });
        return res.payload;
      },
      hasSwap: async (swapId) => {
        const res = await send({
          id: getRandomId(),
          tag,
          type: "SM-HAS_SWAP",
          payload: { swapId }
        });
        return res.payload.has;
      },
      isProcessing: async (swapId) => {
        const res = await send({
          id: getRandomId(),
          tag,
          type: "SM-IS_PROCESSING",
          payload: { swapId }
        });
        return res.payload.processing;
      },
      getStats: async () => {
        const res = await send({
          id: getRandomId(),
          tag,
          type: "SM-GET_STATS"
        });
        return res.payload;
      },
      waitForSwapCompletion: async (swapId) => {
        const res = await send({
          id: getRandomId(),
          tag,
          type: "SM-WAIT_FOR_COMPLETION",
          payload: { swapId }
        });
        return res.payload;
      },
      subscribeToSwapUpdates: async (swapId, callback) => {
        const filteredListener = (swap, oldStatus) => {
          if (swap.id === swapId) {
            callback(swap, oldStatus);
          }
        };
        this.swapUpdateListeners.add(filteredListener);
        return () => this.swapUpdateListeners.delete(filteredListener);
      },
      onSwapUpdate: async (listener) => {
        this.swapUpdateListeners.add(listener);
        return () => this.swapUpdateListeners.delete(listener);
      },
      onSwapCompleted: async (listener) => {
        this.swapCompletedListeners.add(listener);
        return () => this.swapCompletedListeners.delete(listener);
      },
      onSwapFailed: async (listener) => {
        this.swapFailedListeners.add(listener);
        return () => this.swapFailedListeners.delete(listener);
      },
      onActionExecuted: async (listener) => {
        this.actionExecutedListeners.add(listener);
        return () => this.actionExecutedListeners.delete(listener);
      },
      onWebSocketConnected: async (listener) => {
        this.wsConnectedListeners.add(listener);
        return () => this.wsConnectedListeners.delete(listener);
      },
      onWebSocketDisconnected: async (listener) => {
        this.wsDisconnectedListeners.add(listener);
        return () => this.wsDisconnectedListeners.delete(listener);
      },
      offSwapUpdate: (listener) => {
        this.swapUpdateListeners.delete(listener);
      },
      offSwapCompleted: (listener) => {
        this.swapCompletedListeners.delete(listener);
      },
      offSwapFailed: (listener) => {
        this.swapFailedListeners.delete(listener);
      },
      offActionExecuted: (listener) => {
        this.actionExecutedListeners.delete(listener);
      },
      offWebSocketConnected: (listener) => {
        this.wsConnectedListeners.delete(listener);
      },
      offWebSocketDisconnected: (listener) => {
        this.wsDisconnectedListeners.delete(listener);
      }
    };
    return proxy;
  }
  async createLightningInvoice(args) {
    try {
      const res = await this.sendMessage({
        id: getRandomId(),
        tag: this.messageTag,
        type: "CREATE_LIGHTNING_INVOICE",
        payload: args
      });
      return res.payload;
    } catch (e) {
      throw new Error("Cannot create Lightning Invoice", { cause: e });
    }
  }
  async sendLightningPayment(args) {
    try {
      const res = await this.sendMessage({
        id: getRandomId(),
        tag: this.messageTag,
        type: "SEND_LIGHTNING_PAYMENT",
        payload: args
      });
      return res.payload;
    } catch (e) {
      throw new Error("Cannot send Lightning payment", { cause: e });
    }
  }
  async createSubmarineSwap(args) {
    try {
      const res = await this.sendMessage({
        id: getRandomId(),
        tag: this.messageTag,
        type: "CREATE_SUBMARINE_SWAP",
        payload: args
      });
      return res.payload;
    } catch (e) {
      throw new Error("Cannot create submarine swap", { cause: e });
    }
  }
  async createReverseSwap(args) {
    try {
      const res = await this.sendMessage({
        id: getRandomId(),
        tag: this.messageTag,
        type: "CREATE_REVERSE_SWAP",
        payload: args
      });
      return res.payload;
    } catch (e) {
      throw new Error("Cannot create reverse swap", { cause: e });
    }
  }
  async claimVHTLC(pendingSwap) {
    await this.sendMessage({
      id: getRandomId(),
      tag: this.messageTag,
      type: "CLAIM_VHTLC",
      payload: pendingSwap
    });
  }
  async refundVHTLC(pendingSwap) {
    await this.sendMessage({
      id: getRandomId(),
      tag: this.messageTag,
      type: "REFUND_VHTLC",
      payload: pendingSwap
    });
  }
  async waitAndClaim(pendingSwap) {
    try {
      const res = await this.sendMessage({
        id: getRandomId(),
        tag: this.messageTag,
        type: "WAIT_AND_CLAIM",
        payload: pendingSwap
      });
      return res.payload;
    } catch (e) {
      throw new Error("Cannot wait and claim reverse swap", {
        cause: e
      });
    }
  }
  async waitForSwapSettlement(pendingSwap) {
    try {
      const res = await this.sendMessage({
        id: getRandomId(),
        tag: this.messageTag,
        type: "WAIT_FOR_SWAP_SETTLEMENT",
        payload: pendingSwap
      });
      return res.payload;
    } catch (e) {
      throw new Error("Cannot wait for swap settlement", { cause: e });
    }
  }
  async restoreSwaps(boltzFees) {
    try {
      const res = await this.sendMessage({
        id: getRandomId(),
        tag: this.messageTag,
        type: "RESTORE_SWAPS",
        payload: boltzFees
      });
      return res.payload;
    } catch (e) {
      throw new Error("Cannot restore swaps", { cause: e });
    }
  }
  async arkToBtc(args) {
    try {
      const res = await this.sendMessage({
        id: getRandomId(),
        tag: this.messageTag,
        type: "ARK_TO_BTC",
        payload: args
      });
      return res.payload;
    } catch (e) {
      throw new Error("Cannot create ARK -> BTC chain swap", {
        cause: e
      });
    }
  }
  async btcToArk(args) {
    try {
      const res = await this.sendMessage({
        id: getRandomId(),
        tag: this.messageTag,
        type: "BTC_TO_ARK",
        payload: args
      });
      return res.payload;
    } catch (e) {
      throw new Error("Cannot create BTC -> ARK chain swap", {
        cause: e
      });
    }
  }
  async createChainSwap(args) {
    try {
      const res = await this.sendMessage({
        id: getRandomId(),
        tag: this.messageTag,
        type: "CREATE_CHAIN_SWAP",
        payload: args
      });
      return res.payload;
    } catch (e) {
      throw new Error("Cannot create chain swap", { cause: e });
    }
  }
  async waitAndClaimChain(pendingSwap) {
    try {
      const res = await this.sendMessage({
        id: getRandomId(),
        tag: this.messageTag,
        type: "WAIT_AND_CLAIM_CHAIN",
        payload: pendingSwap
      });
      return res.payload;
    } catch (e) {
      throw new Error("Cannot wait and claim chain swap", {
        cause: e
      });
    }
  }
  async waitAndClaimArk(pendingSwap) {
    try {
      const res = await this.sendMessage({
        id: getRandomId(),
        tag: this.messageTag,
        type: "WAIT_AND_CLAIM_ARK",
        payload: pendingSwap
      });
      return res.payload;
    } catch (e) {
      throw new Error("Cannot wait and claim ARK", {
        cause: e
      });
    }
  }
  async waitAndClaimBtc(pendingSwap) {
    try {
      const res = await this.sendMessage({
        id: getRandomId(),
        tag: this.messageTag,
        type: "WAIT_AND_CLAIM_BTC",
        payload: pendingSwap
      });
      return res.payload;
    } catch (e) {
      throw new Error("Cannot wait and claim BTC", {
        cause: e
      });
    }
  }
  async claimArk(pendingSwap) {
    await this.sendMessage({
      id: getRandomId(),
      tag: this.messageTag,
      type: "CLAIM_ARK",
      payload: pendingSwap
    });
  }
  async claimBtc(pendingSwap) {
    await this.sendMessage({
      id: getRandomId(),
      tag: this.messageTag,
      type: "CLAIM_BTC",
      payload: pendingSwap
    });
  }
  async refundArk(pendingSwap) {
    await this.sendMessage({
      id: getRandomId(),
      tag: this.messageTag,
      type: "REFUND_ARK",
      payload: pendingSwap
    });
  }
  async signCooperativeClaimForServer(pendingSwap) {
    await this.sendMessage({
      id: getRandomId(),
      tag: this.messageTag,
      type: "SIGN_SERVER_CLAIM",
      payload: pendingSwap
    });
  }
  async verifyChainSwap(args) {
    try {
      const res = await this.sendMessage({
        id: getRandomId(),
        tag: this.messageTag,
        type: "VERIFY_CHAIN_SWAP",
        payload: {
          to: args.to,
          from: args.from,
          swap: args.swap,
          arkInfo: args.arkInfo
        }
      });
      return res.payload.verified;
    } catch (e) {
      throw new Error("Cannot verify chain swap", { cause: e });
    }
  }
  async quoteSwap(swapId) {
    try {
      const res = await this.sendMessage({
        id: getRandomId(),
        tag: this.messageTag,
        type: "QUOTE_SWAP",
        payload: { swapId }
      });
      return res.payload.amount;
    } catch (e) {
      throw new Error("Cannot quote swap", { cause: e });
    }
  }
  enrichReverseSwapPreimage(swap, preimage) {
    return enrichReverseSwapPreimage(swap, preimage);
  }
  enrichSubmarineSwapInvoice(swap, invoice) {
    return enrichSubmarineSwapInvoice(swap, invoice);
  }
  createVHTLCScript(_args) {
    throw new Error(
      "createVHTLCScript is not supported via service worker"
    );
  }
  async joinBatch(_identity, _input, _output, _arkInfo, _isRecoverable = true) {
    throw new Error("joinBatch is not supported via service worker");
  }
  async getFees(from, to) {
    if (from === void 0 !== (to === void 0)) {
      throw new Error("Both 'from' and 'to' must be provided together");
    }
    try {
      const res = await this.sendMessage({
        id: getRandomId(),
        tag: this.messageTag,
        type: "GET_FEES",
        ...from !== void 0 && to !== void 0 ? { payload: { from, to } } : {}
      });
      return res.payload;
    } catch (e) {
      throw new Error("Cannot get fees", { cause: e });
    }
  }
  async getLimits(from, to) {
    if (from === void 0 !== (to === void 0)) {
      throw new Error("Both 'from' and 'to' must be provided together");
    }
    try {
      const res = await this.sendMessage({
        id: getRandomId(),
        tag: this.messageTag,
        type: "GET_LIMITS",
        ...from !== void 0 && to !== void 0 ? { payload: { from, to } } : {}
      });
      return res.payload;
    } catch (e) {
      throw new Error("Cannot get limits", { cause: e });
    }
  }
  async getSwapStatus(swapId) {
    try {
      const res = await this.sendMessage({
        id: getRandomId(),
        tag: this.messageTag,
        type: "GET_SWAP_STATUS",
        payload: { swapId }
      });
      return res.payload;
    } catch (e) {
      throw new Error("Cannot get swap status", { cause: e });
    }
  }
  async getPendingSubmarineSwaps() {
    try {
      const res = await this.sendMessage({
        id: getRandomId(),
        tag: this.messageTag,
        type: "GET_PENDING_SUBMARINE_SWAPS"
      });
      return res.payload;
    } catch (e) {
      throw new Error("Cannot get pending submarine swaps", {
        cause: e
      });
    }
  }
  async getPendingReverseSwaps() {
    try {
      const res = await this.sendMessage({
        id: getRandomId(),
        tag: this.messageTag,
        type: "GET_PENDING_REVERSE_SWAPS"
      });
      return res.payload;
    } catch (e) {
      throw new Error("Cannot get pending reverse swaps", { cause: e });
    }
  }
  async getPendingChainSwaps() {
    try {
      const res = await this.sendMessage({
        id: getRandomId(),
        tag: this.messageTag,
        type: "GET_PENDING_CHAIN_SWAPS"
      });
      return res.payload;
    } catch (e) {
      throw new Error("Cannot get pending chain swaps", { cause: e });
    }
  }
  async getSwapHistory() {
    try {
      const res = await this.sendMessage({
        id: getRandomId(),
        tag: this.messageTag,
        type: "GET_SWAP_HISTORY"
      });
      return res.payload;
    } catch (e) {
      throw new Error("Cannot get swap history", { cause: e });
    }
  }
  async refreshSwapsStatus() {
    await this.sendMessage({
      id: getRandomId(),
      tag: this.messageTag,
      type: "REFRESH_SWAPS_STATUS"
    });
  }
  async dispose() {
    if (this.withSwapManager) {
      await this.stopSwapManager().catch(() => {
      });
    }
  }
  async [Symbol.asyncDispose]() {
    return this.dispose();
  }
  async sendMessage(request) {
    return new Promise((resolve, reject) => {
      const timeoutMs = 3e4;
      let timeout;
      const cleanup = () => {
        if (timeout) clearTimeout(timeout);
        navigator.serviceWorker.removeEventListener(
          "message",
          messageHandler
        );
      };
      const messageHandler = (event) => {
        const response = event.data;
        if (!response || response.tag !== this.messageTag || response.id !== request.id) {
          return;
        }
        cleanup();
        if (response.error) {
          reject(response.error);
        } else {
          resolve(response);
        }
      };
      timeout = setTimeout(() => {
        cleanup();
        reject(
          new Error(
            `Timed out waiting for service worker response: ${request.type}`
          )
        );
      }, timeoutMs);
      navigator.serviceWorker.addEventListener("message", messageHandler);
      this.serviceWorker.postMessage(request);
    });
  }
  initEventStream() {
    if (this.eventListenerInitialized) return;
    this.eventListenerInitialized = true;
    navigator.serviceWorker.addEventListener(
      "message",
      this.handleEventMessage
    );
  }
  handleEventMessage = (event) => {
    const data = event.data;
    if (!data || data.tag !== this.messageTag) return;
    if (typeof data.type !== "string") return;
    if (!data.type.startsWith("SM-EVENT-")) return;
    switch (data.type) {
      case "SM-EVENT-SWAP_UPDATE":
        this.swapUpdateListeners.forEach((cb) => {
          cb(data.payload.swap, data.payload.oldStatus);
        });
        break;
      case "SM-EVENT-SWAP_COMPLETED":
        this.swapCompletedListeners.forEach((cb) => {
          cb(data.payload.swap);
        });
        break;
      case "SM-EVENT-SWAP_FAILED": {
        const err = new Error(data.payload.error?.message);
        this.swapFailedListeners.forEach((cb) => {
          cb(data.payload.swap, err);
        });
        break;
      }
      case "SM-EVENT-ACTION_EXECUTED":
        this.actionExecutedListeners.forEach((cb) => {
          cb(data.payload.swap, data.payload.action);
        });
        break;
      case "SM-EVENT-WS_CONNECTED":
        this.wsConnectedListeners.forEach((cb) => {
          cb();
        });
        break;
      case "SM-EVENT-WS_DISCONNECTED": {
        const err = data.payload?.errorMessage ? new Error(data.payload.errorMessage) : void 0;
        this.wsDisconnectedListeners.forEach((cb) => {
          cb(err);
        });
        break;
      }
      default:
        break;
    }
  };
};
function getRandomId() {
  return `${Date.now().toString(36)}-${Math.random().toString(36).slice(2, 10)}`;
}

// src/repositories/migrationFromContracts.ts
var MIGRATION_KEY = "migration-from-storage-adapter-swaps";
async function migrateToSwapRepository(storageAdapter, fresh) {
  try {
    const migration = await storageAdapter.getItem(MIGRATION_KEY);
    if (migration === "done") {
      return false;
    }
    const reverseSwaps = await getContractCollection(storageAdapter, "reverseSwaps");
    const submarineSwaps = await getContractCollection(storageAdapter, "submarineSwaps");
    for (const swap of reverseSwaps) {
      await fresh.saveSwap(swap);
    }
    for (const swap of submarineSwaps) {
      await fresh.saveSwap(swap);
    }
    await storageAdapter.setItem(MIGRATION_KEY, "done");
    return true;
  } catch (error) {
    if (error instanceof Error && error.message.includes(
      "One of the specified object stores was not found."
    )) {
      return false;
    }
    throw error;
  }
}
async function getContractCollection(storage, contractType) {
  const stored = await storage.getItem(`collection:${contractType}`);
  if (!stored) return [];
  try {
    return JSON.parse(stored);
  } catch (error) {
    const errMessage = "message" in error ? error.message : "";
    throw new Error(
      `Failed to parse contract collection ${contractType} from storage: ${errMessage}`
    );
  }
}
export {
  ArkadeSwapsMessageHandler as ArkadeLightningMessageHandler,
  ArkadeSwaps,
  ArkadeSwapsMessageHandler,
  BoltzSwapProvider,
  IndexedDbSwapRepository,
  InsufficientFundsError,
  InvoiceExpiredError,
  InvoiceFailedToPayError,
  NetworkError,
  PreimageFetchError,
  SchemaError,
  ServiceWorkerArkadeSwaps as ServiceWorkerArkadeLightning,
  ServiceWorkerArkadeSwaps,
  SwapError,
  SwapExpiredError,
  SwapManager,
  TransactionFailedError,
  decodeInvoice,
  enrichReverseSwapPreimage,
  enrichSubmarineSwapInvoice,
  getInvoicePaymentHash,
  getInvoiceSatoshis,
  isChainClaimableStatus,
  isChainFailedStatus,
  isChainFinalStatus,
  isChainPendingStatus,
  isChainRefundableStatus,
  isChainSignableStatus,
  isChainSuccessStatus,
  isChainSwapClaimable,
  isChainSwapRefundable,
  isPendingChainSwap,
  isPendingReverseSwap,
  isPendingSubmarineSwap,
  isReverseClaimableStatus,
  isReverseFailedStatus,
  isReverseFinalStatus,
  isReversePendingStatus,
  isReverseSuccessStatus,
  isReverseSwapClaimable,
  isSubmarineFailedStatus,
  isSubmarineFinalStatus,
  isSubmarinePendingStatus,
  isSubmarineRefundableStatus,
  isSubmarineSuccessStatus,
  isSubmarineSwapRefundable,
  isValidArkAddress,
  logger,
  migrateToSwapRepository,
  saveSwap,
  setLogger,
  updateChainSwapStatus,
  updateReverseSwapStatus,
  updateSubmarineSwapStatus,
  verifySignatures
};
