// @arkade-os/sdk/dist/cjs/contracts/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContractManager = exports.ContractWatcher = exports.isArkContract = exports.contractFromArkContractWithAddress = exports.contractFromArkContract = exports.decodeArkContract = exports.encodeArkContract = exports.VHTLCContractHandler = exports.DelegateContractHandler = exports.DefaultContractHandler = exports.contractHandlers = void 0;
// Types
__exportStar(require("./types"), exports);
// Contract handlers
var handlers_1 = require("./handlers");
Object.defineProperty(exports, "contractHandlers", { enumerable: true, get: function () { return handlers_1.contractHandlers; } });
var handlers_2 = require("./handlers");
Object.defineProperty(exports, "DefaultContractHandler", { enumerable: true, get: function () { return handlers_2.DefaultContractHandler; } });
var handlers_3 = require("./handlers");
Object.defineProperty(exports, "DelegateContractHandler", { enumerable: true, get: function () { return handlers_3.DelegateContractHandler; } });
var handlers_4 = require("./handlers");
Object.defineProperty(exports, "VHTLCContractHandler", { enumerable: true, get: function () { return handlers_4.VHTLCContractHandler; } });
// arkcontract string codec
var arkcontract_1 = require("./arkcontract");
Object.defineProperty(exports, "encodeArkContract", { enumerable: true, get: function () { return arkcontract_1.encodeArkContract; } });
Object.defineProperty(exports, "decodeArkContract", { enumerable: true, get: function () { return arkcontract_1.decodeArkContract; } });
Object.defineProperty(exports, "contractFromArkContract", { enumerable: true, get: function () { return arkcontract_1.contractFromArkContract; } });
Object.defineProperty(exports, "contractFromArkContractWithAddress", { enumerable: true, get: function () { return arkcontract_1.contractFromArkContractWithAddress; } });
Object.defineProperty(exports, "isArkContract", { enumerable: true, get: function () { return arkcontract_1.isArkContract; } });
// Contract watcher
var contractWatcher_1 = require("./contractWatcher");
Object.defineProperty(exports, "ContractWatcher", { enumerable: true, get: function () { return contractWatcher_1.ContractWatcher; } });
// Contract manager
var contractManager_1 = require("./contractManager");
Object.defineProperty(exports, "ContractManager", { enumerable: true, get: function () { return contractManager_1.ContractManager; } });
