// @arkade-os/sdk/dist/esm/script/delegate.js
import { DefaultVtxo } from './default.js';
import { MultisigTapscript } from './tapscript.js';
import { VtxoScript } from './base.js';
import { hex } from "@scure/base";
/**
 * DelegateVtxo extends DefaultVtxo with an extra delegator path
 */
export var DelegateVtxo;
(function (DelegateVtxo) {
    /**
     * DelegateVtxo.Script extends DefaultVtxo.Script and adds a delegate path.
     * @example
     * ```typescript
     * const vtxoScript = new DelegateVtxo.Script({
     *     pubKey: new Uint8Array(32),
     *     serverPubKey: new Uint8Array(32),
     *     delegatePubKey: new Uint8Array(32),
     * });
     *
     * console.log("script pub key:", vtxoScript.pkScript)
     * ```
     */
    class Script extends VtxoScript {
        constructor(options) {
            const defaultVtxo = new DefaultVtxo.Script(options);
            const { delegatePubKey, pubKey, serverPubKey } = options;
            const delegateScript = MultisigTapscript.encode({
                pubkeys: [pubKey, delegatePubKey, serverPubKey],
            }).script;
            super([...defaultVtxo.scripts, delegateScript]);
            this.options = options;
            this.defaultVtxo = defaultVtxo;
            this.delegateScript = hex.encode(delegateScript);
        }
        forfeit() {
            return this.findLeaf(this.defaultVtxo.forfeitScript);
        }
        exit() {
            return this.findLeaf(this.defaultVtxo.exitScript);
        }
        delegate() {
            return this.findLeaf(this.delegateScript);
        }
    }
    DelegateVtxo.Script = Script;
})(DelegateVtxo || (DelegateVtxo = {}));
