// @arkade-os/boltz-swap/dist/expo/index.js
import {
  SWAP_POLL_TASK_TYPE,
  defineExpoSwapBackgroundTask,
  registerExpoSwapBackgroundTask,
  swapsPollProcessor,
  unregisterExpoSwapBackgroundTask
} from "../chunk-WNTQXNGX.js";
import {
  ArkadeSwaps
} from "../chunk-XVJ47GNJ.js";
import "../chunk-3RG5ZIWI.js";

// src/expo/arkade-lightning.ts
function getRandomId() {
  return Math.random().toString(36).slice(2) + Date.now().toString(36);
}
var ExpoArkadeSwaps = class _ExpoArkadeSwaps {
  constructor(inner, config) {
    this.inner = inner;
    this.config = config;
    this.taskName = config.background.taskName;
    this.swapRepository = inner.swapRepository;
  }
  swapRepository;
  foregroundIntervalId;
  taskName;
  /**
   * Create an ExpoArkadeSwaps with background task support.
   *
   * 1. Creates the inner {@link ArkadeSwaps} with SwapManager enabled.
   * 2. Persists {@link PersistedSwapBackgroundConfig} for background rehydration.
   * 3. Seeds the task queue with a swap-poll task.
   * 4. Registers the background task with the OS scheduler (if configured).
   * 5. Starts foreground interval (if configured).
   */
  static async setup(config) {
    const inner = new ArkadeSwaps({
      ...config,
      swapManager: config.swapManager ?? true
    });
    const { taskQueue } = config.background;
    const derivedArkServerUrl = inner.arkProvider.serverUrl;
    const arkServerUrl = config.arkServerUrl ?? derivedArkServerUrl;
    if (!arkServerUrl) {
      throw new Error(
        "Ark server URL is required for Expo background rehydration. Pass `arkServerUrl` to ExpoArkadeSwaps.setup()."
      );
    }
    const bgConfig = {
      boltzApiUrl: config.swapProvider.getApiUrl(),
      arkServerUrl,
      network: config.swapProvider.getNetwork()
    };
    await taskQueue.persistConfig(bgConfig);
    const instance = new _ExpoArkadeSwaps(inner, config);
    await instance.seedSwapPollTask();
    if (config.background.minimumBackgroundInterval) {
      try {
        const { registerExpoSwapBackgroundTask: registerExpoSwapBackgroundTask2 } = await import("../background-OI6ONB3C.js");
        await registerExpoSwapBackgroundTask2(
          config.background.taskName,
          {
            minimumInterval: config.background.minimumBackgroundInterval
          }
        );
      } catch (err) {
        const message = err instanceof Error ? err.message : String(err);
        const code = typeof err === "object" && err !== null && "code" in err ? err.code : void 0;
        const codeString = typeof code === "string" ? code : void 0;
        const isModuleNotFound = codeString === "MODULE_NOT_FOUND" || /cannot find module/i.test(message) || /module not found/i.test(message);
        if (!isModuleNotFound) {
          console.warn(
            `[boltz-swap] Failed to register background task "${config.background.taskName}":`,
            err
          );
        }
      }
    }
    if (config.background.foregroundIntervalMs && config.background.foregroundIntervalMs > 0) {
      instance.startForegroundPolling(
        config.background.foregroundIntervalMs
      );
    }
    return instance;
  }
  // ── Foreground polling ───────────────────────────────────────────
  startForegroundPolling(intervalMs) {
    this.foregroundIntervalId = setInterval(() => {
      this.runForegroundPoll().catch(console.error);
    }, intervalMs);
  }
  async runForegroundPoll() {
    const { taskQueue } = this.config.background;
    const results = await taskQueue.getResults();
    if (results.length > 0) {
      await taskQueue.acknowledgeResults(
        results.map((r) => r.id)
      );
    }
    await this.seedSwapPollTask();
  }
  async seedSwapPollTask() {
    const { taskQueue } = this.config.background;
    const existing = await taskQueue.getTasks(SWAP_POLL_TASK_TYPE);
    if (existing.length > 0) return;
    const task = {
      id: getRandomId(),
      type: SWAP_POLL_TASK_TYPE,
      data: {},
      createdAt: Date.now()
    };
    await taskQueue.addTask(task);
  }
  // ── Lifecycle ────────────────────────────────────────────────────
  async dispose() {
    if (this.foregroundIntervalId) {
      clearInterval(this.foregroundIntervalId);
      this.foregroundIntervalId = void 0;
    }
    await this.inner.dispose();
    try {
      const { unregisterExpoSwapBackgroundTask: unregisterExpoSwapBackgroundTask2 } = await import("../background-OI6ONB3C.js");
      await unregisterExpoSwapBackgroundTask2(this.taskName);
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      const code = typeof err === "object" && err !== null && "code" in err ? err.code : void 0;
      const codeString = typeof code === "string" ? code : void 0;
      const isModuleNotFound = codeString === "MODULE_NOT_FOUND" || /cannot find module/i.test(message) || /module not found/i.test(message);
      if (!isModuleNotFound) {
        console.warn(
          `[boltz-swap] Failed to unregister background task "${this.taskName}":`,
          err
        );
      }
    }
  }
  async [Symbol.asyncDispose]() {
    await this.dispose();
  }
  // ── IArkadeSwaps delegation ──────────────────────────────────────
  startSwapManager() {
    return this.inner.startSwapManager();
  }
  stopSwapManager() {
    return this.inner.stopSwapManager();
  }
  getSwapManager() {
    return this.inner.getSwapManager();
  }
  createLightningInvoice(args) {
    return this.inner.createLightningInvoice(args);
  }
  sendLightningPayment(args) {
    return this.inner.sendLightningPayment(args);
  }
  createSubmarineSwap(args) {
    return this.inner.createSubmarineSwap(args);
  }
  createReverseSwap(args) {
    return this.inner.createReverseSwap(args);
  }
  claimVHTLC(pendingSwap) {
    return this.inner.claimVHTLC(pendingSwap);
  }
  refundVHTLC(pendingSwap) {
    return this.inner.refundVHTLC(pendingSwap);
  }
  waitAndClaim(pendingSwap) {
    return this.inner.waitAndClaim(pendingSwap);
  }
  waitForSwapSettlement(pendingSwap) {
    return this.inner.waitForSwapSettlement(pendingSwap);
  }
  restoreSwaps(boltzFees) {
    return this.inner.restoreSwaps(boltzFees);
  }
  enrichReverseSwapPreimage(swap, preimage) {
    return this.inner.enrichReverseSwapPreimage(swap, preimage);
  }
  enrichSubmarineSwapInvoice(swap, invoice) {
    return this.inner.enrichSubmarineSwapInvoice(swap, invoice);
  }
  // ── Chain swap delegation ────────────────────────────────────────
  arkToBtc(args) {
    return this.inner.arkToBtc(args);
  }
  waitAndClaimBtc(pendingSwap) {
    return this.inner.waitAndClaimBtc(pendingSwap);
  }
  claimBtc(pendingSwap) {
    return this.inner.claimBtc(pendingSwap);
  }
  refundArk(pendingSwap) {
    return this.inner.refundArk(pendingSwap);
  }
  btcToArk(args) {
    return this.inner.btcToArk(args);
  }
  waitAndClaimArk(pendingSwap) {
    return this.inner.waitAndClaimArk(pendingSwap);
  }
  claimArk(pendingSwap) {
    return this.inner.claimArk(pendingSwap);
  }
  signCooperativeClaimForServer(pendingSwap) {
    return this.inner.signCooperativeClaimForServer(pendingSwap);
  }
  waitAndClaimChain(pendingSwap) {
    return this.inner.waitAndClaimChain(pendingSwap);
  }
  createChainSwap(args) {
    return this.inner.createChainSwap(args);
  }
  verifyChainSwap(args) {
    return this.inner.verifyChainSwap(args);
  }
  quoteSwap(swapId) {
    return this.inner.quoteSwap(swapId);
  }
  joinBatch(identity, input, output, arkInfo, isRecoverable) {
    return this.inner.joinBatch(
      identity,
      input,
      output,
      arkInfo,
      isRecoverable
    );
  }
  createVHTLCScript(params) {
    return this.inner.createVHTLCScript(params);
  }
  getFees(from, to) {
    if (from !== void 0 && to !== void 0) {
      return this.inner.getFees(from, to);
    }
    return this.inner.getFees();
  }
  getLimits(from, to) {
    if (from !== void 0 && to !== void 0) {
      return this.inner.getLimits(from, to);
    }
    return this.inner.getLimits();
  }
  getSwapStatus(swapId) {
    return this.inner.getSwapStatus(swapId);
  }
  getPendingSubmarineSwaps() {
    return this.inner.getPendingSubmarineSwaps();
  }
  getPendingReverseSwaps() {
    return this.inner.getPendingReverseSwaps();
  }
  getPendingChainSwaps() {
    return this.inner.getPendingChainSwaps();
  }
  getSwapHistory() {
    return this.inner.getSwapHistory();
  }
  refreshSwapsStatus() {
    return this.inner.refreshSwapsStatus();
  }
};
var ExpoArkadeLightning = ExpoArkadeSwaps;
export {
  ExpoArkadeLightning,
  ExpoArkadeSwaps,
  SWAP_POLL_TASK_TYPE,
  defineExpoSwapBackgroundTask,
  registerExpoSwapBackgroundTask,
  swapsPollProcessor,
  unregisterExpoSwapBackgroundTask
};
