// @arkade-os/sdk/dist/esm/wallet/utils.js
import { ArkAddress, } from '../index.js';
import { contractHandlers } from '../contracts/handlers/index.js';
import { hex } from "@scure/base";
export const DUST_AMOUNT = 546; // sats
export function extendVirtualCoin(wallet, vtxo) {
    return {
        ...vtxo,
        forfeitTapLeafScript: wallet.offchainTapscript.forfeit(),
        intentTapLeafScript: wallet.offchainTapscript.forfeit(),
        tapTree: wallet.offchainTapscript.encode(),
    };
}
export function extendCoin(wallet, utxo) {
    return {
        ...utxo,
        forfeitTapLeafScript: wallet.boardingTapscript.forfeit(),
        intentTapLeafScript: wallet.boardingTapscript.forfeit(),
        tapTree: wallet.boardingTapscript.encode(),
    };
}
export function extendVtxoFromContract(vtxo, contract) {
    const handler = contractHandlers.get(contract.type);
    if (!handler) {
        throw new Error(`No handler for contract type '${contract.type}'`);
    }
    const script = handler.createScript(contract.params);
    return {
        ...vtxo,
        forfeitTapLeafScript: script.forfeit(),
        intentTapLeafScript: script.forfeit(),
        tapTree: script.encode(),
    };
}
export function getRandomId() {
    const randomValue = crypto.getRandomValues(new Uint8Array(16));
    return hex.encode(randomValue);
}
export function isValidArkAddress(address) {
    try {
        ArkAddress.decode(address);
        return true;
    }
    catch (e) {
        return false;
    }
}
export function validateRecipients(recipients, dustAmount) {
    const validatedRecipients = [];
    for (const recipient of recipients) {
        let address;
        try {
            address = ArkAddress.decode(recipient.address);
        }
        catch (e) {
            throw new Error(`Invalid Ark address: ${recipient.address}`);
        }
        const amount = recipient.amount || dustAmount;
        if (amount <= 0) {
            throw new Error("Amount must be positive");
        }
        validatedRecipients.push({
            address: recipient.address,
            assets: recipient.assets ?? [],
            amount,
            script: amount < dustAmount
                ? address.subdustPkScript
                : address.pkScript,
        });
    }
    return validatedRecipients;
}
