// @arkade-os/sdk/dist/esm/wallet/wallet.js
import { base64, hex } from "@scure/base";
import { tapLeafHash } from "@scure/btc-signer/payment.js";
import { Address, OutScript, SigHash, Transaction } from "@scure/btc-signer";
import { sha256 } from "@scure/btc-signer/utils.js";
import { ArkAddress } from '../script/address.js';
import { DefaultVtxo } from '../script/default.js';
import { getNetwork } from '../networks.js';
import { ESPLORA_URL, EsploraProvider, } from '../providers/onchain.js';
import { RestArkProvider, } from '../providers/ark.js';
import { buildForfeitTx } from '../forfeit.js';
import { validateConnectorsTxGraph, validateVtxoTxGraph, } from '../tree/validation.js';
import { validateBatchRecipients } from './validation.js';
import { isExpired, isRecoverable, isSpendable, isSubdust, TxType, } from './index.js';
import { createAssetPacket, selectedCoinsToAssetInputs, selectCoinsWithAsset, } from './asset.js';
import { VtxoScript } from '../script/base.js';
import { CSVMultisigTapscript } from '../script/tapscript.js';
import { buildOffchainTx, hasBoardingTxExpired, isValidArkAddress, } from '../utils/arkTransaction.js';
import { DEFAULT_RENEWAL_CONFIG } from './vtxo-manager.js';
import { ArkNote } from '../arknote/index.js';
import { Intent } from '../intent/index.js';
import { RestIndexerProvider } from '../providers/indexer.js';
import { extendCoin, extendVirtualCoin, validateRecipients } from './utils.js';
import { ArkError } from '../providers/errors.js';
import { Batch } from './batch.js';
import { Estimator } from '../arkfee/index.js';
import { buildTransactionHistory } from '../utils/transactionHistory.js';
import { AssetManager, ReadonlyAssetManager } from './asset-manager.js';
import { Extension } from '../extension/index.js';
import { DelegateVtxo } from '../script/delegate.js';
import { DelegatorManagerImpl } from './delegator.js';
import { IndexedDBContractRepository, IndexedDBWalletRepository, } from '../repositories/index.js';
import { ContractManager } from '../contracts/contractManager.js';
import { contractHandlers } from '../contracts/handlers/index.js';
import { timelockToSequence } from '../contracts/handlers/helpers.js';
/**
 * Type guard function to check if an identity has a toReadonly method.
 */
function hasToReadonly(identity) {
    return (typeof identity === "object" &&
        identity !== null &&
        "toReadonly" in identity &&
        typeof identity.toReadonly === "function");
}
export class ReadonlyWallet {
    get assetManager() {
        return this._assetManager;
    }
    constructor(identity, network, onchainProvider, indexerProvider, arkServerPublicKey, offchainTapscript, boardingTapscript, dustAmount, walletRepository, contractRepository, delegatorProvider, watcherConfig) {
        this.identity = identity;
        this.network = network;
        this.onchainProvider = onchainProvider;
        this.indexerProvider = indexerProvider;
        this.arkServerPublicKey = arkServerPublicKey;
        this.offchainTapscript = offchainTapscript;
        this.boardingTapscript = boardingTapscript;
        this.dustAmount = dustAmount;
        this.walletRepository = walletRepository;
        this.contractRepository = contractRepository;
        this.delegatorProvider = delegatorProvider;
        this.watcherConfig = watcherConfig;
        this._assetManager = new ReadonlyAssetManager(this.indexerProvider);
    }
    /**
     * Protected helper to set up shared wallet configuration.
     * Extracts common logic used by both ReadonlyWallet.create() and Wallet.create().
     */
    static async setupWalletConfig(config, pubKey) {
        // Use provided arkProvider instance or create a new one from arkServerUrl
        const arkProvider = config.arkProvider ||
            (() => {
                if (!config.arkServerUrl) {
                    throw new Error("Either arkProvider or arkServerUrl must be provided");
                }
                return new RestArkProvider(config.arkServerUrl);
            })();
        // Extract arkServerUrl from provider if not explicitly provided
        const arkServerUrl = config.arkServerUrl || arkProvider.serverUrl;
        if (!arkServerUrl) {
            throw new Error("Could not determine arkServerUrl from provider");
        }
        // Use provided indexerProvider instance or create a new one
        // indexerUrl defaults to arkServerUrl if not provided
        const indexerUrl = config.indexerUrl || arkServerUrl;
        const indexerProvider = config.indexerProvider || new RestIndexerProvider(indexerUrl);
        const info = await arkProvider.getInfo();
        const network = getNetwork(info.network);
        // Extract esploraUrl from provider if not explicitly provided
        const esploraUrl = config.esploraUrl || ESPLORA_URL[info.network];
        // Use provided onchainProvider instance or create a new one
        const onchainProvider = config.onchainProvider || new EsploraProvider(esploraUrl);
        // validate unilateral exit timelock passed in config if any
        if (config.exitTimelock) {
            const { value, type } = config.exitTimelock;
            if ((value < 512n && type !== "blocks") ||
                (value >= 512n && type !== "seconds")) {
                throw new Error("invalid exitTimelock");
            }
        }
        // create unilateral exit timelock
        const exitTimelock = config.exitTimelock ?? {
            value: info.unilateralExitDelay,
            type: info.unilateralExitDelay < 512n ? "blocks" : "seconds",
        };
        // validate boarding timelock passed in config if any
        if (config.boardingTimelock) {
            const { value, type } = config.boardingTimelock;
            if ((value < 512n && type !== "blocks") ||
                (value >= 512n && type !== "seconds")) {
                throw new Error("invalid boardingTimelock");
            }
        }
        // create boarding timelock
        const boardingTimelock = config.boardingTimelock ?? {
            value: info.boardingExitDelay,
            type: info.boardingExitDelay < 512n ? "blocks" : "seconds",
        };
        // Generate tapscripts for offchain and boarding address
        const serverPubKey = hex.decode(info.signerPubkey).slice(1);
        const delegatePubKey = config.delegatorProvider
            ? await config.delegatorProvider
                .getDelegateInfo()
                .then((info) => hex.decode(info.pubkey).slice(1))
            : undefined;
        const offchainOptions = {
            pubKey,
            serverPubKey,
            csvTimelock: exitTimelock,
        };
        const offchainTapscript = !delegatePubKey
            ? new DefaultVtxo.Script(offchainOptions)
            : new DelegateVtxo.Script({ ...offchainOptions, delegatePubKey });
        const boardingTapscript = new DefaultVtxo.Script({
            ...offchainOptions,
            csvTimelock: boardingTimelock,
        });
        const walletRepository = config.storage?.walletRepository ?? new IndexedDBWalletRepository();
        const contractRepository = config.storage?.contractRepository ??
            new IndexedDBContractRepository();
        return {
            arkProvider,
            indexerProvider,
            onchainProvider,
            network,
            networkName: info.network,
            serverPubKey,
            offchainTapscript,
            boardingTapscript,
            dustAmount: info.dust,
            walletRepository,
            contractRepository,
            info,
            delegatorProvider: config.delegatorProvider,
        };
    }
    static async create(config) {
        const pubkey = await config.identity.xOnlyPublicKey();
        if (!pubkey) {
            throw new Error("Invalid configured public key");
        }
        const setup = await ReadonlyWallet.setupWalletConfig(config, pubkey);
        return new ReadonlyWallet(config.identity, setup.network, setup.onchainProvider, setup.indexerProvider, setup.serverPubKey, setup.offchainTapscript, setup.boardingTapscript, setup.dustAmount, setup.walletRepository, setup.contractRepository, setup.delegatorProvider, config.watcherConfig);
    }
    get arkAddress() {
        return this.offchainTapscript.address(this.network.hrp, this.arkServerPublicKey);
    }
    /**
     * Get the contract script for the wallet's default address.
     * This is the pkScript hex, used to identify the wallet in ContractManager.
     */
    get defaultContractScript() {
        return hex.encode(this.offchainTapscript.pkScript);
    }
    async getAddress() {
        return this.arkAddress.encode();
    }
    async getBoardingAddress() {
        return this.boardingTapscript.onchainAddress(this.network);
    }
    async getBalance() {
        const [boardingUtxos, vtxos] = await Promise.all([
            this.getBoardingUtxos(),
            this.getVtxos(),
        ]);
        // boarding
        let confirmed = 0;
        let unconfirmed = 0;
        for (const utxo of boardingUtxos) {
            if (utxo.status.confirmed) {
                confirmed += utxo.value;
            }
            else {
                unconfirmed += utxo.value;
            }
        }
        // offchain
        let settled = 0;
        let preconfirmed = 0;
        let recoverable = 0;
        settled = vtxos
            .filter((coin) => coin.virtualStatus.state === "settled")
            .reduce((sum, coin) => sum + coin.value, 0);
        preconfirmed = vtxos
            .filter((coin) => coin.virtualStatus.state === "preconfirmed")
            .reduce((sum, coin) => sum + coin.value, 0);
        recoverable = vtxos
            .filter((coin) => isSpendable(coin) && coin.virtualStatus.state === "swept")
            .reduce((sum, coin) => sum + coin.value, 0);
        const totalBoarding = confirmed + unconfirmed;
        const totalOffchain = settled + preconfirmed + recoverable;
        // aggregate asset balances from spendable vtxos
        const assetBalances = new Map();
        for (const vtxo of vtxos) {
            if (!isSpendable(vtxo))
                continue;
            if (vtxo.assets) {
                for (const a of vtxo.assets) {
                    const current = assetBalances.get(a.assetId) ?? 0;
                    assetBalances.set(a.assetId, current + a.amount);
                }
            }
        }
        const assets = Array.from(assetBalances.entries()).map(([assetId, amount]) => ({
            assetId,
            amount,
        }));
        return {
            boarding: {
                confirmed,
                unconfirmed,
                total: totalBoarding,
            },
            settled,
            preconfirmed,
            available: settled + preconfirmed,
            recoverable,
            total: totalBoarding + totalOffchain,
            assets,
        };
    }
    async getVtxos(filter) {
        const address = await this.getAddress();
        const scriptMap = await this.getScriptMap();
        const f = filter ?? { withRecoverable: true, withUnrolled: false };
        const allExtended = [];
        // Query each script separately so we can extend VTXOs with the correct tapscript
        for (const [scriptHex, vtxoScript] of scriptMap) {
            const response = await this.indexerProvider.getVtxos({
                scripts: [scriptHex],
            });
            let vtxos = response.vtxos.filter(isSpendable);
            if (!f.withRecoverable) {
                vtxos = vtxos.filter((vtxo) => !isRecoverable(vtxo) && !isExpired(vtxo));
            }
            if (f.withUnrolled) {
                const spentVtxos = response.vtxos.filter((vtxo) => !isSpendable(vtxo));
                vtxos.push(...spentVtxos.filter((vtxo) => vtxo.isUnrolled));
            }
            for (const vtxo of vtxos) {
                allExtended.push({
                    ...vtxo,
                    forfeitTapLeafScript: vtxoScript.forfeit(),
                    intentTapLeafScript: vtxoScript.forfeit(),
                    tapTree: vtxoScript.encode(),
                });
            }
        }
        // Update cache with fresh data
        await this.walletRepository.saveVtxos(address, allExtended);
        return allExtended;
    }
    async getTransactionHistory() {
        const scripts = await this.getWalletScripts();
        const response = await this.indexerProvider.getVtxos({ scripts });
        const { boardingTxs, commitmentsToIgnore } = await this.getBoardingTxs();
        const getTxCreatedAt = (txid) => this.indexerProvider
            .getVtxos({ outpoints: [{ txid, vout: 0 }] })
            .then((res) => res.vtxos[0]?.createdAt.getTime() || 0);
        return buildTransactionHistory(response.vtxos, boardingTxs, commitmentsToIgnore, getTxCreatedAt);
    }
    async getBoardingTxs() {
        const utxos = [];
        const commitmentsToIgnore = new Set();
        const boardingAddress = await this.getBoardingAddress();
        const txs = await this.onchainProvider.getTransactions(boardingAddress);
        for (const tx of txs) {
            for (let i = 0; i < tx.vout.length; i++) {
                const vout = tx.vout[i];
                if (vout.scriptpubkey_address === boardingAddress) {
                    const spentStatuses = await this.onchainProvider.getTxOutspends(tx.txid);
                    const spentStatus = spentStatuses[i];
                    if (spentStatus?.spent) {
                        commitmentsToIgnore.add(spentStatus.txid);
                    }
                    utxos.push({
                        txid: tx.txid,
                        vout: i,
                        value: Number(vout.value),
                        status: {
                            confirmed: tx.status.confirmed,
                            block_time: tx.status.block_time,
                        },
                        isUnrolled: true,
                        virtualStatus: {
                            state: spentStatus?.spent ? "spent" : "settled",
                            commitmentTxIds: spentStatus?.spent
                                ? [spentStatus.txid]
                                : undefined,
                        },
                        createdAt: tx.status.confirmed
                            ? new Date(tx.status.block_time * 1000)
                            : new Date(0),
                    });
                }
            }
        }
        const unconfirmedTxs = [];
        const confirmedTxs = [];
        for (const utxo of utxos) {
            const tx = {
                key: {
                    boardingTxid: utxo.txid,
                    commitmentTxid: "",
                    arkTxid: "",
                },
                amount: utxo.value,
                type: TxType.TxReceived,
                settled: utxo.virtualStatus.state === "spent",
                createdAt: utxo.status.block_time
                    ? new Date(utxo.status.block_time * 1000).getTime()
                    : 0,
            };
            if (!utxo.status.block_time) {
                unconfirmedTxs.push(tx);
            }
            else {
                confirmedTxs.push(tx);
            }
        }
        return {
            boardingTxs: [...unconfirmedTxs, ...confirmedTxs],
            commitmentsToIgnore,
        };
    }
    async getBoardingUtxos() {
        const boardingAddress = await this.getBoardingAddress();
        const boardingUtxos = await this.onchainProvider.getCoins(boardingAddress);
        const utxos = boardingUtxos.map((utxo) => {
            return extendCoin(this, utxo);
        });
        // Save boardingUtxos using unified repository
        await this.walletRepository.saveUtxos(boardingAddress, utxos);
        return utxos;
    }
    async notifyIncomingFunds(eventCallback) {
        const arkAddress = await this.getAddress();
        const boardingAddress = await this.getBoardingAddress();
        let onchainStopFunc;
        let indexerStopFunc;
        if (this.onchainProvider && boardingAddress) {
            const findVoutOnTx = (tx) => {
                return tx.vout.findIndex((v) => v.scriptpubkey_address === boardingAddress);
            };
            onchainStopFunc = await this.onchainProvider.watchAddresses([boardingAddress], (txs) => {
                // find all utxos belonging to our boarding address
                const coins = txs
                    // filter txs where address is in output
                    .filter((tx) => findVoutOnTx(tx) !== -1)
                    // return utxo as Coin
                    .map((tx) => {
                    const { txid, status } = tx;
                    const vout = findVoutOnTx(tx);
                    const value = Number(tx.vout[vout].value);
                    return { txid, vout, value, status };
                });
                // and notify via callback
                eventCallback({
                    type: "utxo",
                    coins,
                });
            });
        }
        if (this.indexerProvider && arkAddress) {
            const walletScripts = await this.getWalletScripts();
            const subscriptionId = await this.indexerProvider.subscribeForScripts(walletScripts);
            const abortController = new AbortController();
            const subscription = this.indexerProvider.getSubscription(subscriptionId, abortController.signal);
            indexerStopFunc = async () => {
                abortController.abort();
                await this.indexerProvider?.unsubscribeForScripts(subscriptionId);
            };
            // Handle subscription updates asynchronously without blocking.
            // Note: subscription covers all wallet scripts (default + delegate),
            // but we can't determine which script each VTXO belongs to from the
            // subscription event. VTXOs are extended with the current offchainTapscript;
            // this is for notification/display only — not for spending.
            // For correct extension metadata, use getVtxos() which queries per-script.
            (async () => {
                try {
                    for await (const update of subscription) {
                        if (update.newVtxos?.length > 0 ||
                            update.spentVtxos?.length > 0) {
                            eventCallback({
                                type: "vtxo",
                                newVtxos: update.newVtxos.map((vtxo) => extendVirtualCoin(this, vtxo)),
                                spentVtxos: update.spentVtxos.map((vtxo) => extendVirtualCoin(this, vtxo)),
                            });
                        }
                    }
                }
                catch (error) {
                    console.error("Subscription error:", error);
                }
            })();
        }
        const stopFunc = () => {
            onchainStopFunc?.();
            indexerStopFunc?.();
        };
        return stopFunc;
    }
    async fetchPendingTxs() {
        // get non-swept VTXOs, rely on the indexer only in case DB doesn't have the right state
        const scripts = await this.getWalletScripts();
        let { vtxos } = await this.indexerProvider.getVtxos({
            scripts,
        });
        return vtxos
            .filter((vtxo) => vtxo.virtualStatus.state !== "swept" &&
            vtxo.virtualStatus.state !== "settled" &&
            vtxo.arkTxId !== undefined)
            .map((_) => _.arkTxId);
    }
    // ========================================================================
    // Multi-script support (default + delegate addresses)
    // ========================================================================
    /**
     * Get all pkScript hex strings for the wallet's own addresses
     * (both delegate and non-delegate, current and historical).
     * Falls back to only the current script if ContractManager is not yet initialized.
     */
    async getWalletScripts() {
        // Only use the contract manager if it's already initialized or
        // currently initializing — never trigger initialization here to
        // avoid blocking callers that don't need it.
        if (this._contractManager || this._contractManagerInitializing) {
            try {
                const manager = await this.getContractManager();
                const contracts = await manager.getContracts({
                    type: ["default", "delegate"],
                });
                if (contracts.length > 0) {
                    return contracts.map((c) => c.script);
                }
            }
            catch {
                // fall through to current script only
            }
        }
        return [hex.encode(this.offchainTapscript.pkScript)];
    }
    /**
     * Build a map of scriptHex → VtxoScript for all wallet contracts,
     * so VTXOs can be extended with the correct tapscript per contract.
     */
    async getScriptMap() {
        const map = new Map();
        // Always include the current script
        const currentScriptHex = hex.encode(this.offchainTapscript.pkScript);
        map.set(currentScriptHex, this.offchainTapscript);
        if (this._contractManager) {
            try {
                const contracts = await this._contractManager.getContracts({
                    type: ["default", "delegate"],
                });
                for (const contract of contracts) {
                    if (map.has(contract.script))
                        continue;
                    const handler = contractHandlers.get(contract.type);
                    if (handler) {
                        const script = handler.createScript(contract.params);
                        map.set(contract.script, script);
                    }
                }
            }
            catch {
                // ContractManager error — only current script in map
            }
        }
        return map;
    }
    // ========================================================================
    // Contract Management
    // ========================================================================
    /**
     * Get the ContractManager for managing contracts including the wallet's default address.
     *
     * The ContractManager handles:
     * - The wallet's default receiving address (as a "default" contract)
     * - External contracts (Boltz swaps, HTLCs, etc.)
     * - Multi-contract watching with resilient connections
     *
     * @example
     * ```typescript
     * const manager = await wallet.getContractManager();
     *
     * // Create a contract for a Boltz swap
     * const contract = await manager.createContract({
     *   label: "Boltz Swap",
     *   type: "vhtlc",
     *   params: { ... },
     *   script: swapScript,
     *   address: swapAddress,
     * });
     *
     * // Start watching for events (includes wallet's default address)
     * const stop = await manager.onContractEvent((event) => {
     *   console.log(`${event.type} on ${event.contractScript}`);
     * });
     * ```
     */
    async getContractManager() {
        // Return existing manager if already initialized
        if (this._contractManager) {
            return this._contractManager;
        }
        // If initialization is in progress, wait for it
        if (this._contractManagerInitializing) {
            return this._contractManagerInitializing;
        }
        // Start initialization and store the promise
        this._contractManagerInitializing = this.initializeContractManager();
        try {
            const manager = await this._contractManagerInitializing;
            this._contractManager = manager;
            return manager;
        }
        catch (error) {
            // Clear the initializing promise so subsequent calls can retry
            this._contractManagerInitializing = undefined;
            throw error;
        }
        finally {
            // Clear the initializing promise after completion
            this._contractManagerInitializing = undefined;
        }
    }
    async initializeContractManager() {
        const manager = await ContractManager.create({
            indexerProvider: this.indexerProvider,
            contractRepository: this.contractRepository,
            walletRepository: this.walletRepository,
            getDefaultAddress: () => this.getAddress(),
            watcherConfig: this.watcherConfig,
        });
        // Register the wallet's current address as a contract
        const csvTimelock = this.offchainTapscript.options.csvTimelock ??
            DefaultVtxo.Script.DEFAULT_TIMELOCK;
        const csvTimelockStr = timelockToSequence(csvTimelock).toString();
        const isDelegateScript = this.offchainTapscript instanceof DelegateVtxo.Script;
        if (isDelegateScript) {
            const delegateScript = this
                .offchainTapscript;
            // Register the delegate contract (current address)
            await manager.createContract({
                type: "delegate",
                params: {
                    pubKey: hex.encode(delegateScript.options.pubKey),
                    serverPubKey: hex.encode(delegateScript.options.serverPubKey),
                    delegatePubKey: hex.encode(delegateScript.options.delegatePubKey),
                    csvTimelock: csvTimelockStr,
                },
                script: this.defaultContractScript,
                address: await this.getAddress(),
                state: "active",
            });
            // Also register the non-delegate version so old VTXOs remain visible
            const nonDelegateScript = new DefaultVtxo.Script({
                pubKey: delegateScript.options.pubKey,
                serverPubKey: delegateScript.options.serverPubKey,
                csvTimelock,
            });
            await manager.createContract({
                type: "default",
                params: {
                    pubKey: hex.encode(delegateScript.options.pubKey),
                    serverPubKey: hex.encode(delegateScript.options.serverPubKey),
                    csvTimelock: csvTimelockStr,
                },
                script: hex.encode(nonDelegateScript.pkScript),
                address: nonDelegateScript
                    .address(this.network.hrp, this.arkServerPublicKey)
                    .encode(),
                state: "active",
            });
        }
        else {
            // Register the default contract (current address)
            await manager.createContract({
                type: "default",
                params: {
                    pubKey: hex.encode(this.offchainTapscript.options.pubKey),
                    serverPubKey: hex.encode(this.offchainTapscript.options.serverPubKey),
                    csvTimelock: csvTimelockStr,
                },
                script: this.defaultContractScript,
                address: await this.getAddress(),
                state: "active",
            });
            // Any old "delegate" contract from a prior wallet incarnation
            // is already loaded by ContractManager.initialize() from ContractRepository
        }
        return manager;
    }
}
/**
 * Main wallet implementation for Bitcoin transactions with Ark protocol support.
 * The wallet does not store any data locally and relies on Ark and onchain
 * providers to fetch UTXOs and VTXOs.
 *
 * @example
 * ```typescript
 * // Create a wallet with URL configuration
 * const wallet = await Wallet.create({
 *   identity: SingleKey.fromHex('your_private_key'),
 *   arkServerUrl: 'https://ark.example.com',
 *   esploraUrl: 'https://mempool.space/api'
 * });
 *
 * // Or with custom provider instances (e.g., for Expo/React Native)
 * const wallet = await Wallet.create({
 *   identity: SingleKey.fromHex('your_private_key'),
 *   arkProvider: new ExpoArkProvider('https://ark.example.com'),
 *   indexerProvider: new ExpoIndexerProvider('https://ark.example.com'),
 *   esploraUrl: 'https://mempool.space/api'
 * });
 *
 * // Get addresses
 * const arkAddress = await wallet.getAddress();
 * const boardingAddress = await wallet.getBoardingAddress();
 *
 * // Send bitcoin
 * const txid = await wallet.sendBitcoin({
 *   address: 'tb1...',
 *   amount: 50000
 * });
 * ```
 */
export class Wallet extends ReadonlyWallet {
    constructor(identity, network, networkName, onchainProvider, arkProvider, indexerProvider, arkServerPublicKey, offchainTapscript, boardingTapscript, serverUnrollScript, forfeitOutputScript, forfeitPubkey, dustAmount, walletRepository, contractRepository, renewalConfig, delegatorProvider, watcherConfig) {
        super(identity, network, onchainProvider, indexerProvider, arkServerPublicKey, offchainTapscript, boardingTapscript, dustAmount, walletRepository, contractRepository, delegatorProvider, watcherConfig);
        this.networkName = networkName;
        this.arkProvider = arkProvider;
        this.serverUnrollScript = serverUnrollScript;
        this.forfeitOutputScript = forfeitOutputScript;
        this.forfeitPubkey = forfeitPubkey;
        this.identity = identity;
        this.renewalConfig = {
            enabled: renewalConfig?.enabled ?? false,
            ...DEFAULT_RENEWAL_CONFIG,
            ...renewalConfig,
        };
        this._delegatorManager = delegatorProvider
            ? new DelegatorManagerImpl(delegatorProvider, arkProvider, identity)
            : undefined;
    }
    get assetManager() {
        this._walletAssetManager ?? (this._walletAssetManager = new AssetManager(this));
        return this._walletAssetManager;
    }
    static async create(config) {
        const pubkey = await config.identity.xOnlyPublicKey();
        if (!pubkey) {
            throw new Error("Invalid configured public key");
        }
        const setup = await ReadonlyWallet.setupWalletConfig(config, pubkey);
        // Compute Wallet-specific forfeit and unroll scripts
        // the serverUnrollScript is the one used to create output scripts of the checkpoint transactions
        let serverUnrollScript;
        try {
            const raw = hex.decode(setup.info.checkpointTapscript);
            serverUnrollScript = CSVMultisigTapscript.decode(raw);
        }
        catch (e) {
            throw new Error("Invalid checkpointTapscript from server");
        }
        // parse the server forfeit address
        // server is expecting funds to be sent to this address
        const forfeitPubkey = hex.decode(setup.info.forfeitPubkey).slice(1);
        const forfeitAddress = Address(setup.network).decode(setup.info.forfeitAddress);
        const forfeitOutputScript = OutScript.encode(forfeitAddress);
        return new Wallet(config.identity, setup.network, setup.networkName, setup.onchainProvider, setup.arkProvider, setup.indexerProvider, setup.serverPubKey, setup.offchainTapscript, setup.boardingTapscript, serverUnrollScript, forfeitOutputScript, forfeitPubkey, setup.dustAmount, setup.walletRepository, setup.contractRepository, config.renewalConfig, config.delegatorProvider, config.watcherConfig);
    }
    /**
     * Convert this wallet to a readonly wallet.
     *
     * @returns A readonly wallet with the same configuration but readonly identity
     * @example
     * ```typescript
     * const wallet = await Wallet.create({ identity: SingleKey.fromHex('...'), ... });
     * const readonlyWallet = await wallet.toReadonly();
     *
     * // Can query balance and addresses
     * const balance = await readonlyWallet.getBalance();
     * const address = await readonlyWallet.getAddress();
     *
     * // But cannot send transactions (type error)
     * // readonlyWallet.sendBitcoin(...); // TypeScript error
     * ```
     */
    async toReadonly() {
        // Check if the identity has a toReadonly method using type guard
        const readonlyIdentity = hasToReadonly(this.identity)
            ? await this.identity.toReadonly()
            : this.identity; // Identity extends ReadonlyIdentity, so this is safe
        return new ReadonlyWallet(readonlyIdentity, this.network, this.onchainProvider, this.indexerProvider, this.arkServerPublicKey, this.offchainTapscript, this.boardingTapscript, this.dustAmount, this.walletRepository, this.contractRepository, this.delegatorProvider, this.watcherConfig);
    }
    async getDelegatorManager() {
        return this._delegatorManager;
    }
    /**
     * @deprecated Use `send`
     * @param params
     */
    async sendBitcoin(params) {
        if (params.amount <= 0) {
            throw new Error("Amount must be positive");
        }
        if (!isValidArkAddress(params.address)) {
            throw new Error("Invalid Ark address " + params.address);
        }
        if (params.selectedVtxos && params.selectedVtxos.length > 0) {
            const selectedVtxoSum = params.selectedVtxos
                .map((v) => v.value)
                .reduce((a, b) => a + b, 0);
            if (selectedVtxoSum < params.amount) {
                throw new Error("Selected VTXOs do not cover specified amount");
            }
            const changeAmount = selectedVtxoSum - params.amount;
            const selected = {
                inputs: params.selectedVtxos,
                changeAmount: BigInt(changeAmount),
            };
            const outputAddress = ArkAddress.decode(params.address);
            const outputScript = BigInt(params.amount) < this.dustAmount
                ? outputAddress.subdustPkScript
                : outputAddress.pkScript;
            const outputs = [
                {
                    script: outputScript,
                    amount: BigInt(params.amount),
                },
            ];
            // add change output if needed
            if (selected.changeAmount > 0n) {
                const changeOutputScript = selected.changeAmount < this.dustAmount
                    ? this.arkAddress.subdustPkScript
                    : this.arkAddress.pkScript;
                outputs.push({
                    script: changeOutputScript,
                    amount: BigInt(selected.changeAmount),
                });
            }
            const { arkTxid, signedCheckpointTxs } = await this.buildAndSubmitOffchainTx(selected.inputs, outputs);
            await this.updateDbAfterOffchainTx(selected.inputs, arkTxid, signedCheckpointTxs, params.amount, selected.changeAmount, selected.changeAmount > 0n ? outputs.length - 1 : 0);
            return arkTxid;
        }
        return this.send({
            address: params.address,
            amount: params.amount,
        });
    }
    async settle(params, eventCallback) {
        if (params?.inputs) {
            for (const input of params.inputs) {
                // validate arknotes inputs
                if (typeof input === "string") {
                    try {
                        ArkNote.fromString(input);
                    }
                    catch (e) {
                        throw new Error(`Invalid arknote "${input}"`);
                    }
                }
            }
        }
        // if no params are provided, use all non expired boarding utxos and offchain vtxos as inputs
        // and send all to the offchain address
        if (!params) {
            const { fees } = await this.arkProvider.getInfo();
            const estimator = new Estimator(fees.intentFee);
            let amount = 0;
            const exitScript = CSVMultisigTapscript.decode(hex.decode(this.boardingTapscript.exitScript));
            const boardingTimelock = exitScript.params.timelock;
            const boardingUtxos = (await this.getBoardingUtxos()).filter((utxo) => !hasBoardingTxExpired(utxo, boardingTimelock));
            const filteredBoardingUtxos = [];
            for (const utxo of boardingUtxos) {
                const inputFee = estimator.evalOnchainInput({
                    amount: BigInt(utxo.value),
                });
                if (inputFee.value >= utxo.value) {
                    // skip if fees are greater than the utxo value
                    continue;
                }
                filteredBoardingUtxos.push(utxo);
                amount += utxo.value - inputFee.satoshis;
            }
            const vtxos = await this.getVtxos({ withRecoverable: true });
            const filteredVtxos = [];
            for (const vtxo of vtxos) {
                const inputFee = estimator.evalOffchainInput({
                    amount: BigInt(vtxo.value),
                    type: vtxo.virtualStatus.state === "swept"
                        ? "recoverable"
                        : "vtxo",
                    weight: 0,
                    birth: vtxo.createdAt,
                    expiry: vtxo.virtualStatus.batchExpiry
                        ? new Date(vtxo.virtualStatus.batchExpiry * 1000)
                        : new Date(),
                });
                if (inputFee.value >= vtxo.value) {
                    // skip if fees are greater than the vtxo value
                    continue;
                }
                filteredVtxos.push(vtxo);
                amount += vtxo.value - inputFee.satoshis;
            }
            const inputs = [...filteredBoardingUtxos, ...filteredVtxos];
            if (inputs.length === 0) {
                throw new Error("No inputs found");
            }
            const output = {
                address: await this.getAddress(),
                amount: BigInt(amount),
            };
            const outputFee = estimator.evalOffchainOutput({
                amount: output.amount,
                script: hex.encode(ArkAddress.decode(output.address).pkScript),
            });
            output.amount -= BigInt(outputFee.satoshis);
            if (output.amount <= this.dustAmount) {
                throw new Error("Output amount is below dust limit");
            }
            params = {
                inputs,
                outputs: [output],
            };
        }
        const onchainOutputIndexes = [];
        const outputs = [];
        let hasOffchainOutputs = false;
        for (const [index, output] of params.outputs.entries()) {
            let script;
            try {
                // offchain
                const addr = ArkAddress.decode(output.address);
                script = addr.pkScript;
                hasOffchainOutputs = true;
            }
            catch {
                // onchain
                const addr = Address(this.network).decode(output.address);
                script = OutScript.encode(addr);
                onchainOutputIndexes.push(index);
            }
            outputs.push({
                amount: output.amount,
                script,
            });
        }
        // if some of the inputs hold assets, build the asset packet and append as output
        // in the intent proof tx, there is a "fake" input at index 0
        // so the real coin indices are offset by +1
        const assetInputs = new Map();
        for (let i = 0; i < params.inputs.length; i++) {
            if ("assets" in params.inputs[i]) {
                const assets = params.inputs[i]
                    .assets;
                if (assets && assets.length > 0) {
                    assetInputs.set(i + 1, assets);
                }
            }
        }
        let outputAssets;
        let assetOutputIndex; // where to send the assets
        if (assetInputs.size > 0) {
            // collect all input assets and assign them to the first offchain output
            const allAssets = new Map();
            for (const [, assets] of assetInputs) {
                for (const asset of assets) {
                    const existing = allAssets.get(asset.assetId) ?? 0n;
                    allAssets.set(asset.assetId, existing + BigInt(asset.amount));
                }
            }
            outputAssets = [];
            for (const [assetId, amount] of allAssets) {
                outputAssets.push({ assetId, amount: Number(amount) });
            }
            const firstOffchainIndex = params.outputs.findIndex((_, i) => !onchainOutputIndexes.includes(i));
            if (firstOffchainIndex === -1) {
                throw new Error("Cannot settle assets without an offchain output");
            }
            assetOutputIndex = firstOffchainIndex;
        }
        const recipients = params.outputs.map((output, i) => ({
            address: output.address,
            amount: Number(output.amount),
            assets: i === assetOutputIndex ? outputAssets : undefined,
        }));
        if (outputAssets && outputAssets.length > 0) {
            const assetPacket = createAssetPacket(assetInputs, recipients);
            outputs.push(Extension.create([assetPacket]).txOut());
        }
        // session holds the state of the musig2 signing process of the vtxo tree
        let session;
        const signingPublicKeys = [];
        if (hasOffchainOutputs) {
            session = this.identity.signerSession();
            signingPublicKeys.push(hex.encode(await session.getPublicKey()));
        }
        const [intent, deleteIntent] = await Promise.all([
            this.makeRegisterIntentSignature(params.inputs, outputs, onchainOutputIndexes, signingPublicKeys),
            this.makeDeleteIntentSignature(params.inputs),
        ]);
        const topics = [
            ...signingPublicKeys,
            ...params.inputs.map((input) => `${input.txid}:${input.vout}`),
        ];
        const abortController = new AbortController();
        try {
            const stream = this.arkProvider.getEventStream(abortController.signal, topics);
            const intentId = await this.safeRegisterIntent(intent);
            const handler = this.createBatchHandler(intentId, params.inputs, recipients, session);
            const commitmentTxid = await Batch.join(stream, handler, {
                abortController,
                skipVtxoTreeSigning: !hasOffchainOutputs,
                eventCallback: eventCallback
                    ? (event) => Promise.resolve(eventCallback(event))
                    : undefined,
            });
            await this.updateDbAfterSettle(params.inputs, commitmentTxid);
            return commitmentTxid;
        }
        catch (error) {
            // delete the intent to not be stuck in the queue
            await this.arkProvider.deleteIntent(deleteIntent).catch(() => { });
            throw error;
        }
        finally {
            // close the stream
            abortController.abort();
        }
    }
    async handleSettlementFinalizationEvent(event, inputs, forfeitOutputScript, connectorsGraph) {
        // the signed forfeits transactions to submit
        const signedForfeits = [];
        const vtxos = await this.getVtxos();
        let settlementPsbt = Transaction.fromPSBT(base64.decode(event.commitmentTx));
        let hasBoardingUtxos = false;
        let connectorIndex = 0;
        const connectorsLeaves = connectorsGraph?.leaves() || [];
        for (const input of inputs) {
            // check if the input is an offchain "virtual" coin
            const vtxo = vtxos.find((vtxo) => vtxo.txid === input.txid && vtxo.vout === input.vout);
            // boarding utxo, we need to sign the settlement tx
            if (!vtxo) {
                for (let i = 0; i < settlementPsbt.inputsLength; i++) {
                    const settlementInput = settlementPsbt.getInput(i);
                    if (!settlementInput.txid ||
                        settlementInput.index === undefined) {
                        throw new Error("The server returned incomplete data. No settlement input found in the PSBT");
                    }
                    const inputTxId = hex.encode(settlementInput.txid);
                    if (inputTxId !== input.txid)
                        continue;
                    if (settlementInput.index !== input.vout)
                        continue;
                    // input found in the settlement tx, sign it
                    settlementPsbt.updateInput(i, {
                        tapLeafScript: [input.forfeitTapLeafScript],
                    });
                    settlementPsbt = await this.identity.sign(settlementPsbt, [
                        i,
                    ]);
                    hasBoardingUtxos = true;
                    break;
                }
                continue;
            }
            if (isRecoverable(vtxo) || isSubdust(vtxo, this.dustAmount)) {
                // recoverable or subdust coin, we don't need to create a forfeit tx
                continue;
            }
            if (connectorsLeaves.length === 0) {
                throw new Error("connectors not received");
            }
            if (connectorIndex >= connectorsLeaves.length) {
                throw new Error("not enough connectors received");
            }
            const connectorLeaf = connectorsLeaves[connectorIndex];
            const connectorTxId = connectorLeaf.id;
            const connectorOutput = connectorLeaf.getOutput(0);
            if (!connectorOutput) {
                throw new Error("connector output not found");
            }
            const connectorAmount = connectorOutput.amount;
            const connectorPkScript = connectorOutput.script;
            if (!connectorAmount || !connectorPkScript) {
                throw new Error("invalid connector output");
            }
            connectorIndex++;
            let forfeitTx = buildForfeitTx([
                {
                    txid: input.txid,
                    index: input.vout,
                    witnessUtxo: {
                        amount: BigInt(vtxo.value),
                        script: VtxoScript.decode(input.tapTree).pkScript,
                    },
                    sighashType: SigHash.DEFAULT,
                    tapLeafScript: [input.forfeitTapLeafScript],
                },
                {
                    txid: connectorTxId,
                    index: 0,
                    witnessUtxo: {
                        amount: connectorAmount,
                        script: connectorPkScript,
                    },
                },
            ], forfeitOutputScript);
            // do not sign the connector input
            forfeitTx = await this.identity.sign(forfeitTx, [0]);
            signedForfeits.push(base64.encode(forfeitTx.toPSBT()));
        }
        if (signedForfeits.length > 0 || hasBoardingUtxos) {
            await this.arkProvider.submitSignedForfeitTxs(signedForfeits, hasBoardingUtxos
                ? base64.encode(settlementPsbt.toPSBT())
                : undefined);
        }
    }
    /**
     * @implements Batch.Handler interface.
     * @param intentId - The intent ID.
     * @param inputs - The inputs of the intent.
     * @param session - The musig2 signing session, if not provided, the signing will be skipped.
     * @param expectedRecipients - Expected recipients to validate in the vtxo tree.
     */
    createBatchHandler(intentId, inputs, expectedRecipients, session) {
        let sweepTapTreeRoot;
        return {
            onBatchStarted: async (event) => {
                const utf8IntentId = new TextEncoder().encode(intentId);
                const intentIdHash = sha256(utf8IntentId);
                const intentIdHashStr = hex.encode(intentIdHash);
                let skip = true;
                // check if our intent ID hash matches any in the event
                for (const idHash of event.intentIdHashes) {
                    if (idHash === intentIdHashStr) {
                        if (!this.arkProvider) {
                            throw new Error("Ark provider not configured");
                        }
                        await this.arkProvider.confirmRegistration(intentId);
                        skip = false;
                    }
                }
                if (skip) {
                    return { skip };
                }
                const sweepTapscript = CSVMultisigTapscript.encode({
                    timelock: {
                        value: event.batchExpiry,
                        type: event.batchExpiry >= 512n ? "seconds" : "blocks",
                    },
                    pubkeys: [this.forfeitPubkey],
                }).script;
                sweepTapTreeRoot = tapLeafHash(sweepTapscript);
                return { skip: false };
            },
            onTreeSigningStarted: async (event, vtxoTree) => {
                if (!session) {
                    return { skip: true };
                }
                if (!sweepTapTreeRoot) {
                    throw new Error("Sweep tap tree root not set");
                }
                const xOnlyPublicKeys = event.cosignersPublicKeys.map((k) => k.slice(2));
                const signerPublicKey = await session.getPublicKey();
                const xonlySignerPublicKey = signerPublicKey.subarray(1);
                if (!xOnlyPublicKeys.includes(hex.encode(xonlySignerPublicKey))) {
                    // not a cosigner, skip the signing
                    return { skip: true };
                }
                // validate the unsigned vtxo tree
                const commitmentTx = Transaction.fromPSBT(base64.decode(event.unsignedCommitmentTx));
                validateVtxoTxGraph(vtxoTree, commitmentTx, sweepTapTreeRoot);
                // validate that all expected receivers are in the vtxo tree with correct amounts and assets
                if (expectedRecipients && expectedRecipients.length > 0) {
                    validateBatchRecipients(commitmentTx, vtxoTree.leaves(), expectedRecipients, this.network);
                }
                const sharedOutput = commitmentTx.getOutput(0);
                if (!sharedOutput?.amount) {
                    throw new Error("Shared output not found");
                }
                await session.init(vtxoTree, sweepTapTreeRoot, sharedOutput.amount);
                const pubkey = hex.encode(await session.getPublicKey());
                const nonces = await session.getNonces();
                await this.arkProvider.submitTreeNonces(event.id, pubkey, nonces);
                return { skip: false };
            },
            onTreeNonces: async (event) => {
                if (!session) {
                    return { fullySigned: true }; // Signing complete (no signing needed)
                }
                const { hasAllNonces } = await session.aggregatedNonces(event.txid, event.nonces);
                // wait to receive and aggregate all nonces before sending signatures
                if (!hasAllNonces)
                    return { fullySigned: false };
                const signatures = await session.sign();
                const pubkey = hex.encode(await session.getPublicKey());
                await this.arkProvider.submitTreeSignatures(event.id, pubkey, signatures);
                return { fullySigned: true };
            },
            onBatchFinalization: async (event, _, connectorTree) => {
                if (!this.forfeitOutputScript) {
                    throw new Error("Forfeit output script not set");
                }
                if (connectorTree) {
                    validateConnectorsTxGraph(event.commitmentTx, connectorTree);
                }
                await this.handleSettlementFinalizationEvent(event, inputs, this.forfeitOutputScript, connectorTree);
            },
        };
    }
    async safeRegisterIntent(intent) {
        try {
            return await this.arkProvider.registerIntent(intent);
        }
        catch (error) {
            // catch the "already registered by another intent" error
            if (error instanceof ArkError &&
                error.code === 0 &&
                error.message.includes("duplicated input")) {
                // delete all intents spending one of the wallet coins
                const allSpendableCoins = await this.getVtxos({
                    withRecoverable: true,
                });
                const deleteIntent = await this.makeDeleteIntentSignature(allSpendableCoins);
                await this.arkProvider.deleteIntent(deleteIntent);
                // try again
                return this.arkProvider.registerIntent(intent);
            }
            throw error;
        }
    }
    async makeRegisterIntentSignature(coins, outputs, onchainOutputsIndexes, cosignerPubKeys, validAt) {
        const message = {
            type: "register",
            onchain_output_indexes: onchainOutputsIndexes,
            valid_at: validAt ? Math.floor(validAt) : 0,
            expire_at: 0,
            cosigners_public_keys: cosignerPubKeys,
        };
        const proof = Intent.create(message, coins, outputs);
        const signedProof = await this.identity.sign(proof);
        return {
            proof: base64.encode(signedProof.toPSBT()),
            message,
        };
    }
    async makeDeleteIntentSignature(coins) {
        const message = {
            type: "delete",
            expire_at: 0,
        };
        const proof = Intent.create(message, coins, []);
        const signedProof = await this.identity.sign(proof);
        return {
            proof: base64.encode(signedProof.toPSBT()),
            message,
        };
    }
    async makeGetPendingTxIntentSignature(coins) {
        const message = {
            type: "get-pending-tx",
            expire_at: 0,
        };
        const proof = Intent.create(message, coins, []);
        const signedProof = await this.identity.sign(proof);
        return {
            proof: base64.encode(signedProof.toPSBT()),
            message,
        };
    }
    /**
     * Finalizes pending transactions by retrieving them from the server and finalizing each one.
     * @param vtxos - Optional list of VTXOs to use instead of retrieving them from the server
     * @returns Array of transaction IDs that were finalized
     */
    async finalizePendingTxs(vtxos) {
        const MAX_INPUTS_PER_INTENT = 20;
        if (!vtxos || vtxos.length === 0) {
            // Query per-script so each VTXO is extended with the correct tapscript
            const scriptMap = await this.getScriptMap();
            const allExtended = [];
            for (const [scriptHex, vtxoScript] of scriptMap) {
                const { vtxos: fetchedVtxos } = await this.indexerProvider.getVtxos({
                    scripts: [scriptHex],
                });
                const pending = fetchedVtxos.filter((vtxo) => vtxo.virtualStatus.state !== "swept" &&
                    vtxo.virtualStatus.state !== "settled");
                for (const vtxo of pending) {
                    allExtended.push({
                        ...vtxo,
                        forfeitTapLeafScript: vtxoScript.forfeit(),
                        intentTapLeafScript: vtxoScript.forfeit(),
                        tapTree: vtxoScript.encode(),
                    });
                }
            }
            if (allExtended.length === 0) {
                return { finalized: [], pending: [] };
            }
            vtxos = allExtended;
        }
        const finalized = [];
        const pending = [];
        for (let i = 0; i < vtxos.length; i += MAX_INPUTS_PER_INTENT) {
            const batch = vtxos.slice(i, i + MAX_INPUTS_PER_INTENT);
            const intent = await this.makeGetPendingTxIntentSignature(batch);
            const pendingTxs = await this.arkProvider.getPendingTxs(intent);
            // finalize each transaction by signing the checkpoints
            for (const pendingTx of pendingTxs) {
                pending.push(pendingTx.arkTxid);
                try {
                    // sign the checkpoints
                    const finalCheckpoints = await Promise.all(pendingTx.signedCheckpointTxs.map(async (c) => {
                        const tx = Transaction.fromPSBT(base64.decode(c));
                        const signedCheckpoint = await this.identity.sign(tx);
                        return base64.encode(signedCheckpoint.toPSBT());
                    }));
                    await this.arkProvider.finalizeTx(pendingTx.arkTxid, finalCheckpoints);
                    finalized.push(pendingTx.arkTxid);
                }
                catch (error) {
                    console.error(`Failed to finalize transaction ${pendingTx.arkTxid}:`, error);
                    // continue with other transactions even if one fails
                }
            }
        }
        return { finalized, pending };
    }
    /**
     * Send BTC and/or assets to one or more recipients.
     *
     * @param recipients - Array of recipients with their addresses, BTC amounts, and assets
     * @returns Promise resolving to the ark transaction ID
     *
     * @example
     * ```typescript
     * const txid = await wallet.send({
     *     address: 'ark1...',
     *     amount: 1000, // (optional, default to dust) btc amount to send to the output
     *     assets: [{ assetId: 'abc123...', amount: 50 }] // (optional) list of assets to send
     * });
     * ```
     */
    async send(...args) {
        if (args.length === 0) {
            throw new Error("At least one receiver is required");
        }
        // validate recipients and populate undefined amount with dust amount
        const recipients = validateRecipients(args, Number(this.dustAmount));
        const address = await this.getAddress();
        const outputAddress = ArkAddress.decode(address);
        const virtualCoins = await this.getVtxos({
            withRecoverable: false,
        });
        // keep track of asset changes
        const assetChanges = new Map();
        let selectedCoins = [];
        let btcAmountToSelect = 0;
        for (const recipient of recipients) {
            btcAmountToSelect += Math.max(recipient.amount, Number(this.dustAmount));
        }
        // select assets
        for (const recipient of recipients) {
            if (!recipient.assets) {
                continue;
            }
            for (const receiverAsset of recipient.assets) {
                let amountToSelect = BigInt(receiverAsset.amount);
                // check if existing change covers the needed amount
                const existingChange = assetChanges.get(receiverAsset.assetId) ?? 0n;
                if (existingChange >= amountToSelect) {
                    assetChanges.set(receiverAsset.assetId, existingChange - amountToSelect);
                    if (assetChanges.get(receiverAsset.assetId) === 0n) {
                        assetChanges.delete(receiverAsset.assetId);
                    }
                    continue;
                }
                if (existingChange > 0n) {
                    amountToSelect -= existingChange;
                    assetChanges.delete(receiverAsset.assetId);
                }
                const availableCoins = virtualCoins.filter((c) => !selectedCoins.find((sc) => sc.txid === c.txid && sc.vout === c.vout));
                const { selected, totalAssetAmount } = selectCoinsWithAsset(availableCoins, receiverAsset.assetId, amountToSelect);
                for (const coin of selected) {
                    selectedCoins.push(coin);
                    // asset coins contain btc, subtract from total amount to select
                    btcAmountToSelect -= coin.value;
                    // coin may contain other assets, add them to asset changes
                    if (coin.assets) {
                        for (const a of coin.assets) {
                            if (a.assetId === receiverAsset.assetId) {
                                continue;
                            }
                            const existing = assetChanges.get(a.assetId) ?? 0n;
                            assetChanges.set(a.assetId, existing + BigInt(a.amount));
                        }
                    }
                }
                const assetChangeAmount = totalAssetAmount - amountToSelect;
                if (assetChangeAmount > 0n) {
                    const existing = assetChanges.get(receiverAsset.assetId) ?? 0n;
                    assetChanges.set(receiverAsset.assetId, existing + assetChangeAmount);
                }
            }
        }
        // select remaining btc
        if (btcAmountToSelect > 0) {
            const availableCoins = virtualCoins.filter((c) => !selectedCoins.find((sc) => sc.txid === c.txid && sc.vout === c.vout));
            const { inputs: btcCoins } = selectVirtualCoins(availableCoins, btcAmountToSelect);
            // some coins may contain assets, add them to asset changes
            for (const coin of btcCoins) {
                if (coin.assets) {
                    for (const asset of coin.assets) {
                        const existing = assetChanges.get(asset.assetId) ?? 0n;
                        assetChanges.set(asset.assetId, existing + BigInt(asset.amount));
                    }
                }
            }
            selectedCoins = [...selectedCoins, ...btcCoins];
        }
        let totalBtcSelected = selectedCoins.reduce((sum, c) => sum + c.value, 0);
        // build tx outputs
        const outputs = recipients.map((recipient) => ({
            script: recipient.script,
            amount: BigInt(recipient.amount),
        }));
        const totalBtcOutput = outputs.reduce((sum, o) => sum + Number(o.amount), 0);
        let changeAmount = totalBtcSelected - totalBtcOutput;
        // enforce minimum change amount when there are asset changes
        if (assetChanges.size > 0 && changeAmount < Number(this.dustAmount)) {
            const availableCoins = virtualCoins.filter((c) => !selectedCoins.find((sc) => sc.txid === c.txid && sc.vout === c.vout));
            const { inputs: extraCoins } = selectVirtualCoins(availableCoins, Number(this.dustAmount) - changeAmount);
            for (const coin of extraCoins) {
                if (coin.assets) {
                    for (const asset of coin.assets) {
                        const existing = assetChanges.get(asset.assetId) ?? 0n;
                        assetChanges.set(asset.assetId, existing + BigInt(asset.amount));
                    }
                }
            }
            selectedCoins = [...selectedCoins, ...extraCoins];
            totalBtcSelected += extraCoins.reduce((sum, c) => sum + c.value, 0);
            changeAmount = totalBtcSelected - totalBtcOutput;
        }
        // build change receiver with BTC change and all asset changes
        let changeReceiver;
        let changeIndex = 0;
        if (changeAmount > 0) {
            const changeAssets = [];
            for (const [assetId, amount] of assetChanges) {
                if (amount > 0n) {
                    changeAssets.push({ assetId, amount: Number(amount) });
                }
            }
            changeIndex = outputs.length;
            outputs.push({
                script: BigInt(changeAmount) < this.dustAmount
                    ? outputAddress.subdustPkScript
                    : outputAddress.pkScript,
                amount: BigInt(changeAmount),
            });
            changeReceiver = {
                address: address,
                amount: changeAmount,
                assets: changeAssets.length > 0 ? changeAssets : undefined,
            };
        }
        // create asset packet only if there are assets involved
        const assetInputs = selectedCoinsToAssetInputs(selectedCoins);
        const hasAssets = assetInputs.size > 0 ||
            recipients.some((r) => r.assets && r.assets.length > 0);
        if (hasAssets) {
            const assetPacket = createAssetPacket(assetInputs, recipients, changeReceiver);
            outputs.push(Extension.create([assetPacket]).txOut());
        }
        const sentAmount = recipients.reduce((sum, r) => sum + r.amount, 0);
        const { arkTxid, signedCheckpointTxs } = await this.buildAndSubmitOffchainTx(selectedCoins, outputs);
        await this.updateDbAfterOffchainTx(selectedCoins, arkTxid, signedCheckpointTxs, sentAmount, BigInt(changeAmount), changeReceiver ? changeIndex : 0, changeReceiver?.assets);
        return arkTxid;
    }
    /**
     * Build an offchain transaction from the given inputs and outputs,
     * sign it, submit to the ark provider, and finalize.
     * @returns The ark transaction id and server-signed checkpoint PSBTs (for bookkeeping)
     */
    async buildAndSubmitOffchainTx(inputs, outputs) {
        const offchainTx = buildOffchainTx(inputs.map((input) => {
            return {
                ...input,
                tapLeafScript: input.forfeitTapLeafScript,
            };
        }), outputs, this.serverUnrollScript);
        const signedVirtualTx = await this.identity.sign(offchainTx.arkTx);
        const { arkTxid, signedCheckpointTxs } = await this.arkProvider.submitTx(base64.encode(signedVirtualTx.toPSBT()), offchainTx.checkpoints.map((c) => base64.encode(c.toPSBT())));
        const finalCheckpoints = await Promise.all(signedCheckpointTxs.map(async (c) => {
            const tx = Transaction.fromPSBT(base64.decode(c));
            const signedCheckpoint = await this.identity.sign(tx);
            return base64.encode(signedCheckpoint.toPSBT());
        }));
        await this.arkProvider.finalizeTx(arkTxid, finalCheckpoints);
        return { arkTxid, signedCheckpointTxs };
    }
    // mark vtxo spent and save change vtxo if any
    async updateDbAfterOffchainTx(inputs, arkTxid, signedCheckpointTxs, sentAmount, changeAmount, changeVout, changeAssets) {
        try {
            const spentVtxos = [];
            const commitmentTxIds = new Set();
            let batchExpiry = Number.MAX_SAFE_INTEGER;
            if (inputs.length !== signedCheckpointTxs.length) {
                console.warn(`updateDbAfterOffchainTx: inputs length (${inputs.length}) differs from signedCheckpointTxs length (${signedCheckpointTxs.length})`);
            }
            const safeLength = Math.min(inputs.length, signedCheckpointTxs.length);
            for (const [inputIndex, input] of inputs.entries()) {
                const vtxo = extendVirtualCoin(this, input);
                if (inputIndex < safeLength &&
                    signedCheckpointTxs[inputIndex]) {
                    const checkpoint = Transaction.fromPSBT(base64.decode(signedCheckpointTxs[inputIndex]));
                    spentVtxos.push({
                        ...vtxo,
                        virtualStatus: {
                            ...vtxo.virtualStatus,
                            state: "spent",
                        },
                        spentBy: checkpoint.id,
                        arkTxId: arkTxid,
                        isSpent: true,
                    });
                }
                else {
                    spentVtxos.push({
                        ...vtxo,
                        virtualStatus: {
                            ...vtxo.virtualStatus,
                            state: "spent",
                        },
                        arkTxId: arkTxid,
                        isSpent: true,
                    });
                }
                if (vtxo.virtualStatus.commitmentTxIds) {
                    for (const id of vtxo.virtualStatus.commitmentTxIds) {
                        commitmentTxIds.add(id);
                    }
                }
                if (vtxo.virtualStatus.batchExpiry) {
                    batchExpiry = Math.min(batchExpiry, vtxo.virtualStatus.batchExpiry);
                }
            }
            const createdAt = Date.now();
            const addr = this.arkAddress.encode();
            // Only save a change VTXO for preconfirmed coins (those with a batchExpiry).
            // Inputs without a batchExpiry are already settled/unrolled and don't need tracking.
            let changeVtxo;
            if (changeAmount > 0n && batchExpiry !== Number.MAX_SAFE_INTEGER) {
                changeVtxo = {
                    txid: arkTxid,
                    vout: changeVout,
                    createdAt: new Date(createdAt),
                    forfeitTapLeafScript: this.offchainTapscript.forfeit(),
                    intentTapLeafScript: this.offchainTapscript.forfeit(),
                    isUnrolled: false,
                    isSpent: false,
                    tapTree: this.offchainTapscript.encode(),
                    value: Number(changeAmount),
                    virtualStatus: {
                        state: "preconfirmed",
                        commitmentTxIds: Array.from(commitmentTxIds),
                        batchExpiry,
                    },
                    status: {
                        confirmed: false,
                    },
                    assets: changeAssets,
                };
            }
            await this.walletRepository.saveVtxos(addr, changeVtxo ? [...spentVtxos, changeVtxo] : spentVtxos);
            await this.walletRepository.saveTransactions(addr, [
                {
                    key: {
                        boardingTxid: "",
                        commitmentTxid: "",
                        arkTxid: arkTxid,
                    },
                    amount: sentAmount,
                    type: TxType.TxSent,
                    settled: false,
                    createdAt,
                },
            ]);
        }
        catch (e) {
            console.warn("error saving offchain tx to repository", e);
        }
    }
    // mark vtxo spent & settled, remove boarding utxo
    async updateDbAfterSettle(inputs, commitmentTxid) {
        try {
            const addr = this.arkAddress.encode();
            const boardingAddress = await this.getBoardingAddress();
            const spentVtxos = [];
            const inputArkTxIds = new Set();
            const boardingUtxoToRemove = new Set();
            const isVtxo = (input) => "virtualStatus" in input;
            for (const input of inputs) {
                if (isVtxo(input)) {
                    // vtxo = mark it settled
                    const vtxo = extendVirtualCoin(this, input);
                    if (vtxo.arkTxId) {
                        inputArkTxIds.add(vtxo.arkTxId);
                    }
                    spentVtxos.push({
                        ...vtxo,
                        virtualStatus: {
                            ...vtxo.virtualStatus,
                            state: "settled",
                        },
                        settledBy: commitmentTxid,
                        isSpent: true,
                    });
                }
                else {
                    // boarding utxo = remove it
                    boardingUtxoToRemove.add(`${input.txid}:${input.vout}`);
                }
            }
            if (spentVtxos.length > 0) {
                await this.walletRepository.saveVtxos(addr, spentVtxos);
            }
            if (boardingUtxoToRemove.size > 0) {
                const currentUtxos = await this.walletRepository.getUtxos(boardingAddress);
                const filtered = currentUtxos.filter((u) => !boardingUtxoToRemove.has(`${u.txid}:${u.vout}`));
                // Clear and re-save the filtered list
                await this.walletRepository.deleteUtxos(boardingAddress);
                if (filtered.length > 0) {
                    await this.walletRepository.saveUtxos(boardingAddress, filtered);
                }
            }
        }
        catch (e) {
            console.warn("error updating repository after settle", e);
        }
    }
}
Wallet.MIN_FEE_RATE = 1; // sats/vbyte
/**
 * Select virtual coins to reach a target amount, prioritizing those closer to expiry
 * @param coins List of virtual coins to select from
 * @param targetAmount Target amount to reach in satoshis
 * @returns Selected coins and change amount
 */
export function selectVirtualCoins(coins, targetAmount) {
    // Sort VTXOs by expiry (ascending) and amount (descending)
    const sortedCoins = [...coins].sort((a, b) => {
        // First sort by expiry if available
        const expiryA = a.virtualStatus.batchExpiry || Number.MAX_SAFE_INTEGER;
        const expiryB = b.virtualStatus.batchExpiry || Number.MAX_SAFE_INTEGER;
        if (expiryA !== expiryB) {
            return expiryA - expiryB; // Earlier expiry first
        }
        // Then sort by amount
        return b.value - a.value; // Larger amount first
    });
    const selectedCoins = [];
    let selectedAmount = 0;
    // Select coins until we have enough
    for (const coin of sortedCoins) {
        selectedCoins.push(coin);
        selectedAmount += coin.value;
        if (selectedAmount >= targetAmount) {
            break;
        }
    }
    if (selectedAmount === targetAmount) {
        return { inputs: selectedCoins, changeAmount: 0n };
    }
    // Check if we have enough
    if (selectedAmount < targetAmount) {
        throw new Error("Insufficient funds");
    }
    const changeAmount = BigInt(selectedAmount - targetAmount);
    return {
        inputs: selectedCoins,
        changeAmount,
    };
}
/**
 * Wait for incoming funds to the wallet
 * @param wallet - The wallet to wait for incoming funds
 * @returns A promise that resolves the next new coins received by the wallet's address
 */
export async function waitForIncomingFunds(wallet) {
    let stopFunc;
    const promise = new Promise((resolve) => {
        wallet
            .notifyIncomingFunds((coins) => {
            resolve(coins);
            if (stopFunc)
                stopFunc();
        })
            .then((stop) => {
            stopFunc = stop;
        });
    });
    return promise;
}
