// @arkade-os/sdk/dist/esm/repositories/indexedDB/walletRepository.js
import { STORE_VTXOS, STORE_UTXOS, STORE_TRANSACTIONS, STORE_WALLET_STATE, serializeVtxo, serializeUtxo, deserializeVtxo, deserializeUtxo, DB_VERSION, } from './db.js';
import { closeDatabase, openDatabase } from './manager.js';
import { initDatabase } from './schema.js';
import { DEFAULT_DB_NAME } from '../../worker/browser/utils.js';
/**
 * IndexedDB-based implementation of WalletRepository.
 */
export class IndexedDBWalletRepository {
    constructor(dbName = DEFAULT_DB_NAME) {
        this.dbName = dbName;
        this.version = 1;
        this.db = null;
    }
    async clear() {
        try {
            const db = await this.getDB();
            return new Promise((resolve, reject) => {
                const transaction = db.transaction([
                    STORE_VTXOS,
                    STORE_UTXOS,
                    STORE_TRANSACTIONS,
                    STORE_WALLET_STATE,
                ], "readwrite");
                const vtxosStore = transaction.objectStore(STORE_VTXOS);
                const utxosStore = transaction.objectStore(STORE_UTXOS);
                const transactionsStore = transaction.objectStore(STORE_TRANSACTIONS);
                const walletStateStore = transaction.objectStore(STORE_WALLET_STATE);
                const requests = [
                    vtxosStore.clear(),
                    utxosStore.clear(),
                    transactionsStore.clear(),
                    walletStateStore.clear(),
                ];
                let completed = 0;
                const checkComplete = () => {
                    completed++;
                    if (completed === requests.length) {
                        resolve();
                    }
                };
                requests.forEach((request) => {
                    request.onsuccess = checkComplete;
                    request.onerror = () => reject(request.error);
                });
            });
        }
        catch (error) {
            console.error("Failed to clear wallet data:", error);
            throw error;
        }
    }
    async [Symbol.asyncDispose]() {
        if (!this.db)
            return;
        await closeDatabase(this.dbName);
        this.db = null;
    }
    async getVtxos(address) {
        try {
            const db = await this.getDB();
            return new Promise((resolve, reject) => {
                const transaction = db.transaction([STORE_VTXOS], "readonly");
                const store = transaction.objectStore(STORE_VTXOS);
                const index = store.index("address");
                const request = index.getAll(address);
                request.onerror = () => reject(request.error);
                request.onsuccess = () => {
                    const results = request.result || [];
                    const vtxos = results.map(deserializeVtxo);
                    resolve(vtxos);
                };
            });
        }
        catch (error) {
            console.error(`Failed to get VTXOs for address ${address}:`, error);
            return [];
        }
    }
    async saveVtxos(address, vtxos) {
        try {
            const db = await this.getDB();
            return new Promise((resolve, reject) => {
                const transaction = db.transaction([STORE_VTXOS], "readwrite");
                const store = transaction.objectStore(STORE_VTXOS);
                const promises = vtxos.map((vtxo) => {
                    return new Promise((resolveItem, rejectItem) => {
                        const serialized = serializeVtxo(vtxo);
                        const item = {
                            address,
                            ...serialized,
                        };
                        const request = store.put(item);
                        request.onerror = () => rejectItem(request.error);
                        request.onsuccess = () => resolveItem();
                    });
                });
                Promise.all(promises)
                    .then(() => resolve())
                    .catch(reject);
                transaction.onerror = () => reject(transaction.error);
            });
        }
        catch (error) {
            console.error(`Failed to save VTXOs for address ${address}:`, error);
            throw error;
        }
    }
    async deleteVtxos(address) {
        try {
            const db = await this.getDB();
            return new Promise((resolve, reject) => {
                const transaction = db.transaction([STORE_VTXOS], "readwrite");
                const store = transaction.objectStore(STORE_VTXOS);
                const index = store.index("address");
                const request = index.openCursor(IDBKeyRange.only(address));
                request.onerror = () => reject(request.error);
                request.onsuccess = () => {
                    const cursor = request.result;
                    if (cursor) {
                        cursor.delete();
                        cursor.continue();
                    }
                    else {
                        resolve();
                    }
                };
            });
        }
        catch (error) {
            console.error(`Failed to clear VTXOs for address ${address}:`, error);
            throw error;
        }
    }
    async getUtxos(address) {
        try {
            const db = await this.getDB();
            return new Promise((resolve, reject) => {
                const transaction = db.transaction([STORE_UTXOS], "readonly");
                const store = transaction.objectStore(STORE_UTXOS);
                const index = store.index("address");
                const request = index.getAll(address);
                request.onerror = () => reject(request.error);
                request.onsuccess = () => {
                    const results = request.result || [];
                    const utxos = results.map(deserializeUtxo);
                    resolve(utxos);
                };
            });
        }
        catch (error) {
            console.error(`Failed to get UTXOs for address ${address}:`, error);
            return [];
        }
    }
    async saveUtxos(address, utxos) {
        try {
            const db = await this.getDB();
            return new Promise((resolve, reject) => {
                const transaction = db.transaction([STORE_UTXOS], "readwrite");
                const store = transaction.objectStore(STORE_UTXOS);
                const promises = utxos.map((utxo) => {
                    return new Promise((resolveItem, rejectItem) => {
                        const serialized = serializeUtxo(utxo);
                        const item = {
                            address,
                            ...serialized,
                        };
                        const request = store.put(item);
                        request.onerror = () => rejectItem(request.error);
                        request.onsuccess = () => resolveItem();
                    });
                });
                Promise.all(promises)
                    .then(() => resolve())
                    .catch(reject);
                transaction.onerror = () => reject(transaction.error);
            });
        }
        catch (error) {
            console.error(`Failed to save UTXOs for address ${address}:`, error);
            throw error;
        }
    }
    async deleteUtxos(address) {
        try {
            const db = await this.getDB();
            return new Promise((resolve, reject) => {
                const transaction = db.transaction([STORE_UTXOS], "readwrite");
                const store = transaction.objectStore(STORE_UTXOS);
                const index = store.index("address");
                const request = index.openCursor(IDBKeyRange.only(address));
                request.onerror = () => reject(request.error);
                request.onsuccess = () => {
                    const cursor = request.result;
                    if (cursor) {
                        cursor.delete();
                        cursor.continue();
                    }
                    else {
                        resolve();
                    }
                };
            });
        }
        catch (error) {
            console.error(`Failed to clear UTXOs for address ${address}:`, error);
            throw error;
        }
    }
    async getTransactionHistory(address) {
        try {
            const db = await this.getDB();
            return new Promise((resolve, reject) => {
                const transaction = db.transaction([STORE_TRANSACTIONS], "readonly");
                const store = transaction.objectStore(STORE_TRANSACTIONS);
                const index = store.index("address");
                const request = index.getAll(address);
                request.onerror = () => reject(request.error);
                request.onsuccess = () => {
                    const results = request.result || [];
                    resolve(results.sort((a, b) => a.createdAt - b.createdAt));
                };
            });
        }
        catch (error) {
            console.error(`Failed to get transaction history for address ${address}:`, error);
            return [];
        }
    }
    async saveTransactions(address, txs) {
        try {
            const db = await this.getDB();
            return new Promise((resolve, reject) => {
                const transaction = db.transaction([STORE_TRANSACTIONS], "readwrite");
                const store = transaction.objectStore(STORE_TRANSACTIONS);
                // Queue all put operations
                txs.forEach((tx) => {
                    const item = {
                        address,
                        ...tx,
                        keyBoardingTxid: tx.key.boardingTxid,
                        keyCommitmentTxid: tx.key.commitmentTxid,
                        keyArkTxid: tx.key.arkTxid,
                    };
                    store.put(item);
                });
                // Handle transaction completion
                transaction.oncomplete = () => resolve();
                transaction.onerror = () => reject(transaction.error);
                transaction.onabort = () => reject(new Error("Transaction aborted"));
            });
        }
        catch (error) {
            console.error(`Failed to save transactions for address ${address}:`, error);
            throw error;
        }
    }
    async deleteTransactions(address) {
        try {
            const db = await this.getDB();
            return new Promise((resolve, reject) => {
                const transaction = db.transaction([STORE_TRANSACTIONS], "readwrite");
                const store = transaction.objectStore(STORE_TRANSACTIONS);
                const index = store.index("address");
                const request = index.openCursor(IDBKeyRange.only(address));
                request.onerror = () => reject(request.error);
                request.onsuccess = () => {
                    const cursor = request.result;
                    if (cursor) {
                        cursor.delete();
                        cursor.continue();
                    }
                    else {
                        resolve();
                    }
                };
            });
        }
        catch (error) {
            console.error(`Failed to clear transactions for address ${address}:`, error);
            throw error;
        }
    }
    async getWalletState() {
        try {
            const db = await this.getDB();
            return new Promise((resolve, reject) => {
                const transaction = db.transaction([STORE_WALLET_STATE], "readonly");
                const store = transaction.objectStore(STORE_WALLET_STATE);
                const request = store.get("state");
                request.onerror = () => reject(request.error);
                request.onsuccess = () => {
                    const result = request.result;
                    if (result && result.data) {
                        resolve(result.data);
                    }
                    else {
                        resolve(null);
                    }
                };
            });
        }
        catch (error) {
            console.error("Failed to get wallet state:", error);
            return null;
        }
    }
    async saveWalletState(state) {
        try {
            const db = await this.getDB();
            return new Promise((resolve, reject) => {
                const transaction = db.transaction([STORE_WALLET_STATE], "readwrite");
                const store = transaction.objectStore(STORE_WALLET_STATE);
                const item = {
                    key: "state",
                    data: state,
                };
                const request = store.put(item);
                request.onerror = () => reject(request.error);
                request.onsuccess = () => resolve();
            });
        }
        catch (error) {
            console.error("Failed to save wallet state:", error);
            throw error;
        }
    }
    async getDB() {
        if (this.db)
            return this.db;
        this.db = await openDatabase(this.dbName, DB_VERSION, initDatabase);
        return this.db;
    }
}
