// @arkade-os/sdk/dist/cjs/repositories/migrations/walletRepositoryImpl.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WalletRepositoryImpl = void 0;
const base_1 = require("@scure/base");
const btc_signer_1 = require("@scure/btc-signer");
const getVtxosStorageKey = (address) => `vtxos:${address}`;
const getUtxosStorageKey = (address) => `utxos:${address}`;
const getTransactionsStorageKey = (address) => `tx:${address}`;
const walletStateStorageKey = "wallet:state";
// Utility functions for (de)serializing complex structures
const serializeVtxo = (v) => ({
    ...v,
    tapTree: base_1.hex.encode(v.tapTree),
    forfeitTapLeafScript: serializeTapLeaf(v.forfeitTapLeafScript),
    intentTapLeafScript: serializeTapLeaf(v.intentTapLeafScript),
    extraWitness: v.extraWitness?.map(base_1.hex.encode),
});
const serializeUtxo = (u) => ({
    ...u,
    tapTree: base_1.hex.encode(u.tapTree),
    forfeitTapLeafScript: serializeTapLeaf(u.forfeitTapLeafScript),
    intentTapLeafScript: serializeTapLeaf(u.intentTapLeafScript),
    extraWitness: u.extraWitness?.map(base_1.hex.encode),
});
const deserializeVtxo = (o) => ({
    ...o,
    createdAt: new Date(o.createdAt),
    tapTree: base_1.hex.decode(o.tapTree),
    forfeitTapLeafScript: deserializeTapLeaf(o.forfeitTapLeafScript),
    intentTapLeafScript: deserializeTapLeaf(o.intentTapLeafScript),
    extraWitness: o.extraWitness?.map(base_1.hex.decode),
});
const deserializeUtxo = (o) => ({
    ...o,
    tapTree: base_1.hex.decode(o.tapTree),
    forfeitTapLeafScript: deserializeTapLeaf(o.forfeitTapLeafScript),
    intentTapLeafScript: deserializeTapLeaf(o.intentTapLeafScript),
    extraWitness: o.extraWitness?.map(base_1.hex.decode),
});
const serializeTapLeaf = ([cb, s]) => ({
    cb: base_1.hex.encode(btc_signer_1.TaprootControlBlock.encode(cb)),
    s: base_1.hex.encode(s),
});
const deserializeTapLeaf = (t) => {
    const cb = btc_signer_1.TaprootControlBlock.decode(base_1.hex.decode(t.cb));
    const s = base_1.hex.decode(t.s);
    return [cb, s];
};
/**
 * @deprecated This is only to be used in migration from storage V1
 */
class WalletRepositoryImpl {
    constructor(storage) {
        this.version = 1;
        this.storage = storage;
    }
    async getVtxos(address) {
        const stored = await this.storage.getItem(getVtxosStorageKey(address));
        if (!stored)
            return [];
        try {
            const parsed = JSON.parse(stored);
            return parsed.map(deserializeVtxo);
        }
        catch (error) {
            console.error(`Failed to parse VTXOs for address ${address}:`, error);
            return [];
        }
    }
    async saveVtxos(address, vtxos) {
        const storedVtxos = await this.getVtxos(address);
        for (const vtxo of vtxos) {
            const existing = storedVtxos.findIndex((v) => v.txid === vtxo.txid && v.vout === vtxo.vout);
            if (existing !== -1) {
                storedVtxos[existing] = vtxo;
            }
            else {
                storedVtxos.push(vtxo);
            }
        }
        await this.storage.setItem(getVtxosStorageKey(address), JSON.stringify(storedVtxos.map(serializeVtxo)));
    }
    async clearVtxos(address) {
        return this.deleteVtxos(address);
    }
    async deleteVtxos(address) {
        await this.storage.removeItem(getVtxosStorageKey(address));
    }
    async getUtxos(address) {
        const stored = await this.storage.getItem(getUtxosStorageKey(address));
        if (!stored)
            return [];
        try {
            const parsed = JSON.parse(stored);
            return parsed.map(deserializeUtxo);
        }
        catch (error) {
            console.error(`Failed to parse UTXOs for address ${address}:`, error);
            return [];
        }
    }
    async saveUtxos(address, utxos) {
        const storedUtxos = await this.getUtxos(address);
        utxos.forEach((utxo) => {
            const existing = storedUtxos.findIndex((u) => u.txid === utxo.txid && u.vout === utxo.vout);
            if (existing !== -1) {
                storedUtxos[existing] = utxo;
            }
            else {
                storedUtxos.push(utxo);
            }
        });
        await this.storage.setItem(getUtxosStorageKey(address), JSON.stringify(storedUtxos.map(serializeUtxo)));
    }
    async clearUtxos(address) {
        return this.deleteVtxos(address);
    }
    async deleteUtxos(address) {
        await this.storage.removeItem(getUtxosStorageKey(address));
    }
    async getTransactionHistory(address) {
        const storageKey = getTransactionsStorageKey(address);
        const stored = await this.storage.getItem(storageKey);
        if (!stored)
            return [];
        try {
            return JSON.parse(stored);
        }
        catch (error) {
            console.error(`Failed to parse transactions for address ${address}:`, error);
            return [];
        }
    }
    async saveTransactions(address, txs) {
        const storedTransactions = await this.getTransactionHistory(address);
        for (const tx of txs) {
            const existing = storedTransactions.findIndex((t) => t.key.boardingTxid === tx.key.boardingTxid &&
                t.key.commitmentTxid === tx.key.commitmentTxid &&
                t.key.arkTxid === tx.key.arkTxid);
            if (existing !== -1) {
                storedTransactions[existing] = tx;
            }
            else {
                storedTransactions.push(tx);
            }
        }
        await this.storage.setItem(getTransactionsStorageKey(address), JSON.stringify(storedTransactions));
    }
    async clearTransactions(address) {
        return this.deleteTransactions(address);
    }
    async deleteTransactions(address) {
        await this.storage.removeItem(getTransactionsStorageKey(address));
    }
    async getWalletState() {
        const stored = await this.storage.getItem(walletStateStorageKey);
        if (!stored)
            return null;
        try {
            const state = JSON.parse(stored);
            return state;
        }
        catch (error) {
            console.error("Failed to parse wallet state:", error);
            return null;
        }
    }
    async saveWalletState(state) {
        await this.storage.setItem(walletStateStorageKey, JSON.stringify(state));
    }
    // New method added in V2, not implemented for legacy
    async clear() {
        throw new Error("Method not implemented.");
    }
    async [Symbol.asyncDispose]() {
        // deprecated StorageAdapter doesn't have a `close()` method
        return;
    }
}
exports.WalletRepositoryImpl = WalletRepositoryImpl;
