// @arkade-os/sdk/dist/esm/repositories/inMemory/contractRepository.js
/**
 * In-memory implementation of ContractRepository.
 * Data is ephemeral and scoped to the instance.
 */
export class InMemoryContractRepository {
    constructor() {
        this.version = 1;
        this.contractData = new Map();
        this.collections = new Map();
        this.contractsByScript = new Map();
    }
    async clear() {
        this.contractData.clear();
        this.collections.clear();
        this.contractsByScript.clear();
    }
    // Contract entity management methods
    async getContracts(filter) {
        const contracts = this.contractsByScript.values();
        if (!filter) {
            return [...contracts];
        }
        const matches = (value, criterion) => {
            if (criterion === undefined) {
                return true;
            }
            return Array.isArray(criterion)
                ? criterion.includes(value)
                : value === criterion;
        };
        const results = [];
        for (const contract of contracts) {
            if (matches(contract.script, filter.script) &&
                matches(contract.state, filter.state) &&
                matches(contract.type, filter.type)) {
                results.push(contract);
            }
        }
        return results;
    }
    async saveContract(contract) {
        this.contractsByScript.set(contract.script, contract);
    }
    async deleteContract(script) {
        this.contractsByScript.delete(script);
    }
    async [Symbol.asyncDispose]() {
        // nothing to dispose, data is ephemeral and scoped to the instance
        return;
    }
}
