// @arkade-os/sdk/dist/esm/repositories/inMemory/walletRepository.js
/**
 * In-memory implementation of WalletRepository.
 * Data is ephemeral and scoped to the instance.
 */
export class InMemoryWalletRepository {
    constructor() {
        this.version = 1;
        this.vtxosByAddress = new Map();
        this.utxosByAddress = new Map();
        this.txsByAddress = new Map();
        this.walletState = null;
    }
    async getVtxos(address) {
        return this.vtxosByAddress.get(address) ?? [];
    }
    async saveVtxos(address, vtxos) {
        const existing = this.vtxosByAddress.get(address) ?? [];
        const next = mergeByKey(existing, vtxos, (item) => `${item.txid}:${item.vout}`);
        this.vtxosByAddress.set(address, next);
    }
    async deleteVtxos(address) {
        this.vtxosByAddress.delete(address);
    }
    async getUtxos(address) {
        return this.utxosByAddress.get(address) ?? [];
    }
    async saveUtxos(address, utxos) {
        const existing = this.utxosByAddress.get(address) ?? [];
        const next = mergeByKey(existing, utxos, (item) => `${item.txid}:${item.vout}`);
        this.utxosByAddress.set(address, next);
    }
    async deleteUtxos(address) {
        this.utxosByAddress.delete(address);
    }
    async getTransactionHistory(address) {
        return this.txsByAddress.get(address) ?? [];
    }
    async saveTransactions(address, txs) {
        const existing = this.txsByAddress.get(address) ?? [];
        const next = mergeByKey(existing, txs, serializeTxKey);
        this.txsByAddress.set(address, next);
    }
    async deleteTransactions(address) {
        this.txsByAddress.delete(address);
    }
    async getWalletState() {
        return this.walletState;
    }
    async saveWalletState(state) {
        this.walletState = state;
    }
    async clear() {
        this.vtxosByAddress.clear();
        this.utxosByAddress.clear();
        this.txsByAddress.clear();
        this.walletState = null;
    }
    async [Symbol.asyncDispose]() {
        // nothing to dispose, data is ephemeral and scoped to the instance
        return;
    }
}
function serializeTxKey(tx) {
    const key = tx.key;
    return `${key.boardingTxid}:${key.commitmentTxid}:${key.arkTxid}`;
}
function mergeByKey(existing, incoming, toKey) {
    const next = new Map();
    existing.forEach((item) => {
        next.set(toKey(item), item);
    });
    incoming.forEach((item) => {
        next.set(toKey(item), item);
    });
    return Array.from(next.values());
}
