// @arkade-os/sdk/dist/esm/worker/browser/service-worker-manager.js
const registrations = new Map();
let handshakes = new WeakSet();
function ensureServiceWorkerSupport() {
    if (!("serviceWorker" in navigator)) {
        throw new Error("Service workers are not supported in this browser");
    }
}
function debugLog(debug, ...args) {
    if (debug) {
        // eslint-disable-next-line no-console
        console.debug(...args);
    }
}
function normalizeOptions(pathOrOptions) {
    if (typeof pathOrOptions === "string") {
        return {
            path: pathOrOptions,
            updateViaCache: "none",
            autoReload: true,
            debug: false,
            activationTimeoutMs: 10000,
        };
    }
    return {
        path: pathOrOptions.path,
        updateViaCache: pathOrOptions.updateViaCache ?? "none",
        autoReload: pathOrOptions.autoReload ?? true,
        onNeedRefresh: pathOrOptions.onNeedRefresh,
        onUpdated: pathOrOptions.onUpdated,
        debug: pathOrOptions.debug ?? false,
        activationTimeoutMs: pathOrOptions.activationTimeoutMs ?? 10000,
    };
}
function sendSkipWaiting(worker, debug) {
    if (!worker)
        return;
    try {
        worker.postMessage({ type: "SKIP_WAITING" });
        debugLog(debug, "Sent SKIP_WAITING to waiting service worker");
    }
    catch (error) {
        console.warn("Failed to post SKIP_WAITING to service worker", error);
    }
}
function attachUpdateHandlers(registration, options) {
    // Guard: only the first caller per registration attaches handlers.
    // Subsequent calls with different options are silently ignored.
    if (handshakes.has(registration))
        return;
    handshakes.add(registration);
    const { autoReload, onNeedRefresh, onUpdated, activationTimeoutMs, debug } = options;
    let reloadTriggered = false;
    const maybeReload = () => {
        if (reloadTriggered)
            return;
        reloadTriggered = true;
        debugLog(debug, "Service worker controller change detected");
        onUpdated?.();
        if (autoReload &&
            typeof window !== "undefined" &&
            typeof window.location?.reload === "function") {
            window.location.reload();
        }
    };
    const handleWaiting = (worker) => {
        if (!worker)
            return;
        onNeedRefresh?.();
        sendSkipWaiting(worker, debug);
        if (activationTimeoutMs > 0 && typeof window !== "undefined") {
            window.setTimeout(() => {
                if (registration.waiting) {
                    debugLog(debug, "Waiting worker still pending; re-sending SKIP_WAITING");
                    sendSkipWaiting(registration.waiting, debug);
                    registration
                        .update()
                        .catch(() => debugLog(debug, "Service worker update retry failed (timeout path)"));
                }
            }, activationTimeoutMs);
        }
    };
    // Handle an already waiting worker at startup.
    if (registration.waiting) {
        handleWaiting(registration.waiting);
    }
    // Listen for newly installed workers becoming waiting.
    registration.addEventListener("updatefound", () => {
        const installing = registration.installing;
        if (!installing)
            return;
        installing.addEventListener("statechange", () => {
            if (installing.state === "installed") {
                handleWaiting(registration.waiting);
            }
        });
    });
    // Reload (or callback) once the new controller takes over.
    navigator.serviceWorker.addEventListener("controllerchange", maybeReload, {
        once: true,
    });
}
function registerOnce(options) {
    const { path, updateViaCache } = options;
    if (!registrations.has(path)) {
        const registrationPromise = navigator.serviceWorker
            .register(path, { updateViaCache })
            .then(async (registration) => {
            try {
                await registration.update();
            }
            catch (error) {
                console.warn("Service worker update failed; continuing with registration", error);
            }
            return registration;
        })
            .catch((error) => {
            // delete failed registration to allow retrials
            registrations.delete(path);
            throw error;
        });
        registrations.set(path, registrationPromise);
    }
    return registrations.get(path).then((registration) => {
        attachUpdateHandlers(registration, options);
        return registration;
    });
}
/**
 * Registers a service worker for the given path only once, attaches an
 * update/activation handshake (SKIP_WAITING + controllerchange reload), and
 * caches the registration promise for subsequent calls.
 *
 * @param pathOrOptions - Service worker script path or a configuration object.
 * @throws if service workers are not supported or registration fails.
 */
export async function setupServiceWorkerOnce(pathOrOptions) {
    ensureServiceWorkerSupport();
    const options = normalizeOptions(pathOrOptions);
    return registerOnce(options);
}
/**
 * Returns an active service worker instance, optionally ensuring a specific
 * script path is registered before resolving.
 *
 * @param path - Optional service worker script path to register and prefer.
 * @throws if service workers are not supported or no active worker is available.
 */
export async function getActiveServiceWorker(path) {
    ensureServiceWorkerSupport();
    const registration = path
        ? await registerOnce(normalizeOptions(path))
        : await navigator.serviceWorker.ready;
    let serviceWorker = registration.active ||
        registration.waiting ||
        registration.installing ||
        navigator.serviceWorker.controller;
    if (!serviceWorker && path) {
        const readyRegistration = await navigator.serviceWorker.ready;
        serviceWorker =
            readyRegistration.active ||
                readyRegistration.waiting ||
                readyRegistration.installing ||
                navigator.serviceWorker.controller;
    }
    if (!serviceWorker) {
        throw new Error("Service worker not ready yet");
    }
    return serviceWorker;
}
/**
 * Clears the cached registration promises.
 * Intended for tests to reset state between runs.
 */
export const __resetServiceWorkerManager = () => {
    registrations.clear();
    handshakes = new WeakSet();
};
